{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "f5d772b6",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Improved and updated DEG_analysis_062020.ipynb\n",
    "##Ver.031721\n",
    "import pandas as pd\n",
    "import rpy2\n",
    "import rpy2.robjects as ro\n",
    "from rpy2.robjects import pandas2ri, Formula\n",
    "pandas2ri.activate()\n",
    "from rpy2.robjects.packages import importr\n",
    "from rpy2.robjects.conversion import localconverter\n",
    "'''\n",
    "Adopted from:  https://stackoverflow.com/questions/41821100/running-deseq2-through-rpy2\n",
    "and also from: https://gist.github.com/wckdouglas/3f8fb27a3d7a1eb24c598aa04f70fb25\n",
    "Using DESeq2 in python\n",
    "'''\n",
    "\n",
    "import os, sys, time,re\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import matplotlib as mpl\n",
    "import matplotlib.pyplot as plt\n",
    "import matplotlib.font_manager as fm\n",
    "import matplotlib.lines as mlines\n",
    "from matplotlib.gridspec import  GridSpec\n",
    "import seaborn as sns\n",
    "\n",
    "from scipy import stats\n",
    "from scipy.cluster import hierarchy\n",
    "from statsmodels.stats.multitest import multipletests\n",
    "from collections import defaultdict\n",
    "import pysam\n",
    "from collections import defaultdict\n",
    "import statsmodels.api as sm\n",
    "from class_gene import *\n",
    "\n",
    "REFFLAT_hg38      = '/extdata1/baeklab/Doyeon/SARSCOV2/ref/hg38/annotations/r171015.txt'\n",
    "REFFLAT_chlSab2   = '/extdata1/baeklab/Doyeon/SARSCOV2/ref/chlSab2.genes.refFlat.rearranged.txt' #green monkey Vero cell\n",
    "REFFLAT_SARSCOV2      = '/extdata1/baeklab/Doyeon/SARSCOV2/ref/SARSCOV2_annotation_021921.txt' ##updated: correct end position for ORF1a\n",
    "\n",
    "BAMDIR_hostmapped = '/extdata1/baeklab/Doyeon/SARSCOV2/data/%s_hostalign_021721/%s.bam' #e.g. %('mRNASeq','mRNA_2h_rep1') %('RPFSeq','QTI_0h_rep2')\n",
    "BAMDIR_cov2mapped = '/extdata1/baeklab/Doyeon/SARSCOV2/data/%s_SARSCOV2align_021721/%s.bam' \n",
    "RPKMDIR           = '/extdata1/baeklab/Doyeon/SARSCOV2/data/rpkm_021721/%s.txt' #%('mRNA_2h_rep1')\n",
    "\n",
    "GENELISTDIR  = '/extdata1/baeklab/Doyeon/SARSCOV2/ref/%s.txt'\n",
    "\n",
    "RESULTDIR    = '/extdata1/baeklab/Doyeon/SARSCOV2/result/%s' #e.g. %('timecourse_SARSCOV2exp/021721_mRNA_quantification.tsv')\n",
    "FIGDIR       = '/extdata1/baeklab/Doyeon/SARSCOV2/fig/%s' #e.g. %('revision_QC/021721_QTISeq_12h.pdf')\n",
    "fontfiles    = fm.findSystemFonts(fontpaths='/extdata1/baeklab/Doyeon/fonts', fontext='ttf')\n",
    "fontlist     = fm.createFontList(fontfiles)\n",
    "fm.fontManager.ttflist.extend(fontlist)\n",
    "plt.style.use('file:///home/sonic/.config/matplotlib/Doyeon.mplstyle')\n",
    "\n",
    "%matplotlib inline\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1b4bd004",
   "metadata": {},
   "source": [
    "## DESeq R package"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "e82789c5",
   "metadata": {},
   "outputs": [],
   "source": [
    "'''\n",
    "For install DESeq2 packages:\n",
    "open R: ~/baeklab/Hyeonseong/anaconda3/envs/Doyeon-tf2/bin/R\n",
    "> install.packages(\"BiocManager\")\n",
    "> BiocManager::install(\"DESeq2\")\n",
    "loooooooong time required for installing dependent packages\n",
    "'''\n",
    "\n",
    "deseq2 = importr('DESeq2')\n",
    "to_dataframe = ro.r('function(x) data.frame(x)')\n",
    "deseq1 = importr('DESeq')\n",
    "BiocGenerics = importr('BiocGenerics')\n",
    "#print(sorted(list(dir(deseq1))))\n",
    "#sorted(list(dir(BiocGenerics)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "7a72e94a",
   "metadata": {},
   "outputs": [],
   "source": [
    "class py_DESeq1:\n",
    "    '''\n",
    "    DESeq (version 1) object through rpy2\n",
    "    input:\n",
    "    count_matrix: should be a pandas dataframe with each column as count, and a id column for gene id\n",
    "        example:\n",
    "        id    sampleA    sampleB\n",
    "        geneA    5    1\n",
    "        geneB    4    5\n",
    "        geneC    1    2\n",
    "    design_matrix: an design matrix in the form of pandas dataframe, see DESeq1 manual, samplenames as rownames\n",
    "                treatment\n",
    "    sampleA1        A\n",
    "    sampleA2        A\n",
    "    sampleB1        B\n",
    "    sampleB2        B\n",
    "    design_formula: see DESeq1 manual, example: \"~ treatment\"\"\n",
    "    gene_column: column name of gene id columns, example \"id\"\n",
    "    '''\n",
    "    def __init__(self, count_matrix, design_matrix, design_formula, gene_column='id'):\n",
    "        try:\n",
    "            assert gene_column in count_matrix.columns, 'Wrong gene id column name'\n",
    "            gene_id = count_matrix[gene_column]\n",
    "        except AttributeError:\n",
    "            sys.exit('Wrong Pandas dataframe?')\n",
    "\n",
    "        self.cds = None\n",
    "        self.deseq_result = None\n",
    "        self.deseq_result_Robj = None\n",
    "        self.resLFC = None\n",
    "        \n",
    "        self.normalized_count_matrix = None\n",
    "        self.gene_column = gene_column\n",
    "        self.gene_id = count_matrix[self.gene_column]\n",
    "        #renamed_matrix = count_matrix.rename(columns={index: row['treatment']for index,row in design_matrix.iterrows()})\n",
    "        #display(renamed_matrix.drop(gene_column,axis=1).head())\n",
    "        #display(count_matrix.drop(gene_column,axis=1).head())\n",
    "        #display(design_matrix)\n",
    "        with localconverter(ro.default_converter + pandas2ri.converter):\n",
    "            self.count_matrix = ro.conversion.py2rpy(count_matrix.drop(gene_column,axis=1))\n",
    "            self.design_matrix = ro.conversion.py2rpy(design_matrix['treatment'])\n",
    "        self.design_formula = Formula(design_formula)\n",
    "\n",
    "\n",
    "    def run_deseq(self, **kwargs):\n",
    "        \n",
    "        self.cds = deseq1.newCountDataSet(self.count_matrix,self.design_matrix)\n",
    "        self.cds = BiocGenerics.estimateSizeFactors(self.cds)\n",
    "        self.cds = BiocGenerics.estimateDispersions(self.cds,modelFrame=self.design_matrix)\n",
    "        #print(self.cds)\n",
    "        #print(BiocGenerics.counts(self.cds))\n",
    "        #print(BiocGenerics.conditions(self.cds))\n",
    "        self.deseq_result_Robj = deseq1.nbinomTest( self.cds, \"A\", \"B\" )\n",
    "\n",
    "        \n",
    "    def get_deseq_result(self,lfc_shrink=True, **kwargs):\n",
    "        ##For DESeq1, lfc shrink is not available\n",
    "        self.deseq_result = to_dataframe(self.deseq_result_Robj)\n",
    "        #print(type(self.deseq_result))\n",
    "        with localconverter(ro.default_converter + pandas2ri.converter):\n",
    "            self.deseq_result = ro.conversion.rpy2py(self.deseq_result) ## back to pandas dataframe\n",
    "        \n",
    "        self.deseq_result[self.gene_column] = self.gene_id.values\n",
    "        \n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "c6ef5883",
   "metadata": {},
   "outputs": [],
   "source": [
    "class py_DESeq2:\n",
    "    '''\n",
    "    DESeq2 object through rpy2\n",
    "    input:\n",
    "    count_matrix: should be a pandas dataframe with each column as count, and a id column for gene id\n",
    "        example:\n",
    "        id    sampleA    sampleB\n",
    "        geneA    5    1\n",
    "        geneB    4    5\n",
    "        geneC    1    2\n",
    "    design_matrix: an design matrix in the form of pandas dataframe, see DESeq2 manual, samplenames as rownames\n",
    "                treatment\n",
    "    sampleA1        A\n",
    "    sampleA2        A\n",
    "    sampleB1        B\n",
    "    sampleB2        B\n",
    "    design_formula: see DESeq2 manual, example: \"~ treatment\"\"\n",
    "    gene_column: column name of gene id columns, exmplae \"id\"\n",
    "    '''\n",
    "    def __init__(self, count_matrix, design_matrix, design_formula, gene_column='id'):\n",
    "        try:\n",
    "            assert gene_column in count_matrix.columns, 'Wrong gene id column name'\n",
    "            gene_id = count_matrix[gene_column]\n",
    "        except AttributeError:\n",
    "            sys.exit('Wrong Pandas dataframe?')\n",
    "\n",
    "        self.dds = None\n",
    "        self.deseq_result = None\n",
    "        self.deseq_result_Robj = None\n",
    "        self.resLFC = None\n",
    "        self.comparison = None\n",
    "        self.normalized_count_matrix = None\n",
    "        self.gene_column = gene_column\n",
    "        self.gene_id = count_matrix[self.gene_column]\n",
    "        with localconverter(ro.default_converter + pandas2ri.converter):\n",
    "            self.count_matrix = ro.conversion.py2rpy(count_matrix.drop(gene_column,axis=1))\n",
    "            self.design_matrix = ro.conversion.py2rpy(design_matrix)\n",
    "        self.design_formula = Formula(design_formula)\n",
    "\n",
    "\n",
    "    def run_deseq(self, **kwargs):\n",
    "        self.dds = deseq2.DESeqDataSetFromMatrix(countData=self.count_matrix, \n",
    "                                        colData=self.design_matrix,\n",
    "                                        design=self.design_formula)\n",
    "        \n",
    "        \n",
    "        self.dds = deseq2.DESeq(self.dds, **kwargs)\n",
    "        \n",
    "        \n",
    "        \n",
    "\n",
    "    def get_deseq_result(self,lfcshrink=True, **kwargs):\n",
    "\n",
    "        self.comparison = deseq2.resultsNames(self.dds)\n",
    "        if lfcshrink:\n",
    "            self.deseq_result_Robj = deseq2.lfcShrink(self.dds,coef='treatment_B_vs_A',type='apeglm',**kwargs)\n",
    "        else:\n",
    "            self.deseq_result_Robj = deseq2.results(self.dds, **kwargs)\n",
    "        self.deseq_result = to_dataframe(self.deseq_result_Robj)\n",
    "        #print(type(self.deseq_result))\n",
    "        with localconverter(ro.default_converter + pandas2ri.converter):\n",
    "            self.deseq_result = ro.conversion.rpy2py(self.deseq_result) ## back to pandas dataframe\n",
    "        \n",
    "        self.deseq_result[self.gene_column] = self.gene_id.values\n",
    "        \n",
    "#@ri2py.register(DataFrame)\n",
    "#def ri2py_dataframe(obj):\n",
    "#    # items = tuple((k, ri2py(v)) for k, v in obj.items())\n",
    "#    # res = PandasDataFrame.from_items(items)\n",
    "#    items = OrderedDict((k, ri2py(v)) for k, v in obj.items())\n",
    "#    res = PandasDataFrame.from_dict(items)\n",
    "#    return res    "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b38fbc96",
   "metadata": {},
   "source": [
    "## Basic functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "52e8d6af",
   "metadata": {},
   "outputs": [],
   "source": [
    "def load_RPKM(samplename):\n",
    "    '''\n",
    "    id\tsymbol\tef_length\tnsegments1\tnsegments2\tnsegments3\treadcount\treadcount_uniq\texlev\texrnk\texpct\n",
    "    NM_006004\tUQCRH\t276\t18702\t18702\t18494\t9250.416666666679\t32\t27048.868164275482\t1.0\t0.005263989050902774\n",
    "    '''\n",
    "    df = pd.read_csv(RPKMDIR %samplename,sep='\\t')\n",
    "    df = df.rename(columns={'id':'NMID','symbol':'gsym',\n",
    "                            'readcount':'nreads','exlev':'RPKM'})\n",
    "    \n",
    "    df['gsym'] = df['gsym'].str.upper()\n",
    "    \n",
    "    df['nreads'] = df['nreads'].round(0)\n",
    "    df = df[['NMID','gsym','nreads','RPKM','exrnk','expct']]\n",
    "    df = df.set_index('NMID')\n",
    "    return df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "dcb8b33e",
   "metadata": {},
   "source": [
    "## Fig. 5a (Functions for DEG identification and plotting)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f9aa064a",
   "metadata": {},
   "source": [
    "### DEG identification"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "75288b90",
   "metadata": {},
   "outputs": [],
   "source": [
    "## previously identify_DEG_v2 in DEG_analysis_062020.ipynb\n",
    "def identify_DEG(infected=[],uninfected=[], DEG_method=py_DESeq2, verbose=False,\n",
    "                 OutF = ''): #default: DESeq2.\n",
    "    ## infected samples\n",
    "    nread_df      = load_RPKM(infected[0])\n",
    "    infected_nreads_df         = nread_df[['gsym','nreads']]\n",
    "    infected_nreads_df.columns = ['gsym',infected[0]]\n",
    "    infected_expct_df          = nread_df[['gsym','expct']]\n",
    "    infected_expct_df.columns  = ['gsym',infected[0]+'_expct']\n",
    "    for samplename in infected[1:]:\n",
    "        nread_df = load_RPKM(samplename)\n",
    "        infected_nreads_df[samplename]         = nread_df['nreads']\n",
    "        infected_expct_df[samplename+'_expct'] = nread_df['expct']\n",
    "    infected_expct_df = infected_expct_df.set_index(['gsym']).mean(axis=1) #average expct\n",
    "    ##\n",
    "    ## uninfected samples\n",
    "    nread_df      = load_RPKM(uninfected[0])\n",
    "    uninfected_nreads_df         = nread_df[['gsym','nreads']]\n",
    "    uninfected_nreads_df.columns = ['gsym',uninfected[0]]\n",
    "    uninfected_expct_df          = nread_df[['gsym','expct']]\n",
    "    uninfected_expct_df.columns  = ['gsym',uninfected[0]+'_expct']\n",
    "    for samplename in uninfected[1:]:\n",
    "        nread_df = load_RPKM(samplename)\n",
    "        uninfected_nreads_df[samplename]         = nread_df['nreads']\n",
    "        uninfected_expct_df[samplename+'_expct'] = nread_df['expct']\n",
    "    uninfected_expct_df = uninfected_expct_df.set_index(['gsym']).mean(axis=1) #average expct\n",
    "    ##\n",
    "    \n",
    "    #for convenience, replace gsym to index and drop nmid\n",
    "    infected_nreads_df   = infected_nreads_df.reset_index(drop=True)\n",
    "    infected_nreads_df   = infected_nreads_df.set_index(['gsym'])\n",
    "    uninfected_nreads_df = uninfected_nreads_df.reset_index(drop=True)\n",
    "    uninfected_nreads_df = uninfected_nreads_df.set_index(['gsym'])\n",
    "    \n",
    "    count_mat = pd.concat([infected_nreads_df,uninfected_nreads_df],axis=1)\n",
    "    count_mat = count_mat.reset_index()\n",
    "    count_mat = count_mat.rename(columns={'index':'id'})\n",
    "    #display(count_mat.head())\n",
    "    design_mat= pd.DataFrame(['B' for i in range(len(infected))]+['A' for i in range(len(uninfected))],\n",
    "                          index=infected+uninfected)\n",
    "    design_mat.columns = ['treatment']\n",
    "    #display(design_mat)\n",
    "    \n",
    "    DESeq = DEG_method(count_mat,design_mat,design_formula='~ treatment',gene_column='id')\n",
    "    DESeq.run_deseq()\n",
    "    DESeq.get_deseq_result(lfcshrink=True)\n",
    "    \n",
    "    result_df = DESeq.deseq_result\n",
    "    if verbose:\n",
    "        display(result_df.head())\n",
    "    result_df = result_df.set_index(['id'])\n",
    "    result_df['infected_expct']   = infected_expct_df\n",
    "    result_df['uninfected_expct'] = uninfected_expct_df\n",
    "    \n",
    "    if verbose:\n",
    "        display(result_df.head(10))\n",
    "    if OutF != '':\n",
    "        result_df.to_csv(RESULTDIR %OutF, sep = '\\t')\n",
    "    return result_df\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c5eae758",
   "metadata": {},
   "source": [
    "### Plots (MA, volcano)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "615a0b3c",
   "metadata": {},
   "outputs": [],
   "source": [
    "def plotMA(sample_prefix='mRNA',hpi='0h', padj_cut=0.01, log2fc_cut=2.0, OutFigname='',\n",
    "           ax = None, show_fig = False):\n",
    "    InFname = f'DEG/{sample_prefix}_DESeq2_{hpi}.tsv'\n",
    "    df = pd.read_csv(RESULTDIR %InFname, sep ='\\t', header=0, index_col=0)\n",
    "    '''\n",
    "    id\tbaseMean\tlog2FoldChange\tlfcSE\tpvalue\tpadj\tinfection_expct\tuninfection_expct\n",
    "    A1BG\t0.0\t\t\t\t\t83.81389166710532\t87.61120176870033\n",
    "    A1CF\t0.20868225617457992\t0.06273705450357624\t0.5463246183313462\t0.4157015259052005\t\t79.67310627993893\t87.61120176870033\n",
    "    '''\n",
    "    up_df       = df[(df['padj']<padj_cut) & (df['log2FoldChange']>0)]\n",
    "    up_lfc_df   = up_df[(up_df['log2FoldChange']> log2fc_cut)]\n",
    "    up_x_df     = up_df[(up_df['log2FoldChange']<=log2fc_cut)]\n",
    "    down_df     = df[(df['padj']<padj_cut) & (df['log2FoldChange']<0)]\n",
    "    down_lfc_df = down_df[(down_df['log2FoldChange']<  -log2fc_cut)]\n",
    "    down_x_df   = down_df[(down_df['log2FoldChange']>= -log2fc_cut)]\n",
    "    nonsig_df   = df[df['padj']>=padj_cut]\n",
    "    \n",
    "    if ax == None:\n",
    "        fig, ax = plt.subplots(figsize=(4,3))\n",
    "    dot_size = 3\n",
    "    ax.scatter(nonsig_df['baseMean']   ,nonsig_df['log2FoldChange'],   color='grey',   rasterized = True, s = dot_size)\n",
    "    ax.scatter(up_x_df['baseMean']     ,up_x_df['log2FoldChange'],     color='pink',   rasterized = True, s = dot_size)\n",
    "    ax.scatter(down_x_df['baseMean']   ,down_x_df['log2FoldChange'],   color='skyblue',rasterized = True, s = dot_size)\n",
    "    ax.scatter(up_lfc_df['baseMean']   ,up_lfc_df['log2FoldChange'],   color='red',    rasterized = True, s = dot_size)\n",
    "    ax.scatter(down_lfc_df['baseMean'] ,down_lfc_df['log2FoldChange'], color='blue',   rasterized = True, s = dot_size)\n",
    "    ax.set_ylim(-5,5)\n",
    "    ax.set_xscale('log')\n",
    "    ax.set_xlabel('Num. reads')\n",
    "    ax.set_ylabel('log2(fold change)')\n",
    "    ax.set_title('Criteria: P.adj<%.02f, abs(log2fc)>%.01f\\n# upregulated: %d, #downregulated: %d' %(padj_cut,\n",
    "                                                                                                   log2fc_cut,\n",
    "                                                                                                   len(up_lfc_df),\n",
    "                                                                                                   len(down_lfc_df)\n",
    "                                                                                                  ))\n",
    "    if OutFigname != '':\n",
    "        plt.tight_layout()\n",
    "        plt.savefig(FIGDIR %OutFigname)\n",
    "        plt.close()\n",
    "    elif show_fig:\n",
    "        plt.tight_layout()\n",
    "        plt.show()\n",
    "        plt.close()\n",
    "    \n",
    "    return None\n",
    "\n",
    "def plotVolcano(sample_prefix='mRNA',hpi='0h',padj_cut=0.01,log2fc_cut=2.0,\n",
    "                OutFigname='',ax = None, show_fig = False):\n",
    "    InFname = f'DEG/{sample_prefix}_DESeq2_{hpi}.tsv'\n",
    "    df = pd.read_csv(RESULTDIR %InFname, sep ='\\t', header=0, index_col=0)\n",
    "    '''\n",
    "    id\tbaseMean\tlog2FoldChange\tlfcSE\tpvalue\tpadj\tinfection_expct\tuninfection_expct\n",
    "    A1BG\t0.0\t\t\t\t\t83.81389166710532\t87.61120176870033\n",
    "    A1CF\t0.20868225617457992\t0.06273705450357624\t0.5463246183313462\t0.4157015259052005\t\t79.67310627993893\t87.61120176870033\n",
    "    '''\n",
    "    up_df       = df[(df['padj']<padj_cut) & (df['log2FoldChange']>0)]\n",
    "    up_lfc_df   = up_df[(up_df['log2FoldChange']> log2fc_cut)]\n",
    "    up_x_df     = up_df[(up_df['log2FoldChange']<=log2fc_cut)]\n",
    "    down_df     = df[(df['padj']<padj_cut) & (df['log2FoldChange']<0)]\n",
    "    down_lfc_df = down_df[(down_df['log2FoldChange']<  -log2fc_cut)]\n",
    "    down_x_df   = down_df[(down_df['log2FoldChange']>= -log2fc_cut)]\n",
    "    nonsig_df   = df[df['padj']>=padj_cut]\n",
    "    \n",
    "    if ax == None:\n",
    "        fig, ax = plt.subplots(figsize=(3,3))\n",
    "    dot_size = 3\n",
    "    ax.scatter(nonsig_df['log2FoldChange'],  -np.log10(nonsig_df['padj']),    color='k',      rasterized = True, s = dot_size)\n",
    "    ax.scatter(up_x_df['log2FoldChange'],    -np.log10(up_x_df['padj']    ),  color='pink',   rasterized = True, s = dot_size)\n",
    "    ax.scatter(down_x_df['log2FoldChange'],  -np.log10(down_x_df['padj']  ),  color='skyblue',rasterized = True, s = dot_size)\n",
    "    ax.scatter(up_lfc_df['log2FoldChange'],  -np.log10(up_lfc_df['padj']    ),color='red',    rasterized = True, s = dot_size)\n",
    "    ax.scatter(down_lfc_df['log2FoldChange'],-np.log10(down_lfc_df['padj']  ),color='blue',   rasterized = True, s = dot_size)\n",
    "    \n",
    "    #plt.xlabel(r'''$\\mathregular{log_{2}}$ fold changes''')\n",
    "    #plt.ylabel(r'''$\\mathregular{-log_{10}}$ ($\\it{q}$ value)''' )\n",
    "    ax.set_xlabel('log2(fold change)')\n",
    "    ax.set_ylabel('-log10(q value)')\n",
    "    ax.set_xlim(-10,10)\n",
    "    ax.set_ylim(0,100)\n",
    "    '''plt.title('Criteria: P.adj<%.02f, abs(log2fc)>%.01f\\n# upregulated: %d, #downregulated: %d' %(padj_cut,\n",
    "                                                                                                   log2fc_cut,\n",
    "                                                                                                   len(up_lfc_df),\n",
    "                                                                                                   len(down_lfc_df)\n",
    "                                                                                                  ))'''\n",
    "    ax.set_title(hpi)\n",
    "    if OutFigname != '':\n",
    "        plt.tight_layout()\n",
    "        plt.savefig(FIGDIR %OutFigname)\n",
    "        plt.close()\n",
    "    elif show_fig:\n",
    "        plt.tight_layout()\n",
    "        plt.show()\n",
    "        plt.close()\n",
    "    \n",
    "    return None\n",
    "\n",
    "def plot_volcano_multipanel(sample_prefix, hpi_list_list = [['0h','1h','2h','4h'],['12h','16h','24h','36h']],\n",
    "                            padj_cut=0.01,log2fc_cut=2.0,OutFigname=''):\n",
    "    \n",
    "    fig, axes  = plt.subplots(nrows=len(hpi_list_list),           ncols=len(hpi_list_list[0]), \n",
    "                              figsize=(3*(len(hpi_list_list[0])), 3*len(hpi_list_list)))\n",
    "    for idx_row, hpi_list in enumerate(hpi_list_list):\n",
    "        for idx_col, hpi in enumerate(hpi_list):\n",
    "            ax = axes[idx_row,idx_col]\n",
    "            plotVolcano(sample_prefix=sample_prefix,hpi=hpi, padj_cut=padj_cut,log2fc_cut=log2fc_cut,\n",
    "                        OutFigname='',ax = ax, show_fig = False)\n",
    "    if OutFigname != None:\n",
    "        plt.tight_layout()\n",
    "        plt.savefig(FIGDIR %OutFigname)\n",
    "        plt.close()\n",
    "    else:\n",
    "        plt.tight_layout()\n",
    "        plt.show()\n",
    "        plt.close()\n",
    "    return None"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "64b93803",
   "metadata": {},
   "source": [
    "### Fig. 5 and ExtDatFig. 7 (DEG identification and plotting for mRNA, RPF, and QTI)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "b51fce30",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "mRNA\t0h\tWed Mar 31 16:25:19 2021\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/sonic/baeklab/Hyeonseong/anaconda3/envs/Doyeon-tf2/lib/python3.6/site-packages/ipykernel_launcher.py:35: FutureWarning: Sorting because non-concatenation axis is not aligned. A future version\n",
      "of pandas will change to not sort by default.\n",
      "\n",
      "To accept the future behavior, pass 'sort=False'.\n",
      "\n",
      "To retain the current behavior and silence the warning, pass 'sort=True'.\n",
      "\n",
      "R[write to console]: converting counts to integer mode\n",
      "\n",
      "R[write to console]: estimating size factors\n",
      "\n",
      "R[write to console]: estimating dispersions\n",
      "\n",
      "R[write to console]: gene-wise dispersion estimates\n",
      "\n",
      "R[write to console]: mean-dispersion relationship\n",
      "\n",
      "R[write to console]: final dispersion estimates\n",
      "\n",
      "R[write to console]: fitting model and testing\n",
      "\n",
      "R[write to console]: using 'apeglm' for LFC shrinkage. If used in published research, please cite:\n",
      "    Zhu, A., Ibrahim, J.G., Love, M.I. (2018) Heavy-tailed prior distributions for\n",
      "    sequence count data: removing the noise and preserving large differences.\n",
      "    Bioinformatics. https://doi.org/10.1093/bioinformatics/bty895\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "mRNA\t1h\tWed Mar 31 16:25:46 2021\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "R[write to console]: converting counts to integer mode\n",
      "\n",
      "R[write to console]: estimating size factors\n",
      "\n",
      "R[write to console]: estimating dispersions\n",
      "\n",
      "R[write to console]: gene-wise dispersion estimates\n",
      "\n",
      "R[write to console]: mean-dispersion relationship\n",
      "\n",
      "R[write to console]: final dispersion estimates\n",
      "\n",
      "R[write to console]: fitting model and testing\n",
      "\n",
      "R[write to console]: using 'apeglm' for LFC shrinkage. If used in published research, please cite:\n",
      "    Zhu, A., Ibrahim, J.G., Love, M.I. (2018) Heavy-tailed prior distributions for\n",
      "    sequence count data: removing the noise and preserving large differences.\n",
      "    Bioinformatics. https://doi.org/10.1093/bioinformatics/bty895\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "mRNA\t2h\tWed Mar 31 16:26:08 2021\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "R[write to console]: converting counts to integer mode\n",
      "\n",
      "R[write to console]: estimating size factors\n",
      "\n",
      "R[write to console]: estimating dispersions\n",
      "\n",
      "R[write to console]: gene-wise dispersion estimates\n",
      "\n",
      "R[write to console]: mean-dispersion relationship\n",
      "\n",
      "R[write to console]: final dispersion estimates\n",
      "\n",
      "R[write to console]: fitting model and testing\n",
      "\n",
      "R[write to console]: using 'apeglm' for LFC shrinkage. If used in published research, please cite:\n",
      "    Zhu, A., Ibrahim, J.G., Love, M.I. (2018) Heavy-tailed prior distributions for\n",
      "    sequence count data: removing the noise and preserving large differences.\n",
      "    Bioinformatics. https://doi.org/10.1093/bioinformatics/bty895\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "mRNA\t4h\tWed Mar 31 16:26:29 2021\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "R[write to console]: converting counts to integer mode\n",
      "\n",
      "R[write to console]: estimating size factors\n",
      "\n",
      "R[write to console]: estimating dispersions\n",
      "\n",
      "R[write to console]: gene-wise dispersion estimates\n",
      "\n",
      "R[write to console]: mean-dispersion relationship\n",
      "\n",
      "R[write to console]: final dispersion estimates\n",
      "\n",
      "R[write to console]: fitting model and testing\n",
      "\n",
      "R[write to console]: using 'apeglm' for LFC shrinkage. If used in published research, please cite:\n",
      "    Zhu, A., Ibrahim, J.G., Love, M.I. (2018) Heavy-tailed prior distributions for\n",
      "    sequence count data: removing the noise and preserving large differences.\n",
      "    Bioinformatics. https://doi.org/10.1093/bioinformatics/bty895\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "mRNA\t12h\tWed Mar 31 16:26:50 2021\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "R[write to console]: converting counts to integer mode\n",
      "\n",
      "R[write to console]: estimating size factors\n",
      "\n",
      "R[write to console]: estimating dispersions\n",
      "\n",
      "R[write to console]: gene-wise dispersion estimates\n",
      "\n",
      "R[write to console]: mean-dispersion relationship\n",
      "\n",
      "R[write to console]: final dispersion estimates\n",
      "\n",
      "R[write to console]: fitting model and testing\n",
      "\n",
      "R[write to console]: using 'apeglm' for LFC shrinkage. If used in published research, please cite:\n",
      "    Zhu, A., Ibrahim, J.G., Love, M.I. (2018) Heavy-tailed prior distributions for\n",
      "    sequence count data: removing the noise and preserving large differences.\n",
      "    Bioinformatics. https://doi.org/10.1093/bioinformatics/bty895\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "mRNA\t16h\tWed Mar 31 16:27:10 2021\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "R[write to console]: converting counts to integer mode\n",
      "\n",
      "R[write to console]: estimating size factors\n",
      "\n",
      "R[write to console]: estimating dispersions\n",
      "\n",
      "R[write to console]: gene-wise dispersion estimates\n",
      "\n",
      "R[write to console]: mean-dispersion relationship\n",
      "\n",
      "R[write to console]: final dispersion estimates\n",
      "\n",
      "R[write to console]: fitting model and testing\n",
      "\n",
      "R[write to console]: using 'apeglm' for LFC shrinkage. If used in published research, please cite:\n",
      "    Zhu, A., Ibrahim, J.G., Love, M.I. (2018) Heavy-tailed prior distributions for\n",
      "    sequence count data: removing the noise and preserving large differences.\n",
      "    Bioinformatics. https://doi.org/10.1093/bioinformatics/bty895\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "mRNA\t24h\tWed Mar 31 16:27:31 2021\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "R[write to console]: converting counts to integer mode\n",
      "\n",
      "R[write to console]: estimating size factors\n",
      "\n",
      "R[write to console]: estimating dispersions\n",
      "\n",
      "R[write to console]: gene-wise dispersion estimates\n",
      "\n",
      "R[write to console]: mean-dispersion relationship\n",
      "\n",
      "R[write to console]: final dispersion estimates\n",
      "\n",
      "R[write to console]: fitting model and testing\n",
      "\n",
      "R[write to console]: using 'apeglm' for LFC shrinkage. If used in published research, please cite:\n",
      "    Zhu, A., Ibrahim, J.G., Love, M.I. (2018) Heavy-tailed prior distributions for\n",
      "    sequence count data: removing the noise and preserving large differences.\n",
      "    Bioinformatics. https://doi.org/10.1093/bioinformatics/bty895\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "mRNA\t36h\tWed Mar 31 16:27:52 2021\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "R[write to console]: converting counts to integer mode\n",
      "\n",
      "R[write to console]: estimating size factors\n",
      "\n",
      "R[write to console]: estimating dispersions\n",
      "\n",
      "R[write to console]: gene-wise dispersion estimates\n",
      "\n",
      "R[write to console]: mean-dispersion relationship\n",
      "\n",
      "R[write to console]: final dispersion estimates\n",
      "\n",
      "R[write to console]: fitting model and testing\n",
      "\n",
      "R[write to console]: using 'apeglm' for LFC shrinkage. If used in published research, please cite:\n",
      "    Zhu, A., Ibrahim, J.G., Love, M.I. (2018) Heavy-tailed prior distributions for\n",
      "    sequence count data: removing the noise and preserving large differences.\n",
      "    Bioinformatics. https://doi.org/10.1093/bioinformatics/bty895\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "mRNA\t48h\tWed Mar 31 16:28:14 2021\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "R[write to console]: converting counts to integer mode\n",
      "\n",
      "R[write to console]: estimating size factors\n",
      "\n",
      "R[write to console]: estimating dispersions\n",
      "\n",
      "R[write to console]: gene-wise dispersion estimates\n",
      "\n",
      "R[write to console]: mean-dispersion relationship\n",
      "\n",
      "R[write to console]: final dispersion estimates\n",
      "\n",
      "R[write to console]: fitting model and testing\n",
      "\n",
      "R[write to console]: using 'apeglm' for LFC shrinkage. If used in published research, please cite:\n",
      "    Zhu, A., Ibrahim, J.G., Love, M.I. (2018) Heavy-tailed prior distributions for\n",
      "    sequence count data: removing the noise and preserving large differences.\n",
      "    Bioinformatics. https://doi.org/10.1093/bioinformatics/bty895\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "RPF\t0h\tWed Mar 31 16:28:38 2021\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "R[write to console]: converting counts to integer mode\n",
      "\n",
      "R[write to console]: estimating size factors\n",
      "\n",
      "R[write to console]: estimating dispersions\n",
      "\n",
      "R[write to console]: gene-wise dispersion estimates\n",
      "\n",
      "R[write to console]: mean-dispersion relationship\n",
      "\n",
      "R[write to console]: final dispersion estimates\n",
      "\n",
      "R[write to console]: fitting model and testing\n",
      "\n",
      "R[write to console]: using 'apeglm' for LFC shrinkage. If used in published research, please cite:\n",
      "    Zhu, A., Ibrahim, J.G., Love, M.I. (2018) Heavy-tailed prior distributions for\n",
      "    sequence count data: removing the noise and preserving large differences.\n",
      "    Bioinformatics. https://doi.org/10.1093/bioinformatics/bty895\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "RPF\t1h\tWed Mar 31 16:28:56 2021\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "R[write to console]: converting counts to integer mode\n",
      "\n",
      "R[write to console]: estimating size factors\n",
      "\n",
      "R[write to console]: estimating dispersions\n",
      "\n",
      "R[write to console]: gene-wise dispersion estimates\n",
      "\n",
      "R[write to console]: mean-dispersion relationship\n",
      "\n",
      "R[write to console]: final dispersion estimates\n",
      "\n",
      "R[write to console]: fitting model and testing\n",
      "\n",
      "R[write to console]: using 'apeglm' for LFC shrinkage. If used in published research, please cite:\n",
      "    Zhu, A., Ibrahim, J.G., Love, M.I. (2018) Heavy-tailed prior distributions for\n",
      "    sequence count data: removing the noise and preserving large differences.\n",
      "    Bioinformatics. https://doi.org/10.1093/bioinformatics/bty895\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "RPF\t2h\tWed Mar 31 16:29:12 2021\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "R[write to console]: converting counts to integer mode\n",
      "\n",
      "R[write to console]: estimating size factors\n",
      "\n",
      "R[write to console]: estimating dispersions\n",
      "\n",
      "R[write to console]: gene-wise dispersion estimates\n",
      "\n",
      "R[write to console]: mean-dispersion relationship\n",
      "\n",
      "R[write to console]: final dispersion estimates\n",
      "\n",
      "R[write to console]: fitting model and testing\n",
      "\n",
      "R[write to console]: using 'apeglm' for LFC shrinkage. If used in published research, please cite:\n",
      "    Zhu, A., Ibrahim, J.G., Love, M.I. (2018) Heavy-tailed prior distributions for\n",
      "    sequence count data: removing the noise and preserving large differences.\n",
      "    Bioinformatics. https://doi.org/10.1093/bioinformatics/bty895\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "RPF\t4h\tWed Mar 31 16:29:29 2021\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "R[write to console]: converting counts to integer mode\n",
      "\n",
      "R[write to console]: estimating size factors\n",
      "\n",
      "R[write to console]: estimating dispersions\n",
      "\n",
      "R[write to console]: gene-wise dispersion estimates\n",
      "\n",
      "R[write to console]: mean-dispersion relationship\n",
      "\n",
      "R[write to console]: final dispersion estimates\n",
      "\n",
      "R[write to console]: fitting model and testing\n",
      "\n",
      "R[write to console]: using 'apeglm' for LFC shrinkage. If used in published research, please cite:\n",
      "    Zhu, A., Ibrahim, J.G., Love, M.I. (2018) Heavy-tailed prior distributions for\n",
      "    sequence count data: removing the noise and preserving large differences.\n",
      "    Bioinformatics. https://doi.org/10.1093/bioinformatics/bty895\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "RPF\t12h\tWed Mar 31 16:29:44 2021\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "R[write to console]: converting counts to integer mode\n",
      "\n",
      "R[write to console]: estimating size factors\n",
      "\n",
      "R[write to console]: estimating dispersions\n",
      "\n",
      "R[write to console]: gene-wise dispersion estimates\n",
      "\n",
      "R[write to console]: mean-dispersion relationship\n",
      "\n",
      "R[write to console]: final dispersion estimates\n",
      "\n",
      "R[write to console]: fitting model and testing\n",
      "\n",
      "R[write to console]: using 'apeglm' for LFC shrinkage. If used in published research, please cite:\n",
      "    Zhu, A., Ibrahim, J.G., Love, M.I. (2018) Heavy-tailed prior distributions for\n",
      "    sequence count data: removing the noise and preserving large differences.\n",
      "    Bioinformatics. https://doi.org/10.1093/bioinformatics/bty895\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "RPF\t16h\tWed Mar 31 16:30:00 2021\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "R[write to console]: converting counts to integer mode\n",
      "\n",
      "R[write to console]: estimating size factors\n",
      "\n",
      "R[write to console]: estimating dispersions\n",
      "\n",
      "R[write to console]: gene-wise dispersion estimates\n",
      "\n",
      "R[write to console]: mean-dispersion relationship\n",
      "\n",
      "R[write to console]: final dispersion estimates\n",
      "\n",
      "R[write to console]: fitting model and testing\n",
      "\n",
      "R[write to console]: using 'apeglm' for LFC shrinkage. If used in published research, please cite:\n",
      "    Zhu, A., Ibrahim, J.G., Love, M.I. (2018) Heavy-tailed prior distributions for\n",
      "    sequence count data: removing the noise and preserving large differences.\n",
      "    Bioinformatics. https://doi.org/10.1093/bioinformatics/bty895\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "RPF\t24h\tWed Mar 31 16:30:16 2021\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "R[write to console]: converting counts to integer mode\n",
      "\n",
      "R[write to console]: estimating size factors\n",
      "\n",
      "R[write to console]: estimating dispersions\n",
      "\n",
      "R[write to console]: gene-wise dispersion estimates\n",
      "\n",
      "R[write to console]: mean-dispersion relationship\n",
      "\n",
      "R[write to console]: final dispersion estimates\n",
      "\n",
      "R[write to console]: fitting model and testing\n",
      "\n",
      "R[write to console]: using 'apeglm' for LFC shrinkage. If used in published research, please cite:\n",
      "    Zhu, A., Ibrahim, J.G., Love, M.I. (2018) Heavy-tailed prior distributions for\n",
      "    sequence count data: removing the noise and preserving large differences.\n",
      "    Bioinformatics. https://doi.org/10.1093/bioinformatics/bty895\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "RPF\t36h\tWed Mar 31 16:30:32 2021\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "R[write to console]: converting counts to integer mode\n",
      "\n",
      "R[write to console]: estimating size factors\n",
      "\n",
      "R[write to console]: estimating dispersions\n",
      "\n",
      "R[write to console]: gene-wise dispersion estimates\n",
      "\n",
      "R[write to console]: mean-dispersion relationship\n",
      "\n",
      "R[write to console]: final dispersion estimates\n",
      "\n",
      "R[write to console]: fitting model and testing\n",
      "\n",
      "R[write to console]: using 'apeglm' for LFC shrinkage. If used in published research, please cite:\n",
      "    Zhu, A., Ibrahim, J.G., Love, M.I. (2018) Heavy-tailed prior distributions for\n",
      "    sequence count data: removing the noise and preserving large differences.\n",
      "    Bioinformatics. https://doi.org/10.1093/bioinformatics/bty895\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "RPF\t48h\tWed Mar 31 16:30:48 2021\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "R[write to console]: converting counts to integer mode\n",
      "\n",
      "R[write to console]: estimating size factors\n",
      "\n",
      "R[write to console]: estimating dispersions\n",
      "\n",
      "R[write to console]: gene-wise dispersion estimates\n",
      "\n",
      "R[write to console]: mean-dispersion relationship\n",
      "\n",
      "R[write to console]: final dispersion estimates\n",
      "\n",
      "R[write to console]: fitting model and testing\n",
      "\n",
      "R[write to console]: using 'apeglm' for LFC shrinkage. If used in published research, please cite:\n",
      "    Zhu, A., Ibrahim, J.G., Love, M.I. (2018) Heavy-tailed prior distributions for\n",
      "    sequence count data: removing the noise and preserving large differences.\n",
      "    Bioinformatics. https://doi.org/10.1093/bioinformatics/bty895\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "QTI\t0h\tWed Mar 31 16:31:07 2021\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "R[write to console]: converting counts to integer mode\n",
      "\n",
      "R[write to console]: estimating size factors\n",
      "\n",
      "R[write to console]: estimating dispersions\n",
      "\n",
      "R[write to console]: gene-wise dispersion estimates\n",
      "\n",
      "R[write to console]: mean-dispersion relationship\n",
      "\n",
      "R[write to console]: final dispersion estimates\n",
      "\n",
      "R[write to console]: fitting model and testing\n",
      "\n",
      "R[write to console]: using 'apeglm' for LFC shrinkage. If used in published research, please cite:\n",
      "    Zhu, A., Ibrahim, J.G., Love, M.I. (2018) Heavy-tailed prior distributions for\n",
      "    sequence count data: removing the noise and preserving large differences.\n",
      "    Bioinformatics. https://doi.org/10.1093/bioinformatics/bty895\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "QTI\t1h\tWed Mar 31 16:31:25 2021\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "R[write to console]: converting counts to integer mode\n",
      "\n",
      "R[write to console]: estimating size factors\n",
      "\n",
      "R[write to console]: estimating dispersions\n",
      "\n",
      "R[write to console]: gene-wise dispersion estimates\n",
      "\n",
      "R[write to console]: mean-dispersion relationship\n",
      "\n",
      "R[write to console]: final dispersion estimates\n",
      "\n",
      "R[write to console]: fitting model and testing\n",
      "\n",
      "R[write to console]: using 'apeglm' for LFC shrinkage. If used in published research, please cite:\n",
      "    Zhu, A., Ibrahim, J.G., Love, M.I. (2018) Heavy-tailed prior distributions for\n",
      "    sequence count data: removing the noise and preserving large differences.\n",
      "    Bioinformatics. https://doi.org/10.1093/bioinformatics/bty895\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "QTI\t2h\tWed Mar 31 16:31:40 2021\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "R[write to console]: converting counts to integer mode\n",
      "\n",
      "R[write to console]: estimating size factors\n",
      "\n",
      "R[write to console]: estimating dispersions\n",
      "\n",
      "R[write to console]: gene-wise dispersion estimates\n",
      "\n",
      "R[write to console]: mean-dispersion relationship\n",
      "\n",
      "R[write to console]: final dispersion estimates\n",
      "\n",
      "R[write to console]: fitting model and testing\n",
      "\n",
      "R[write to console]: using 'apeglm' for LFC shrinkage. If used in published research, please cite:\n",
      "    Zhu, A., Ibrahim, J.G., Love, M.I. (2018) Heavy-tailed prior distributions for\n",
      "    sequence count data: removing the noise and preserving large differences.\n",
      "    Bioinformatics. https://doi.org/10.1093/bioinformatics/bty895\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "QTI\t4h\tWed Mar 31 16:31:56 2021\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "R[write to console]: converting counts to integer mode\n",
      "\n",
      "R[write to console]: estimating size factors\n",
      "\n",
      "R[write to console]: estimating dispersions\n",
      "\n",
      "R[write to console]: gene-wise dispersion estimates\n",
      "\n",
      "R[write to console]: mean-dispersion relationship\n",
      "\n",
      "R[write to console]: final dispersion estimates\n",
      "\n",
      "R[write to console]: fitting model and testing\n",
      "\n",
      "R[write to console]: using 'apeglm' for LFC shrinkage. If used in published research, please cite:\n",
      "    Zhu, A., Ibrahim, J.G., Love, M.I. (2018) Heavy-tailed prior distributions for\n",
      "    sequence count data: removing the noise and preserving large differences.\n",
      "    Bioinformatics. https://doi.org/10.1093/bioinformatics/bty895\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "QTI\t12h\tWed Mar 31 16:32:12 2021\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "R[write to console]: converting counts to integer mode\n",
      "\n",
      "R[write to console]: estimating size factors\n",
      "\n",
      "R[write to console]: estimating dispersions\n",
      "\n",
      "R[write to console]: gene-wise dispersion estimates\n",
      "\n",
      "R[write to console]: mean-dispersion relationship\n",
      "\n",
      "R[write to console]: final dispersion estimates\n",
      "\n",
      "R[write to console]: fitting model and testing\n",
      "\n",
      "R[write to console]: using 'apeglm' for LFC shrinkage. If used in published research, please cite:\n",
      "    Zhu, A., Ibrahim, J.G., Love, M.I. (2018) Heavy-tailed prior distributions for\n",
      "    sequence count data: removing the noise and preserving large differences.\n",
      "    Bioinformatics. https://doi.org/10.1093/bioinformatics/bty895\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "QTI\t16h\tWed Mar 31 16:32:28 2021\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "R[write to console]: converting counts to integer mode\n",
      "\n",
      "R[write to console]: estimating size factors\n",
      "\n",
      "R[write to console]: estimating dispersions\n",
      "\n",
      "R[write to console]: gene-wise dispersion estimates\n",
      "\n",
      "R[write to console]: mean-dispersion relationship\n",
      "\n",
      "R[write to console]: final dispersion estimates\n",
      "\n",
      "R[write to console]: fitting model and testing\n",
      "\n",
      "R[write to console]: using 'apeglm' for LFC shrinkage. If used in published research, please cite:\n",
      "    Zhu, A., Ibrahim, J.G., Love, M.I. (2018) Heavy-tailed prior distributions for\n",
      "    sequence count data: removing the noise and preserving large differences.\n",
      "    Bioinformatics. https://doi.org/10.1093/bioinformatics/bty895\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "QTI\t24h\tWed Mar 31 16:32:44 2021\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "R[write to console]: converting counts to integer mode\n",
      "\n",
      "R[write to console]: estimating size factors\n",
      "\n",
      "R[write to console]: estimating dispersions\n",
      "\n",
      "R[write to console]: gene-wise dispersion estimates\n",
      "\n",
      "R[write to console]: mean-dispersion relationship\n",
      "\n",
      "R[write to console]: final dispersion estimates\n",
      "\n",
      "R[write to console]: fitting model and testing\n",
      "\n",
      "R[write to console]: using 'apeglm' for LFC shrinkage. If used in published research, please cite:\n",
      "    Zhu, A., Ibrahim, J.G., Love, M.I. (2018) Heavy-tailed prior distributions for\n",
      "    sequence count data: removing the noise and preserving large differences.\n",
      "    Bioinformatics. https://doi.org/10.1093/bioinformatics/bty895\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "QTI\t36h\tWed Mar 31 16:33:02 2021\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "R[write to console]: converting counts to integer mode\n",
      "\n",
      "R[write to console]: estimating size factors\n",
      "\n",
      "R[write to console]: estimating dispersions\n",
      "\n",
      "R[write to console]: gene-wise dispersion estimates\n",
      "\n",
      "R[write to console]: mean-dispersion relationship\n",
      "\n",
      "R[write to console]: final dispersion estimates\n",
      "\n",
      "R[write to console]: fitting model and testing\n",
      "\n",
      "R[write to console]: using 'apeglm' for LFC shrinkage. If used in published research, please cite:\n",
      "    Zhu, A., Ibrahim, J.G., Love, M.I. (2018) Heavy-tailed prior distributions for\n",
      "    sequence count data: removing the noise and preserving large differences.\n",
      "    Bioinformatics. https://doi.org/10.1093/bioinformatics/bty895\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "QTI\t48h\tWed Mar 31 16:33:18 2021\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "R[write to console]: converting counts to integer mode\n",
      "\n",
      "R[write to console]: estimating size factors\n",
      "\n",
      "R[write to console]: estimating dispersions\n",
      "\n",
      "R[write to console]: gene-wise dispersion estimates\n",
      "\n",
      "R[write to console]: mean-dispersion relationship\n",
      "\n",
      "R[write to console]: final dispersion estimates\n",
      "\n",
      "R[write to console]: fitting model and testing\n",
      "\n",
      "R[write to console]: using 'apeglm' for LFC shrinkage. If used in published research, please cite:\n",
      "    Zhu, A., Ibrahim, J.G., Love, M.I. (2018) Heavy-tailed prior distributions for\n",
      "    sequence count data: removing the noise and preserving large differences.\n",
      "    Bioinformatics. https://doi.org/10.1093/bioinformatics/bty895\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "\"plotMA(sample_prefix='RPF',hpi='36h', padj_cut=0.01, log2fc_cut=2.0, OutFigname='',\\n           ax = None, show_fig = True)\\nplotVolcano(sample_prefix='RPF',hpi='36h',padj_cut=0.01,log2fc_cut=2.0,\\n                OutFigname='',ax = None, show_fig = True)\""
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#DESeq2 for mRNA, RPF, and QTI\n",
    "tmp_hpi_list = ['%dh' %i for i in [0,1,2,4,12,16,24,36,48]]\n",
    "tmp_sample_prefix_list = ['mRNA','RPF','QTI']\n",
    "\n",
    "for sample_prefix in tmp_sample_prefix_list:\n",
    "    if sample_prefix == 'mRNA':\n",
    "        enumerated_prefix = ['RPFpaired','QTIpaired']\n",
    "    else:\n",
    "        enumerated_prefix = [sample_prefix]\n",
    "    for hpi in tmp_hpi_list:\n",
    "        print(sample_prefix,hpi,time.ctime(),sep='\\t')\n",
    "        if hpi == '48h':\n",
    "            reps = 3\n",
    "            uninfected_name = 'uninfected48h'\n",
    "        else:\n",
    "            reps = 2\n",
    "            uninfected_name = 'uninfected'\n",
    "        infected   = []\n",
    "        uninfected = []\n",
    "        for prefix in enumerated_prefix:\n",
    "            for rep in range(1, reps+1):\n",
    "                infected.append(  f'{prefix}_{hpi}_rep{rep}')\n",
    "                uninfected.append(f'{prefix}_{uninfected_name}_rep{rep}')\n",
    "        OutF = f'DEG/{sample_prefix}_DESeq2_{hpi}.tsv'\n",
    "        _= identify_DEG(infected=infected,uninfected=uninfected, \n",
    "                        DEG_method=py_DESeq2, verbose= False, OutF = OutF)\n",
    "                \n",
    "'''plotMA(sample_prefix='RPF',hpi='36h', padj_cut=0.01, log2fc_cut=2.0, OutFigname='',\n",
    "           ax = None, show_fig = True)\n",
    "plotVolcano(sample_prefix='RPF',hpi='36h',padj_cut=0.01,log2fc_cut=2.0,\n",
    "                OutFigname='',ax = None, show_fig = True)'''\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "17a1d309",
   "metadata": {},
   "outputs": [],
   "source": [
    "#volcano plot\n",
    "'''Fig. 5: RPF'''\n",
    "plot_volcano_multipanel('RPF', hpi_list_list = [['0h','1h','2h','4h'],['12h','16h','24h','36h']],\n",
    "                        padj_cut=0.01,log2fc_cut=2.0,OutFigname='DEG/Fig5_RPF_volcano.pdf')\n",
    "'''ExtDatFig. 7: mRNA, QTI'''\n",
    "plot_volcano_multipanel('mRNA', hpi_list_list = [['0h','1h','2h','4h'],['12h','16h','24h','36h']],\n",
    "                        padj_cut=0.01,log2fc_cut=2.0,OutFigname='DEG/ExtDatFig7a_mRNA_volcano.pdf')\n",
    "plot_volcano_multipanel('QTI', hpi_list_list = [['0h','1h','2h','4h'],['12h','16h','24h','36h']],\n",
    "                        padj_cut=0.01,log2fc_cut=2.0,OutFigname='DEG/ExtDatFig7a_QTI_volcano.pdf')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bb84c486",
   "metadata": {},
   "source": [
    "## Fig. 5b-g (hierarchical clustering of DEGs, and time-course expression)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "eea777b3",
   "metadata": {},
   "outputs": [],
   "source": [
    "def timecourse_plot(DEG_df_list,hpi_list, padj_cut=0.01,log2fc_cut=2, clip = 3,\n",
    "                    ylim = 8, n_cluster=5, as_average=False, \n",
    "                    load_precalc_clusters=False, clusterFname='',\n",
    "                    clusters_to_show=[],clusters_priority=[], highlight_gene_info_df = pd.DataFrame(),\n",
    "                    cl_method='ward',cl_metric='euclidean',\n",
    "                    ax = None, OutFigname='',show_fig=True, verbose=False):\n",
    "    if OutFigname != '':\n",
    "        assert ('timecourse.' in OutFigname)\n",
    "    if not(type(ylim) in [list,tuple]):\n",
    "        ylim = (-ylim,ylim)\n",
    "    up_DEG_list   = []\n",
    "    down_DEG_list = []\n",
    "    timecourse_df = pd.DataFrame(columns=hpi_list,     index=DEG_df_list[0].index) #index: gsym (id)\n",
    "    DEG_summary_df= pd.DataFrame(columns=['up','down'],index = hpi_list)\n",
    "    \n",
    "    for hpi,DEG_df in zip(hpi_list,DEG_df_list):\n",
    "        df = DEG_df.copy()\n",
    "        up_DEG   = list(df[(df['padj']<padj_cut) & \\\n",
    "                           (df['log2FoldChange']>log2fc_cut) &\\\n",
    "                           ((df['infected_expct']<50) |(df['uninfected_expct']<50))\n",
    "                          ].index)\n",
    "        down_DEG = list(df[(df['padj']<padj_cut) & \\\n",
    "                           (df['log2FoldChange']<-log2fc_cut) &\\\n",
    "                           ((df['infected_expct']<50) |(df['uninfected_expct']<50))\n",
    "                          ].index)\n",
    "        up_DEG_list   += up_DEG\n",
    "        down_DEG_list += down_DEG\n",
    "        DEG_summary_df.loc[hpi] = [len(up_DEG),len(down_DEG)]\n",
    "        timecourse_df[hpi]   = df['log2FoldChange']\n",
    "    \n",
    "    DEG_list = list(set(up_DEG_list)|set(down_DEG_list))\n",
    "    if verbose:\n",
    "        print('summary of DEGs')\n",
    "        print('total #:', len(DEG_list), sep = '\\t')\n",
    "        display(DEG_summary_df)\n",
    "        print('Expression pattern of ACE2')\n",
    "        display(timecourse_df.loc['ACE2'])\n",
    "    \n",
    "    timecourse_df = timecourse_df.fillna(0.0)\n",
    "    if load_precalc_clusters:\n",
    "        ID_cl_df = pd.read_csv(RESULTDIR %clusterFname, sep = '\\t', header = 0, index_col = 0)\n",
    "    else:\n",
    "        ID_cl_df = cluster_DEGs(timecourse_df,DEG_list, cl_method=cl_method,############################\n",
    "                                  cl_metric=cl_metric, n_cluster=n_cluster, clip=clip,\n",
    "                                  OutFigname=OutFigname.replace('timecourse.pdf','hclust.pdf'),show_fig=show_fig)\n",
    "        ID_cl_df.to_csv(RESULTDIR %clusterFname,\n",
    "                        sep = '\\t')\n",
    "    np.random.seed(2020)\n",
    "    sampled_cloud = np.random.choice(sorted(list(set(timecourse_df.index)- (set(up_DEG_list)|set(down_DEG_list)))), 100)\n",
    "    \n",
    "    ####Colors for different clusters\n",
    "    if ax == None:\n",
    "        fig, ax = plt.subplots(figsize=(3.5,3))\n",
    "    \n",
    "    palette = sns.color_palette('bright',10) ###########CAUTION: HARD-CODED (max 6 colors)\n",
    "    palette = [color for i,color in enumerate(palette) if not(i in [0,3,5,7])]\n",
    "    \n",
    "    if clusters_to_show == []:\n",
    "        clusters_to_show = list(range(1,1+n_cluster))\n",
    "    \n",
    "    \n",
    "    if as_average:\n",
    "        ylim = (-5,5)\n",
    "        for DEG in DEG_list:\n",
    "            ax.plot(timecourse_df.loc[DEG],color='grey',alpha=0.2,linewidth=0.5)\n",
    "            \n",
    "        for cluster in clusters_to_show:\n",
    "            color   = palette[cluster-1]\n",
    "            DEGS_within_cluster  = [DEG for DEG in DEG_list if ID_cl_df.loc[DEG,'cluster'] == cluster]\n",
    "            lfc_df  = timecourse_df.loc[DEGS_within_cluster]\n",
    "            avg_lfc = lfc_df.mean(axis=0)\n",
    "            std_lfc = lfc_df.std( axis=0)\n",
    "            ax.plot(avg_lfc,color=color,linewidth=1)\n",
    "            ax.fill_between(np.arange(len(avg_lfc)), \n",
    "                            avg_lfc-std_lfc,avg_lfc+std_lfc, color=color, alpha=0.7)\n",
    "        \n",
    "    else:\n",
    "        if len(highlight_gene_info_df) == 0:\n",
    "            grey_color = False\n",
    "        else:\n",
    "            grey_color = True\n",
    "        for DEG in DEG_list:\n",
    "            cluster = ID_cl_df.loc[DEG,'cluster']\n",
    "            if grey_color:\n",
    "                color  = 'grey'\n",
    "                alpha  = 0.2\n",
    "            else:\n",
    "                color  = palette[cluster-1]\n",
    "                alpha  = 1.0\n",
    "                \n",
    "            if cluster in clusters_to_show:\n",
    "                if not (cluster in clusters_priority):\n",
    "                    ax.plot(timecourse_df.loc[DEG],color=color,linewidth=0.5, alpha = alpha)#, rasterized = True)\n",
    "                    \n",
    "        ## clusters to draw last\n",
    "        for p_cluster in clusters_priority:\n",
    "            for DEG in DEG_list:\n",
    "                cluster = ID_cl_df.loc[DEG,'cluster']\n",
    "                if grey_color:\n",
    "                    color  = 'grey'\n",
    "                    alpha  = 0.2\n",
    "                else:\n",
    "                    color  = palette[cluster-1]\n",
    "                    alpha  = 1.0\n",
    "                if cluster == p_cluster:\n",
    "                    ax.plot(timecourse_df.loc[DEG],color=color,linewidth=0.5, alpha = alpha)#, rasterized = True)\n",
    "        \n",
    "        highlight_summary_df = pd.DataFrame(columns = ['lfc_36h','lfc_maxabs']) \n",
    "        #highlight_gene_info_df = pd.DataFrame(columns = ['color','linewidth','print_cutoff','genelist']) index: genelist_name\n",
    "        for genelist_name, row in highlight_gene_info_df.iterrows():\n",
    "            color        = row['color']\n",
    "            linewidth    = row['linewidth']\n",
    "            print_cutoff = row['print_cutoff']\n",
    "            highlight_genelist = row['genelist']\n",
    "            \n",
    "            for gsym in highlight_genelist: ##for special markup, Fig. 6\n",
    "                if gsym in timecourse_df.index:\n",
    "                    ax.plot(timecourse_df.loc[gsym], color=color, linewidth=linewidth)#, rasterized = True)\n",
    "                    lfc_maxabs = timecourse_df.loc[gsym].abs().max()\n",
    "                    if lfc_maxabs>print_cutoff:\n",
    "                        lfc_36h = timecourse_df.loc[gsym,str(hpi_list[-1])]\n",
    "                        loc_y   = np.clip(lfc_36h,*ylim)\n",
    "                        \n",
    "                        ax.text(len(hpi_list)-1,loc_y,gsym)\n",
    "                        highlight_summary_df.loc[gsym] = [lfc_36h, lfc_maxabs]\n",
    "                    \n",
    "        \n",
    "        if verbose:\n",
    "            print('highlight genes with >1.5 max(|log2fc|)')\n",
    "            display(highlight_summary_df)\n",
    "    \n",
    "    ax.set_ylim(*ylim)\n",
    "    \n",
    "    ax.set_ylabel('log2(fold change)')\n",
    "    ax.set_xlabel('hpi')\n",
    "    \n",
    "    if OutFigname != '':\n",
    "        plt.tight_layout()\n",
    "        plt.savefig(FIGDIR %OutFigname)\n",
    "        plt.close()\n",
    "    elif show_fig:\n",
    "        plt.tight_layout()\n",
    "        plt.show()\n",
    "        plt.close()\n",
    "    \n",
    "    return timecourse_df\n",
    "\n",
    "\n",
    "def to_hex(r, g, b,float_scale = True):\n",
    "    if float_scale:\n",
    "        r = int(r*255)\n",
    "        g = int(g*255)\n",
    "        b = int(b*255)\n",
    "    rgb = r<<16 | g<<8 | b\n",
    "    return f'{rgb:06x}'\n",
    "\n",
    "\n",
    "def cluster_DEGs(timecourse_df,DEG_list, cl_method='ward',cl_metric='euclidean',\n",
    "                 n_cluster=5, clip=2.8,  OutFigname='', show_fig=False):\n",
    "    \n",
    "    df = timecourse_df.loc[DEG_list].clip(-clip,clip)\n",
    "    linkage = hierarchy.linkage(df, method=cl_method,metric=cl_metric)\n",
    "    cluster = hierarchy.fcluster(linkage,n_cluster,criterion='maxclust')\n",
    "    palette = sns.color_palette('bright',10)\n",
    "    palette = [color for i,color in enumerate(palette) if not(i in [0,3,5,7])]  ##CAUTION: HARD-CODED\n",
    "    \n",
    "    lut         = {(i+1): col for i,col in enumerate(palette)}\n",
    "    ID_cl_df    = pd.DataFrame(np.array([ [cluster[i] for i,ID in enumerate(df.index)] ]).T, \n",
    "                               index=df.index, columns=['cluster'])\n",
    "    \n",
    "    colors      = pd.Series([lut[ID_cl_df.loc[ID,'cluster']] for ID in ID_cl_df.index],index = ID_cl_df.index)\n",
    "    colors.name = 'Cluster'\n",
    "    max_lfc = np.ceil(clip)\n",
    "    clustergrid = sns.clustermap(df,figsize=(10,10),method=cl_method,metric=cl_metric,\n",
    "                                 row_colors = colors,row_linkage= linkage,col_cluster=False,\n",
    "                                 cmap='coolwarm',vmin=-max_lfc,vmax=max_lfc,\n",
    "                                 cbar_kws={'label': 'log2(fold change)'},\n",
    "                                 rasterized = True)\n",
    "    \n",
    "    clustergrid.ax_heatmap.set_xlabel('hpi')\n",
    "    clustergrid.ax_heatmap.set_ylabel('')\n",
    "    \n",
    "    ## print list of clustered DEGs, for cytoscape\n",
    "    for i in range(1,1+n_cluster):\n",
    "        IDs = list(ID_cl_df[ID_cl_df['cluster'] == i].index)\n",
    "        print('Cluster%d (%d)' %(i, len(IDs)),','.join(IDs))\n",
    "    \n",
    "    if OutFigname != '':\n",
    "        \n",
    "        plt.savefig(FIGDIR %OutFigname)\n",
    "        plt.close()\n",
    "    elif show_fig:\n",
    "        \n",
    "        plt.show()\n",
    "        plt.close()\n",
    "        \n",
    "    return ID_cl_df\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f1bba32a",
   "metadata": {},
   "source": [
    "### Fig. 5b,c (ExtDatFig. 7b,c)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "afe20d10",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "-------mRNA-------\n",
      "summary of DEGs\n",
      "total #:\t542\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>up</th>\n",
       "      <th>down</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0h</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1h</th>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2h</th>\n",
       "      <td>5</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4h</th>\n",
       "      <td>22</td>\n",
       "      <td>26</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12h</th>\n",
       "      <td>100</td>\n",
       "      <td>85</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16h</th>\n",
       "      <td>193</td>\n",
       "      <td>87</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24h</th>\n",
       "      <td>209</td>\n",
       "      <td>80</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36h</th>\n",
       "      <td>216</td>\n",
       "      <td>58</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      up  down\n",
       "0h     0     0\n",
       "1h     1     0\n",
       "2h     5     4\n",
       "4h    22    26\n",
       "12h  100    85\n",
       "16h  193    87\n",
       "24h  209    80\n",
       "36h  216    58"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Expression pattern of ACE2\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "0h     0.000001\n",
       "1h     0.001296\n",
       "2h     0.215375\n",
       "4h     0.037792\n",
       "12h   -0.683587\n",
       "16h   -1.133465\n",
       "24h   -0.901907\n",
       "36h   -0.340775\n",
       "Name: ACE2, dtype: float64"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Cluster1 (23) BCOR,BCORL1,TIGD7,CXXC5,FOS,PRKCD,FAM110C,MIDN,CDC42EP4,ZNF792,NOL4L,EGR1,SPSB2,PLEKHH3,TOX3,IER5L,SPRED3,NKX6-1,GSE1,SOCS3,BCL3,GRB7,PIM1\n",
      "Cluster2 (134) CRYM,PHF7,TP53I11,UGT1A1,TMEM176A,UGT1A6,GCNT4,FSTL1,SYT7,NANOS1,AREG,BCAS1,IGF2BP2,CABYR,CYP1B1,RTN4RL2,WIPF3,FSCN1,NR1H4,PLAC8,HLA-DOA,TUBB4A,NMT2,REEP2,SLC29A4,KLF9,ST6GALNAC2,BCL2L14,COLCA2,TXNIP,RDM1,SULF2,TTYH1,KLK11,C9ORF135,ARRDC4,IQCH,NDRG4,AKR1B10,SERPINA4,RARRES1,AKR1C1,MCAM,MYZAP,H3F3C,TMPRSS3,RASSF5,FBLN1,ATP2B4,TLE2,TRIM15,SCARA3,FOSL1,ID3,GNAZ,CDKN3,AKR1B15,ANKRD33B,CCDC110,MOK,CACNG4,AQP4,VSIR,ZNF488,RCAN1,ERP27,ASS1,IHH,ADGRB2,DBP,ID1,BCAM,ABI3BP,UNC5A,IFITM2,PLXND1,ALDH1L1,TFAP2C,TSPAN7,MYLK,FGD3,KIF5C,ACTA2,ACTBL2,PSG4,SLC16A14,TCEA2,VSIG1,ADSSL1,PPP1R1B,ARHGAP23,GMCL1,ZNF285,SCNN1A,TMPRSS2,GSDMB,ZNF816-ZNF321P,CPM,SMARCD3,YPEL4,GDPD5,KRT13,BBOF1,TCEA3,NCCRP1,GOLT1A,ARHGEF25,TMEM156,AKR1C2,TNFAIP2,SPIRE2,TIMP3,AQP3,HRASLS2,WNT7B,KREMEN1,MAOA,HLF,RASL11A,TMCC3,ALDH1A3,PTGES,ALDH1A1,FAM131B,SH3BP2,FAM171B,CYBRD1,BPIFB1,ANKRD1,CLDN16,CCL2,APOL3,MUC5B,TNNC1\n",
      "Cluster3 (103) IFIT3,HELZ2,RASGEF1B,MX1,STAT2,SLC1A3,IFNL3,JAK2,TNFSF13B,ISG20,CMPK2,TREX1,SLC25A28,TNFAIP3,HERC5,CLEC7A,SP140L,NFKBIA,NLRC5,ZNFX1,PDGFRL,IFI16,ZC3HAV1,RNF19B,TMEM140,PARP10,SOCS1,USP18,UBE2L6,LMO2,BST2,VAMP5,IFNL2,SP110,LAP3,RASGRP3,GBP4,PLA1A,PARP9,MX2,MUC13,PNPT1,DHX58,UBA7,PARP14,CD274,PSMB9,CFHR1,IFITM1,PARP12,TYMP,TRIM22,IFI44,PER1,EPSTI1,TRANK1,PPM1K,CEACAM1,DDX60,TAP1,CCL5,RNF122,NT5C3A,SAMHD1,TRIM21,STAT1,IRF7,OAS2,THEMIS2,LGALS9,IFI27,ETV7,IFI44L,FAM46A,IFI35,IDO1,GBP1,MYD88,IFIH1,SAMD9L,LIFR,MT2A,SAMD9,LAMP3,SLC15A3,OAS3,RTP4,CARD16,IFNL1,ZBP1,TMEM27,CH25H,CSF1,FAM122C,EPHA7,APOL4,APOL2,CFH,NMI,DDX60L,TLR3,C5ORF56,OAS1\n",
      "Cluster4 (16) ICAM4,DNAJB9,LANCL2,NDP,ATF5,TRMO,KBTBD8,SERPINB2,DIEXF,FOSB,HSPA6,EDN2,CD3EAP,SERPINB8,C3ORF58,NCOA5\n",
      "Cluster5 (266) FANCL,CD69,ZNF511,FBXL13,IFIT1,STARD4,TFRC,HIST1H3I,LRRC37A2,UXT,ARSE,RPSA,NDUFAF8,UBA52,TSTD1,RNASEH2C,SNRPF,HIST1H2AH,HIST1H4C,HIST1H4K,HIST1H2BB,CCDC7,HIST3H2BB,HIST1H1A,CAP2,MOV10,RPS29,HIST1H2BM,PCDHA6,SLC26A2,RPL17-C18ORF32,HIST2H2AB,RPL36,GAR1,HIST1H4D,POMP,SLC25A3,FASN,CLN6,HIST1H4H,C9ORF152,MCTP1,CEMIP,OASL,RPS26,POLR2I,RPL34,HIST1H2AK,NDUFB1,ACYP2,ITGA1,HSD17B2,ST20-MTHFS,INSIG1,COX7A2,HIST1H2BL,CCDC134,PCDHA3,MVK,DHCR7,ATP5J2,HIST1H2BJ,LINC01835,C2CD4A,HIST1H4A,ATP5D,ADI1,PRH1,CAVIN3,HERC6,PCDHA1,PTX3,IFI6,HIST1H2AG,PET100,RPS27A,HIST1H3B,RPS16,TMA7,TNFRSF11B,TMEM62,MCRIP1,BATF2,HLA-B,USMG5,HSPA8,RPL23,EIF4EBP3,MSRA,THBS1,SCD,PCDHA2,RPLP1,TNFSF15,CTSA,RPLP2,UBL5,DDX56,AVEN,AMIGO2,GLRX2,NME1,SLC22A4,ISG15,HCFC1R1,RPS21,HIST1H2BI,CXCL10,HIST1H3H,QTRT1,EMC7,HIST1H2BF,ITIH2,CATSPERB,PFKFB4,ACTL8,EEFSEC,MEMO1,CEACAM16,RPL26,PCDHA8,RSAD2,HIST1H2BC,ISPD,FABP5,HIST1H3D,PCDHA5,HIST4H4,SLC22A18AS,CYP2B6,CAPN8,EMC10,S100A2,HMGA2,PRR15,TAF15,ADAM32,CCL20,FAM13B,C12ORF57,TRIR,TFF1,PCDHA7,SNCA,IFIT2,HIST2H2AC,CXCL1,EMC6,LAMA3,ADAM19,PCDHA11,SLC28A3,HIST1H4I,NDUFB2,HIST1H2AM,PDSS1,POLE4,MRPS6,RPL30,RPL27,NUDCD2,FLRT3,IFNB1,CXCR4,RPLP0,RPL17,PRKAR1B,PRH1-TAS2R14,MTHFS,FZD8,PCDHA10,AP1G2,C2CD4B,RCL1,FADS2,RPL35A,POLR2F,RANBP17,FUS,SLC39A10,NPC1L1,HIST1H3F,SLIRP,IL1A,UFD1,HIST1H2AJ,C1ORF122,CAPN6,PTMS,HIST1H2AI,LYPD3,HIST1H2BH,SLC2A1,LMAN2L,RPL22L1,HIST1H2BO,LRRC37A,DDX58,FAM185A,COMMD6,HIST1H2AE,ADAMTS17,SF3B4,ANKRD37,CRNDE,RPS27,TNFAIP6,EMG1,SMDT1,HIST2H3D,RPS19,HIST1H3E,EEF1B2,HIST1H2BG,HIST1H3J,BRI3,PCDHA13,KRT6B,PIN1,HSD17B7,LSS,ADAM28,ORC4,HIST1H2BN,DNLZ,HIST1H4B,RPL41,RPSAP58,FERMT1,HIST1H3A,ZNF430,HIST1H4J,CXCL11,RPS13,BAAT,PRSS1,PCDHA12,ARMC10,CCR1,MIA,MSMO1,MRPL12,IGFBP7,ADRA1B,ADSL,LRRC37A3,CDKN1A,TXNL4A,TP53I3,PKIB,SFMBT2,HIST1H4E,RPS15A,PCSK9,ATP5G2,PCGF2,HMGCS1,HIST1H3G,CDH6,NDUFA7,TMEM154,FTL,VIM,MMP1,OSBPL10,PRSS2\n",
      "highlight genes with >1.5 max(|log2fc|)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>lfc_36h</th>\n",
       "      <th>lfc_maxabs</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [lfc_36h, lfc_maxabs]\n",
       "Index: []"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "-------RPF-------\n",
      "summary of DEGs\n",
      "total #:\t677\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>up</th>\n",
       "      <th>down</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0h</th>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1h</th>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2h</th>\n",
       "      <td>3</td>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4h</th>\n",
       "      <td>24</td>\n",
       "      <td>35</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12h</th>\n",
       "      <td>91</td>\n",
       "      <td>146</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16h</th>\n",
       "      <td>150</td>\n",
       "      <td>201</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24h</th>\n",
       "      <td>164</td>\n",
       "      <td>168</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36h</th>\n",
       "      <td>191</td>\n",
       "      <td>100</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      up  down\n",
       "0h     1     0\n",
       "1h     0     2\n",
       "2h     3     8\n",
       "4h    24    35\n",
       "12h   91   146\n",
       "16h  150   201\n",
       "24h  164   168\n",
       "36h  191   100"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Expression pattern of ACE2\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "0h    -0.000007\n",
       "1h    -0.000702\n",
       "2h    -0.055994\n",
       "4h    -0.319392\n",
       "12h   -2.566465\n",
       "16h   -3.545599\n",
       "24h   -3.782672\n",
       "36h   -4.101040\n",
       "Name: ACE2, dtype: float64"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Cluster1 (37) PRKD2,CTTNBP2NL,ZFP36,SHF,CDK18,BCOR,BCORL1,TFEB,KLF9,IRF5,MAFF,NFATC4,RTN4R,TBC1D10A,TRIM31,ZSWIM4,MYZAP,TRIM15,LIPE,ADM2,NOL4L,PLEKHG6,TFAP2C,PLEKHG3,CDC42EP2,GLIS2,CDKN1C,PIK3IP1,FBXL8,SESN2,IER5L,KLF2,GSE1,BCL3,CHAC1,GRB7,YPEL3\n",
      "Cluster2 (292) TMEM173,MFSD4A,AARS2,CTSV,CERCAM,FAM234B,RPL32,FGF19,PPP1R10,SLC26A9,C3,UGT1A1,ODAM,TLR5,TMEM176A,RARRES3,UGT1A6,CILP2,HLA-DPB1,FSTL1,HLA-DQA2,GALNT4,CNIH4,UGT1A10,AREG,UGT1A4,ACTA1,MYSM1,SLC34A2,CYP1B1,HLA-DRB5,GALNT12,ADAM10,GPRC5A,ANG,HLA-DQA1,CYP24A1,VWA8,AGT,MGLL,HBEGF,NR1H4,PLAC8,GGH,TSPAN8,HPSE,TSKU,EMP1,ACE2,PPM1L,TGFB2,ST6GALNAC2,CPD,ALDH18A1,CHST4,ALG6,AQP5,RBM12,TXNIP,TTYH1,SULF2,UCP2,PIGN,MDK,GLRX,CGRRF1,SMIM26,ALG8,F2RL1,ALDH1L2,TMEM161B,MOXD1,TMEM206,STRA6,DEGS2,FCGBP,UGCG,PTTG2,CD47,UGT1A7,CP,NDRG4,MRC2,AKR1B10,SERPINA4,RARRES1,NID1,INTS8,AKR1C1,CLU,GFM2,MUT,MCAM,RASGRP2,FOS,MPZL2,NNT,MANSC1,DPP4,LRP1,MAN2A1,MT1X,ERAP2,DNAAF3,DIRC2,FBLN1,LCN2,ATP2B4,GDA,HIST1H2BI,PDPR,SLC19A3,LETM2,SLC52A3,ID3,TMEM176B,CYP1A1,LAMA5,SEC22A,TMC5,SRPX2,AKR1B15,ALG10,FAM168B,PLAT,FBXO6,HLA-DPA1,ERO1B,PPARG,ITM2A,AASS,VSIR,FAM171A2,ACAD10,MUC1,CEACAM1,ZNF226,GCNT3,COL1A1,RCAN1,UGT1A3,SLC40A1,ASS1,NPNT,ERP44,SEMA3B,CD24,MAPT,CXCL5,HLA-DMB,HLTF,ID1,CHURC1,KIF20A,BCAM,ABI3BP,MUC5AC,NECTIN2,POFUT2,SAT1,UGGT2,UNC5A,SDC4,SFTA2,SAMD8,CTSH,PRTG,UGT1A8,HLA-DRA,DHDH,NRGN,WFDC2,LMF1,HSPG2,UPK3B,IFITM10,TSPAN7,MME,RHBDD1,PRCP,FGD3,ATP6AP2,PLBD1,TTC13,SLC38A2,MTERF3,IGF1R,SLC16A7,GPR155,VSIG1,LY75-CD302,CD55,REEP6,AKR1C3,AKR1C4,FRAS1,ADSSL1,C5,PPP1R1B,GMCL1,LETM1,SLC7A11,SCNN1A,NEK2,PRKCA,MGAM2,PAQR8,FUCA2,TMPRSS2,CDC25C,IFNAR2,GSDMB,DRAM2,EMP2,TM4SF4,MBTPS2,TMEM151A,TMEM131L,C4BPB,FBP1,WLS,TXNDC12,CPM,MYADM,IARS2,LCORL,PTPRB,ENPP5,PROCR,MAGED1,TCTN3,HSD17B11,NRP2,KRT13,ARHGDIB,IL4R,UGT1A5,LY75,SERPINA5,TCEA3,HACL1,SGSH,SECTM1,AKR1C2,UGT1A9,PODXL,TIMP3,LAPTM4A,AQP3,PRR11,WNT7B,CD40,CTSE,HAVCR1,STT3B,ARSK,ALDH1A3,TMEM168,ASNS,PLA2G10,NFE2L3,CYP2C18,PTGES,SERPING1,DEFB1,ACSM3,BTD,ALDH1A1,SYTL3,LAMB2,ZSCAN18,KDELR3,SLC16A5,ADRB2,CXORF38,APOL1,BPIFB1,ANKRD1,TMEM45A,UGGT1,LRG1,IGFBP3,ADGRF1,SLC22A5,PLAU,MUC5B,EPHX1,VASN,ITGB4,FTL,LBR,SLC30A1\n",
      "Cluster3 (100) IFIT3,SMIM29,HELZ2,BCL2L2,MX1,STAT2,SLC1A3,IFNL3,ISG20,CMPK2,TREX1,HERC5,OASL,NLRC5,ZNFX1,PDGFRL,IFI16,IFIT1B,IFI6,RNF19B,TMEM140,PARP10,USP18,UBE2L6,LMO2,SP140,BST2,IFNL2,SP110,BATF2,HLA-B,LAP3,GBP4,CEBPD,MX2,DHX58,CD274,CXCL10,PSMB9,PARP12,IFITM1,HPS1,TRIM22,IFI44,PML,EPSTI1,APOL6,TRANK1,PPM1K,DDX60,TAP1,PRR15,RNF122,SAMHD1,STAT1,IFIT2,HES4,MARCKSL1,OAS2,THEMIS2,IFITM2,APOBEC3G,LGALS9,ETV7,IFNB1,IFI44L,ZNF618,LGALS9C,DENND5A,IFI35,IDO1,PCGF5,GBP1,IFIH1,SAMD9L,CIB2,BBC3,MT2A,SAMD9,LAMP3,C19ORF66,SLC15A3,OAS3,OPTN,RTP4,DBF4B,IFNL1,ZBP1,GBP7,LGALS9B,CXCL11,CCR1,METRNL,GMPR,PSMB8,APOL3,DDX60L,MOB3C,MAP2K6,OAS1\n",
      "Cluster4 (16) LRRC8B,TFRC,DSEL,LANCL2,ST3GAL1,LUC7L,SERPINB2,ARMC5,OSGIN1,FOSB,HSPA6,LYPD1,EGR1,FJX1,HSPH1,NCOA5\n",
      "Cluster5 (232) PPP1R14B,KRT10,ZNF511,RAMP1,METTL21A,IFIT1,CHMP4A,DVL1,KANSL1,RNASEH2C,KHNYN,TMEM189,DPH3,NHP2,PEBP1,KDM4B,CST6,CLDN2,XPOT,POLR2J,RPS6KA2,ZC2HC1A,CBWD5,IDI1,RAB3IP,PLEKHJ1,BET1L,CEMIP,HRAS,PSPH,BLOC1S2,SETD7,HSPA1B,PEX10,S100A16,PFDN2,SMIM15,HSPA1L,HSPA1A,ICAM3,C19ORF25,GTPBP2,ADI1,HIRA,GFAP,ENSA,NSDHL,CAVIN3,JUND,S100A13,PPIA,BCL7C,TYW3,PEA15,SAP30L,AGTRAP,TMA7,CTXN1,TNFRSF11B,MCRIP1,FXYD1,DUS1L,SARNP,SUMO3,SREBF1,LAMTOR1,DAP,THBS1,PDCD5,R3HDM4,IRAK1,PHACTR2,RSRC2,RBCK1,NCL,WSB2,COA4,OGFR,DDIT3,C19ORF24,ISG15,GPBP1L1,CHCHD5,PALMD,MON1A,PPIAL4A,CTNNBIP1,ULK3,ATF2,ITIH2,PFKFB4,ACTL8,IL1RN,NCBP2-AS2,DHFR2,GTF3C6,WDR77,TYMP,RSAD2,ATXN7L3B,HOMER1,INA,DPP3,SLC22A18AS,CNP,ZNF146,PRELID1,C4ORF3,DPH2,CBWD1,VMA21,BTBD6,HMGA2,HMGN3,PIK3R2,DNTTIP1,TCEAL3,HLA-G,TCEAL9,UBTD2,DHFR,IRF7,CXCL1,CSNK1G2,CA9,TPM2,ZNF664,ADAM19,C6ORF223,NUDT4,MYO5C,HMGB3,POLE4,HMGN1,C1ORF50,MBD3,C6ORF62,TMSB10,DNAJB2,PRKAR1B,MTHFS,ATAD3C,ZNF33B,CBWD3,UBALD2,RNASEH1,POLR2F,GPATCH4,PYURF,TMEM163,BEX3,CD3EAP,NUAK1,LRRFIP2,IL1A,HDGFL2,MAPK6,GUCY1A3,SRM,C1ORF122,HYPK,PTMS,C5ORF24,JADE2,RPS6KB2,UPP1,PAM16,GALNT10,NAPA,SLC25A39,SH3BP5L,DDX58,JAZF1,NTAN1,ANKRD37,CDR2L,PTRH1,ACSS2,FAM13A,CLTB,C3ORF52,FBXO27,TRIM7,PXMP2,SRF,NAPEPLD,PIN1,LSS,BNIP3,PPP5D1,MMAB,PHACTR4,POLR3H,RPS6KA4,GAS2L1,PDAP1,EPC2,C19ORF53,MPND,FERMT1,TRIM14,BRMS1L,LUC7L3,CARD11,TOMM5,IGFBP7,C19ORF48,CBWD2,C6ORF141,ZFAND1,TP53I3,MYH11,HMGA1,PCSK9,AZI2,DBI,ATAD3B,CBWD6,CDH6,DBNDD1,TPM1,RFX5,DDX41,FURIN,MYCBP,TMEM154,ALKBH5,CASP7,ARIH2,KHDRBS3,UTP18,MMP1\n",
      "highlight genes with >1.5 max(|log2fc|)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>lfc_36h</th>\n",
       "      <th>lfc_maxabs</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [lfc_36h, lfc_maxabs]\n",
       "Index: []"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "-------QTI-------\n",
      "summary of DEGs\n",
      "total #:\t861\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>up</th>\n",
       "      <th>down</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0h</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1h</th>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2h</th>\n",
       "      <td>6</td>\n",
       "      <td>16</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4h</th>\n",
       "      <td>104</td>\n",
       "      <td>54</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12h</th>\n",
       "      <td>126</td>\n",
       "      <td>149</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16h</th>\n",
       "      <td>232</td>\n",
       "      <td>216</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24h</th>\n",
       "      <td>197</td>\n",
       "      <td>185</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36h</th>\n",
       "      <td>148</td>\n",
       "      <td>112</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      up  down\n",
       "0h     0     0\n",
       "1h     1     0\n",
       "2h     6    16\n",
       "4h   104    54\n",
       "12h  126   149\n",
       "16h  232   216\n",
       "24h  197   185\n",
       "36h  148   112"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Expression pattern of ACE2\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "0h    -8.358979e-07\n",
       "1h     5.180962e-03\n",
       "2h    -6.349377e-02\n",
       "4h    -4.636234e-01\n",
       "12h   -1.471266e+00\n",
       "16h   -1.872763e+00\n",
       "24h   -2.101192e+00\n",
       "36h   -2.338524e+00\n",
       "Name: ACE2, dtype: float64"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Cluster1 (38) ARID5B,ANKMY2,TTC33,SERTAD2,TBC1D10A,PLSCR4,IFRD1,KLHL22,B3GALT5,TMEM139,ING4,SRGAP2,RALB,IER5L,PEX11B,GRB7,STYK1,TGFBR3,PRR15L,MYORG,PEF1,ARRDC3,VPS25,BCL6,TAF7,VRK3,NOL4L,TFAP2C,CDC42EP2,IRF2,MID1IP1,IER2,BTN3A3,HOXB4,SOCS3,BLCAP,TLDC1,RARA\n",
      "Cluster2 (349) MFSD4A,RPL32,PPP1R10,EPS8L1,LACTB2,RARRES3,CILP2,HSD17B8,HLA-DPB1,HLA-DQA2,UGT1A10,UGT1A4,RPL37,SLC34A2,PEX6,RAB11FIP1,GLIS3,CYP24A1,RNF19A,DDIT4,TMEM37,NAB2,PLAC8,CLDN10,GSTM3,TSKU,TACSTD2,TUBB4A,NNMT,ACE2,POTEF,KRT15,IRF5,CHST4,MAFF,FUCA1,COLCA2,DYNC1LI2,TXNIP,SERPINH1,EBPL,ICAM5,VAPA,DTX2,CD47,PEX11A,EML2,SGK2,SERPINA4,HSPB1,CNOT9,AKR1C1,CLU,TM4SF4,FOS,FAM98C,DNAAF3,TGFA,TRIM15,FOSL1,RDH13,CLDN1,MUC1,SIDT2,GCNT3,UGT1A3,KREMEN2,ASS1,SCIN,SPATA5L1,DBP,CRABP2,CD24,CXCL5,HEXA,ID1,BCAM,MUC5AC,ABI3BP,RPL18,PDZK1IP1,SAT1,UNC5A,SFTA2,EGR1,CTSH,UGT1A8,DHDH,UPK3B,B3GNT7,CXCL16,IFITM10,FGD3,TGFB1I1,RTL10,TMEM161A,PLPP1,ACTBL2,DUSP5,SLC16A7,ZNF217,ZC3H12A,REEP6,C5,GTPBP1,CSTB,GCOM1,MYO18A,EMP2,AHCYL1,VNN2,IFNGR1,RARB,FBXO2,PROCR,KRT13,KRTCAP3,ECI2,UGT1A5,NDRG2,CHCHD10,WDR72,KLF2,UGT1A9,SPINK1,AQP3,WNT7B,NEDD9,C1ORF116,MFSD10,PTGES,TPM4,DEFB1,ACSM3,SNAPC2,ICAM1,ZNF274,IFFO2,KDELR3,SLC16A5,ATXN10,RPL15,TMEM45A,LRG1,ADGRF1,RPL9,TMEM59,ADAP1,PLAU,TMED8,CYR61,MRPL41,VASN,NR4A1,FTL,DGAT1,GMCL1,TMEM173,METTL15,DMAC2,NAGA,NUAK2,COMTD1,ODAM,C3,UGT1A1,DALRD3,TMEM176A,UGT1A6,CAPN5,VGLL1,AREG,ACTA1,BCAS1,TAPBP,CYP1B1,HLA-DRB5,MUC6,CLDN4,GPRC5A,PLCD1,MGLL,NR1H4,AIG1,MFGE8,EFNA1,TSPAN8,HLA-F,EMP1,MRPL10,TGFB2,DNASE1L1,PARL,ACADS,SSR3,SLC37A3,P3H3,RBM12,SULF2,TTYH1,NFATC4,GNE,EFNA5,KLK11,SWI5,FCGBP,DMBT1,UGT1A7,CP,AKR1B10,MRC2,FBXO31,TMEM115,RARRES1,C15ORF39,ALDH2,MIEF2,MYZAP,RASGRP2,PLPP6,DUSP1,LRP1,MT1X,GPR37,CSN3,RASSF5,TKT,FBLN1,LCN2,ATP2B4,COQ5,FUT3,PMM1,PCYOX1L,ZMIZ1,TMEM45B,THEM6,LETM2,SLC52A3,HLA-DMA,AKR1B1,ID3,DAZAP2,MRPS7,MANBAL,TMC5,AKR1B15,D2HGDH,PER1,FBXO6,HLA-DPA1,PPARG,FAM110C,ITM2A,VSIR,MIDN,ABTB1,CEACAM1,PRAME,COL1A1,CITED4,CLN3,SLC40A1,SEMA3B,UBE2W,MAPT,GBP2,CHURC1,NECTIN2,TMPRSS4,PLD3,SULT2B1,SDC4,PLK2,ABHD11,ALDH1L1,FAM234A,HLA-DRA,NRGN,WFDC2,STX8,MISP,RPL17,MTSS1,ENTPD2,CD74,PHYH,ARHGAP45,SLC25A20,LITAF,RND3,SERTAD1,PPP1R1B,SLC7A11,SCNN1A,EHHADH,PSAP,MGAM2,SALL2,GSDMB,IER3IP1,FBP1,NR3C1,MYADM,TMEM51,SLPI,MAGED1,ARHGDIB,MAPK8IP3,GALK1,SECTM1,RHOB,AKR1C2,TNFAIP2,KDELR1,RHPN2,TIMP3,LAPTM4A,TNFRSF21,PIM3,CTSE,FBXL20,LTBP4,EIF3F,TADA1,ALDH1A3,ASNS,PDK4,VILL,BCL3,C1R,RELA,SLC9A3R2,EPN2,ECHS1,CXORF38,BPIFB1,CPT2,ANKRD1,TMEM9,IGFBP3,SMC5,AMOTL2,PIGR,MUC5B,SHISA5,ITGB4,YPEL3,SLC30A1\n",
      "Cluster3 (86) FOXC2,CREBBP,DNAJB9,PFKFB3,ESF1,DNTTIP2,UCP2,ADI1,C6ORF106,FAM111B,LUC7L,THBS1,RPLP1,YLPM1,PDE3A,STC2,ODF3B,SPTSSA,ERCC6,C1QBP,FJX1,MAT2A,AMD1,TEX264,HSBP1L1,EEF1B2,ING5,C3ORF58,BAAT,NDUFS6,MRPL18,DKK1,MRPS17,DDX5,NDUFS7,RBM3,FST,RPS12,SS18L2,U2AF2,METTL1,AKAP12,BORA,COQ10B,GGCT,RPP25,REEP2,ZSWIM9,ZNF639,ASTE1,ZSCAN2,KBTBD8,SERPINB2,C19ORF24,CAPZB,LBH,SRSF2,STX3,RNF126,CDC25A,ZNF316,GLOD4,C2CD4B,PYURF,CD3EAP,SFPQ,TTK,ZDHHC18,DDX39A,ADM,UBA2,HSPH1,C18ORF32,RPS19,SERPINB8,SKAP1,SEC11C,SS18,NCOA5,UTP20,SURF6,PSMA3,BTAF1,ATAD3B,H2AFX,URB2\n",
      "Cluster4 (88) MX1,ISG20,TSC22D3,IFI6,USP18,UBE2L6,BST2,LAP3,GBP4,FMR1,OGFR,ISG15,CD274,PSMB9,IFITM1,PARP12,TRIM22,PML,PPM1K,DDX60,RNF122,OAS2,GPT2,THEMIS2,ETV7,LGALS9C,IFI35,PCGF5,GBP1,IFIH1,NAPA,SAMD9,LAMP3,OPTN,RTP4,IFNL1,NMI,SAA2-SAA4,IFIT3,IFIT1,SLC1A3,CMPK2,HERC5,OASL,NLRC5,PDGFRL,HIRA,PARP10,CD68,SP110,BATF2,MX2,DHX58,CXCL10,SAA2,IFI44,CNP,EPSTI1,APOL6,TRANK1,DKK3,MASTL,TAP1,NT5C3A,SAMHD1,STAT1,IFIT2,IRF7,LGALS9,IFI44L,IFNB1,IDO1,TNFSF10,SAMD9L,MT2A,DDX58,C19ORF66,OAS3,ORC6,TMEM27,LGALS9B,TRIM14,CCR1,APOL3,DDX60L,MOB3C,MAP2K6,OAS1\n",
      "Cluster5 (300) PPP1R14B,BHLHA15,APOBEC3D,TCEA1,RNASEH2C,DNA2,TCEAL2,HIST1H4K,NHP2,RBM4B,SSR1,C17ORF100,POLR2J,HNRNPH3,BCL7B,TPGS2,IDI1,BET1L,DEK,MAX,EIF5A,NEDD8,REEP3,HSD17B2,PHLDA2,IFI16,ANP32E,OTUD6B,QKI,ENSA,CAVIN3,CKLF-CMTM1,PITPNA,NDUFA5,MFSD1,PSD3,IQCK,BCL7C,PEA15,COA6,COX17,FXYD1,CDC37L1,DYNLRB1,TMEM56-RWDD3,SUMO3,SREBF1,ACSL1,CRIPT,PDCD5,BORCS7,SCD,KCTD3,IRAK1,ZDHHC2,NCL,AMIGO2,MLKL,PTMA,TMEM14C,PALMD,UBE2D2,ATF2,ITIH2,PIK3AP1,GTF3C6,RSAD2,ENOX2,TPT1,SREK1IP1,HIST1H3D,GRB10,SLC22A18AS,RBM34,C4ORF3,S100A2,TMEM258,TRIM8,CBWD1,PABPN1,RNF103-CHMP3,UBTD2,FAM213A,CSNK1G2,ZCCHC2,LAMA3,TBC1D4,ARMC9,CLK1,NUDT4,MYO5C,HMGN1,RAB43,C6ORF62,MRPL35,APOBEC3C,TMSB10,PRKAR1B,RNASEH1,POLR2F,DMRTA1,FXN,SLC39A10,NEDD8-MDP1,LRRFIP2,IL1A,COPRS,CC2D1B,EIF3J,HNF4G,TMEM243,FAM129A,GALNT10,RPS27L,UBTF,TNFAIP6,CLTB,C3ORF52,TLX3,RBM25,HIST1H3J,TIMM17B,ACAT2,EXOSC8,WBP1L,PDAP1,LSM3,FERMT1,EIF3A,ABT1,TOMM5,CBX5,FANCM,CDKN1A,KPNA4,YWHAQ,GMIP,CDH6,FURIN,LSM1,NPAS2,DYRK4,MYCBP,PRELID3B,CNEP1R1,KRT10,CDC45,SZRD1,EBNA1BP2,ZNF519,GALNT1,MRPL32,SNX7,LSM8,C1ORF52,DSC2,TMEM189,HIST1H4C,COX20,ACOT4,RTKN2,PRPF18,TCEAL4,ARL6IP4,XPOT,DYNLL1,SLC26A2,RAB11B,DCTN6,DCBLD2,PLEKHJ1,C16ORF87,FASN,MCTP1,CEMIP,KIF1C,S100A16,TAF9B,STUB1,ZNFX1,ITGA2,LRP8,YAP1,MAP1LC3B,NSDHL,HERC6,S100A13,TTC3,ZNF428,MED10,NT5DC2,TMA7,TUSC1,CTXN1,OPHN1,MCRIP1,SARNP,SLC20A1,CPSF6,ZNF749,CHIC2,LAMTOR1,DAP,SGTA,MAP1LC3B2,R3HDM4,EI24,SET,RSRC2,DDX56,WSB2,PFKFB4,YKT6,VCAN,DHCR24,CYP2B6,CFDP1,VMA21,HMGA2,HMGN3,CCL20,GIGYF2,C1ORF21,BLOC1S1,TMEM263,CXCL1,GSK3A,MAOB,MYBL2,IPO4,ADAM19,POLR2E,MYL12A,TMEM265,TRAPPC1,UMPS,PRMT1,NUDT3,SEC61B,WASHC3,LIPG,PFDN1,CASD1,RWDD1,FDCSP,SNX10,SMIM7,SVIP,NELFCD,CISD2,ATP1B3,SRM,C1ORF122,PTMS,TIMM8B,UPP1,DENND1B,AGPAT2,GTF3C4,IFT43,ANAPC15,NAA10,HEATR3,GPALPP1,HINT1,ATP5G1,RAB27B,BNIP3,HSD17B7,POLR3G,RPS6KA4,FAM104A,LUC7L3,HIST1H4J,DPH7,MACROD2,HSPE1,PMEPA1,PCNP,IGFBP7,CBWD2,C6ORF141,BTBD1,ZFAND1,MYH10,COL7A1,MSI2,PAICS,PNN,HMGA1,UBE2G1,DBI,CLTA,C12ORF75,CHMP3,TPM1,TMSB4X,MAP4K4,KHDRBS3,MMP1\n",
      "highlight genes with >1.5 max(|log2fc|)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>lfc_36h</th>\n",
       "      <th>lfc_maxabs</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [lfc_36h, lfc_maxabs]\n",
       "Index: []"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "for tmp_sample_prefix in ['mRNA','RPF','QTI']:\n",
    "    print('-------'+tmp_sample_prefix+'-------')\n",
    "    tmp_DEG_df_list   = []\n",
    "    tmp_hpi_list      = ['%dh' %i for i in [0,1,2,4,12,16,24,36]]\n",
    "    for hpi in tmp_hpi_list:\n",
    "        InFname    = f'DEG/{tmp_sample_prefix}_DESeq2_{hpi}.tsv'\n",
    "        tmp_DEG_df = pd.read_csv(RESULTDIR %(InFname), sep ='\\t', header=0, index_col=0)\n",
    "        tmp_DEG_df_list.append(tmp_DEG_df)\n",
    "\n",
    "    timecourse_df = timecourse_plot(tmp_DEG_df_list,tmp_hpi_list, padj_cut=0.01, log2fc_cut=2, clip = 2.8,\n",
    "                                    ylim = 8, n_cluster=5, as_average=False, \n",
    "                                    load_precalc_clusters=False, clusterFname=f'timecourse_hostexp/{tmp_sample_prefix}_cluster.tsv',\n",
    "                                    clusters_to_show=[],clusters_priority=[], \n",
    "                                    cl_method='ward',cl_metric='euclidean',ax = None, \n",
    "                                    OutFigname=f'Fig5_timecourse_hostexp/Fig5_{tmp_sample_prefix}_individual_timecourse.pdf',\n",
    "                                    show_fig=False, verbose=True)\n",
    "    timecourse_df.to_csv(RESULTDIR %(f'timecourse_hostexp/{tmp_sample_prefix}_timecourse.tsv'),sep='\\t')\n",
    "    _             = timecourse_plot(tmp_DEG_df_list,tmp_hpi_list, padj_cut=0.01, log2fc_cut=2, clip = 2.8,\n",
    "                                    ylim = 8, n_cluster=5, as_average=True, \n",
    "                                    load_precalc_clusters=True, clusterFname=f'timecourse_hostexp/{tmp_sample_prefix}_cluster.tsv',\n",
    "                                    clusters_to_show=[],clusters_priority=[], \n",
    "                                    cl_method='ward',cl_metric='euclidean',ax = None, \n",
    "                                    OutFigname=f'Fig5_timecourse_hostexp/Fig5_{tmp_sample_prefix}_average_timecourse.pdf',\n",
    "                                    show_fig=False, verbose=False)\n",
    "    "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2fa952fc",
   "metadata": {},
   "source": [
    "### Fig. 5d-g, ExtDatFigs. 7d, 8a-e: individual clusters"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "27aeede8",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "-------mRNA-------\n",
      "-------RPF-------\n",
      "-------QTI-------\n"
     ]
    }
   ],
   "source": [
    "for tmp_sample_prefix in ['mRNA','RPF','QTI']:\n",
    "    print('-------'+tmp_sample_prefix+'-------')\n",
    "    tmp_DEG_df_list   = []\n",
    "    tmp_hpi_list      = ['%dh' %i for i in [0,1,2,4,12,16,24,36]]\n",
    "    for hpi in tmp_hpi_list:\n",
    "        InFname    = f'DEG/{tmp_sample_prefix}_DESeq2_{hpi}.tsv'\n",
    "        tmp_DEG_df = pd.read_csv(RESULTDIR %(InFname), sep ='\\t', header=0, index_col=0)\n",
    "        tmp_DEG_df_list.append(tmp_DEG_df)\n",
    "\n",
    "    for cluster in range(1,1+5):\n",
    "        _             = timecourse_plot(tmp_DEG_df_list,tmp_hpi_list, padj_cut=0.01, log2fc_cut=2, clip = 2.8,\n",
    "                                        ylim = 8, n_cluster=5, as_average=True, \n",
    "                                        load_precalc_clusters=True, clusterFname=f'timecourse_hostexp/{tmp_sample_prefix}_cluster.tsv',\n",
    "                                        clusters_to_show=[cluster],clusters_priority=[], \n",
    "                                        cl_method='ward',cl_metric='euclidean',ax = None, \n",
    "                                        OutFigname=f'Fig5_timecourse_hostexp/Fig5_{tmp_sample_prefix}_average_cluster{cluster}_timecourse.pdf',\n",
    "                                        show_fig=False, verbose=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6423760e",
   "metadata": {},
   "source": [
    "### Source data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "3ef1686a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "-------mRNA-------\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0h</th>\n",
       "      <th>1h</th>\n",
       "      <th>2h</th>\n",
       "      <th>4h</th>\n",
       "      <th>12h</th>\n",
       "      <th>16h</th>\n",
       "      <th>24h</th>\n",
       "      <th>36h</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cluster</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>green</th>\n",
       "      <td>-3.509958e-06</td>\n",
       "      <td>-0.010155</td>\n",
       "      <td>-0.144303</td>\n",
       "      <td>-0.484754</td>\n",
       "      <td>-2.243646</td>\n",
       "      <td>-2.242855</td>\n",
       "      <td>-2.144924</td>\n",
       "      <td>-1.812561</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>orange</th>\n",
       "      <td>-6.853996e-06</td>\n",
       "      <td>-0.002369</td>\n",
       "      <td>-1.310808</td>\n",
       "      <td>-2.187750</td>\n",
       "      <td>-0.766063</td>\n",
       "      <td>-0.977212</td>\n",
       "      <td>-0.836389</td>\n",
       "      <td>-0.157820</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>pink</th>\n",
       "      <td>1.043716e-01</td>\n",
       "      <td>0.541693</td>\n",
       "      <td>1.714755</td>\n",
       "      <td>2.559446</td>\n",
       "      <td>0.819257</td>\n",
       "      <td>0.721431</td>\n",
       "      <td>0.457308</td>\n",
       "      <td>-0.017288</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>purple</th>\n",
       "      <td>1.279531e-07</td>\n",
       "      <td>0.000466</td>\n",
       "      <td>0.020821</td>\n",
       "      <td>-0.109426</td>\n",
       "      <td>-0.082964</td>\n",
       "      <td>0.135543</td>\n",
       "      <td>1.898897</td>\n",
       "      <td>3.283364</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>yellow</th>\n",
       "      <td>-5.827553e-08</td>\n",
       "      <td>0.017184</td>\n",
       "      <td>0.252235</td>\n",
       "      <td>0.551300</td>\n",
       "      <td>1.950007</td>\n",
       "      <td>2.346395</td>\n",
       "      <td>2.327602</td>\n",
       "      <td>2.032264</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                   0h        1h        2h        4h       12h       16h  \\\n",
       "cluster                                                                   \n",
       "green   -3.509958e-06 -0.010155 -0.144303 -0.484754 -2.243646 -2.242855   \n",
       "orange  -6.853996e-06 -0.002369 -1.310808 -2.187750 -0.766063 -0.977212   \n",
       "pink     1.043716e-01  0.541693  1.714755  2.559446  0.819257  0.721431   \n",
       "purple   1.279531e-07  0.000466  0.020821 -0.109426 -0.082964  0.135543   \n",
       "yellow  -5.827553e-08  0.017184  0.252235  0.551300  1.950007  2.346395   \n",
       "\n",
       "              24h       36h  \n",
       "cluster                      \n",
       "green   -2.144924 -1.812561  \n",
       "orange  -0.836389 -0.157820  \n",
       "pink     0.457308 -0.017288  \n",
       "purple   1.898897  3.283364  \n",
       "yellow   2.327602  2.032264  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0h</th>\n",
       "      <th>1h</th>\n",
       "      <th>2h</th>\n",
       "      <th>4h</th>\n",
       "      <th>12h</th>\n",
       "      <th>16h</th>\n",
       "      <th>24h</th>\n",
       "      <th>36h</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cluster</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>green</th>\n",
       "      <td>0.000046</td>\n",
       "      <td>0.082845</td>\n",
       "      <td>0.339071</td>\n",
       "      <td>0.733603</td>\n",
       "      <td>0.869172</td>\n",
       "      <td>0.815891</td>\n",
       "      <td>0.816753</td>\n",
       "      <td>1.181286</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>orange</th>\n",
       "      <td>0.000052</td>\n",
       "      <td>0.001858</td>\n",
       "      <td>0.572643</td>\n",
       "      <td>0.603702</td>\n",
       "      <td>0.473002</td>\n",
       "      <td>0.614208</td>\n",
       "      <td>0.688474</td>\n",
       "      <td>0.434079</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>pink</th>\n",
       "      <td>0.417476</td>\n",
       "      <td>0.662753</td>\n",
       "      <td>0.581041</td>\n",
       "      <td>0.835050</td>\n",
       "      <td>0.670795</td>\n",
       "      <td>0.752898</td>\n",
       "      <td>0.540694</td>\n",
       "      <td>0.477038</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>purple</th>\n",
       "      <td>0.000007</td>\n",
       "      <td>0.001413</td>\n",
       "      <td>0.212299</td>\n",
       "      <td>0.478282</td>\n",
       "      <td>0.539707</td>\n",
       "      <td>0.506086</td>\n",
       "      <td>1.201066</td>\n",
       "      <td>1.467521</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>yellow</th>\n",
       "      <td>0.000014</td>\n",
       "      <td>0.117209</td>\n",
       "      <td>0.374098</td>\n",
       "      <td>0.647315</td>\n",
       "      <td>0.681110</td>\n",
       "      <td>0.749524</td>\n",
       "      <td>0.945279</td>\n",
       "      <td>1.265501</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "               0h        1h        2h        4h       12h       16h       24h  \\\n",
       "cluster                                                                         \n",
       "green    0.000046  0.082845  0.339071  0.733603  0.869172  0.815891  0.816753   \n",
       "orange   0.000052  0.001858  0.572643  0.603702  0.473002  0.614208  0.688474   \n",
       "pink     0.417476  0.662753  0.581041  0.835050  0.670795  0.752898  0.540694   \n",
       "purple   0.000007  0.001413  0.212299  0.478282  0.539707  0.506086  1.201066   \n",
       "yellow   0.000014  0.117209  0.374098  0.647315  0.681110  0.749524  0.945279   \n",
       "\n",
       "              36h  \n",
       "cluster            \n",
       "green    1.181286  \n",
       "orange   0.434079  \n",
       "pink     0.477038  \n",
       "purple   1.467521  \n",
       "yellow   1.265501  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "-------RPF-------\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0h</th>\n",
       "      <th>1h</th>\n",
       "      <th>2h</th>\n",
       "      <th>4h</th>\n",
       "      <th>12h</th>\n",
       "      <th>16h</th>\n",
       "      <th>24h</th>\n",
       "      <th>36h</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cluster</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>green</th>\n",
       "      <td>2.172120e-03</td>\n",
       "      <td>-0.017243</td>\n",
       "      <td>-0.068478</td>\n",
       "      <td>-0.287732</td>\n",
       "      <td>-2.044621</td>\n",
       "      <td>-2.287178</td>\n",
       "      <td>-2.228490</td>\n",
       "      <td>-1.750059</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>orange</th>\n",
       "      <td>3.545986e-02</td>\n",
       "      <td>-0.139185</td>\n",
       "      <td>-0.961587</td>\n",
       "      <td>-2.290647</td>\n",
       "      <td>-0.701016</td>\n",
       "      <td>-0.782072</td>\n",
       "      <td>-0.449461</td>\n",
       "      <td>-0.190107</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>pink</th>\n",
       "      <td>2.733832e-01</td>\n",
       "      <td>0.380496</td>\n",
       "      <td>0.831175</td>\n",
       "      <td>2.025800</td>\n",
       "      <td>0.662366</td>\n",
       "      <td>0.305421</td>\n",
       "      <td>0.085255</td>\n",
       "      <td>0.031234</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>purple</th>\n",
       "      <td>3.359293e-07</td>\n",
       "      <td>0.000109</td>\n",
       "      <td>-0.028956</td>\n",
       "      <td>-0.224386</td>\n",
       "      <td>0.005149</td>\n",
       "      <td>0.298347</td>\n",
       "      <td>2.025824</td>\n",
       "      <td>3.527006</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>yellow</th>\n",
       "      <td>1.162663e-06</td>\n",
       "      <td>0.006363</td>\n",
       "      <td>0.060510</td>\n",
       "      <td>0.396599</td>\n",
       "      <td>1.832983</td>\n",
       "      <td>2.151903</td>\n",
       "      <td>2.131167</td>\n",
       "      <td>1.750147</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                   0h        1h        2h        4h       12h       16h  \\\n",
       "cluster                                                                   \n",
       "green    2.172120e-03 -0.017243 -0.068478 -0.287732 -2.044621 -2.287178   \n",
       "orange   3.545986e-02 -0.139185 -0.961587 -2.290647 -0.701016 -0.782072   \n",
       "pink     2.733832e-01  0.380496  0.831175  2.025800  0.662366  0.305421   \n",
       "purple   3.359293e-07  0.000109 -0.028956 -0.224386  0.005149  0.298347   \n",
       "yellow   1.162663e-06  0.006363  0.060510  0.396599  1.832983  2.151903   \n",
       "\n",
       "              24h       36h  \n",
       "cluster                      \n",
       "green   -2.228490 -1.750059  \n",
       "orange  -0.449461 -0.190107  \n",
       "pink     0.085255  0.031234  \n",
       "purple   2.025824  3.527006  \n",
       "yellow   2.131167  1.750147  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0h</th>\n",
       "      <th>1h</th>\n",
       "      <th>2h</th>\n",
       "      <th>4h</th>\n",
       "      <th>12h</th>\n",
       "      <th>16h</th>\n",
       "      <th>24h</th>\n",
       "      <th>36h</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cluster</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>green</th>\n",
       "      <td>0.083781</td>\n",
       "      <td>0.216179</td>\n",
       "      <td>0.333276</td>\n",
       "      <td>0.486959</td>\n",
       "      <td>0.800649</td>\n",
       "      <td>0.787601</td>\n",
       "      <td>0.900510</td>\n",
       "      <td>1.012195</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>orange</th>\n",
       "      <td>0.215698</td>\n",
       "      <td>0.435826</td>\n",
       "      <td>0.820356</td>\n",
       "      <td>0.499250</td>\n",
       "      <td>0.516529</td>\n",
       "      <td>0.700262</td>\n",
       "      <td>0.596190</td>\n",
       "      <td>0.577111</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>pink</th>\n",
       "      <td>0.785708</td>\n",
       "      <td>0.700544</td>\n",
       "      <td>1.018561</td>\n",
       "      <td>0.964031</td>\n",
       "      <td>0.675493</td>\n",
       "      <td>0.353016</td>\n",
       "      <td>0.623697</td>\n",
       "      <td>0.390405</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>purple</th>\n",
       "      <td>0.000003</td>\n",
       "      <td>0.001090</td>\n",
       "      <td>0.154070</td>\n",
       "      <td>0.453901</td>\n",
       "      <td>0.551558</td>\n",
       "      <td>0.592402</td>\n",
       "      <td>1.416105</td>\n",
       "      <td>1.532069</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>yellow</th>\n",
       "      <td>0.000009</td>\n",
       "      <td>0.068009</td>\n",
       "      <td>0.281283</td>\n",
       "      <td>0.758919</td>\n",
       "      <td>0.786472</td>\n",
       "      <td>0.736528</td>\n",
       "      <td>0.843241</td>\n",
       "      <td>1.139114</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "               0h        1h        2h        4h       12h       16h       24h  \\\n",
       "cluster                                                                         \n",
       "green    0.083781  0.216179  0.333276  0.486959  0.800649  0.787601  0.900510   \n",
       "orange   0.215698  0.435826  0.820356  0.499250  0.516529  0.700262  0.596190   \n",
       "pink     0.785708  0.700544  1.018561  0.964031  0.675493  0.353016  0.623697   \n",
       "purple   0.000003  0.001090  0.154070  0.453901  0.551558  0.592402  1.416105   \n",
       "yellow   0.000009  0.068009  0.281283  0.758919  0.786472  0.736528  0.843241   \n",
       "\n",
       "              36h  \n",
       "cluster            \n",
       "green    1.012195  \n",
       "orange   0.577111  \n",
       "pink     0.390405  \n",
       "purple   1.532069  \n",
       "yellow   1.139114  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "-------QTI-------\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0h</th>\n",
       "      <th>1h</th>\n",
       "      <th>2h</th>\n",
       "      <th>4h</th>\n",
       "      <th>12h</th>\n",
       "      <th>16h</th>\n",
       "      <th>24h</th>\n",
       "      <th>36h</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cluster</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>green</th>\n",
       "      <td>1.389483e-04</td>\n",
       "      <td>-0.065349</td>\n",
       "      <td>-0.309868</td>\n",
       "      <td>-0.590083</td>\n",
       "      <td>-2.037440</td>\n",
       "      <td>-2.310588</td>\n",
       "      <td>-2.249368</td>\n",
       "      <td>-1.677603</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>orange</th>\n",
       "      <td>-1.390877e-06</td>\n",
       "      <td>-0.329683</td>\n",
       "      <td>-1.311438</td>\n",
       "      <td>-2.285586</td>\n",
       "      <td>-0.681984</td>\n",
       "      <td>-1.146140</td>\n",
       "      <td>-1.001340</td>\n",
       "      <td>-0.172914</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>pink</th>\n",
       "      <td>-4.725912e-07</td>\n",
       "      <td>0.004275</td>\n",
       "      <td>-0.067231</td>\n",
       "      <td>-0.202450</td>\n",
       "      <td>0.037802</td>\n",
       "      <td>0.224839</td>\n",
       "      <td>1.870903</td>\n",
       "      <td>3.229478</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>purple</th>\n",
       "      <td>7.477085e-07</td>\n",
       "      <td>0.402028</td>\n",
       "      <td>1.159288</td>\n",
       "      <td>2.423515</td>\n",
       "      <td>0.357786</td>\n",
       "      <td>0.779682</td>\n",
       "      <td>0.643652</td>\n",
       "      <td>-0.046434</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>yellow</th>\n",
       "      <td>1.431711e-07</td>\n",
       "      <td>0.044575</td>\n",
       "      <td>0.236530</td>\n",
       "      <td>0.583076</td>\n",
       "      <td>1.766701</td>\n",
       "      <td>2.428972</td>\n",
       "      <td>2.147937</td>\n",
       "      <td>1.265620</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                   0h        1h        2h        4h       12h       16h  \\\n",
       "cluster                                                                   \n",
       "green    1.389483e-04 -0.065349 -0.309868 -0.590083 -2.037440 -2.310588   \n",
       "orange  -1.390877e-06 -0.329683 -1.311438 -2.285586 -0.681984 -1.146140   \n",
       "pink    -4.725912e-07  0.004275 -0.067231 -0.202450  0.037802  0.224839   \n",
       "purple   7.477085e-07  0.402028  1.159288  2.423515  0.357786  0.779682   \n",
       "yellow   1.431711e-07  0.044575  0.236530  0.583076  1.766701  2.428972   \n",
       "\n",
       "              24h       36h  \n",
       "cluster                      \n",
       "green   -2.249368 -1.677603  \n",
       "orange  -1.001340 -0.172914  \n",
       "pink     1.870903  3.229478  \n",
       "purple   0.643652 -0.046434  \n",
       "yellow   2.147937  1.265620  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0h</th>\n",
       "      <th>1h</th>\n",
       "      <th>2h</th>\n",
       "      <th>4h</th>\n",
       "      <th>12h</th>\n",
       "      <th>16h</th>\n",
       "      <th>24h</th>\n",
       "      <th>36h</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cluster</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>green</th>\n",
       "      <td>0.045633</td>\n",
       "      <td>0.244818</td>\n",
       "      <td>0.580654</td>\n",
       "      <td>0.813805</td>\n",
       "      <td>1.020183</td>\n",
       "      <td>0.902601</td>\n",
       "      <td>1.060589</td>\n",
       "      <td>1.033601</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>orange</th>\n",
       "      <td>0.000003</td>\n",
       "      <td>0.504529</td>\n",
       "      <td>0.729597</td>\n",
       "      <td>0.431167</td>\n",
       "      <td>0.446472</td>\n",
       "      <td>0.769800</td>\n",
       "      <td>0.656605</td>\n",
       "      <td>0.403276</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>pink</th>\n",
       "      <td>0.000003</td>\n",
       "      <td>0.195975</td>\n",
       "      <td>0.289415</td>\n",
       "      <td>0.548362</td>\n",
       "      <td>0.697557</td>\n",
       "      <td>1.002305</td>\n",
       "      <td>1.628003</td>\n",
       "      <td>1.256850</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>purple</th>\n",
       "      <td>0.000002</td>\n",
       "      <td>0.572350</td>\n",
       "      <td>0.601575</td>\n",
       "      <td>0.461895</td>\n",
       "      <td>0.777596</td>\n",
       "      <td>0.698840</td>\n",
       "      <td>0.651005</td>\n",
       "      <td>0.631839</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>yellow</th>\n",
       "      <td>0.000002</td>\n",
       "      <td>0.152688</td>\n",
       "      <td>0.429353</td>\n",
       "      <td>0.748959</td>\n",
       "      <td>0.910318</td>\n",
       "      <td>0.909728</td>\n",
       "      <td>1.070832</td>\n",
       "      <td>0.992498</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "               0h        1h        2h        4h       12h       16h       24h  \\\n",
       "cluster                                                                         \n",
       "green    0.045633  0.244818  0.580654  0.813805  1.020183  0.902601  1.060589   \n",
       "orange   0.000003  0.504529  0.729597  0.431167  0.446472  0.769800  0.656605   \n",
       "pink     0.000003  0.195975  0.289415  0.548362  0.697557  1.002305  1.628003   \n",
       "purple   0.000002  0.572350  0.601575  0.461895  0.777596  0.698840  0.651005   \n",
       "yellow   0.000002  0.152688  0.429353  0.748959  0.910318  0.909728  1.070832   \n",
       "\n",
       "              36h  \n",
       "cluster            \n",
       "green    1.033601  \n",
       "orange   0.403276  \n",
       "pink     1.256850  \n",
       "purple   0.631839  \n",
       "yellow   0.992498  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAV0AAABECAYAAAAiJuZQAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADh0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uMy4xLjEsIGh0dHA6Ly9tYXRwbG90bGliLm9yZy8QZhcZAAABwUlEQVR4nO3YoUolcRjG4e+IcNBkMApisggGk2g1mQ16NcdkM26RzVr3ChYM2k0nCCJqExGLnl1xvAJNO+9/0eepX3nL/GaYQdd1BUDGVOsBAN+J6AIEiS5AkOgCBIkuQND0p9fTg65eJ6Epeb9WN+ry5br1jN5sX+zVw9Xf1jN6sTb3WFMLM/V2+9x6Si+GW+dVz+PWM3pzMjeq8Z/WK/ozmq/BR7fPo/s6qfq9/88H/S8ul4/q8O5n6xm9Wb/aqbMfT61n9GJl6bqGu4s1Of6aL83h5rjq5us+e+PZUe3ft17Rn9H8xze/FwCCRBcgSHQBgkQXIEh0AYJEFyBIdAGCRBcgSHQBgkQXIEh0AYJEFyBIdAGCRBcgSHQBgkQXIEh0AYJEFyBIdAGCRBcgSHQBgkQXIEh0AYJEFyBIdAGCRBcgSHQBgkQXIEh0AYJEFyBIdAGCRBcgSHQBgkQXIEh0AYJEFyBIdAGCRBcgSHQBgkQXIEh0AYJEFyBIdAGCRBcgSHQBgkQXIEh0AYJEFyBIdAGCRBcgSHQBgkQXIEh0AYIGXde13gDwbfjSBQgSXYAg0QUIEl2AINEFCBJdgKB3/MgyM1UIEBMAAAAASUVORK5CYII=\n",
      "text/plain": [
       "<Figure size 432x72 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "##For sourcedata\n",
    "palette = sns.color_palette('bright',10) ###########CAUTION: HARD-CODED (max 6 colors)\n",
    "palette = [color for i,color in enumerate(palette) if not(i in [0,3,5,7])]\n",
    "sns.palplot(palette)\n",
    "palette = ['orange','green','purple','pink','yellow','skyblue']\n",
    "for tmp_sample_prefix in ['mRNA','RPF','QTI']:\n",
    "    print('-------'+tmp_sample_prefix+'-------')\n",
    "    tmp_DEG_df_list   = []\n",
    "    tmp_hpi_list      = ['%dh' %i for i in [0,1,2,4,12,16,24,36]]\n",
    "    clusterFname=f'timecourse_hostexp/{tmp_sample_prefix}_cluster.tsv'\n",
    "    ID_cl_df          = pd.read_csv(RESULTDIR %clusterFname, sep = '\\t', index_col = 0)\n",
    "    ID_cl_df['color'] = [palette[i-1] for i in ID_cl_df['cluster']]\n",
    "    timecourse_df     = pd.read_csv(RESULTDIR %(f'timecourse_hostexp/{tmp_sample_prefix}_timecourse.tsv'), sep = '\\t',index_col = 0)\n",
    "    timecourse_DEG_df = timecourse_df.loc[ID_cl_df.index]\n",
    "    timecourse_DEG_df['cluster'] = ID_cl_df['color']\n",
    "    timecourse_DEG_df.to_csv(RESULTDIR %(f'timecourse_hostexp/{tmp_sample_prefix}_timecourse_DEG.tsv'), sep = '\\t')\n",
    "    cluster_mean = timecourse_DEG_df.groupby('cluster').mean()\n",
    "    cluster_std  = timecourse_DEG_df.groupby('cluster').std()\n",
    "    display(cluster_mean)\n",
    "    display(cluster_std)\n",
    "    #display(cluster_n)\n",
    "    #cluster_CI95 = cluster_std/np.sqrt(cluster_n)*1.96\n",
    "    #display(cluster_CI95)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a9d75aee",
   "metadata": {},
   "source": [
    "### Fig. 6a-c: Inspection of the genes of interest"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "411aaada",
   "metadata": {},
   "outputs": [],
   "source": [
    "def plot_Fig6abc_multipanel(OutFigname=''):\n",
    "    #Fig.6a: Gordon et al + ACE2, TMPRSS2\n",
    "    Fig6a_df = pd.DataFrame(columns = ['color','linewidth','print_cutoff','genelist'])\n",
    "    Fig6a_df.loc['Gordonetal']  = ['#a52a2a', 0.5, 100.0, [i.strip().upper() for i in open(GENELISTDIR %'Gordonetal_interacting_proteins')]]\n",
    "    Fig6a_df.loc['ACE2+TMPRSS2']= ['blue',    1.0, 0.0,   ['ACE2', 'TMPRSS2']]\n",
    "    #Fig. 6b: IFN_typeI_III+Blancomelo_IFN1response_DEGs_sig\n",
    "    Fig6b_df = pd.DataFrame(columns = ['color','linewidth','print_cutoff','genelist'])\n",
    "    IFN_response_list      = [i.strip().upper() for i in open(GENELISTDIR %'IFN_typeI_III+Blancomelo_IFN1response_DEGs_sig')]\n",
    "    IFN_list               = [gsym for gsym in IFN_response_list if 'IFN' in gsym]\n",
    "    others_list            = [gsym for gsym in IFN_response_list if not('IFN' in gsym)]\n",
    "    Fig6b_df.loc['others'] = ['#a52a2a',    0.5, 100.0, others_list]\n",
    "    Fig6b_df.loc['IFN']    = ['green',    1.0,   100.0, IFN_list]\n",
    "    #Fig. 6c: Blancomelo_cytokineDEGs_sig_Fig2+4\n",
    "    Fig6c_df = pd.DataFrame(columns = ['color','linewidth','print_cutoff','genelist'])\n",
    "    Fig6c_df.loc['cytokine']  = ['#a52a2a', 0.5, 100.0, [i.strip().upper() for i in open(GENELISTDIR %'Blancomelo_cytokineDEGs_sig_Fig2+4')]]\n",
    "    \n",
    "    highlight_gene_info_df_list = [Fig6a_df,Fig6b_df,Fig6c_df]\n",
    "    tmp_sample_prefix_list = ['mRNA','RPF','QTI']\n",
    "    \n",
    "    fig, axes  = plt.subplots(nrows=len(highlight_gene_info_df_list), ncols=len(tmp_sample_prefix_list), \n",
    "                              figsize=(3.5*len(tmp_sample_prefix_list), 3*len(highlight_gene_info_df_list)))\n",
    "    \n",
    "    for idx_col, tmp_sample_prefix in enumerate(tmp_sample_prefix_list):\n",
    "        print('------'+tmp_sample_prefix+'------')\n",
    "        tmp_DEG_df_list   = []\n",
    "        tmp_hpi_list      = ['%dh' %i for i in [0,1,2,4,12,16,24,36]]\n",
    "        for hpi in tmp_hpi_list:\n",
    "            InFname    = f'DEG/{tmp_sample_prefix}_DESeq2_{hpi}.tsv'\n",
    "            tmp_DEG_df = pd.read_csv(RESULTDIR %(InFname), sep ='\\t', header=0, index_col=0)\n",
    "            tmp_DEG_df_list.append(tmp_DEG_df)\n",
    "        \n",
    "        for idx_row, highlight_gene_info_df in enumerate(highlight_gene_info_df_list):\n",
    "            display(highlight_gene_info_df)\n",
    "            ax = axes[idx_row,idx_col]\n",
    "            _             = timecourse_plot(tmp_DEG_df_list,tmp_hpi_list, padj_cut=0.01, log2fc_cut=2, clip = 2.8,\n",
    "                                        ylim = 5, n_cluster=5, as_average=False, \n",
    "                                        load_precalc_clusters=True, clusterFname=f'timecourse_hostexp/{tmp_sample_prefix}_cluster.tsv',\n",
    "                                        clusters_to_show=[], clusters_priority=[], highlight_gene_info_df = highlight_gene_info_df,\n",
    "                                        cl_method='ward',cl_metric='euclidean',ax = ax, \n",
    "                                        OutFigname='', show_fig=False, verbose=False)\n",
    "    if OutFigname != '':\n",
    "        #plt.tight_layout()\n",
    "        plt.savefig(FIGDIR %OutFigname)\n",
    "        plt.close()\n",
    "    else:\n",
    "        #plt.tight_layout()\n",
    "        plt.show()\n",
    "        plt.close()\n",
    "    return None\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "e404fae1",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "-------mRNA-------\n",
      "-------RPF-------\n",
      "-------QTI-------\n",
      "-------mRNA-------\n",
      "-------RPF-------\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/sonic/baeklab/Hyeonseong/anaconda3/envs/Doyeon-tf2/lib/python3.6/site-packages/ipykernel_launcher.py:47: FutureWarning: \n",
      "Passing list-likes to .loc or [] with any missing label will raise\n",
      "KeyError in the future, you can use .reindex() as an alternative.\n",
      "\n",
      "See the documentation here:\n",
      "https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#deprecate-loc-reindex-listlike\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "-------QTI-------\n",
      "-------mRNA-------\n",
      "-------RPF-------\n",
      "-------QTI-------\n",
      "-------mRNA-------\n",
      "-------RPF-------\n",
      "-------QTI-------\n",
      "-------mRNA-------\n",
      "-------RPF-------\n",
      "-------QTI-------\n"
     ]
    }
   ],
   "source": [
    "##source data for Fig. 6a\n",
    "Fig6a_df = pd.DataFrame(columns = ['color','linewidth','print_cutoff','genelist'])\n",
    "Fig6a_df.loc['Gordonetal']  = ['#a52a2a', 0.5, 100.0, [i.strip().upper() for i in open(GENELISTDIR %'Gordonetal_interacting_proteins')]]\n",
    "Fig6a_df.loc['ACE2+TMPRSS2']= ['blue',    1.0, 0.0,   ['ACE2', 'TMPRSS2']]\n",
    "#Fig. 6b: Daniloski targets\n",
    "Fig6b_df = pd.DataFrame(columns = ['color','linewidth','print_cutoff','genelist'])\n",
    "hostfactors_drugtargets = pd.read_csv(GENELISTDIR %'Daniloski_MOI03_dgidb',sep='\\t').iloc[:,0].str.upper().tolist()\n",
    "hostfactors = pd.read_csv(GENELISTDIR %'Daniloski_MOI03_rank',sep='\\t').iloc[:,0].str.upper().tolist()\n",
    "hostfactors_nontargets  = list(set(hostfactors)-set(hostfactors_drugtargets))\n",
    "Fig6b_df.loc['Daniloski_hostfactors_nontargets'] = ['#a52a2a', 0.5, 1.5, hostfactors_nontargets]\n",
    "Fig6b_df.loc['Daniloski_hostfactors_drugtargets']= ['magenta', 1.0, 1.5, hostfactors_drugtargets]\n",
    "\n",
    "#Fig. 6d: IFN_typeI_III+Blancomelo_IFN1response_DEGs_sig\n",
    "Fig6d_df = pd.DataFrame(columns = ['color','linewidth','print_cutoff','genelist'])\n",
    "IFN_response_list      = [i.strip().upper() for i in open(GENELISTDIR %'IFN_typeI_III+Blancomelo_IFN1response_DEGs_sig')]\n",
    "IFN_list               = [gsym for gsym in IFN_response_list if 'IFN' in gsym]\n",
    "others_list            = [gsym for gsym in IFN_response_list if not('IFN' in gsym)]\n",
    "Fig6d_df.loc['others'] = ['#a52a2a',    0.5, 100.0, others_list]\n",
    "Fig6d_df.loc['IFN']    = ['green',    1.0,   100.0, IFN_list]\n",
    "#Fig. 6e: Blancomelo_cytokineDEGs_sig_Fig2+4\n",
    "Fig6e_df = pd.DataFrame(columns = ['color','linewidth','print_cutoff','genelist'])\n",
    "Fig6e_df.loc['cytokine']  = ['#a52a2a', 0.5, 100.0, [i.strip().upper() for i in open(GENELISTDIR %'Blancomelo_cytokineDEGs_sig_Fig2+4')]]\n",
    "\n",
    "#Fig. 6i: eIFs related to the start codon recognition fidelity and non-AUC translation\n",
    "Fig6i_df = pd.DataFrame(columns = ['color','linewidth','print_cutoff','genelist'])\n",
    "\n",
    "Fig6i_df.loc['nonAUG_enhancers']       = ['red', 0.5, 1.5, ['EIF5B','EIF2A','EIF2D']]\n",
    "Fig6i_df.loc['nonAUG_enhancers_eIF5']  = ['red', 1.0, 1.5, ['EIF5']] #eIF5 alone for thicker line\n",
    "Fig6i_df.loc['nonAUG_repressors']      = ['blue', 0.5, 1.5, ['EIF1','EIF1AX','EIF1AY']]\n",
    "\n",
    "highlight_gene_info_df_list = [Fig6a_df, Fig6b_df, Fig6d_df, Fig6e_df, Fig6i_df]\n",
    "highlight_gene_name_list    = ['Fig6a', 'Fig6b', 'Fig6d', 'Fig6e', 'Fig6i']\n",
    "for highlight_gene_name, highlight_gene_info_df in zip(highlight_gene_name_list, highlight_gene_info_df_list):\n",
    "    ID_subgroup_df = pd.DataFrame(columns=['subgroup'], index = [i for genelist in highlight_gene_info_df['genelist'] for i in genelist ])\n",
    "    for subgroup, row in highlight_gene_info_df.iterrows():\n",
    "        ID_subgroup_df.loc[row['genelist']] = subgroup\n",
    "    #display(ID_subgroup_df)\n",
    "    for tmp_sample_prefix in ['mRNA','RPF','QTI']:\n",
    "        print('-------'+tmp_sample_prefix+'-------')\n",
    "        tmp_DEG_df_list   = []\n",
    "        tmp_hpi_list      = ['%dh' %i for i in [0,1,2,4,12,16,24,36]]\n",
    "        clusterFname=f'timecourse_hostexp/{tmp_sample_prefix}_cluster.tsv'\n",
    "        timecourse_df     = pd.read_csv(RESULTDIR %(f'timecourse_hostexp/{tmp_sample_prefix}_timecourse.tsv'), sep = '\\t',index_col = 0)\n",
    "\n",
    "\n",
    "\n",
    "        timecourse_DEG_df = timecourse_df.loc[ID_subgroup_df.index]\n",
    "        timecourse_DEG_df['subgroup'] = ID_subgroup_df['subgroup']\n",
    "        timecourse_DEG_df = timecourse_DEG_df.dropna()\n",
    "        timecourse_DEG_df.to_csv(RESULTDIR %(f'timecourse_hostexp/{highlight_gene_name}_{tmp_sample_prefix}_timecourse_Fig6genes.tsv'), sep = '\\t')\n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "af683b28",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "------mRNA------\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>color</th>\n",
       "      <th>linewidth</th>\n",
       "      <th>print_cutoff</th>\n",
       "      <th>genelist</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Gordonetal</th>\n",
       "      <td>#a52a2a</td>\n",
       "      <td>0.5</td>\n",
       "      <td>100.0</td>\n",
       "      <td>[PCNT, PVR, POLA1, FASTKD5, PRIM2, ITGB1, CNTR...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ACE2+TMPRSS2</th>\n",
       "      <td>blue</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>[ACE2, TMPRSS2]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                color  linewidth  print_cutoff  \\\n",
       "Gordonetal    #a52a2a        0.5         100.0   \n",
       "ACE2+TMPRSS2     blue        1.0           0.0   \n",
       "\n",
       "                                                       genelist  \n",
       "Gordonetal    [PCNT, PVR, POLA1, FASTKD5, PRIM2, ITGB1, CNTR...  \n",
       "ACE2+TMPRSS2                                    [ACE2, TMPRSS2]  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>color</th>\n",
       "      <th>linewidth</th>\n",
       "      <th>print_cutoff</th>\n",
       "      <th>genelist</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>others</th>\n",
       "      <td>#a52a2a</td>\n",
       "      <td>0.5</td>\n",
       "      <td>100.0</td>\n",
       "      <td>[MX1, XAF1, OAS1, IFITM1, OAS3, IRF9, OAS2, IR...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>IFN</th>\n",
       "      <td>green</td>\n",
       "      <td>1.0</td>\n",
       "      <td>100.0</td>\n",
       "      <td>[IFNA1, IFNA2, IFNA4, IFNA5, IFNA6, IFNA7, IFN...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          color  linewidth  print_cutoff  \\\n",
       "others  #a52a2a        0.5         100.0   \n",
       "IFN       green        1.0         100.0   \n",
       "\n",
       "                                                 genelist  \n",
       "others  [MX1, XAF1, OAS1, IFITM1, OAS3, IRF9, OAS2, IR...  \n",
       "IFN     [IFNA1, IFNA2, IFNA4, IFNA5, IFNA6, IFNA7, IFN...  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>color</th>\n",
       "      <th>linewidth</th>\n",
       "      <th>print_cutoff</th>\n",
       "      <th>genelist</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>cytokine</th>\n",
       "      <td>#a52a2a</td>\n",
       "      <td>0.5</td>\n",
       "      <td>100.0</td>\n",
       "      <td>[CCL2, CCL20, CCL8, CSF2, CSF3, CXCL1, CXCL14,...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            color  linewidth  print_cutoff  \\\n",
       "cytokine  #a52a2a        0.5         100.0   \n",
       "\n",
       "                                                   genelist  \n",
       "cytokine  [CCL2, CCL20, CCL8, CSF2, CSF3, CXCL1, CXCL14,...  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "------RPF------\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>color</th>\n",
       "      <th>linewidth</th>\n",
       "      <th>print_cutoff</th>\n",
       "      <th>genelist</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Gordonetal</th>\n",
       "      <td>#a52a2a</td>\n",
       "      <td>0.5</td>\n",
       "      <td>100.0</td>\n",
       "      <td>[PCNT, PVR, POLA1, FASTKD5, PRIM2, ITGB1, CNTR...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ACE2+TMPRSS2</th>\n",
       "      <td>blue</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>[ACE2, TMPRSS2]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                color  linewidth  print_cutoff  \\\n",
       "Gordonetal    #a52a2a        0.5         100.0   \n",
       "ACE2+TMPRSS2     blue        1.0           0.0   \n",
       "\n",
       "                                                       genelist  \n",
       "Gordonetal    [PCNT, PVR, POLA1, FASTKD5, PRIM2, ITGB1, CNTR...  \n",
       "ACE2+TMPRSS2                                    [ACE2, TMPRSS2]  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>color</th>\n",
       "      <th>linewidth</th>\n",
       "      <th>print_cutoff</th>\n",
       "      <th>genelist</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>others</th>\n",
       "      <td>#a52a2a</td>\n",
       "      <td>0.5</td>\n",
       "      <td>100.0</td>\n",
       "      <td>[MX1, XAF1, OAS1, IFITM1, OAS3, IRF9, OAS2, IR...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>IFN</th>\n",
       "      <td>green</td>\n",
       "      <td>1.0</td>\n",
       "      <td>100.0</td>\n",
       "      <td>[IFNA1, IFNA2, IFNA4, IFNA5, IFNA6, IFNA7, IFN...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          color  linewidth  print_cutoff  \\\n",
       "others  #a52a2a        0.5         100.0   \n",
       "IFN       green        1.0         100.0   \n",
       "\n",
       "                                                 genelist  \n",
       "others  [MX1, XAF1, OAS1, IFITM1, OAS3, IRF9, OAS2, IR...  \n",
       "IFN     [IFNA1, IFNA2, IFNA4, IFNA5, IFNA6, IFNA7, IFN...  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>color</th>\n",
       "      <th>linewidth</th>\n",
       "      <th>print_cutoff</th>\n",
       "      <th>genelist</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>cytokine</th>\n",
       "      <td>#a52a2a</td>\n",
       "      <td>0.5</td>\n",
       "      <td>100.0</td>\n",
       "      <td>[CCL2, CCL20, CCL8, CSF2, CSF3, CXCL1, CXCL14,...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            color  linewidth  print_cutoff  \\\n",
       "cytokine  #a52a2a        0.5         100.0   \n",
       "\n",
       "                                                   genelist  \n",
       "cytokine  [CCL2, CCL20, CCL8, CSF2, CSF3, CXCL1, CXCL14,...  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "------QTI------\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>color</th>\n",
       "      <th>linewidth</th>\n",
       "      <th>print_cutoff</th>\n",
       "      <th>genelist</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Gordonetal</th>\n",
       "      <td>#a52a2a</td>\n",
       "      <td>0.5</td>\n",
       "      <td>100.0</td>\n",
       "      <td>[PCNT, PVR, POLA1, FASTKD5, PRIM2, ITGB1, CNTR...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ACE2+TMPRSS2</th>\n",
       "      <td>blue</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>[ACE2, TMPRSS2]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                color  linewidth  print_cutoff  \\\n",
       "Gordonetal    #a52a2a        0.5         100.0   \n",
       "ACE2+TMPRSS2     blue        1.0           0.0   \n",
       "\n",
       "                                                       genelist  \n",
       "Gordonetal    [PCNT, PVR, POLA1, FASTKD5, PRIM2, ITGB1, CNTR...  \n",
       "ACE2+TMPRSS2                                    [ACE2, TMPRSS2]  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>color</th>\n",
       "      <th>linewidth</th>\n",
       "      <th>print_cutoff</th>\n",
       "      <th>genelist</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>others</th>\n",
       "      <td>#a52a2a</td>\n",
       "      <td>0.5</td>\n",
       "      <td>100.0</td>\n",
       "      <td>[MX1, XAF1, OAS1, IFITM1, OAS3, IRF9, OAS2, IR...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>IFN</th>\n",
       "      <td>green</td>\n",
       "      <td>1.0</td>\n",
       "      <td>100.0</td>\n",
       "      <td>[IFNA1, IFNA2, IFNA4, IFNA5, IFNA6, IFNA7, IFN...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          color  linewidth  print_cutoff  \\\n",
       "others  #a52a2a        0.5         100.0   \n",
       "IFN       green        1.0         100.0   \n",
       "\n",
       "                                                 genelist  \n",
       "others  [MX1, XAF1, OAS1, IFITM1, OAS3, IRF9, OAS2, IR...  \n",
       "IFN     [IFNA1, IFNA2, IFNA4, IFNA5, IFNA6, IFNA7, IFN...  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>color</th>\n",
       "      <th>linewidth</th>\n",
       "      <th>print_cutoff</th>\n",
       "      <th>genelist</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>cytokine</th>\n",
       "      <td>#a52a2a</td>\n",
       "      <td>0.5</td>\n",
       "      <td>100.0</td>\n",
       "      <td>[CCL2, CCL20, CCL8, CSF2, CSF3, CXCL1, CXCL14,...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            color  linewidth  print_cutoff  \\\n",
       "cytokine  #a52a2a        0.5         100.0   \n",
       "\n",
       "                                                   genelist  \n",
       "cytokine  [CCL2, CCL20, CCL8, CSF2, CSF3, CXCL1, CXCL14,...  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "plot_Fig6abc_multipanel(OutFigname='Fig5_timecourse_hostexp/Fig6abc_reported_genes_timecourse.pdf')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9af4fae8",
   "metadata": {},
   "source": [
    "### Fig. R: Inspection of the genes of pathophysiological relevance"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "c193040c",
   "metadata": {},
   "outputs": [],
   "source": [
    "def plot_FigR_pathophysio_multipanel(OutFigname=''):\n",
    "    #Fig.Rx: Daniloski et al. MOI0.3,CRISPR log2fc>1 host factors (drugtargets: subgroup)\n",
    "    FigRx_df = pd.DataFrame(columns = ['color','linewidth','print_cutoff','genelist'])\n",
    "    hostfactors_drugtargets = pd.read_csv(GENELISTDIR %'Daniloski_MOI03_dgidb',sep='\\t').iloc[:,0].str.upper().tolist()\n",
    "    hostfactors = pd.read_csv(GENELISTDIR %'Daniloski_MOI03_rank',sep='\\t').iloc[:,0].str.upper().tolist()\n",
    "    hostfactors_nontargets  = list(set(hostfactors)-set(hostfactors_drugtargets))\n",
    "    FigRx_df.loc['Daniloski_hostfactors_nontargets'] = ['#a52a2a', 0.5, 1.5, hostfactors_nontargets]\n",
    "    FigRx_df.loc['Daniloski_hostfactors_drugtargets']= ['magenta', 1.0, 1.5, hostfactors_drugtargets]\n",
    "    \n",
    "    #Fig.Ry: eIFs related to the start codon recognition fidelity and non-AUC translation\n",
    "    FigRy_df = pd.DataFrame(columns = ['color','linewidth','print_cutoff','genelist'])\n",
    "    \n",
    "    FigRy_df.loc['nonAUG_enhancers']       = ['red', 0.5, 1.5, ['EIF5B','EIF2A','EIF2D']]\n",
    "    FigRy_df.loc['nonAUG_enhancers_eIF5']  = ['red', 1.0, 1.5, ['EIF5']] #eIF5 alone for thicker line\n",
    "    FigRy_df.loc['nonAUG_repressors']      = ['blue', 0.5, 1.5, ['EIF1','EIF1AX','EIF1AY']]\n",
    "    \n",
    "    highlight_gene_info_df_list = [FigRx_df,FigRy_df]\n",
    "    tmp_sample_prefix_list = ['mRNA','RPF','QTI']\n",
    "    \n",
    "    fig, axes  = plt.subplots(nrows=len(highlight_gene_info_df_list), ncols=len(tmp_sample_prefix_list), \n",
    "                              figsize=(3.5*len(tmp_sample_prefix_list), 3*len(highlight_gene_info_df_list)))\n",
    "    \n",
    "    for idx_col, tmp_sample_prefix in enumerate(tmp_sample_prefix_list):\n",
    "        print('------'+tmp_sample_prefix+'------')\n",
    "        tmp_DEG_df_list   = []\n",
    "        tmp_hpi_list      = ['%dh' %i for i in [0,1,2,4,12,16,24,36]]\n",
    "        for hpi in tmp_hpi_list:\n",
    "            InFname    = f'DEG/{tmp_sample_prefix}_DESeq2_{hpi}.tsv'\n",
    "            tmp_DEG_df = pd.read_csv(RESULTDIR %(InFname), sep ='\\t', header=0, index_col=0)\n",
    "            tmp_DEG_df_list.append(tmp_DEG_df)\n",
    "        \n",
    "        for idx_row, highlight_gene_info_df in enumerate(highlight_gene_info_df_list):\n",
    "            if idx_row == 0:\n",
    "                ylim = (-5,2.5)\n",
    "            else:\n",
    "                ylim = 2.5\n",
    "            display(highlight_gene_info_df)\n",
    "            ax = axes[idx_row,idx_col]\n",
    "            _             = timecourse_plot(tmp_DEG_df_list,tmp_hpi_list, padj_cut=0.01, log2fc_cut=2, clip = 2.8,\n",
    "                                        ylim = ylim, n_cluster=5, as_average=False, \n",
    "                                        load_precalc_clusters=True, clusterFname=f'timecourse_hostexp/{tmp_sample_prefix}_cluster.tsv',\n",
    "                                        clusters_to_show=[], clusters_priority=[], highlight_gene_info_df = highlight_gene_info_df,\n",
    "                                        cl_method='ward',cl_metric='euclidean',ax = ax, \n",
    "                                        OutFigname='', show_fig=False, verbose=False)\n",
    "    if OutFigname != '':\n",
    "        #plt.tight_layout()\n",
    "        plt.savefig(FIGDIR %OutFigname)\n",
    "        plt.close()\n",
    "    else:\n",
    "        #plt.tight_layout()\n",
    "        plt.show()\n",
    "        plt.close()\n",
    "    return None\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "7bda8a4c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "------mRNA------\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>color</th>\n",
       "      <th>linewidth</th>\n",
       "      <th>print_cutoff</th>\n",
       "      <th>genelist</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Daniloski_hostfactors_nontargets</th>\n",
       "      <td>#a52a2a</td>\n",
       "      <td>0.5</td>\n",
       "      <td>1.5</td>\n",
       "      <td>[GNGT1, ATP6V1G1, SPEN, KIAA1033, UVRAG, PGBD4...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Daniloski_hostfactors_drugtargets</th>\n",
       "      <td>magenta</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.5</td>\n",
       "      <td>[VPS35, ACE2, ATP6V1C1, ATP6V1B2, CTSL, HDAC9,...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                     color  linewidth  print_cutoff  \\\n",
       "Daniloski_hostfactors_nontargets   #a52a2a        0.5           1.5   \n",
       "Daniloski_hostfactors_drugtargets  magenta        1.0           1.5   \n",
       "\n",
       "                                                                            genelist  \n",
       "Daniloski_hostfactors_nontargets   [GNGT1, ATP6V1G1, SPEN, KIAA1033, UVRAG, PGBD4...  \n",
       "Daniloski_hostfactors_drugtargets  [VPS35, ACE2, ATP6V1C1, ATP6V1B2, CTSL, HDAC9,...  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>color</th>\n",
       "      <th>linewidth</th>\n",
       "      <th>print_cutoff</th>\n",
       "      <th>genelist</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>nonAUG_enhancers</th>\n",
       "      <td>red</td>\n",
       "      <td>0.5</td>\n",
       "      <td>1.5</td>\n",
       "      <td>[EIF5B, EIF2A, EIF2D]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>nonAUG_enhancers_eIF5</th>\n",
       "      <td>red</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.5</td>\n",
       "      <td>[EIF5]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>nonAUG_repressors</th>\n",
       "      <td>blue</td>\n",
       "      <td>0.5</td>\n",
       "      <td>1.5</td>\n",
       "      <td>[EIF1, EIF1AX, EIF1AY]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                      color  linewidth  print_cutoff                genelist\n",
       "nonAUG_enhancers        red        0.5           1.5   [EIF5B, EIF2A, EIF2D]\n",
       "nonAUG_enhancers_eIF5   red        1.0           1.5                  [EIF5]\n",
       "nonAUG_repressors      blue        0.5           1.5  [EIF1, EIF1AX, EIF1AY]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "------RPF------\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>color</th>\n",
       "      <th>linewidth</th>\n",
       "      <th>print_cutoff</th>\n",
       "      <th>genelist</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Daniloski_hostfactors_nontargets</th>\n",
       "      <td>#a52a2a</td>\n",
       "      <td>0.5</td>\n",
       "      <td>1.5</td>\n",
       "      <td>[GNGT1, ATP6V1G1, SPEN, KIAA1033, UVRAG, PGBD4...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Daniloski_hostfactors_drugtargets</th>\n",
       "      <td>magenta</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.5</td>\n",
       "      <td>[VPS35, ACE2, ATP6V1C1, ATP6V1B2, CTSL, HDAC9,...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                     color  linewidth  print_cutoff  \\\n",
       "Daniloski_hostfactors_nontargets   #a52a2a        0.5           1.5   \n",
       "Daniloski_hostfactors_drugtargets  magenta        1.0           1.5   \n",
       "\n",
       "                                                                            genelist  \n",
       "Daniloski_hostfactors_nontargets   [GNGT1, ATP6V1G1, SPEN, KIAA1033, UVRAG, PGBD4...  \n",
       "Daniloski_hostfactors_drugtargets  [VPS35, ACE2, ATP6V1C1, ATP6V1B2, CTSL, HDAC9,...  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>color</th>\n",
       "      <th>linewidth</th>\n",
       "      <th>print_cutoff</th>\n",
       "      <th>genelist</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>nonAUG_enhancers</th>\n",
       "      <td>red</td>\n",
       "      <td>0.5</td>\n",
       "      <td>1.5</td>\n",
       "      <td>[EIF5B, EIF2A, EIF2D]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>nonAUG_enhancers_eIF5</th>\n",
       "      <td>red</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.5</td>\n",
       "      <td>[EIF5]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>nonAUG_repressors</th>\n",
       "      <td>blue</td>\n",
       "      <td>0.5</td>\n",
       "      <td>1.5</td>\n",
       "      <td>[EIF1, EIF1AX, EIF1AY]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                      color  linewidth  print_cutoff                genelist\n",
       "nonAUG_enhancers        red        0.5           1.5   [EIF5B, EIF2A, EIF2D]\n",
       "nonAUG_enhancers_eIF5   red        1.0           1.5                  [EIF5]\n",
       "nonAUG_repressors      blue        0.5           1.5  [EIF1, EIF1AX, EIF1AY]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "------QTI------\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>color</th>\n",
       "      <th>linewidth</th>\n",
       "      <th>print_cutoff</th>\n",
       "      <th>genelist</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Daniloski_hostfactors_nontargets</th>\n",
       "      <td>#a52a2a</td>\n",
       "      <td>0.5</td>\n",
       "      <td>1.5</td>\n",
       "      <td>[GNGT1, ATP6V1G1, SPEN, KIAA1033, UVRAG, PGBD4...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Daniloski_hostfactors_drugtargets</th>\n",
       "      <td>magenta</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.5</td>\n",
       "      <td>[VPS35, ACE2, ATP6V1C1, ATP6V1B2, CTSL, HDAC9,...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                     color  linewidth  print_cutoff  \\\n",
       "Daniloski_hostfactors_nontargets   #a52a2a        0.5           1.5   \n",
       "Daniloski_hostfactors_drugtargets  magenta        1.0           1.5   \n",
       "\n",
       "                                                                            genelist  \n",
       "Daniloski_hostfactors_nontargets   [GNGT1, ATP6V1G1, SPEN, KIAA1033, UVRAG, PGBD4...  \n",
       "Daniloski_hostfactors_drugtargets  [VPS35, ACE2, ATP6V1C1, ATP6V1B2, CTSL, HDAC9,...  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>color</th>\n",
       "      <th>linewidth</th>\n",
       "      <th>print_cutoff</th>\n",
       "      <th>genelist</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>nonAUG_enhancers</th>\n",
       "      <td>red</td>\n",
       "      <td>0.5</td>\n",
       "      <td>1.5</td>\n",
       "      <td>[EIF5B, EIF2A, EIF2D]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>nonAUG_enhancers_eIF5</th>\n",
       "      <td>red</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.5</td>\n",
       "      <td>[EIF5]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>nonAUG_repressors</th>\n",
       "      <td>blue</td>\n",
       "      <td>0.5</td>\n",
       "      <td>1.5</td>\n",
       "      <td>[EIF1, EIF1AX, EIF1AY]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                      color  linewidth  print_cutoff                genelist\n",
       "nonAUG_enhancers        red        0.5           1.5   [EIF5B, EIF2A, EIF2D]\n",
       "nonAUG_enhancers_eIF5   red        1.0           1.5                  [EIF5]\n",
       "nonAUG_repressors      blue        0.5           1.5  [EIF1, EIF1AX, EIF1AY]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "plot_FigR_pathophysio_multipanel(OutFigname='Fig5_timecourse_hostexp/FigR_pathophysiology_timecourse.pdf')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c6720e1e",
   "metadata": {},
   "source": [
    "### Fig.R: correlation b/w max(|log2fc|) ~ Daniloski et al. physiological impact"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "4605c846",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "vs. maxabs SpearmanrResult(correlation=0.387003406631181, pvalue=0.0007844012325431527)\n",
      "vs. 36h SpearmanrResult(correlation=-0.16277839064274974, pvalue=0.17188299305160892)\n",
      "vs. max SpearmanrResult(correlation=0.04946490811923513, pvalue=0.679877249165559)\n",
      "vs. min SpearmanrResult(correlation=-0.2555169426117058, pvalue=0.030289291368957135)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/sonic/baeklab/Hyeonseong/anaconda3/envs/Doyeon-tf2/lib/python3.6/site-packages/ipykernel_launcher.py:19: FutureWarning: \n",
      "Passing list-likes to .loc or [] with any missing label will raise\n",
      "KeyError in the future, you can use .reindex() as an alternative.\n",
      "\n",
      "See the documentation here:\n",
      "https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#deprecate-loc-reindex-listlike\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0h</th>\n",
       "      <th>1h</th>\n",
       "      <th>2h</th>\n",
       "      <th>4h</th>\n",
       "      <th>12h</th>\n",
       "      <th>16h</th>\n",
       "      <th>24h</th>\n",
       "      <th>36h</th>\n",
       "      <th>max</th>\n",
       "      <th>rank</th>\n",
       "      <th>lfc</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ACE2</th>\n",
       "      <td>-6.607765e-06</td>\n",
       "      <td>-0.000702</td>\n",
       "      <td>-0.055994</td>\n",
       "      <td>-0.319392</td>\n",
       "      <td>-2.566465</td>\n",
       "      <td>-3.545599</td>\n",
       "      <td>-3.782672</td>\n",
       "      <td>-4.101040</td>\n",
       "      <td>4.101040</td>\n",
       "      <td>12</td>\n",
       "      <td>1.6266</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ATP6AP2</th>\n",
       "      <td>-8.086250e-06</td>\n",
       "      <td>0.001427</td>\n",
       "      <td>0.116543</td>\n",
       "      <td>0.130180</td>\n",
       "      <td>-1.579491</td>\n",
       "      <td>-2.076438</td>\n",
       "      <td>-2.049512</td>\n",
       "      <td>-1.855499</td>\n",
       "      <td>2.076438</td>\n",
       "      <td>5</td>\n",
       "      <td>2.0384</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRSF1</th>\n",
       "      <td>1.806937e-06</td>\n",
       "      <td>0.004399</td>\n",
       "      <td>0.659602</td>\n",
       "      <td>1.374653</td>\n",
       "      <td>1.379408</td>\n",
       "      <td>1.851016</td>\n",
       "      <td>1.163534</td>\n",
       "      <td>0.166559</td>\n",
       "      <td>1.851016</td>\n",
       "      <td>30</td>\n",
       "      <td>1.2700</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>COMMD4</th>\n",
       "      <td>-9.434869e-07</td>\n",
       "      <td>-0.000646</td>\n",
       "      <td>-0.057539</td>\n",
       "      <td>-0.217448</td>\n",
       "      <td>-1.021387</td>\n",
       "      <td>-0.324847</td>\n",
       "      <td>-0.372262</td>\n",
       "      <td>-1.518002</td>\n",
       "      <td>1.518002</td>\n",
       "      <td>38</td>\n",
       "      <td>1.1835</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ATP6V1G1</th>\n",
       "      <td>-2.726768e-06</td>\n",
       "      <td>-0.000727</td>\n",
       "      <td>0.044275</td>\n",
       "      <td>0.437669</td>\n",
       "      <td>1.441861</td>\n",
       "      <td>1.310660</td>\n",
       "      <td>1.383131</td>\n",
       "      <td>1.245990</td>\n",
       "      <td>1.441861</td>\n",
       "      <td>15</td>\n",
       "      <td>1.5741</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                    0h        1h        2h        4h       12h       16h  \\\n",
       "id                                                                         \n",
       "ACE2     -6.607765e-06 -0.000702 -0.055994 -0.319392 -2.566465 -3.545599   \n",
       "ATP6AP2  -8.086250e-06  0.001427  0.116543  0.130180 -1.579491 -2.076438   \n",
       "SRSF1     1.806937e-06  0.004399  0.659602  1.374653  1.379408  1.851016   \n",
       "COMMD4   -9.434869e-07 -0.000646 -0.057539 -0.217448 -1.021387 -0.324847   \n",
       "ATP6V1G1 -2.726768e-06 -0.000727  0.044275  0.437669  1.441861  1.310660   \n",
       "\n",
       "               24h       36h       max  rank     lfc  \n",
       "id                                                    \n",
       "ACE2     -3.782672 -4.101040  4.101040    12  1.6266  \n",
       "ATP6AP2  -2.049512 -1.855499  2.076438     5  2.0384  \n",
       "SRSF1     1.163534  0.166559  1.851016    30  1.2700  \n",
       "COMMD4   -0.372262 -1.518002  1.518002    38  1.1835  \n",
       "ATP6V1G1  1.383131  1.245990  1.441861    15  1.5741  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "def calc_plot_assoc_lfc_physioimpact(sample_prefix='RPF', OutFigname=''):\n",
    "    tmp_DEG_df_list   = []\n",
    "    tmp_hpi_list      = ['%dh' %i for i in [0,1,2,4,12,16,24,36]]\n",
    "    for hpi in tmp_hpi_list:\n",
    "        InFname    = f'DEG/{sample_prefix}_DESeq2_{hpi}.tsv'\n",
    "        tmp_DEG_df = pd.read_csv(RESULTDIR %(InFname), sep ='\\t', header=0, index_col=0)\n",
    "        tmp_DEG_df_list.append(tmp_DEG_df)\n",
    "        \n",
    "    \n",
    "    hostfactors_df = pd.read_csv(GENELISTDIR %'Daniloski_MOI03_rank',sep='\\t')\n",
    "    hostfactors_df['gsym'] = hostfactors_df['gsym'].str.upper()\n",
    "    hostfactors    = hostfactors_df['gsym']\n",
    "    hostfactors_df = hostfactors_df.set_index('gsym')\n",
    "    \n",
    "    hostfactors_drugtargets = pd.read_csv(GENELISTDIR %'Daniloski_MOI03_dgidb',sep='\\t').iloc[:,0].str.upper().tolist()\n",
    "    \n",
    "    timecourse_df = pd.read_csv(RESULTDIR %(f'timecourse_hostexp/{sample_prefix}_timecourse.tsv'), \n",
    "                                sep = '\\t', index_col = 0)\n",
    "    timecourse_df = timecourse_df.loc[hostfactors]\n",
    "    timecourse_df['max'] = timecourse_df.abs().max(axis=1)\n",
    "    timecourse_df['rank']= hostfactors_df['rank']\n",
    "    timecourse_df['lfc'] = hostfactors_df['lfc']\n",
    "    timecourse_df = timecourse_df.dropna(how='any')\n",
    "    colors = ['magenta' if i in hostfactors_drugtargets else '#a52a2a' for i in timecourse_df.index]\n",
    "    print('vs. maxabs',stats.spearmanr(timecourse_df['lfc'],timecourse_df['max']))\n",
    "    print('vs. 36h',   stats.spearmanr(timecourse_df['lfc'],timecourse_df['36h']))\n",
    "    print('vs. max',   stats.spearmanr(timecourse_df['lfc'],timecourse_df[tmp_hpi_list].max(axis=1)))\n",
    "    print('vs. min',   stats.spearmanr(timecourse_df['lfc'],timecourse_df[tmp_hpi_list].min(axis=1)))\n",
    "    \n",
    "    xvar = 'max'\n",
    "    yvar = 'rank'\n",
    "    fig  = plt.figure(figsize=(3.5,3.5))\n",
    "    display(timecourse_df.sort_values(xvar,ascending=False).head(5))\n",
    "    plt.scatter(timecourse_df[xvar],timecourse_df[yvar], c = colors)\n",
    "    plt.ylabel('CRISPR screening rank')\n",
    "    plt.xlabel('max(|log2fc|)')\n",
    "    plt.ylim(bottom=0)\n",
    "    timecourse_df[[xvar,yvar]].to_csv(RESULTDIR %(f'timecourse_hostexp/Fig6i_correlation.tsv'), \n",
    "                                      sep = '\\t')\n",
    "    z = np.polyfit( timecourse_df[xvar], timecourse_df[yvar],1)\n",
    "    p = np.poly1d(z)\n",
    "    plt.plot(timecourse_df[xvar].sort_values(),p(timecourse_df[xvar].sort_values()),\"k--\")\n",
    "    sprho, spp = stats.spearmanr(timecourse_df[xvar],timecourse_df[yvar])\n",
    "    plt.text(timecourse_df[xvar].max()/2,timecourse_df[yvar].max()/1.1,f'Spearman rho={sprho:.03f}\\nP={spp:.02e}')\n",
    "    \n",
    "    plt.tight_layout()\n",
    "    if OutFigname != '':\n",
    "        plt.savefig(FIGDIR %OutFigname)\n",
    "    else:\n",
    "        plt.show()\n",
    "    plt.close()\n",
    "    \n",
    "    return None\n",
    "\n",
    "calc_plot_assoc_lfc_physioimpact(sample_prefix = 'RPF', OutFigname='Fig5_timecourse_hostexp/FigR_phyiological_assoc.pdf')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e875aee5",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "533baba9",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d3e80734",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e4d0d43c",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5611732f",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6cbfa57a",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Doyeon-tf2",
   "language": "python",
   "name": "doyeon-tf2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
