{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "82e63bfc",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import rpy2\n",
    "import rpy2.robjects as ro\n",
    "from rpy2.robjects import pandas2ri, Formula\n",
    "pandas2ri.activate()\n",
    "from rpy2.robjects.packages import importr\n",
    "from rpy2.robjects.conversion import localconverter\n",
    "'''\n",
    "Using DESeq2 in python\n",
    "Adopted from:  https://stackoverflow.com/questions/41821100/running-deseq2-through-rpy2\n",
    "and also from: https://gist.github.com/wckdouglas/3f8fb27a3d7a1eb24c598aa04f70fb25\n",
    "'''\n",
    "\n",
    "import os, sys, time,re\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import matplotlib as mpl\n",
    "import matplotlib.pyplot as plt\n",
    "import matplotlib.font_manager as fm\n",
    "import matplotlib.lines as mlines\n",
    "from matplotlib.gridspec import  GridSpec\n",
    "import seaborn as sns\n",
    "\n",
    "from scipy import stats\n",
    "from scipy.cluster import hierarchy\n",
    "from statsmodels.stats.multitest import multipletests\n",
    "from collections import defaultdict\n",
    "import pysam\n",
    "from collections import defaultdict\n",
    "import statsmodels.api as sm\n",
    "\n",
    "REFFLAT_hg38      = '../ref/refFlat_hg38_repiso.txt'\n",
    "REFFLAT_chlSab2   = '../ref/refFlat_chlSab2.txt'     # Green monkey genome, for Vero cell data.\n",
    "REFFLAT_SARSCOV2  = '../ref/annot_SARSCOV2.txt'      # Not exactly refFlat, but similar format. Used ORF start-end information.\n",
    "\n",
    "BAMDIR_hostmapped = '/extdata1/baeklab/Doyeon/SARSCOV2/data/%s_hostalign_021721/%s.bam'     #e.g. %('mRNASeq','mRNA_2h_rep1')\n",
    "BAMDIR_cov2mapped = '/extdata1/baeklab/Doyeon/SARSCOV2/data/%s_SARSCOV2align_021721/%s.bam' #e.g. %('mRNASeq','mRNA_2h_rep1')\n",
    "RPKMDIR           = '/extdata1/baeklab/Doyeon/SARSCOV2/data/rpkm_081820/%s.txt'             #e.g. %'RPF_2h_rep1'\n",
    "'''\n",
    "Sequencing data can be downloaded from https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE157490\n",
    "'''\n",
    "\n",
    "GENELISTDIR  = '../ref/%s.txt' #e.g. %'Gordonetal_interacting_proteins'\n",
    "RESULTDIR    = '../result/%s'  #e.g. %('mRNA_quantification.tsv')\n",
    "FIGDIR       = '../fig/%s'     #e.g. %('revision_QC/021721_QTISeq_12h.pdf')\n",
    "\n",
    "%matplotlib inline"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "76646c3d",
   "metadata": {},
   "source": [
    "## DESeq R package"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "6a36e2af",
   "metadata": {},
   "outputs": [],
   "source": [
    "'''\n",
    "For install DESeq2 packages:\n",
    "open R: ~/baeklab/Hyeonseong/anaconda3/envs/Doyeon-tf2/bin/R\n",
    "> install.packages(\"BiocManager\")\n",
    "> BiocManager::install(\"DESeq2\")\n",
    "loooooooong time required for installing dependent packages\n",
    "'''\n",
    "\n",
    "deseq2 = importr('DESeq2')\n",
    "to_dataframe = ro.r('function(x) data.frame(x)')\n",
    "BiocGenerics = importr('BiocGenerics')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "a957c0e2",
   "metadata": {},
   "outputs": [],
   "source": [
    "class py_DESeq2:\n",
    "    '''\n",
    "    DESeq2 object through rpy2\n",
    "    input:\n",
    "    count_matrix: should be a pandas dataframe with each column as count, and a id column for gene id\n",
    "        example:\n",
    "        id    sampleA    sampleB\n",
    "        geneA    5    1\n",
    "        geneB    4    5\n",
    "        geneC    1    2\n",
    "    design_matrix: an design matrix in the form of pandas dataframe, see DESeq2 manual, samplenames as rownames\n",
    "                treatment\n",
    "    sampleA1        A\n",
    "    sampleA2        A\n",
    "    sampleB1        B\n",
    "    sampleB2        B\n",
    "    design_formula: see DESeq2 manual, example: \"~ treatment\"\"\n",
    "    gene_column: column name of gene id columns, exmplae \"id\"\n",
    "    '''\n",
    "    def __init__(self, count_matrix, design_matrix, design_formula, gene_column='id'):\n",
    "        try:\n",
    "            assert gene_column in count_matrix.columns, 'Wrong gene id column name'\n",
    "            gene_id = count_matrix[gene_column]\n",
    "        except AttributeError:\n",
    "            sys.exit('Wrong Pandas dataframe?')\n",
    "\n",
    "        self.dds = None\n",
    "        self.deseq_result = None\n",
    "        self.deseq_result_Robj = None\n",
    "        self.resLFC = None\n",
    "        self.comparison = None\n",
    "        self.normalized_count_matrix = None\n",
    "        self.gene_column = gene_column\n",
    "        self.gene_id = count_matrix[self.gene_column]\n",
    "        with localconverter(ro.default_converter + pandas2ri.converter):\n",
    "            self.count_matrix = ro.conversion.py2rpy(count_matrix.drop(gene_column,axis=1))\n",
    "            self.design_matrix = ro.conversion.py2rpy(design_matrix)\n",
    "        self.design_formula = Formula(design_formula)\n",
    "\n",
    "    def run_deseq(self, **kwargs):\n",
    "        self.dds = deseq2.DESeqDataSetFromMatrix(countData=self.count_matrix, \n",
    "                                        colData=self.design_matrix,\n",
    "                                        design=self.design_formula)        \n",
    "        self.dds = deseq2.DESeq(self.dds, **kwargs)\n",
    "        \n",
    "    def get_deseq_result(self,lfcshrink=True, **kwargs):\n",
    "        self.comparison = deseq2.resultsNames(self.dds)\n",
    "        if lfcshrink:\n",
    "            self.deseq_result_Robj = deseq2.lfcShrink(self.dds,coef='treatment_B_vs_A',type='apeglm',**kwargs)\n",
    "        else:\n",
    "            self.deseq_result_Robj = deseq2.results(self.dds, **kwargs)\n",
    "        self.deseq_result = to_dataframe(self.deseq_result_Robj)\n",
    "        \n",
    "        with localconverter(ro.default_converter + pandas2ri.converter):\n",
    "            self.deseq_result = ro.conversion.rpy2py(self.deseq_result) ## back to pandas dataframe\n",
    "        \n",
    "        self.deseq_result[self.gene_column] = self.gene_id.values\n",
    "        "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1bed18fc",
   "metadata": {},
   "source": [
    "## Basic functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "109df0d7",
   "metadata": {},
   "outputs": [],
   "source": [
    "class gene:\n",
    "    def __init__(self):\n",
    "        self.sGeneSym       = ''\n",
    "        self.sNMID          = ''\n",
    "        self.sChrID         = ''\n",
    "        self.nExons         = 0\n",
    "        self.nExonStartlist = []\n",
    "        self.nExonEndlist   = []\n",
    "        self.sStrandDir     = ''\n",
    "        self.nORF_5p_gidx   = 0\n",
    "        self.nORF_3p_gidx   = 0\n",
    "        self.nExonlen       = 0\n",
    "        self.nU5len         = 0\n",
    "        self.nU3len         = 0\n",
    "\n",
    "    def parse_refflat(self,refline):\n",
    "        sInfolist = refline.strip().replace(' ','\\t').split('\\t')\n",
    "        self.sGeneSym = sInfolist[0].upper()\n",
    "        self.sNMID    = sInfolist[1]\n",
    "        self.sChrID   = sInfolist[2] ##chr1,,,,chrX,chrY for human\n",
    "        self.sStrandDir   = sInfolist[3]\n",
    "        self.nORF_5p_gidx = int(sInfolist[6])\n",
    "        self.nORF_3p_gidx = int(sInfolist[7])\n",
    "        self.nExons       = int(sInfolist[8])\n",
    "        self.nExonStartlist = [int(i) for i in sInfolist[9].split(',') if i != '']\n",
    "        self.nExonEndlist   = [int(i) for i in sInfolist[10].split(',') if i != '']\n",
    "        assert (self.nExons == len(self.nExonStartlist)) and (self.nExons == len(self.nExonEndlist))\n",
    "        self.nExonlen       = sum([end-start for start,end in zip(self.nExonStartlist, self.nExonEndlist)]) \n",
    "        \n",
    "        tmp_exonlen = 0\n",
    "        for start,end in zip(self.nExonStartlist, self.nExonEndlist):\n",
    "            if start <= self.nORF_5p_gidx <  end:\n",
    "                self.nU5len  = tmp_exonlen + (self.nORF_5p_gidx - start)\n",
    "            if start <= self.nORF_3p_gidx   <= end:\n",
    "                self.nU3len = self.nExonlen - (tmp_exonlen + (self.nORF_3p_gidx - start))\n",
    "            tmp_exonlen += (end-start)\n",
    "        if self.sStrandDir == '-':\n",
    "            tmp_nU3len  = self.nU5len\n",
    "            self.nU5len = self.nU3len\n",
    "            self.nU3len = tmp_nU3len\n",
    "#########################\n",
    "def load_RPKM(samplename):\n",
    "    '''\n",
    "    id\tsymbol\tef_length\tnsegments1\tnsegments2\tnsegments3\treadcount\treadcount_uniq\texlev\texrnk\texpct\n",
    "    NM_006004\tUQCRH\t276\t18702\t18702\t18494\t9250.416666666679\t32\t27048.868164275482\t1.0\t0.005263989050902774\n",
    "    '''\n",
    "    df = pd.read_csv(RPKMDIR %samplename,sep='\\t')\n",
    "    df = df.rename(columns={'id':'NMID','symbol':'gsym',\n",
    "                            'readcount':'nreads','exlev':'RPKM'})\n",
    "    \n",
    "    df['gsym'] = df['gsym'].str.upper()\n",
    "    \n",
    "    df['nreads'] = df['nreads'].round(0)\n",
    "    df = df[['NMID','gsym','nreads','RPKM','exrnk','expct']]\n",
    "    df = df.set_index('NMID')\n",
    "    return df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "71c4da4e",
   "metadata": {},
   "source": [
    "## Figure 5a- DEG identification and plotting"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0f8c4ba6",
   "metadata": {},
   "source": [
    "### DEG identification"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "450bfa7f",
   "metadata": {},
   "outputs": [],
   "source": [
    "def identify_DEG(infected=[],uninfected=[], DEG_method=py_DESeq2, verbose=False,\n",
    "                 OutF = ''):\n",
    "    ## infected samples\n",
    "    nread_df      = load_RPKM(infected[0])\n",
    "    infected_nreads_df         = nread_df[['gsym','nreads']]\n",
    "    infected_nreads_df.columns = ['gsym',infected[0]]\n",
    "    infected_expct_df          = nread_df[['gsym','expct']]\n",
    "    infected_expct_df.columns  = ['gsym',infected[0]+'_expct']\n",
    "    for samplename in infected[1:]:\n",
    "        nread_df = load_RPKM(samplename)\n",
    "        infected_nreads_df[samplename]         = nread_df['nreads']\n",
    "        infected_expct_df[samplename+'_expct'] = nread_df['expct']\n",
    "    infected_expct_df = infected_expct_df.set_index(['gsym']).mean(axis=1) #average expct\n",
    "    ##\n",
    "    ## uninfected samples\n",
    "    nread_df      = load_RPKM(uninfected[0])\n",
    "    uninfected_nreads_df         = nread_df[['gsym','nreads']]\n",
    "    uninfected_nreads_df.columns = ['gsym',uninfected[0]]\n",
    "    uninfected_expct_df          = nread_df[['gsym','expct']]\n",
    "    uninfected_expct_df.columns  = ['gsym',uninfected[0]+'_expct']\n",
    "    for samplename in uninfected[1:]:\n",
    "        nread_df = load_RPKM(samplename)\n",
    "        uninfected_nreads_df[samplename]         = nread_df['nreads']\n",
    "        uninfected_expct_df[samplename+'_expct'] = nread_df['expct']\n",
    "    uninfected_expct_df = uninfected_expct_df.set_index(['gsym']).mean(axis=1) #average expct\n",
    "    ##\n",
    "    \n",
    "    #For convenience, replace gsym to index and drop NMID\n",
    "    infected_nreads_df   = infected_nreads_df.reset_index(drop=True)\n",
    "    infected_nreads_df   = infected_nreads_df.set_index(['gsym'])\n",
    "    uninfected_nreads_df = uninfected_nreads_df.reset_index(drop=True)\n",
    "    uninfected_nreads_df = uninfected_nreads_df.set_index(['gsym'])\n",
    "    \n",
    "    count_mat = pd.concat([infected_nreads_df,uninfected_nreads_df],axis=1)\n",
    "    count_mat = count_mat.reset_index()\n",
    "    count_mat = count_mat.rename(columns={'index':'id'})\n",
    "    design_mat= pd.DataFrame(['B' for i in range(len(infected))]+['A' for i in range(len(uninfected))],\n",
    "                          index=infected+uninfected)\n",
    "    design_mat.columns = ['treatment']\n",
    "    \n",
    "    DESeq = DEG_method(count_mat,design_mat,design_formula='~ treatment',gene_column='id')\n",
    "    DESeq.run_deseq()\n",
    "    DESeq.get_deseq_result(lfcshrink=True)\n",
    "    \n",
    "    result_df = DESeq.deseq_result\n",
    "    if verbose:\n",
    "        display(result_df.head())\n",
    "    result_df = result_df.set_index(['id'])\n",
    "    result_df['infected_expct']   = infected_expct_df\n",
    "    result_df['uninfected_expct'] = uninfected_expct_df\n",
    "    \n",
    "    if verbose:\n",
    "        display(result_df.head(10))\n",
    "    if OutF != '':\n",
    "        result_df.to_csv(RESULTDIR %OutF, sep = '\\t')\n",
    "    return result_df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "41bb4db2",
   "metadata": {},
   "source": [
    "### Plots (MA, volcano)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "a9d9c680",
   "metadata": {},
   "outputs": [],
   "source": [
    "#MAplot: not used in the paper\n",
    "def plotMA(sample_prefix='mRNA',hpi='0h', padj_cut=0.01, log2fc_cut=2.0, OutFigname='',\n",
    "           ax = None, show_fig = False):\n",
    "    InFname = f'{sample_prefix}_DESeq2_{hpi}.tsv'\n",
    "    df = pd.read_csv(RESULTDIR %InFname, sep ='\\t', header=0, index_col=0)\n",
    "    '''\n",
    "    id\tbaseMean\tlog2FoldChange\tlfcSE\tpvalue\tpadj\tinfection_expct\tuninfection_expct\n",
    "    A1BG\t0.0\t\t\t\t\t83.81389166710532\t87.61120176870033\n",
    "    A1CF\t0.20868225617457992\t0.06273705450357624\t0.5463246183313462\t0.4157015259052005\t\t79.67310627993893\t87.61120176870033\n",
    "    '''\n",
    "    up_df       = df[(df['padj']<padj_cut) & (df['log2FoldChange']>0)]\n",
    "    up_lfc_df   = up_df[(up_df['log2FoldChange']> log2fc_cut)]\n",
    "    up_x_df     = up_df[(up_df['log2FoldChange']<=log2fc_cut)]\n",
    "    down_df     = df[(df['padj']<padj_cut) & (df['log2FoldChange']<0)]\n",
    "    down_lfc_df = down_df[(down_df['log2FoldChange']<  -log2fc_cut)]\n",
    "    down_x_df   = down_df[(down_df['log2FoldChange']>= -log2fc_cut)]\n",
    "    nonsig_df   = df[df['padj']>=padj_cut]\n",
    "    \n",
    "    if ax == None:\n",
    "        fig, ax = plt.subplots(figsize=(4,3))\n",
    "    dot_size = 3\n",
    "    ax.scatter(nonsig_df['baseMean']   ,nonsig_df['log2FoldChange'],   color='grey',   rasterized = True, s = dot_size)\n",
    "    ax.scatter(up_x_df['baseMean']     ,up_x_df['log2FoldChange'],     color='pink',   rasterized = True, s = dot_size)\n",
    "    ax.scatter(down_x_df['baseMean']   ,down_x_df['log2FoldChange'],   color='skyblue',rasterized = True, s = dot_size)\n",
    "    ax.scatter(up_lfc_df['baseMean']   ,up_lfc_df['log2FoldChange'],   color='red',    rasterized = True, s = dot_size)\n",
    "    ax.scatter(down_lfc_df['baseMean'] ,down_lfc_df['log2FoldChange'], color='blue',   rasterized = True, s = dot_size)\n",
    "    ax.set_ylim(-5,5)\n",
    "    ax.set_xscale('log')\n",
    "    ax.set_xlabel('Num. reads')\n",
    "    ax.set_ylabel('log2(fold change)')\n",
    "    ax.set_title('Criteria: P.adj<%.02f, abs(log2fc)>%.01f\\n# upregulated: %d, #downregulated: %d' %(padj_cut,\n",
    "                                                                                                   log2fc_cut,\n",
    "                                                                                                   len(up_lfc_df),\n",
    "                                                                                                   len(down_lfc_df)\n",
    "                                                                                                  ))\n",
    "    if OutFigname != '':\n",
    "        plt.tight_layout()\n",
    "        plt.savefig(FIGDIR %OutFigname)\n",
    "        plt.close()\n",
    "    elif show_fig:\n",
    "        plt.tight_layout()\n",
    "        plt.show()\n",
    "        plt.close()\n",
    "    \n",
    "    return None\n",
    "\n",
    "def plotVolcano(sample_prefix='mRNA',hpi='0h',padj_cut=0.01,log2fc_cut=2.0,\n",
    "                OutFigname='',ax = None, show_fig = False):\n",
    "    InFname = f'{sample_prefix}_DESeq2_{hpi}.tsv'\n",
    "    df = pd.read_csv(RESULTDIR %InFname, sep ='\\t', header=0, index_col=0)\n",
    "    '''\n",
    "    id\tbaseMean\tlog2FoldChange\tlfcSE\tpvalue\tpadj\tinfection_expct\tuninfection_expct\n",
    "    A1BG\t0.0\t\t\t\t\t83.81389166710532\t87.61120176870033\n",
    "    A1CF\t0.20868225617457992\t0.06273705450357624\t0.5463246183313462\t0.4157015259052005\t\t79.67310627993893\t87.61120176870033\n",
    "    '''\n",
    "    up_df       = df[(df['padj']<padj_cut) & (df['log2FoldChange']>0)]\n",
    "    up_lfc_df   = up_df[(up_df['log2FoldChange']> log2fc_cut)]\n",
    "    up_x_df     = up_df[(up_df['log2FoldChange']<=log2fc_cut)]\n",
    "    down_df     = df[(df['padj']<padj_cut) & (df['log2FoldChange']<0)]\n",
    "    down_lfc_df = down_df[(down_df['log2FoldChange']<  -log2fc_cut)]\n",
    "    down_x_df   = down_df[(down_df['log2FoldChange']>= -log2fc_cut)]\n",
    "    nonsig_df   = df[df['padj']>=padj_cut]\n",
    "    \n",
    "    if ax == None:\n",
    "        fig, ax = plt.subplots(figsize=(3,3))\n",
    "    dot_size = 3\n",
    "    ax.scatter(nonsig_df['log2FoldChange'],  -np.log10(nonsig_df['padj']),    color='k',      rasterized = True, s = dot_size)\n",
    "    ax.scatter(up_x_df['log2FoldChange'],    -np.log10(up_x_df['padj']    ),  color='pink',   rasterized = True, s = dot_size)\n",
    "    ax.scatter(down_x_df['log2FoldChange'],  -np.log10(down_x_df['padj']  ),  color='skyblue',rasterized = True, s = dot_size)\n",
    "    ax.scatter(up_lfc_df['log2FoldChange'],  -np.log10(up_lfc_df['padj']    ),color='red',    rasterized = True, s = dot_size)\n",
    "    ax.scatter(down_lfc_df['log2FoldChange'],-np.log10(down_lfc_df['padj']  ),color='blue',   rasterized = True, s = dot_size)\n",
    "    \n",
    "    ax.set_xlabel('log2(fold change)')\n",
    "    ax.set_ylabel('-log10(q value)')\n",
    "    ax.set_xlim(-10,10)\n",
    "    ax.set_ylim(0,100)\n",
    "    ax.set_title(hpi)\n",
    "    if OutFigname != '':\n",
    "        plt.tight_layout()\n",
    "        plt.savefig(FIGDIR %OutFigname)\n",
    "        plt.close()\n",
    "    elif show_fig:\n",
    "        plt.tight_layout()\n",
    "        plt.show()\n",
    "        plt.close()\n",
    "    \n",
    "    return None\n",
    "\n",
    "def plot_volcano_multipanel(sample_prefix, hpi_list_list = [['0h','1h','2h','4h'],['12h','16h','24h','36h']],\n",
    "                            padj_cut=0.01,log2fc_cut=2.0,OutFigname=''):\n",
    "    \n",
    "    fig, axes  = plt.subplots(nrows=len(hpi_list_list),           ncols=len(hpi_list_list[0]), \n",
    "                              figsize=(3*(len(hpi_list_list[0])), 3*len(hpi_list_list)))\n",
    "    for idx_row, hpi_list in enumerate(hpi_list_list):\n",
    "        for idx_col, hpi in enumerate(hpi_list):\n",
    "            ax = axes[idx_row,idx_col]\n",
    "            plotVolcano(sample_prefix=sample_prefix,hpi=hpi, padj_cut=padj_cut,log2fc_cut=log2fc_cut,\n",
    "                        OutFigname='',ax = ax, show_fig = False)\n",
    "    if OutFigname != None:\n",
    "        plt.tight_layout()\n",
    "        plt.savefig(FIGDIR %OutFigname)\n",
    "        plt.close()\n",
    "    else:\n",
    "        plt.tight_layout()\n",
    "        plt.show()\n",
    "        plt.close()\n",
    "    return None"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a5643959",
   "metadata": {},
   "source": [
    "### DEG identification and plotting for mRNA, RPF, and QTI (Fig. 5, SFig. 10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "88f06aae",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "#DESeq2 for mRNA, RPF, and QTI\n",
    "tmp_hpi_list = ['%dh' %i for i in [0,1,2,4,12,16,24,36,48]]\n",
    "tmp_sample_prefix_list = ['mRNA','RPF','QTI']\n",
    "\n",
    "for sample_prefix in tmp_sample_prefix_list:\n",
    "    if sample_prefix == 'mRNA':\n",
    "        enumerated_prefix = ['RPFpaired','QTIpaired']\n",
    "    else:\n",
    "        enumerated_prefix = [sample_prefix]\n",
    "    for hpi in tmp_hpi_list:\n",
    "        print(sample_prefix,hpi,time.ctime(),sep='\\t')\n",
    "        if hpi == '48h':\n",
    "            reps = 3\n",
    "            uninfected_name = 'uninfected48h'\n",
    "        else:\n",
    "            reps = 2\n",
    "            uninfected_name = 'uninfected'\n",
    "        infected   = []\n",
    "        uninfected = []\n",
    "        for prefix in enumerated_prefix:\n",
    "            for rep in range(1, reps+1):\n",
    "                infected.append(  f'{prefix}_{hpi}_rep{rep}')\n",
    "                uninfected.append(f'{prefix}_{uninfected_name}_rep{rep}')\n",
    "        OutF = f'{sample_prefix}_DESeq2_{hpi}.tsv'\n",
    "        _= identify_DEG(infected=infected,uninfected=uninfected, \n",
    "                        DEG_method=py_DESeq2, verbose= False, OutF = OutF)\n",
    "                "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "87a86d4b",
   "metadata": {},
   "outputs": [],
   "source": [
    "#volcano plot\n",
    "'''Fig. 5a: RPF'''\n",
    "plot_volcano_multipanel('RPF', hpi_list_list = [['0h','1h','2h','4h'],['12h','16h','24h','36h']],\n",
    "                        padj_cut=0.01,log2fc_cut=2.0,OutFigname='Fig5_RPF_volcano.pdf')\n",
    "'''SFig. 10a: mRNA, QTI'''\n",
    "plot_volcano_multipanel('mRNA', hpi_list_list = [['0h','1h','2h','4h'],['12h','16h','24h','36h']],\n",
    "                        padj_cut=0.01,log2fc_cut=2.0,OutFigname='SFig10a_mRNA_volcano.pdf')\n",
    "plot_volcano_multipanel('QTI', hpi_list_list = [['0h','1h','2h','4h'],['12h','16h','24h','36h']],\n",
    "                        padj_cut=0.01,log2fc_cut=2.0,OutFigname='SFig10a_QTI_volcano.pdf')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5549bc9e",
   "metadata": {},
   "source": [
    "## Fig. 5b-g (hierarchical clustering of DEGs, and time-course expression)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "61cd1f4c",
   "metadata": {},
   "outputs": [],
   "source": [
    "def timecourse_plot(DEG_df_list,hpi_list, padj_cut=0.01,log2fc_cut=2, clip = 3,\n",
    "                    ylim = 8, n_cluster=5, as_average=False, \n",
    "                    load_precalc_clusters=False, clusterFname='',\n",
    "                    clusters_to_show=[],clusters_priority=[], highlight_gene_info_df = pd.DataFrame(),\n",
    "                    cl_method='ward',cl_metric='euclidean',\n",
    "                    ax = None, OutFigname='',show_fig=True, verbose=False):\n",
    "    if OutFigname != '':\n",
    "        assert ('timecourse.' in OutFigname)\n",
    "    if not(type(ylim) in [list,tuple]):\n",
    "        ylim = (-ylim,ylim)\n",
    "    up_DEG_list   = []\n",
    "    down_DEG_list = []\n",
    "    timecourse_df = pd.DataFrame(columns=hpi_list,     index = DEG_df_list[0].index) #index: gsym (id)\n",
    "    DEG_summary_df= pd.DataFrame(columns=['up','down'],index = hpi_list)\n",
    "    \n",
    "    for hpi,DEG_df in zip(hpi_list,DEG_df_list):\n",
    "        df = DEG_df.copy()\n",
    "        up_DEG   = list(df[(df['padj']<padj_cut) & \\\n",
    "                           (df['log2FoldChange']>log2fc_cut) &\\\n",
    "                           ((df['infected_expct']<50) |(df['uninfected_expct']<50))\n",
    "                          ].index)\n",
    "        down_DEG = list(df[(df['padj']<padj_cut) & \\\n",
    "                           (df['log2FoldChange']<-log2fc_cut) &\\\n",
    "                           ((df['infected_expct']<50) |(df['uninfected_expct']<50))\n",
    "                          ].index)\n",
    "        up_DEG_list   += up_DEG\n",
    "        down_DEG_list += down_DEG\n",
    "        DEG_summary_df.loc[hpi] = [len(up_DEG),len(down_DEG)]\n",
    "        timecourse_df[hpi]   = df['log2FoldChange']\n",
    "    \n",
    "    DEG_list = list(set(up_DEG_list)|set(down_DEG_list))\n",
    "    if verbose:\n",
    "        print('summary of DEGs')\n",
    "        print('total #:', len(DEG_list), sep = '\\t')\n",
    "        display(DEG_summary_df.T)\n",
    "        \n",
    "    timecourse_df = timecourse_df.fillna(0.0)\n",
    "    if load_precalc_clusters:\n",
    "        ID_cl_df = pd.read_csv(RESULTDIR %clusterFname, sep = '\\t', header = 0, index_col = 0)\n",
    "    else:\n",
    "        ID_cl_df = cluster_DEGs(timecourse_df,DEG_list, cl_method=cl_method,\n",
    "                                cl_metric=cl_metric, n_cluster=n_cluster, clip=clip,\n",
    "                                OutFigname=OutFigname.replace('timecourse.pdf','hclust.pdf'),show_fig=show_fig)\n",
    "        ID_cl_df.to_csv(RESULTDIR %clusterFname,\n",
    "                        sep = '\\t')\n",
    "    np.random.seed(2020)\n",
    "    sampled_cloud = np.random.choice(sorted(list(set(timecourse_df.index)- (set(up_DEG_list)|set(down_DEG_list)))), 100)\n",
    "    \n",
    "    ####Colors for different clusters\n",
    "    if ax == None:\n",
    "        fig, ax = plt.subplots(figsize=(3.5,3))\n",
    "    \n",
    "    palette = sns.color_palette('bright',10) ###########CAUTION: HARD-CODED (max 6 colors)\n",
    "    palette = [color for i,color in enumerate(palette) if not(i in [0,3,5,7])]\n",
    "    \n",
    "    if clusters_to_show == []:\n",
    "        clusters_to_show = list(range(1,1+n_cluster))\n",
    "    \n",
    "    \n",
    "    if as_average:\n",
    "        ylim = (-5,5)\n",
    "        for DEG in DEG_list:\n",
    "            ax.plot(timecourse_df.loc[DEG],color='grey',alpha=0.2,linewidth=0.5)\n",
    "            \n",
    "        for cluster in clusters_to_show:\n",
    "            color   = palette[cluster-1]\n",
    "            DEGS_within_cluster  = [DEG for DEG in DEG_list if ID_cl_df.loc[DEG,'cluster'] == cluster]\n",
    "            lfc_df  = timecourse_df.loc[DEGS_within_cluster]\n",
    "            avg_lfc = lfc_df.mean(axis=0)\n",
    "            std_lfc = lfc_df.std( axis=0)\n",
    "            ax.plot(avg_lfc,color=color,linewidth=1)\n",
    "            ax.fill_between(np.arange(len(avg_lfc)), \n",
    "                            avg_lfc-std_lfc,avg_lfc+std_lfc, color=color, alpha=0.7)\n",
    "        \n",
    "    else:\n",
    "        if len(highlight_gene_info_df) == 0:\n",
    "            grey_color = False\n",
    "        else:\n",
    "            grey_color = True\n",
    "        for DEG in DEG_list:\n",
    "            cluster = ID_cl_df.loc[DEG,'cluster']\n",
    "            if grey_color:\n",
    "                color  = 'grey'\n",
    "                alpha  = 0.2\n",
    "            else:\n",
    "                color  = palette[cluster-1]\n",
    "                alpha  = 1.0\n",
    "                \n",
    "            if cluster in clusters_to_show:\n",
    "                if not (cluster in clusters_priority):\n",
    "                    ax.plot(timecourse_df.loc[DEG],color=color,linewidth=0.5, alpha = alpha)\n",
    "                    \n",
    "        ## clusters to draw last\n",
    "        for p_cluster in clusters_priority:\n",
    "            for DEG in DEG_list:\n",
    "                cluster = ID_cl_df.loc[DEG,'cluster']\n",
    "                if grey_color:\n",
    "                    color  = 'grey'\n",
    "                    alpha  = 0.2\n",
    "                else:\n",
    "                    color  = palette[cluster-1]\n",
    "                    alpha  = 1.0\n",
    "                if cluster == p_cluster:\n",
    "                    ax.plot(timecourse_df.loc[DEG],color=color,linewidth=0.5, alpha = alpha)\n",
    "        \n",
    "        highlight_summary_df = pd.DataFrame(columns = ['lfc_36h','lfc_maxabs']) \n",
    "        \n",
    "        for genelist_name, row in highlight_gene_info_df.iterrows():\n",
    "            color        = row['color']\n",
    "            linewidth    = row['linewidth']\n",
    "            print_cutoff = row['print_cutoff']\n",
    "            highlight_genelist = row['genelist']\n",
    "            \n",
    "            for gsym in highlight_genelist: ##for special markup, Fig. 6\n",
    "                if gsym in timecourse_df.index:\n",
    "                    ax.plot(timecourse_df.loc[gsym], color=color, linewidth=linewidth)\n",
    "                    lfc_maxabs = timecourse_df.loc[gsym].abs().max()\n",
    "                    if lfc_maxabs>print_cutoff:\n",
    "                        lfc_36h = timecourse_df.loc[gsym,str(hpi_list[-1])]\n",
    "                        loc_y   = np.clip(lfc_36h,*ylim)\n",
    "                        \n",
    "                        ax.text(len(hpi_list)-1,loc_y,gsym)\n",
    "                        highlight_summary_df.loc[gsym] = [lfc_36h, lfc_maxabs]\n",
    "        \n",
    "        if verbose:\n",
    "            print('highlight genes with >1.5 max(|log2fc|)')\n",
    "            display(highlight_summary_df)\n",
    "    \n",
    "    ax.set_ylim(*ylim)\n",
    "    ax.set_ylabel('log2(fold change)')\n",
    "    ax.set_xlabel('hpi')\n",
    "    \n",
    "    if OutFigname != '':\n",
    "        plt.tight_layout()\n",
    "        plt.savefig(FIGDIR %OutFigname)\n",
    "        plt.close()\n",
    "    elif show_fig:\n",
    "        plt.tight_layout()\n",
    "        plt.show()\n",
    "        plt.close()\n",
    "    \n",
    "    return timecourse_df\n",
    "\n",
    "\n",
    "def to_hex(r, g, b,float_scale = True):\n",
    "    if float_scale:\n",
    "        r = int(r*255)\n",
    "        g = int(g*255)\n",
    "        b = int(b*255)\n",
    "    rgb = r<<16 | g<<8 | b\n",
    "    return f'{rgb:06x}'\n",
    "\n",
    "def cluster_DEGs(timecourse_df,DEG_list, cl_method='ward',cl_metric='euclidean',\n",
    "                 n_cluster=5, clip=2.8,  OutFigname='', show_fig=False):\n",
    "    df = timecourse_df.loc[DEG_list].clip(-clip,clip)\n",
    "    linkage = hierarchy.linkage(df, method=cl_method,metric=cl_metric)\n",
    "    cluster = hierarchy.fcluster(linkage,n_cluster,criterion='maxclust')\n",
    "    palette = sns.color_palette('bright',10)\n",
    "    palette = [color for i,color in enumerate(palette) if not(i in [0,3,5,7])]  ##CAUTION: HARD-CODED\n",
    "    \n",
    "    lut         = {(i+1): col for i,col in enumerate(palette)}\n",
    "    ID_cl_df    = pd.DataFrame(np.array([ [cluster[i] for i,ID in enumerate(df.index)] ]).T, \n",
    "                               index=df.index, columns=['cluster'])\n",
    "    \n",
    "    colors      = pd.Series([lut[ID_cl_df.loc[ID,'cluster']] for ID in ID_cl_df.index],index = ID_cl_df.index)\n",
    "    colors.name = 'Cluster'\n",
    "    max_lfc = np.ceil(clip)\n",
    "    clustergrid = sns.clustermap(df,figsize=(10,10),method=cl_method,metric=cl_metric,\n",
    "                                 row_colors = colors,row_linkage= linkage,col_cluster=False,\n",
    "                                 cmap='coolwarm',vmin=-max_lfc,vmax=max_lfc,\n",
    "                                 cbar_kws={'label': 'log2(fold change)'},\n",
    "                                 rasterized = True)\n",
    "    \n",
    "    clustergrid.ax_heatmap.set_xlabel('hpi')\n",
    "    clustergrid.ax_heatmap.set_ylabel('')\n",
    "    \n",
    "    ## print list of clustered DEGs, for cytoscape\n",
    "    for i in range(1,1+n_cluster):\n",
    "        IDs = list(ID_cl_df[ID_cl_df['cluster'] == i].index)\n",
    "        print('Cluster%d (%d)' %(i, len(IDs)),','.join(IDs))\n",
    "    \n",
    "    if OutFigname != '':\n",
    "        plt.savefig(FIGDIR %OutFigname)\n",
    "        plt.close()\n",
    "    elif show_fig:\n",
    "        plt.show()\n",
    "        plt.close()\n",
    "        \n",
    "    return ID_cl_df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a949ab4e",
   "metadata": {},
   "source": [
    "### Clustering and timecourse plot (Fig. 5b,c, SFig. 10b,c)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "16e98347",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "-------mRNA-------\n",
      "summary of DEGs\n",
      "total #:\t542\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0h</th>\n",
       "      <th>1h</th>\n",
       "      <th>2h</th>\n",
       "      <th>4h</th>\n",
       "      <th>12h</th>\n",
       "      <th>16h</th>\n",
       "      <th>24h</th>\n",
       "      <th>36h</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>up</th>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>5</td>\n",
       "      <td>22</td>\n",
       "      <td>100</td>\n",
       "      <td>193</td>\n",
       "      <td>209</td>\n",
       "      <td>216</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>down</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>4</td>\n",
       "      <td>26</td>\n",
       "      <td>85</td>\n",
       "      <td>87</td>\n",
       "      <td>80</td>\n",
       "      <td>58</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      0h  1h  2h  4h  12h  16h  24h  36h\n",
       "up     0   1   5  22  100  193  209  216\n",
       "down   0   0   4  26   85   87   80   58"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Cluster1 (23) IER5L,PIM1,PRKCD,GRB7,TOX3,SOCS3,MIDN,NOL4L,BCORL1,ZNF792,FAM110C,GSE1,BCL3,CDC42EP4,EGR1,SPRED3,FOS,SPSB2,NKX6-1,BCOR,PLEKHH3,CXXC5,TIGD7\n",
      "Cluster2 (134) CRYM,SPIRE2,KRT13,TCEA3,SERPINA4,RASL11A,ZNF488,AREG,ST6GALNAC2,ANKRD1,TXNIP,AKR1C2,FAM171B,VSIG1,ACTA2,IGF2BP2,YPEL4,HLA-DOA,FSCN1,CABYR,RDM1,ZNF816-ZNF321P,ASS1,ADSSL1,KIF5C,RTN4RL2,SCARA3,UNC5A,AQP3,GCNT4,ARHGAP23,PHF7,CPM,SMARCD3,NDRG4,MOK,TNFAIP2,ARRDC4,WNT7B,SULF2,VSIR,TMPRSS3,AKR1C1,MCAM,CCDC110,KLK11,MYLK,ARHGEF25,ALDH1L1,TMCC3,PPP1R1B,TP53I11,ERP27,UGT1A1,TTYH1,SYT7,AQP4,BBOF1,GOLT1A,ACTBL2,GNAZ,KREMEN1,SH3BP2,TMEM176A,TRIM15,AKR1B15,TCEA2,HLF,PSG4,CYP1B1,PLAC8,NANOS1,PLXND1,CLDN16,TUBB4A,TLE2,FOSL1,FBLN1,IQCH,RCAN1,PTGES,SLC29A4,SLC16A14,TIMP3,CDKN3,UGT1A6,FSTL1,ZNF285,APOL3,BCAS1,GSDMB,BCAM,GMCL1,DBP,SCNN1A,MYZAP,BPIFB1,KLF9,CCL2,C9ORF135,FAM131B,ALDH1A1,MUC5B,TSPAN7,COLCA2,MAOA,WIPF3,AKR1B10,RASSF5,NCCRP1,REEP2,NMT2,RARRES1,ADGRB2,CYBRD1,CACNG4,TNNC1,ALDH1A3,ATP2B4,FGD3,ID1,IFITM2,ANKRD33B,NR1H4,ABI3BP,BCL2L14,H3F3C,HRASLS2,ID3,TMPRSS2,GDPD5,TFAP2C,TMEM156,IHH\n",
      "Cluster3 (103) TRIM22,TYMP,TNFSF13B,TMEM140,TAP1,EPSTI1,APOL2,SP110,TRANK1,ZBP1,CH25H,C5ORF56,NFKBIA,ZC3HAV1,SLC15A3,OAS3,IFI27,IFIT3,PLA1A,TLR3,OAS1,STAT1,CFHR1,LAMP3,MUC13,SAMHD1,CMPK2,ETV7,FAM46A,MT2A,CFH,GBP4,PARP10,HERC5,THEMIS2,LMO2,SLC25A28,IFIH1,GBP1,RTP4,IFNL3,NT5C3A,PARP9,OAS2,DDX60,JAK2,PSMB9,CARD16,IFNL1,BST2,IFI35,STAT2,IRF7,TRIM21,LIFR,NMI,EPHA7,UBA7,LGALS9,PDGFRL,VAMP5,RASGRP3,IFI44L,APOL4,TMEM27,CSF1,PPM1K,UBE2L6,IFNL2,RNF19B,HELZ2,ISG20,RASGEF1B,TREX1,DDX60L,SAMD9,CEACAM1,SAMD9L,ZNFX1,FAM122C,SP140L,MX2,DHX58,SOCS1,NLRC5,RNF122,TNFAIP3,IDO1,PER1,CCL5,SLC1A3,IFITM1,LAP3,CLEC7A,MX1,MYD88,CD274,IFI44,PARP12,PARP14,USP18,PNPT1,IFI16\n",
      "Cluster4 (16) SERPINB2,TRMO,HSPA6,DIEXF,DNAJB9,NCOA5,ATF5,LANCL2,CD3EAP,SERPINB8,C3ORF58,ICAM4,NDP,EDN2,FOSB,KBTBD8\n",
      "Cluster5 (266) ARSE,RPL41,HIST1H3D,HIST2H2AB,HIST1H2BN,TXNL4A,HSD17B2,POLR2I,RPL30,CCDC134,ISG15,LSS,EMG1,EMC7,SMDT1,S100A2,FTL,RPL26,HIST1H1A,ACYP2,FLRT3,RPL27,SLIRP,HIST1H4B,GAR1,AP1G2,MVK,RPL34,CD69,RPS26,ATP5D,PCDHA11,FAM185A,DHCR7,C1ORF122,TFRC,CRNDE,SLC26A2,ITIH2,FANCL,HIST1H2BL,RPL35A,CTSA,TMEM62,MSRA,HIST1H2AK,HIST1H2BI,HIST1H3J,EMC10,RPS21,CAPN8,CAP2,ISPD,EMC6,HIST1H4E,SNRPF,TFF1,UXT,ADSL,PCDHA1,USMG5,NUDCD2,HIST1H2AM,HIST1H2BG,HCFC1R1,RPS27A,HIST1H3G,DDX58,CEACAM16,FABP5,PDSS1,RPLP0,COX7A2,PRKAR1B,RSAD2,MMP1,EEFSEC,HIST1H4J,TMA7,RPS19,ADAM19,HIST1H2BJ,HIST1H3F,MIA,SLC39A10,HIST1H3A,PCDHA6,TAF15,DNLZ,TNFRSF11B,COMMD6,ORC4,RPSAP58,HIST1H2BM,HIST1H4D,SLC2A1,PCDHA10,QTRT1,PFKFB4,BRI3,FADS2,HERC6,SLC28A3,HIST4H4,POMP,CATSPERB,RPS29,IL1A,HIST1H2AI,IFIT2,HIST1H4I,AVEN,MOV10,HIST1H4A,ARMC10,HIST1H2BB,RPL17-C18ORF32,NDUFB2,ATP5G2,PTX3,HMGCS1,POLE4,PRSS1,RPS27,RPL22L1,CCR1,ST20-MTHFS,ADRA1B,PCSK9,HIST1H2AE,RPS16,C2CD4B,INSIG1,PRR15,CAPN6,CAVIN3,PCDHA13,ITGA1,ADAM32,PRH1-TAS2R14,RPSA,HIST1H4C,SCD,OASL,SFMBT2,IFI6,ADI1,FZD8,MCTP1,UBA52,C2CD4A,ZNF511,PRSS2,RPS13,LRRC37A,PCGF2,CCDC7,TNFSF15,HIST1H2AH,HSD17B7,RPL17,PIN1,HIST1H2AG,PKIB,RPS15A,PCDHA8,MEMO1,CYP2B6,MRPS6,PCDHA5,RNASEH2C,MTHFS,PTMS,CDKN1A,IFNB1,BATF2,NDUFA7,POLR2F,HIST2H2AC,C9ORF152,HIST1H2BC,IFIT1,HSPA8,MRPL12,HMGA2,FERMT1,HIST1H3H,HIST3H2BB,PRH1,LRRC37A2,ADAMTS17,TSTD1,EIF4EBP3,VIM,CXCL10,HIST1H4H,DDX56,STARD4,CEMIP,ACTL8,LRRC37A3,CXCL1,LINC01835,NDUFAF8,SNCA,CCL20,TRIR,HIST2H3D,MCRIP1,ANKRD37,FASN,HLA-B,C12ORF57,MSMO1,FAM13B,ATP5J2,HIST1H2BF,RCL1,HIST1H3I,LMAN2L,GLRX2,EEF1B2,HIST1H2AJ,FUS,CDH6,RPLP1,HIST1H4K,NME1,TP53I3,SF3B4,PCDHA3,NDUFB1,RPL23,CXCR4,OSBPL10,HIST1H3E,IGFBP7,LAMA3,NPC1L1,LYPD3,RANBP17,ADAM28,THBS1,UFD1,CXCL11,PCDHA12,CLN6,HIST1H2BO,FBXL13,UBL5,RPL36,HIST1H2BH,PCDHA7,TNFAIP6,TMEM154,PET100,SLC22A18AS,SLC25A3,SLC22A4,RPLP2,HIST1H3B,KRT6B,PCDHA2,BAAT,ZNF430,AMIGO2\n",
      "highlight genes with >1.5 max(|log2fc|)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>lfc_36h</th>\n",
       "      <th>lfc_maxabs</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [lfc_36h, lfc_maxabs]\n",
       "Index: []"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "-------RPF-------\n",
      "summary of DEGs\n",
      "total #:\t677\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0h</th>\n",
       "      <th>1h</th>\n",
       "      <th>2h</th>\n",
       "      <th>4h</th>\n",
       "      <th>12h</th>\n",
       "      <th>16h</th>\n",
       "      <th>24h</th>\n",
       "      <th>36h</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>up</th>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>24</td>\n",
       "      <td>91</td>\n",
       "      <td>150</td>\n",
       "      <td>164</td>\n",
       "      <td>191</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>down</th>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>8</td>\n",
       "      <td>35</td>\n",
       "      <td>146</td>\n",
       "      <td>201</td>\n",
       "      <td>168</td>\n",
       "      <td>100</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      0h  1h  2h  4h  12h  16h  24h  36h\n",
       "up     1   0   3  24   91  150  164  191\n",
       "down   0   2   8  35  146  201  168  100"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Cluster1 (37) CHAC1,IER5L,CDKN1C,PRKD2,PLEKHG3,LIPE,TFEB,KLF2,GRB7,CDC42EP2,RTN4R,ADM2,YPEL3,NOL4L,ZSWIM4,TRIM15,BCORL1,GSE1,MAFF,PLEKHG6,BCL3,PIK3IP1,SHF,MYZAP,IRF5,KLF9,CDK18,SESN2,NFATC4,FBXL8,TRIM31,ZFP36,CTTNBP2NL,BCOR,GLIS2,TFAP2C,TBC1D10A\n",
      "Cluster2 (292) SLC40A1,EPHX1,PAQR8,ERO1B,MFSD4A,PPM1L,SYTL3,SECTM1,KRT13,UGT1A5,TCEA3,BTD,HLA-DMB,SERPINA4,ITM2A,ALDH1L2,DEFB1,ODAM,HLA-DQA2,IGF1R,CTSE,AREG,ZNF226,PRTG,ADAM10,ST6GALNAC2,FTL,ANKRD1,ARHGDIB,MGLL,SLC26A9,TXNIP,PLAT,AKR1C2,NRP2,ERAP2,HLA-DQA1,RBM12,TM4SF4,VSIG1,TMEM173,IFNAR2,ACE2,IFITM10,DHDH,ASS1,NID1,TMEM176B,ADSSL1,VWA8,LMF1,CXORF38,LAPTM4A,SRPX2,CD24,UNC5A,AQP3,HSPG2,MANSC1,UPK3B,CPM,LCN2,PLA2G10,PDPR,HIST1H2BI,CYP1A1,ENPP5,UGT1A10,PODXL,NDRG4,HLA-DRB5,SLC16A5,SEC22A,VASN,WNT7B,TMC5,SULF2,SLC22A5,ACAD10,CNIH4,VSIR,FAM234B,CTSH,FBP1,HAVCR1,AKR1C1,C3,AGT,TMEM45A,TSPAN8,NEK2,MYSM1,MCAM,SLC16A7,LAMB2,UGGT1,ALG10,HACL1,PRCP,PPP1R1B,ANG,WLS,SLC7A11,UGT1A1,TTYH1,UGGT2,ZSCAN18,AKR1C3,SLC34A2,CGRRF1,WFDC2,AASS,TLR5,EMP2,RHBDD1,NRGN,HLA-DPA1,APOL1,SERPING1,DEGS2,LCORL,FCGBP,FUCA2,NECTIN2,RARRES3,AARS2,MAPT,STRA6,RASGRP2,IL4R,DRAM2,TMEM151A,TMEM176A,MYADM,HSD17B11,TMEM206,AKR1B15,PRKCA,COL1A1,KDELR3,MOXD1,STT3B,ERP44,NPNT,UCP2,MPZL2,PPARG,ALG8,LRG1,PLAC8,CILP2,CHST4,CDC25C,POFUT2,UGCG,LY75-CD302,ALG6,MGAM2,CERCAM,SLC19A3,SAT1,FBLN1,PTPRB,RCAN1,GLRX,PTGES,DPP4,TIMP3,GFM2,UGT1A6,GDA,FBXO6,FAM168B,FSTL1,CLU,PRR11,SLC52A3,EMP1,SDC4,SLC30A1,SERPINA5,IARS2,UGT1A7,GSDMB,MUC5AC,MUC1,PLAU,TGFB2,ITGB4,BCAM,TXNDC12,GMCL1,AKR1C4,UGT1A3,TMEM168,HPSE,LRP1,MT1X,SCNN1A,MTERF3,PLBD1,FGF19,MRC2,BPIFB1,RPL32,IGFBP3,REEP6,GGH,SLC38A2,ACSM3,CXCL5,LAMA5,MBTPS2,ALDH1A1,MUC5B,TSPAN7,MAN2A1,CHURC1,FOS,CD55,AQP5,CD47,GPRC5A,PTTG2,SAMD8,UGT1A4,LETM2,LY75,C4BPB,FRAS1,GCNT3,CTSV,AKR1B10,CEACAM1,MAGED1,KIF20A,PROCR,ACTA1,UGT1A8,PIGN,FAM171A2,UGT1A9,ARSK,SMIM26,TTC13,MUT,RARRES1,CP,DNAAF3,DIRC2,GPR155,CYP2C18,SEMA3B,HLTF,ATP6AP2,TCTN3,ATP2B4,ALDH1A3,HLA-DPB1,FGD3,ID1,CD40,SFTA2,ALDH18A1,NNT,INTS8,TMEM131L,GALNT4,NFE2L3,CYP1B1,NR1H4,ADGRF1,TSKU,ASNS,HLA-DRA,HBEGF,LBR,ABI3BP,GALNT12,MDK,TMEM161B,LETM1,ADRB2,C5,CPD,CYP24A1,ID3,MME,TMPRSS2,SGSH,PPP1R10,F2RL1\n",
      "Cluster3 (100) TRIM22,ZNF618,OPTN,TMEM140,TAP1,EPSTI1,SMIM29,SP110,TRANK1,CEBPD,HES4,MAP2K6,ZBP1,APOBEC3G,PCGF5,SLC15A3,OAS3,IFIT3,GBP7,DBF4B,C19ORF66,OAS1,STAT1,LAMP3,SAMHD1,CMPK2,ETV7,MT2A,MOB3C,GBP4,IFIT1B,PARP10,HERC5,THEMIS2,LMO2,GBP1,IFIH1,RTP4,IFNL3,OAS2,DDX60,IFIT2,HPS1,PSMB9,BBC3,IFNL1,BST2,IFI35,STAT2,CCR1,LGALS9,PRR15,PDGFRL,LGALS9B,OASL,GMPR,PML,IFI44L,CIB2,IFI6,PPM1K,APOL6,APOL3,UBE2L6,IFNL2,RNF19B,HELZ2,ISG20,IFNB1,METRNL,BATF2,BCL2L2,TREX1,CXCL10,DDX60L,DENND5A,SP140,PSMB8,MARCKSL1,SAMD9,HLA-B,SAMD9L,ZNFX1,MX2,DHX58,NLRC5,RNF122,IDO1,LGALS9C,CXCL11,SLC1A3,IFITM1,LAP3,MX1,IFITM2,CD274,IFI44,PARP12,USP18,IFI16\n",
      "Cluster4 (16) SERPINB2,LYPD1,HSPA6,TFRC,DSEL,FJX1,NCOA5,LUC7L,OSGIN1,LANCL2,ST3GAL1,EGR1,ARMC5,HSPH1,FOSB,LRRC8B\n",
      "Cluster5 (232) TYMP,R3HDM4,HMGN3,DUS1L,TOMM5,ZNF33B,CA9,CHMP4A,SAP30L,KRT10,PPP1R14B,ISG15,TMEM163,LSS,GPATCH4,NAPA,HMGN1,NTAN1,CLTB,CBWD3,BTBD6,PTRH1,PYURF,KHNYN,LAMTOR1,ALKBH5,VMA21,SH3BP5L,RAMP1,AGTRAP,EPC2,PPIA,PHACTR2,GAS2L1,ULK3,C1ORF122,S100A13,FBXO27,ITIH2,CTNNBIP1,CARD11,HOMER1,SARNP,IDI1,RNASEH1,CLDN2,NUDT4,BCL7C,CNP,DNTTIP1,SUMO3,C5ORF24,MAPK6,NAPEPLD,BET1L,IRAK1,PPP5D1,DDX58,PDCD5,PRKAR1B,BEX3,FXYD1,RSAD2,MMP1,TPM2,FAM13A,TMA7,CBWD6,ADAM19,UTP18,ZNF664,BRMS1L,ENSA,C4ORF3,IL1RN,XPOT,UPP1,TNFRSF11B,DBNDD1,C1ORF50,POLR3H,PLEKHJ1,POLR2J,PFKFB4,RPS6KB2,C19ORF24,JAZF1,CST6,ATAD3B,IL1A,MYH11,RAB3IP,PHACTR4,C6ORF141,ZC2HC1A,PEX10,NCBP2-AS2,NCL,WSB2,POLE4,KDM4B,RPS6KA4,JADE2,IRF7,PCSK9,CASP7,TCEAL9,BNIP3,DDX41,GUCY1A3,PPIAL4A,LUC7L3,SLC25A39,MMAB,GALNT10,PIK3R2,CAVIN3,HSPA1A,NUAK1,MYCBP,SREBF1,KHDRBS3,CHCHD5,LRRFIP2,ATXN7L3B,PRELID1,CD3EAP,ADI1,GTF3C6,HDGFL2,ZNF511,COA4,CBWD1,KANSL1,HLA-G,MPND,NHP2,ACSS2,C19ORF48,PEBP1,UBTD2,RFX5,S100A16,PFDN2,DPH2,TPM1,PIN1,ZFAND1,GFAP,DVL1,HMGB3,RBCK1,HMGA1,HIRA,RNASEH2C,MTHFS,TRIM14,PTMS,GTPBP2,POLR2F,IFIT1,RSRC2,SRF,HMGA2,FERMT1,INA,C6ORF62,SRM,GPBP1L1,HRAS,OGFR,ZNF146,PEA15,DHFR2,CEMIP,TMSB10,ACTL8,TMEM189,CXCL1,DAP,ICAM3,METTL21A,CBWD5,PDAP1,PSPH,C6ORF223,CDR2L,MCRIP1,CBWD2,ANKRD37,DDIT3,UBALD2,NSDHL,DBI,DHFR,HSPA1B,DPH3,PALMD,ATAD3C,CDH6,C19ORF53,CTXN1,MON1A,RPS6KA2,TYW3,SMIM15,TP53I3,AZI2,PAM16,TRIM7,IGFBP7,MYO5C,WDR77,C3ORF52,THBS1,C19ORF25,DPP3,HSPA1L,HYPK,SETD7,JUND,FURIN,TMEM154,ARIH2,SLC22A18AS,TCEAL3,BLOC1S2,CSNK1G2,PXMP2,MBD3,DNAJB2,ATF2\n",
      "highlight genes with >1.5 max(|log2fc|)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>lfc_36h</th>\n",
       "      <th>lfc_maxabs</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [lfc_36h, lfc_maxabs]\n",
       "Index: []"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "-------QTI-------\n",
      "summary of DEGs\n",
      "total #:\t861\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0h</th>\n",
       "      <th>1h</th>\n",
       "      <th>2h</th>\n",
       "      <th>4h</th>\n",
       "      <th>12h</th>\n",
       "      <th>16h</th>\n",
       "      <th>24h</th>\n",
       "      <th>36h</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>up</th>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>6</td>\n",
       "      <td>104</td>\n",
       "      <td>126</td>\n",
       "      <td>232</td>\n",
       "      <td>197</td>\n",
       "      <td>148</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>down</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>16</td>\n",
       "      <td>54</td>\n",
       "      <td>149</td>\n",
       "      <td>216</td>\n",
       "      <td>185</td>\n",
       "      <td>112</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      0h  1h  2h   4h  12h  16h  24h  36h\n",
       "up     0   1   6  104  126  232  197  148\n",
       "down   0   0  16   54  149  216  185  112"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Cluster1 (38) STYK1,GRB7,TGFBR3,KLHL22,HOXB4,PLSCR4,BCL6,IFRD1,ARRDC3,TLDC1,TFAP2C,TBC1D10A,PEX11B,TAF7,B3GALT5,IER5L,IER2,MYORG,CDC42EP2,IRF2,VRK3,VPS25,SOCS3,SRGAP2,NOL4L,PEF1,RALB,SERTAD2,MID1IP1,BTN3A3,ING4,BLCAP,ARID5B,PRR15L,ANKMY2,RARA,TMEM139,TTC33\n",
      "Cluster2 (349) GSTM3,EIF3F,SECTM1,UGT1A5,ITM2A,ODAM,HLA-DQA2,KRTCAP3,AREG,HSD17B8,MGLL,TXNIP,RDH13,AKR1C2,RNF19A,RBM12,PIGR,PSAP,GCOM1,FUT3,ACE2,HEXA,IFITM10,DHDH,ASS1,METTL15,EHHADH,DNASE1L1,STX8,KREMEN2,UNC5A,C15ORF39,NAB2,SGK2,PEX6,SLC16A5,IFFO2,RPL18,KDELR1,POTEF,VASN,TMEM59,TMC5,TMEM37,SULF2,TAPBP,LITAF,VSIR,CTSH,CLN3,TSPAN8,SLPI,DUSP5,KLK11,CYR61,SLC16A7,CPT2,ALDH1L1,RHPN2,CLDN1,RELA,PPP1R1B,SLC7A11,MAPK8IP3,MIDN,B3GNT7,ADAP1,PRAME,DAZAP2,HLA-DPA1,NDRG2,ZC3H12A,MRPL41,FCGBP,NECTIN2,GBP2,MAPT,ZMIZ1,FAM234A,ZNF274,ICAM5,TRIM15,MYADM,AKR1B15,GPR37,COL1A1,ICAM1,CXORF38,SHISA5,CYP1B1,PPARG,SPINK1,RND3,LRG1,PLAC8,CHST4,ATXN10,PEX11A,TMPRSS4,HLA-F,SAT1,TMEM161A,MAFF,SLC9A3R2,KRT15,FUCA1,DUSP1,BCL3,PTGES,GALK1,ACADS,RARB,TIMP3,CHCHD10,CLDN4,EGR1,MFGE8,SLC52A3,EMP1,RPL17,BCAS1,ECHS1,MIEF2,GSDMB,MUC5AC,TGFB2,ITGB4,MYO18A,GMCL1,MT1X,SCNN1A,TKT,MYZAP,MRC2,BPIFB1,FBXL20,REEP6,ACSM3,CXCL5,MUC5B,THEM6,DMAC2,COLCA2,FOS,CD47,C1R,UGT1A4,LETM2,EFNA1,HLA-DMA,PLPP1,DMBT1,GCNT3,ENTPD2,AKR1B10,ACTA1,DALRD3,P3H3,SERPINH1,SLC37A3,UGT1A9,NUAK2,RPL37,VAPA,FBXO2,CSTB,CP,DNAAF3,AIG1,EFNA5,ALDH1A3,FGD3,SFTA2,SALL2,ADGRF1,CLDN10,C5,CYP24A1,ID3,PPP1R10,SLC25A20,GNE,SLC40A1,CRABP2,RPL9,FAM98C,MFSD4A,KRT13,SERPINA4,DEFB1,AHCYL1,GLIS3,CTSE,MRPS7,FTL,ANKRD1,ARHGDIB,PLPP6,MTSS1,PMM1,TM4SF4,DGAT1,TMEM173,CD74,NNMT,KLF2,TNFRSF21,SPATA5L1,CSN3,LAPTM4A,CD24,LACTB2,AQP3,SNAPC2,ZNF217,ABHD11,UPK3B,PDZK1IP1,LCN2,UGT1A10,HLA-DRB5,TNFAIP2,WNT7B,VGLL1,TADA1,SIDT2,YPEL3,UBE2W,FBP1,AKR1C1,C3,TMEM45A,AKR1B1,DYNC1LI2,IER3IP1,SCIN,UGT1A1,TTYH1,TMED8,SERTAD1,SLC34A2,TMEM45B,WFDC2,EMP2,CITED4,NRGN,SULT2B1,DTX2,SWI5,ACTBL2,RARRES3,MUC6,MFSD10,RASGRP2,MANBAL,TMEM176A,TGFB1I1,EBPL,COMTD1,KDELR3,RPL15,MISP,FAM110C,PARL,DDIT4,CILP2,IFNGR1,TUBB4A,ARHGAP45,TACSTD2,PCYOX1L,LTBP4,MGAM2,FOSL1,FBLN1,TMEM51,PLK2,PLD3,UGT1A6,FBXO6,CLU,D2HGDH,EML2,SDC4,SLC30A1,PHYH,UGT1A7,AMOTL2,MUC1,PLAU,BCAM,EPN2,PDK4,VNN2,UGT1A3,DBP,LRP1,CAPN5,EPS8L1,NEDD9,RPL32,IRF5,VILL,IGFBP3,C1ORF116,ECI2,FBXO31,RHOB,CNOT9,CHURC1,WDR72,NFATC4,TMEM9,GPRC5A,RTL10,HSPB1,SSR3,ALDH2,CEACAM1,MAGED1,CXCL16,PROCR,RASSF5,UGT1A8,TPM4,GTPBP1,NR3C1,PER1,NAGA,COQ5,RARRES1,SMC5,ABTB1,TGFA,NR4A1,SEMA3B,RAB11FIP1,ATP2B4,HLA-DPB1,ID1,TMEM115,MRPL10,PIM3,NR1H4,TSKU,ASNS,HLA-DRA,PLCD1,ABI3BP\n",
      "Cluster3 (86) ING5,RPP25,RPS12,PYURF,ADM,DKK1,STC2,UBA2,CREBBP,SPTSSA,H2AFX,FST,UTP20,AMD1,C19ORF24,RNF126,GLOD4,ATAD3B,NCOA5,LUC7L,UCP2,YLPM1,ZDHHC18,LBH,CDC25A,ADI1,BTAF1,C3ORF58,SS18,NDUFS6,SFPQ,DNTTIP2,TTK,EEF1B2,HSPH1,PSMA3,SEC11C,PFKFB3,REEP2,ZSCAN2,THBS1,COQ10B,RBM3,KBTBD8,SS18L2,BAAT,SERPINB2,AKAP12,TEX264,NDUFS7,FAM111B,FOXC2,MRPL18,MAT2A,ASTE1,DNAJB9,DDX5,C6ORF106,RPS19,SRSF2,ZNF316,SURF6,ESF1,FJX1,ZNF639,C18ORF32,BORA,C1QBP,SKAP1,PDE3A,C2CD4B,CD3EAP,HSBP1L1,STX3,CAPZB,SERPINB8,MRPS17,U2AF2,URB2,GGCT,ODF3B,ERCC6,RPLP1,DDX39A,ZSWIM9,METTL1\n",
      "Cluster4 (88) TRIM22,TAP1,EPSTI1,SP110,NAPA,MAP2K6,PCGF5,IFIT3,C19ORF66,STAT1,LAMP3,CMPK2,ETV7,MT2A,DDX58,ORC6,IFIH1,GBP1,IFNL1,BST2,DKK3,IRF7,CCR1,NMI,LGALS9,PDGFRL,PML,IFI44L,IFI6,TMEM27,PPM1K,APOL6,APOL3,UBE2L6,HIRA,BATF2,IFIT1,DDX60L,SAA2,SAMD9,SAMD9L,DHX58,RNF122,IDO1,LGALS9C,GPT2,IFITM1,USP18,SAA2-SAA4,OPTN,TRANK1,ISG15,MASTL,TSC22D3,OAS3,OAS1,SAMHD1,CNP,MOB3C,GBP4,PARP10,HERC5,THEMIS2,RTP4,NT5C3A,OAS2,DDX60,IFIT2,PSMB9,IFI35,LGALS9B,OASL,CD68,ISG20,TRIM14,IFNB1,FMR1,CXCL10,OGFR,MX2,NLRC5,TNFSF10,SLC1A3,LAP3,MX1,IFI44,CD274,PARP12\n",
      "Cluster5 (300) OPHN1,CFDP1,BORCS7,CLK1,HMGN3,EIF3A,FANCM,COPRS,IPO4,ENOX2,HMGN1,RTKN2,MAP4K4,TCEAL4,DYNLRB1,LAMTOR1,MAX,GPALPP1,NELFCD,C1ORF122,CPSF6,CISD2,DYRK4,BCL7B,HIST1H3J,RNASEH1,SZRD1,BCL7C,SUMO3,DMRTA1,PITPNA,GSK3A,BET1L,NEDD8,TMEM263,IFT43,DPH7,DEK,NAA10,ITGA2,PRKAR1B,NT5DC2,MMP1,RBM4B,TMA7,MAOB,SLC39A10,SMIM7,NDUFA5,YKT6,ENSA,TCEA1,TLX3,TMEM243,GIGYF2,TPGS2,PLEKHJ1,PFKFB4,CDC37L1,HERC6,MAP1LC3B,C6ORF141,DNA2,ANAPC15,MED10,QKI,ACSL1,WBP1L,CKLF-CMTM1,RPS6KA4,TIMM17B,UMPS,FAM104A,BNIP3,GALNT10,C16ORF87,PCNP,SCD,SREBF1,SVIP,LRRFIP2,ACAT2,COX20,GTF3C6,MCTP1,MSI2,HSD17B7,HNRNPH3,KPNA4,COA6,HMGA1,CYP2B6,RNASEH2C,HSPE1,PMEPA1,PTMS,POLR2F,PHLDA2,EBNA1BP2,RSRC2,HMGA2,FERMT1,TMSB4X,C6ORF62,EI24,SRM,UBE2G1,CEMIP,TMSB10,EIF5A,SSR1,TMEM189,CXCL1,RAB11B,EXOSC8,TAF9B,EIF3J,BHLHA15,SGTA,CCL20,CBWD2,MCRIP1,LSM8,ATP5G1,CTXN1,HIST1H4K,CASD1,ZDHHC2,RAB43,TTC3,CC2D1B,IGFBP7,RBM25,MYO5C,LAMA3,C3ORF52,TMEM14C,PSD3,PIK3AP1,MAP1LC3B2,GRB10,FURIN,DHCR24,GTF3C4,PRMT1,FAM213A,TMEM265,LSM1,IFI16,WASHC3,C17ORF100,C1ORF21,FDCSP,AMIGO2,NEDD8-MDP1,R3HDM4,SREK1IP1,TOMM5,HIST1H3D,PTMA,RWDD1,HSD17B2,KRT10,GALNT1,PPP1R14B,MRPL35,CLTB,S100A2,NUDT3,TUSC1,VMA21,MYH10,ZNF749,POLR3G,S100A13,SLC26A2,ITIH2,DCBLD2,ZCCHC2,AGPAT2,SARNP,PABPN1,IDI1,ACOT4,RPS27L,CHMP3,NUDT4,GMIP,RBM34,DCTN6,SNX10,C1ORF52,COL7A1,IRAK1,MACROD2,PDCD5,FAM129A,FXYD1,RSAD2,FXN,TRIM8,HIST1H4J,RAB27B,ZNF519,ADAM19,CDC45,HINT1,PAICS,MYL12A,C4ORF3,XPOT,UPP1,CNEP1R1,REEP3,HNF4G,POLR2J,SEC61B,TPT1,MLKL,IL1A,IQCK,TCEAL2,ARMC9,ABT1,TMEM56-RWDD3,NCL,DENND1B,WSB2,NPAS2,BLOC1S1,KIF1C,DYNLL1,LUC7L3,VCAN,TMEM258,CAVIN3,HIST1H4C,MYCBP,POLR2E,KHDRBS3,PRPF18,UBTF,TBC1D4,CBWD1,RNF103-CHMP3,NHP2,BTBD1,UBTD2,HEATR3,SET,APOBEC3C,S100A16,TPM1,ZFAND1,PNN,YWHAQ,TRAPPC1,CDKN1A,STUB1,MFSD1,CRIPT,MYBL2,LRP8,PFDN1,DDX56,PEA15,DAP,PRELID3B,PDAP1,LIPG,SLC20A1,YAP1,FASN,NSDHL,DBI,ZNFX1,C12ORF75,CLTA,PALMD,APOBEC3D,CDH6,CHIC2,TIMM8B,ANP32E,CBX5,ZNF428,SNX7,KCTD3,ATP1B3,TNFAIP6,OTUD6B,SLC22A18AS,CSNK1G2,DSC2,MRPL32,LSM3,ARL6IP4,UBE2D2,COX17,ATF2\n",
      "highlight genes with >1.5 max(|log2fc|)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>lfc_36h</th>\n",
       "      <th>lfc_maxabs</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [lfc_36h, lfc_maxabs]\n",
       "Index: []"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "for tmp_sample_prefix in ['mRNA','RPF','QTI']:\n",
    "    print(f'-------{tmp_sample_prefix}-------')\n",
    "    tmp_DEG_df_list   = []\n",
    "    tmp_hpi_list      = ['%dh' %i for i in [0,1,2,4,12,16,24,36]]\n",
    "    for hpi in tmp_hpi_list:\n",
    "        InFname    = f'{tmp_sample_prefix}_DESeq2_{hpi}.tsv'\n",
    "        tmp_DEG_df = pd.read_csv(RESULTDIR %(InFname), sep ='\\t', header=0, index_col=0)\n",
    "        tmp_DEG_df_list.append(tmp_DEG_df)\n",
    "\n",
    "    timecourse_df = timecourse_plot(tmp_DEG_df_list,tmp_hpi_list, padj_cut=0.01, log2fc_cut=2, clip = 2.8,\n",
    "                                    ylim = 8, n_cluster=5, as_average=False, \n",
    "                                    load_precalc_clusters=False, clusterFname=f'{tmp_sample_prefix}_cluster.tsv',\n",
    "                                    clusters_to_show=[],clusters_priority=[], \n",
    "                                    cl_method='ward',cl_metric='euclidean',ax = None, \n",
    "                                    OutFigname=f'Fig5_{tmp_sample_prefix}_individual_timecourse.pdf',\n",
    "                                    show_fig=False, verbose=True)\n",
    "    timecourse_df.to_csv(RESULTDIR %(f'{tmp_sample_prefix}_timecourse.tsv'),sep='\\t')\n",
    "    _             = timecourse_plot(tmp_DEG_df_list,tmp_hpi_list, padj_cut=0.01, log2fc_cut=2, clip = 2.8,\n",
    "                                    ylim = 8, n_cluster=5, as_average=True, \n",
    "                                    load_precalc_clusters=True, clusterFname=f'{tmp_sample_prefix}_cluster.tsv',\n",
    "                                    clusters_to_show=[],clusters_priority=[], \n",
    "                                    cl_method='ward',cl_metric='euclidean',ax = None, \n",
    "                                    OutFigname=f'Fig5_{tmp_sample_prefix}_average_timecourse.pdf',\n",
    "                                    show_fig=False, verbose=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d4ba75be",
   "metadata": {},
   "source": [
    "### Timscourse plots for wach cluster (Fig. 5d-g, SFigs. 10d, 11a-e)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "3b20d87d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "-------mRNA-------\n",
      "-------RPF-------\n",
      "-------QTI-------\n"
     ]
    }
   ],
   "source": [
    "for tmp_sample_prefix in ['mRNA','RPF','QTI']:\n",
    "    print(f'-------{tmp_sample_prefix}-------')\n",
    "    tmp_DEG_df_list   = []\n",
    "    tmp_hpi_list      = ['%dh' %i for i in [0,1,2,4,12,16,24,36]]\n",
    "    for hpi in tmp_hpi_list:\n",
    "        InFname    = f'{tmp_sample_prefix}_DESeq2_{hpi}.tsv'\n",
    "        tmp_DEG_df = pd.read_csv(RESULTDIR %(InFname), sep ='\\t', header=0, index_col=0)\n",
    "        tmp_DEG_df_list.append(tmp_DEG_df)\n",
    "\n",
    "    for cluster in range(1,1+5):\n",
    "        _             = timecourse_plot(tmp_DEG_df_list,tmp_hpi_list, padj_cut=0.01, log2fc_cut=2, clip = 2.8,\n",
    "                                        ylim = 8, n_cluster=5, as_average=True, \n",
    "                                        load_precalc_clusters=True, clusterFname=f'{tmp_sample_prefix}_cluster.tsv',\n",
    "                                        clusters_to_show=[cluster],clusters_priority=[], \n",
    "                                        cl_method='ward',cl_metric='euclidean',ax = None, \n",
    "                                        OutFigname=f'Fig5_{tmp_sample_prefix}_average_cluster{cluster}_timecourse.pdf',\n",
    "                                        show_fig=False, verbose=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "24f6a316",
   "metadata": {},
   "source": [
    "## Figure 6- Inspection of the genes of interest\n",
    "### Timecourse expression of the highlighted genes (Fig. 6a,b,d,e)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "7d2f1e7b",
   "metadata": {},
   "outputs": [],
   "source": [
    "def plot_Fig6abde_multipanel(OutFigname=''):\n",
    "    #Fig.6a: Gordon et al + ACE2, TMPRSS2\n",
    "    Fig6a_df = pd.DataFrame(columns = ['color','linewidth','print_cutoff','genelist'])\n",
    "    Fig6a_df.loc['Gordonetal']  = ['#a52a2a', 0.5, 100.0, [i.strip().upper() for i in open(GENELISTDIR %'Gordonetal_interacting_proteins')]]\n",
    "    Fig6a_df.loc['ACE2+TMPRSS2']= ['blue',    1.0, 0.0,   ['ACE2', 'TMPRSS2']]\n",
    "    \n",
    "    #Fig. 6b: Daniloski et al. MOI0.3,CRISPR log2fc>1 host factors (drugtargets: subgroup)\n",
    "    Fig6b_df = pd.DataFrame(columns = ['color','linewidth','print_cutoff','genelist'])\n",
    "    hostfactors_drugtargets = pd.read_csv(GENELISTDIR %'Daniloski_MOI03_dgidb',sep='\\t').iloc[:,0].str.upper().tolist()\n",
    "    hostfactors             = pd.read_csv(GENELISTDIR %'Daniloski_MOI03_rank',sep='\\t').iloc[:,0].str.upper().tolist()\n",
    "    hostfactors_nontargets  = list(set(hostfactors)-set(hostfactors_drugtargets))\n",
    "    Fig6b_df.loc['Daniloski_hostfactors_nontargets'] = ['#a52a2a', 0.5, 1.5, hostfactors_nontargets]\n",
    "    Fig6b_df.loc['Daniloski_hostfactors_drugtargets']= ['magenta', 1.0, 1.5, hostfactors_drugtargets]\n",
    "    \n",
    "    #Fig. 6d: IFN_typeI_III+Blancomelo_IFN1response_DEGs_sig\n",
    "    Fig6d_df = pd.DataFrame(columns = ['color','linewidth','print_cutoff','genelist'])\n",
    "    IFN_response_list      = [i.strip().upper() for i in open(GENELISTDIR %'IFN_typeI_III+Blancomelo_IFN1response_DEGs_sig')]\n",
    "    IFN_list               = [gsym for gsym in IFN_response_list if 'IFN' in gsym]\n",
    "    others_list            = [gsym for gsym in IFN_response_list if not('IFN' in gsym)]\n",
    "    Fig6d_df.loc['others'] = ['#a52a2a',    0.5, 100.0, others_list]\n",
    "    Fig6d_df.loc['IFN']    = ['green',    1.0,   100.0, IFN_list]\n",
    "    \n",
    "    #Fig. 6e: Blancomelo_cytokineDEGs_sig_Fig2+4\n",
    "    Fig6e_df = pd.DataFrame(columns = ['color','linewidth','print_cutoff','genelist'])\n",
    "    Fig6e_df.loc['cytokine']  = ['#a52a2a', 0.5, 100.0, [i.strip().upper() for i in open(GENELISTDIR %'Blancomelo_cytokineDEGs_sig_Fig2+4')]]\n",
    "    \n",
    "    highlight_gene_info_df_list = [Fig6a_df,Fig6b_df,Fig6d_df, Fig6e_df]\n",
    "    tmp_sample_prefix_list = ['mRNA','RPF','QTI']\n",
    "    \n",
    "    fig, axes  = plt.subplots(nrows=len(highlight_gene_info_df_list), ncols=len(tmp_sample_prefix_list), \n",
    "                              figsize=(3.5*len(tmp_sample_prefix_list), 3*len(highlight_gene_info_df_list)))\n",
    "    \n",
    "    for idx_col, tmp_sample_prefix in enumerate(tmp_sample_prefix_list):\n",
    "        print(f'------{tmp_sample_prefix}------')\n",
    "        tmp_DEG_df_list   = []\n",
    "        tmp_hpi_list      = ['%dh' %i for i in [0,1,2,4,12,16,24,36]]\n",
    "        for hpi in tmp_hpi_list:\n",
    "            InFname    = f'{tmp_sample_prefix}_DESeq2_{hpi}.tsv'\n",
    "            tmp_DEG_df = pd.read_csv(RESULTDIR %(InFname), sep ='\\t', header=0, index_col=0)\n",
    "            tmp_DEG_df_list.append(tmp_DEG_df)\n",
    "        \n",
    "        for idx_row, highlight_gene_info_df in enumerate(highlight_gene_info_df_list):\n",
    "            if idx_row == 1:\n",
    "                ylim = (2.5,-5.0)\n",
    "            else:\n",
    "                ylim = 5.0\n",
    "            display(highlight_gene_info_df)\n",
    "            ax = axes[idx_row,idx_col]\n",
    "            _             = timecourse_plot(tmp_DEG_df_list,tmp_hpi_list, padj_cut=0.01, log2fc_cut=2, clip = 2.8,\n",
    "                                        ylim = ylim, n_cluster=5, as_average=False, \n",
    "                                        load_precalc_clusters=True, clusterFname=f'{tmp_sample_prefix}_cluster.tsv',\n",
    "                                        clusters_to_show=[], clusters_priority=[], highlight_gene_info_df = highlight_gene_info_df,\n",
    "                                        cl_method='ward',cl_metric='euclidean',ax = ax, \n",
    "                                        OutFigname='', show_fig=False, verbose=False)\n",
    "    if OutFigname != '':\n",
    "        plt.savefig(FIGDIR %OutFigname)\n",
    "        plt.close()\n",
    "    else:\n",
    "        plt.show()\n",
    "        plt.close()\n",
    "    return None\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "9156e042",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "------mRNA------\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>color</th>\n",
       "      <th>linewidth</th>\n",
       "      <th>print_cutoff</th>\n",
       "      <th>genelist</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Gordonetal</th>\n",
       "      <td>#a52a2a</td>\n",
       "      <td>0.5</td>\n",
       "      <td>100.0</td>\n",
       "      <td>[PCNT, PVR, POLA1, FASTKD5, PRIM2, ITGB1, CNTR...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ACE2+TMPRSS2</th>\n",
       "      <td>blue</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>[ACE2, TMPRSS2]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                color  linewidth  print_cutoff  \\\n",
       "Gordonetal    #a52a2a        0.5         100.0   \n",
       "ACE2+TMPRSS2     blue        1.0           0.0   \n",
       "\n",
       "                                                       genelist  \n",
       "Gordonetal    [PCNT, PVR, POLA1, FASTKD5, PRIM2, ITGB1, CNTR...  \n",
       "ACE2+TMPRSS2                                    [ACE2, TMPRSS2]  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>color</th>\n",
       "      <th>linewidth</th>\n",
       "      <th>print_cutoff</th>\n",
       "      <th>genelist</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Daniloski_hostfactors_nontargets</th>\n",
       "      <td>#a52a2a</td>\n",
       "      <td>0.5</td>\n",
       "      <td>1.5</td>\n",
       "      <td>[ACTR3, ATP6V1G1, KIAA1033, CCDC93, UVRAG, POD...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Daniloski_hostfactors_drugtargets</th>\n",
       "      <td>magenta</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.5</td>\n",
       "      <td>[VPS35, ACE2, ATP6V1C1, ATP6V1B2, CTSL, HDAC9,...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                     color  linewidth  print_cutoff  \\\n",
       "Daniloski_hostfactors_nontargets   #a52a2a        0.5           1.5   \n",
       "Daniloski_hostfactors_drugtargets  magenta        1.0           1.5   \n",
       "\n",
       "                                                                            genelist  \n",
       "Daniloski_hostfactors_nontargets   [ACTR3, ATP6V1G1, KIAA1033, CCDC93, UVRAG, POD...  \n",
       "Daniloski_hostfactors_drugtargets  [VPS35, ACE2, ATP6V1C1, ATP6V1B2, CTSL, HDAC9,...  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>color</th>\n",
       "      <th>linewidth</th>\n",
       "      <th>print_cutoff</th>\n",
       "      <th>genelist</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>others</th>\n",
       "      <td>#a52a2a</td>\n",
       "      <td>0.5</td>\n",
       "      <td>100.0</td>\n",
       "      <td>[MX1, XAF1, OAS1, IFITM1, OAS3, IRF9, OAS2, IR...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>IFN</th>\n",
       "      <td>green</td>\n",
       "      <td>1.0</td>\n",
       "      <td>100.0</td>\n",
       "      <td>[IFNA1, IFNA2, IFNA4, IFNA5, IFNA6, IFNA7, IFN...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          color  linewidth  print_cutoff  \\\n",
       "others  #a52a2a        0.5         100.0   \n",
       "IFN       green        1.0         100.0   \n",
       "\n",
       "                                                 genelist  \n",
       "others  [MX1, XAF1, OAS1, IFITM1, OAS3, IRF9, OAS2, IR...  \n",
       "IFN     [IFNA1, IFNA2, IFNA4, IFNA5, IFNA6, IFNA7, IFN...  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>color</th>\n",
       "      <th>linewidth</th>\n",
       "      <th>print_cutoff</th>\n",
       "      <th>genelist</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>cytokine</th>\n",
       "      <td>#a52a2a</td>\n",
       "      <td>0.5</td>\n",
       "      <td>100.0</td>\n",
       "      <td>[CCL2, CCL20, CCL8, CSF2, CSF3, CXCL1, CXCL14,...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            color  linewidth  print_cutoff  \\\n",
       "cytokine  #a52a2a        0.5         100.0   \n",
       "\n",
       "                                                   genelist  \n",
       "cytokine  [CCL2, CCL20, CCL8, CSF2, CSF3, CXCL1, CXCL14,...  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "------RPF------\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>color</th>\n",
       "      <th>linewidth</th>\n",
       "      <th>print_cutoff</th>\n",
       "      <th>genelist</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Gordonetal</th>\n",
       "      <td>#a52a2a</td>\n",
       "      <td>0.5</td>\n",
       "      <td>100.0</td>\n",
       "      <td>[PCNT, PVR, POLA1, FASTKD5, PRIM2, ITGB1, CNTR...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ACE2+TMPRSS2</th>\n",
       "      <td>blue</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>[ACE2, TMPRSS2]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                color  linewidth  print_cutoff  \\\n",
       "Gordonetal    #a52a2a        0.5         100.0   \n",
       "ACE2+TMPRSS2     blue        1.0           0.0   \n",
       "\n",
       "                                                       genelist  \n",
       "Gordonetal    [PCNT, PVR, POLA1, FASTKD5, PRIM2, ITGB1, CNTR...  \n",
       "ACE2+TMPRSS2                                    [ACE2, TMPRSS2]  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>color</th>\n",
       "      <th>linewidth</th>\n",
       "      <th>print_cutoff</th>\n",
       "      <th>genelist</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Daniloski_hostfactors_nontargets</th>\n",
       "      <td>#a52a2a</td>\n",
       "      <td>0.5</td>\n",
       "      <td>1.5</td>\n",
       "      <td>[ACTR3, ATP6V1G1, KIAA1033, CCDC93, UVRAG, POD...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Daniloski_hostfactors_drugtargets</th>\n",
       "      <td>magenta</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.5</td>\n",
       "      <td>[VPS35, ACE2, ATP6V1C1, ATP6V1B2, CTSL, HDAC9,...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                     color  linewidth  print_cutoff  \\\n",
       "Daniloski_hostfactors_nontargets   #a52a2a        0.5           1.5   \n",
       "Daniloski_hostfactors_drugtargets  magenta        1.0           1.5   \n",
       "\n",
       "                                                                            genelist  \n",
       "Daniloski_hostfactors_nontargets   [ACTR3, ATP6V1G1, KIAA1033, CCDC93, UVRAG, POD...  \n",
       "Daniloski_hostfactors_drugtargets  [VPS35, ACE2, ATP6V1C1, ATP6V1B2, CTSL, HDAC9,...  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>color</th>\n",
       "      <th>linewidth</th>\n",
       "      <th>print_cutoff</th>\n",
       "      <th>genelist</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>others</th>\n",
       "      <td>#a52a2a</td>\n",
       "      <td>0.5</td>\n",
       "      <td>100.0</td>\n",
       "      <td>[MX1, XAF1, OAS1, IFITM1, OAS3, IRF9, OAS2, IR...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>IFN</th>\n",
       "      <td>green</td>\n",
       "      <td>1.0</td>\n",
       "      <td>100.0</td>\n",
       "      <td>[IFNA1, IFNA2, IFNA4, IFNA5, IFNA6, IFNA7, IFN...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          color  linewidth  print_cutoff  \\\n",
       "others  #a52a2a        0.5         100.0   \n",
       "IFN       green        1.0         100.0   \n",
       "\n",
       "                                                 genelist  \n",
       "others  [MX1, XAF1, OAS1, IFITM1, OAS3, IRF9, OAS2, IR...  \n",
       "IFN     [IFNA1, IFNA2, IFNA4, IFNA5, IFNA6, IFNA7, IFN...  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>color</th>\n",
       "      <th>linewidth</th>\n",
       "      <th>print_cutoff</th>\n",
       "      <th>genelist</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>cytokine</th>\n",
       "      <td>#a52a2a</td>\n",
       "      <td>0.5</td>\n",
       "      <td>100.0</td>\n",
       "      <td>[CCL2, CCL20, CCL8, CSF2, CSF3, CXCL1, CXCL14,...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            color  linewidth  print_cutoff  \\\n",
       "cytokine  #a52a2a        0.5         100.0   \n",
       "\n",
       "                                                   genelist  \n",
       "cytokine  [CCL2, CCL20, CCL8, CSF2, CSF3, CXCL1, CXCL14,...  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "------QTI------\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>color</th>\n",
       "      <th>linewidth</th>\n",
       "      <th>print_cutoff</th>\n",
       "      <th>genelist</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Gordonetal</th>\n",
       "      <td>#a52a2a</td>\n",
       "      <td>0.5</td>\n",
       "      <td>100.0</td>\n",
       "      <td>[PCNT, PVR, POLA1, FASTKD5, PRIM2, ITGB1, CNTR...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ACE2+TMPRSS2</th>\n",
       "      <td>blue</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>[ACE2, TMPRSS2]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                color  linewidth  print_cutoff  \\\n",
       "Gordonetal    #a52a2a        0.5         100.0   \n",
       "ACE2+TMPRSS2     blue        1.0           0.0   \n",
       "\n",
       "                                                       genelist  \n",
       "Gordonetal    [PCNT, PVR, POLA1, FASTKD5, PRIM2, ITGB1, CNTR...  \n",
       "ACE2+TMPRSS2                                    [ACE2, TMPRSS2]  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>color</th>\n",
       "      <th>linewidth</th>\n",
       "      <th>print_cutoff</th>\n",
       "      <th>genelist</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Daniloski_hostfactors_nontargets</th>\n",
       "      <td>#a52a2a</td>\n",
       "      <td>0.5</td>\n",
       "      <td>1.5</td>\n",
       "      <td>[ACTR3, ATP6V1G1, KIAA1033, CCDC93, UVRAG, POD...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Daniloski_hostfactors_drugtargets</th>\n",
       "      <td>magenta</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.5</td>\n",
       "      <td>[VPS35, ACE2, ATP6V1C1, ATP6V1B2, CTSL, HDAC9,...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                     color  linewidth  print_cutoff  \\\n",
       "Daniloski_hostfactors_nontargets   #a52a2a        0.5           1.5   \n",
       "Daniloski_hostfactors_drugtargets  magenta        1.0           1.5   \n",
       "\n",
       "                                                                            genelist  \n",
       "Daniloski_hostfactors_nontargets   [ACTR3, ATP6V1G1, KIAA1033, CCDC93, UVRAG, POD...  \n",
       "Daniloski_hostfactors_drugtargets  [VPS35, ACE2, ATP6V1C1, ATP6V1B2, CTSL, HDAC9,...  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>color</th>\n",
       "      <th>linewidth</th>\n",
       "      <th>print_cutoff</th>\n",
       "      <th>genelist</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>others</th>\n",
       "      <td>#a52a2a</td>\n",
       "      <td>0.5</td>\n",
       "      <td>100.0</td>\n",
       "      <td>[MX1, XAF1, OAS1, IFITM1, OAS3, IRF9, OAS2, IR...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>IFN</th>\n",
       "      <td>green</td>\n",
       "      <td>1.0</td>\n",
       "      <td>100.0</td>\n",
       "      <td>[IFNA1, IFNA2, IFNA4, IFNA5, IFNA6, IFNA7, IFN...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          color  linewidth  print_cutoff  \\\n",
       "others  #a52a2a        0.5         100.0   \n",
       "IFN       green        1.0         100.0   \n",
       "\n",
       "                                                 genelist  \n",
       "others  [MX1, XAF1, OAS1, IFITM1, OAS3, IRF9, OAS2, IR...  \n",
       "IFN     [IFNA1, IFNA2, IFNA4, IFNA5, IFNA6, IFNA7, IFN...  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>color</th>\n",
       "      <th>linewidth</th>\n",
       "      <th>print_cutoff</th>\n",
       "      <th>genelist</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>cytokine</th>\n",
       "      <td>#a52a2a</td>\n",
       "      <td>0.5</td>\n",
       "      <td>100.0</td>\n",
       "      <td>[CCL2, CCL20, CCL8, CSF2, CSF3, CXCL1, CXCL14,...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            color  linewidth  print_cutoff  \\\n",
       "cytokine  #a52a2a        0.5         100.0   \n",
       "\n",
       "                                                   genelist  \n",
       "cytokine  [CCL2, CCL20, CCL8, CSF2, CSF3, CXCL1, CXCL14,...  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "plot_Fig6abde_multipanel(OutFigname='Fig6abde_reported_genes_timecourse.pdf')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "767ec624",
   "metadata": {},
   "source": [
    "### Correlation b/w max(|log2fc|) ~ Daniloski et al. physiological impact (Fig. 6c)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "4651bde3",
   "metadata": {},
   "outputs": [],
   "source": [
    "def calc_plot_assoc_lfc_physioimpact(sample_prefix='RPF', OutFigname=''):\n",
    "    tmp_DEG_df_list   = []\n",
    "    tmp_hpi_list      = ['%dh' %i for i in [0,1,2,4,12,16,24,36]]\n",
    "    for hpi in tmp_hpi_list:\n",
    "        InFname    = f'{sample_prefix}_DESeq2_{hpi}.tsv'\n",
    "        tmp_DEG_df = pd.read_csv(RESULTDIR %(InFname), sep ='\\t', header=0, index_col=0)\n",
    "        tmp_DEG_df_list.append(tmp_DEG_df)\n",
    "        \n",
    "    \n",
    "    hostfactors_df = pd.read_csv(GENELISTDIR %'Daniloski_MOI03_rank',sep='\\t')\n",
    "    hostfactors_df['gsym'] = hostfactors_df['gsym'].str.upper()\n",
    "    hostfactors    = hostfactors_df['gsym']\n",
    "    hostfactors_df = hostfactors_df.set_index('gsym')\n",
    "    \n",
    "    hostfactors_drugtargets = pd.read_csv(GENELISTDIR %'Daniloski_MOI03_dgidb',sep='\\t').iloc[:,0].str.upper().tolist()\n",
    "    \n",
    "    timecourse_df = pd.read_csv(RESULTDIR %(f'{sample_prefix}_timecourse.tsv'), \n",
    "                                sep = '\\t', index_col = 0)\n",
    "    \n",
    "    timecourse_df = timecourse_df.loc[hostfactors]\n",
    "    timecourse_df['max'] = timecourse_df.abs().max(axis=1)\n",
    "    timecourse_df['rank']= hostfactors_df['rank']\n",
    "    timecourse_df['lfc'] = hostfactors_df['lfc']\n",
    "    timecourse_df = timecourse_df.dropna(how='any')\n",
    "    colors = ['magenta' if i in hostfactors_drugtargets else '#a52a2a' for i in timecourse_df.index]\n",
    "    print('vs. maxabs',stats.spearmanr(timecourse_df['lfc'],timecourse_df['max']))\n",
    "        \n",
    "    xvar = 'max'\n",
    "    yvar = 'rank'\n",
    "    fig  = plt.figure(figsize=(3.5,3.5))\n",
    "    display(timecourse_df.sort_values(xvar,ascending=False).head(5))\n",
    "    plt.scatter(timecourse_df[xvar],timecourse_df[yvar], c = colors)\n",
    "    plt.ylabel('CRISPR screening rank')\n",
    "    plt.xlabel('max(|log2fc|)')\n",
    "    plt.ylim(bottom=0)\n",
    "    timecourse_df[[xvar,yvar]].to_csv(RESULTDIR %(f'Fig6c_correlation.tsv'), \n",
    "                                      sep = '\\t')\n",
    "    z = np.polyfit( timecourse_df[xvar], timecourse_df[yvar],1)\n",
    "    p = np.poly1d(z)\n",
    "    plt.plot(timecourse_df[xvar].sort_values(),p(timecourse_df[xvar].sort_values()),\"k--\")\n",
    "    sprho, spp = stats.spearmanr(timecourse_df[xvar],timecourse_df[yvar])\n",
    "    plt.text(timecourse_df[xvar].max()/2,timecourse_df[yvar].max()/1.1,f'Spearman rho={sprho:.03f}\\nP={spp:.02e}')\n",
    "    \n",
    "    plt.tight_layout()\n",
    "    if OutFigname != '':\n",
    "        plt.savefig(FIGDIR %OutFigname)\n",
    "    else:\n",
    "        plt.show()\n",
    "    plt.close()\n",
    "    \n",
    "    return None"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "0dd2c65d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "vs. maxabs SpearmanrResult(correlation=0.387003406631181, pvalue=0.0007844012325431527)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/sonic/baeklab/Hyeonseong/anaconda3/envs/Doyeon-tf2/lib/python3.6/site-packages/ipykernel_launcher.py:20: FutureWarning: \n",
      "Passing list-likes to .loc or [] with any missing label will raise\n",
      "KeyError in the future, you can use .reindex() as an alternative.\n",
      "\n",
      "See the documentation here:\n",
      "https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#deprecate-loc-reindex-listlike\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0h</th>\n",
       "      <th>1h</th>\n",
       "      <th>2h</th>\n",
       "      <th>4h</th>\n",
       "      <th>12h</th>\n",
       "      <th>16h</th>\n",
       "      <th>24h</th>\n",
       "      <th>36h</th>\n",
       "      <th>max</th>\n",
       "      <th>rank</th>\n",
       "      <th>lfc</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ACE2</th>\n",
       "      <td>-6.607765e-06</td>\n",
       "      <td>-0.000702</td>\n",
       "      <td>-0.055994</td>\n",
       "      <td>-0.319392</td>\n",
       "      <td>-2.566465</td>\n",
       "      <td>-3.545599</td>\n",
       "      <td>-3.782672</td>\n",
       "      <td>-4.101040</td>\n",
       "      <td>4.101040</td>\n",
       "      <td>12</td>\n",
       "      <td>1.6266</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ATP6AP2</th>\n",
       "      <td>-8.086250e-06</td>\n",
       "      <td>0.001427</td>\n",
       "      <td>0.116543</td>\n",
       "      <td>0.130180</td>\n",
       "      <td>-1.579491</td>\n",
       "      <td>-2.076438</td>\n",
       "      <td>-2.049512</td>\n",
       "      <td>-1.855499</td>\n",
       "      <td>2.076438</td>\n",
       "      <td>5</td>\n",
       "      <td>2.0384</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRSF1</th>\n",
       "      <td>1.806937e-06</td>\n",
       "      <td>0.004399</td>\n",
       "      <td>0.659602</td>\n",
       "      <td>1.374653</td>\n",
       "      <td>1.379408</td>\n",
       "      <td>1.851016</td>\n",
       "      <td>1.163534</td>\n",
       "      <td>0.166559</td>\n",
       "      <td>1.851016</td>\n",
       "      <td>30</td>\n",
       "      <td>1.2700</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>COMMD4</th>\n",
       "      <td>-9.434869e-07</td>\n",
       "      <td>-0.000646</td>\n",
       "      <td>-0.057539</td>\n",
       "      <td>-0.217448</td>\n",
       "      <td>-1.021387</td>\n",
       "      <td>-0.324847</td>\n",
       "      <td>-0.372262</td>\n",
       "      <td>-1.518002</td>\n",
       "      <td>1.518002</td>\n",
       "      <td>38</td>\n",
       "      <td>1.1835</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ATP6V1G1</th>\n",
       "      <td>-2.726768e-06</td>\n",
       "      <td>-0.000727</td>\n",
       "      <td>0.044275</td>\n",
       "      <td>0.437669</td>\n",
       "      <td>1.441861</td>\n",
       "      <td>1.310660</td>\n",
       "      <td>1.383131</td>\n",
       "      <td>1.245990</td>\n",
       "      <td>1.441861</td>\n",
       "      <td>15</td>\n",
       "      <td>1.5741</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                    0h        1h        2h        4h       12h       16h  \\\n",
       "id                                                                         \n",
       "ACE2     -6.607765e-06 -0.000702 -0.055994 -0.319392 -2.566465 -3.545599   \n",
       "ATP6AP2  -8.086250e-06  0.001427  0.116543  0.130180 -1.579491 -2.076438   \n",
       "SRSF1     1.806937e-06  0.004399  0.659602  1.374653  1.379408  1.851016   \n",
       "COMMD4   -9.434869e-07 -0.000646 -0.057539 -0.217448 -1.021387 -0.324847   \n",
       "ATP6V1G1 -2.726768e-06 -0.000727  0.044275  0.437669  1.441861  1.310660   \n",
       "\n",
       "               24h       36h       max  rank     lfc  \n",
       "id                                                    \n",
       "ACE2     -3.782672 -4.101040  4.101040    12  1.6266  \n",
       "ATP6AP2  -2.049512 -1.855499  2.076438     5  2.0384  \n",
       "SRSF1     1.163534  0.166559  1.851016    30  1.2700  \n",
       "COMMD4   -0.372262 -1.518002  1.518002    38  1.1835  \n",
       "ATP6V1G1  1.383131  1.245990  1.441861    15  1.5741  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "calc_plot_assoc_lfc_physioimpact(sample_prefix = 'RPF', OutFigname='Fig6c_physiological_assoc.pdf')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f249e0e9",
   "metadata": {},
   "source": [
    "### Fig. 6i: Inspection of eIFs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "322d4851",
   "metadata": {},
   "outputs": [],
   "source": [
    "def plot_Fig6i_eIF(OutFigname=''):\n",
    "    Fig6i_df = pd.DataFrame(columns = ['color','linewidth','print_cutoff','genelist'])\n",
    "    Fig6i_df.loc['nonAUG_enhancers']       = ['red', 0.5, 1.5, ['EIF5B','EIF2A','EIF2D']]\n",
    "    Fig6i_df.loc['nonAUG_enhancers_eIF5']  = ['red', 1.0, 1.5, ['EIF5']] #eIF5 alone for thicker line\n",
    "    Fig6i_df.loc['nonAUG_repressors']      = ['blue', 0.5, 1.5, ['EIF1','EIF1AX','EIF1AY']]\n",
    "    \n",
    "    highlight_gene_info_df = Fig6i_df\n",
    "    tmp_sample_prefix      = 'RPF'\n",
    "    \n",
    "    fig, ax  = plt.subplots(figsize=(3.5, 3))\n",
    "    \n",
    "    print('------'+tmp_sample_prefix+'------')\n",
    "    tmp_DEG_df_list   = []\n",
    "    tmp_hpi_list      = ['%dh' %i for i in [0,1,2,4,12,16,24,36]]\n",
    "    for hpi in tmp_hpi_list:\n",
    "        InFname    = f'{tmp_sample_prefix}_DESeq2_{hpi}.tsv'\n",
    "        tmp_DEG_df = pd.read_csv(RESULTDIR %(InFname), sep ='\\t', header=0, index_col=0)\n",
    "        tmp_DEG_df_list.append(tmp_DEG_df)\n",
    "    ylim = 2.5\n",
    "    display(highlight_gene_info_df)\n",
    "    _             = timecourse_plot(tmp_DEG_df_list,tmp_hpi_list, padj_cut=0.01, log2fc_cut=2, clip = 2.8,\n",
    "                                ylim = ylim, n_cluster=5, as_average=False, \n",
    "                                load_precalc_clusters=True, clusterFname=f'{tmp_sample_prefix}_cluster.tsv',\n",
    "                                clusters_to_show=[], clusters_priority=[], highlight_gene_info_df = highlight_gene_info_df,\n",
    "                                cl_method='ward',cl_metric='euclidean',ax = ax, \n",
    "                                OutFigname='', show_fig=False, verbose=False)\n",
    "    if OutFigname != '':\n",
    "        #plt.tight_layout()\n",
    "        plt.savefig(FIGDIR %OutFigname)\n",
    "        plt.close()\n",
    "    else:\n",
    "        #plt.tight_layout()\n",
    "        plt.show()\n",
    "        plt.close()\n",
    "    return None\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "aaed2f57",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "------RPF------\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>color</th>\n",
       "      <th>linewidth</th>\n",
       "      <th>print_cutoff</th>\n",
       "      <th>genelist</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>nonAUG_enhancers</th>\n",
       "      <td>red</td>\n",
       "      <td>0.5</td>\n",
       "      <td>1.5</td>\n",
       "      <td>[EIF5B, EIF2A, EIF2D]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>nonAUG_enhancers_eIF5</th>\n",
       "      <td>red</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.5</td>\n",
       "      <td>[EIF5]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>nonAUG_repressors</th>\n",
       "      <td>blue</td>\n",
       "      <td>0.5</td>\n",
       "      <td>1.5</td>\n",
       "      <td>[EIF1, EIF1AX, EIF1AY]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                      color  linewidth  print_cutoff                genelist\n",
       "nonAUG_enhancers        red        0.5           1.5   [EIF5B, EIF2A, EIF2D]\n",
       "nonAUG_enhancers_eIF5   red        1.0           1.5                  [EIF5]\n",
       "nonAUG_repressors      blue        0.5           1.5  [EIF1, EIF1AX, EIF1AY]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "plot_Fig6i_eIF(OutFigname='Fig6i_eIF_timecourse.pdf')"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Doyeon-tf2",
   "language": "python",
   "name": "doyeon-tf2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
