{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os, sys, time,re\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import matplotlib as mpl\n",
    "import matplotlib.pyplot as plt\n",
    "import matplotlib.font_manager as fm\n",
    "import matplotlib.lines as mlines\n",
    "from matplotlib.patches import Rectangle\n",
    "from matplotlib.gridspec import  GridSpec\n",
    "from scipy import stats\n",
    "from collections import defaultdict\n",
    "import seaborn as sns\n",
    "import pysam\n",
    "from collections import defaultdict\n",
    "\n",
    "REFFLAT_hg38      = '../ref/refFlat_hg38_repiso.txt'\n",
    "REFFLAT_chlSab2   = '../ref/refFlat_chlSab2.txt'     # Green monkey genome, for Vero cell data.\n",
    "REFFLAT_SARSCOV2  = '../ref/annot_SARSCOV2.txt'      # Not exactly refFlat, but similar format. Used ORF start-end information.\n",
    "\n",
    "BAMDIR_hostmapped = '/extdata1/baeklab/Doyeon/SARSCOV2/data/%s_hostalign_021721/%s.bam'     #e.g. %('mRNASeq','mRNA_2h_rep1')\n",
    "BAMDIR_cov2mapped = '/extdata1/baeklab/Doyeon/SARSCOV2/data/%s_SARSCOV2align_021721/%s.bam' #e.g. %('mRNASeq','mRNA_2h_rep1')\n",
    "RPKMDIR           = '/extdata1/baeklab/Doyeon/SARSCOV2/data/rpkm_081820/%s.txt'             #e.g. %'mRNA_2h_rep1'\n",
    "'''\n",
    "Sequencing data can be downloaded from https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE157490\n",
    "'''\n",
    "\n",
    "RESULTDIR    = '../result/%s' \n",
    "FIGDIR       = '../fig/%s' \n",
    "SOURCEDATADIR= '../source_data/%s' \n",
    "\n",
    "\n",
    "### Global variables ###\n",
    "FASTA_SARSCOV2   = '../ref/covid19.fa'\n",
    "SARSCOV2_genome  = ''.join([i.strip() for i in open(FASTA_SARSCOV2)][1:])\n",
    "CANONICAL_sgRNA_junc_dict = {\n",
    "                            'S':\t[65,21551],\n",
    "                            'ORF3a':[66,25381],\n",
    "                            'E':\t[69,26236],\n",
    "                            'M':\t[64,26467],\n",
    "                            'ORF6': [69,27040],\n",
    "                            'ORF7a':[66,27384],\n",
    "                            'ORF7b':[71,27761],\n",
    "                            'ORF8': [65,27883],\n",
    "                            'N':    [64,28254]\n",
    "                            }\n",
    "ORF_palette = {'5\\'UTR': '#707070',\n",
    "               'ORF1a':  '#fbde8e',\n",
    "               'ORF1b':  '#cce8ff',\n",
    "               'S':      '#d62728',\n",
    "               'ORF3a':  '#8c564b',\n",
    "               'E':      '#1b1464',\n",
    "               'M':      '#1f77b4',\n",
    "               'ORF6':   '#9467bd',\n",
    "               'ORF7a':  '#ff7f0e',\n",
    "               'ORF7b':  '#067f4c',\n",
    "               'ORF8':   '#e377c2',\n",
    "               'N':      '#bcbd22',\n",
    "               'ORF10':  '#3bcf49',\n",
    "               '3\\'UTR': '#707070',\n",
    "               'ORF1ab': '#fbde8e' \n",
    "              }\n",
    "#########################\n",
    "\n",
    "%matplotlib inline"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Basic functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "class gene:\n",
    "    def __init__(self):\n",
    "        self.sGeneSym       = ''\n",
    "        self.sNMID          = ''\n",
    "        self.sChrID         = ''\n",
    "        self.nExons         = 0\n",
    "        self.nExonStartlist = []\n",
    "        self.nExonEndlist   = []\n",
    "        self.sStrandDir     = ''\n",
    "        self.nORF_5p_gidx   = 0\n",
    "        self.nORF_3p_gidx   = 0\n",
    "        self.nExonlen       = 0\n",
    "        self.nU5len         = 0\n",
    "        self.nU3len         = 0\n",
    "\n",
    "    def parse_refflat(self,refline):\n",
    "        sInfolist = refline.strip().replace(' ','\\t').split('\\t')\n",
    "        self.sGeneSym = sInfolist[0].upper()\n",
    "        self.sNMID    = sInfolist[1]\n",
    "        self.sChrID   = sInfolist[2] ##chr1,,,,chrX,chrY for human\n",
    "        self.sStrandDir   = sInfolist[3]\n",
    "        self.nORF_5p_gidx = int(sInfolist[6])\n",
    "        self.nORF_3p_gidx = int(sInfolist[7])\n",
    "        self.nExons       = int(sInfolist[8])\n",
    "        self.nExonStartlist = [int(i) for i in sInfolist[9].split(',') if i != '']\n",
    "        self.nExonEndlist   = [int(i) for i in sInfolist[10].split(',') if i != '']\n",
    "        assert (self.nExons == len(self.nExonStartlist)) and (self.nExons == len(self.nExonEndlist))\n",
    "        self.nExonlen       = sum([end-start for start,end in zip(self.nExonStartlist, self.nExonEndlist)]) \n",
    "        \n",
    "        tmp_exonlen = 0\n",
    "        for start,end in zip(self.nExonStartlist, self.nExonEndlist):\n",
    "            if start <= self.nORF_5p_gidx <  end:\n",
    "                self.nU5len  = tmp_exonlen + (self.nORF_5p_gidx - start)\n",
    "            if start <= self.nORF_3p_gidx   <= end:\n",
    "                self.nU3len = self.nExonlen - (tmp_exonlen + (self.nORF_3p_gidx - start))\n",
    "            tmp_exonlen += (end-start)\n",
    "        if self.sStrandDir == '-':\n",
    "            tmp_nU3len  = self.nU5len\n",
    "            self.nU5len = self.nU3len\n",
    "            self.nU3len = tmp_nU3len\n",
    "#########################\n",
    "\n",
    "def get_df_SARSCOV2_annot(exc_UTR = True):\n",
    "    df = pd.read_csv(REFFLAT_SARSCOV2,sep='\\t',header=None)\n",
    "    df = df.iloc[:,[0,2,3,6,7]]\n",
    "    df.columns = ['ORF','chromosome','strand','ORFstart','ORFend']\n",
    "    df = df.set_index('ORF')\n",
    "    if exc_UTR:\n",
    "        df = df[~df.index.str.contains('UTR')]\n",
    "    return df\n",
    "\n",
    "def get_dict_refgenes(refdir):\n",
    "    ref_lines  = [line.strip() for line in open(refdir)]\n",
    "    chrdict = defaultdict(list)\n",
    "    \n",
    "    for line in ref_lines:\n",
    "        refgene  = gene()\n",
    "        refgene.parse_refflat(line)\n",
    "        chrdict['%s%s'%(refgene.sChrID, refgene.sStrandDir)].append(refgene)\n",
    "    chrdict = dict(chrdict)\n",
    "    return chrdict\n",
    "\n",
    "def get_readpos_arr(samplename, assay_prefix, norm_by_nh=True, pos_5end_range=(46,52), \n",
    "                    f2_5p_range=None, fetch_range=(0,265), offset_RPF = 12, junction_spanning_only=True, \n",
    "                    min_rlen=(32,31,30,29,28,27), as_RPM=False, regions_exclude = [], verbose=False):\n",
    "    #offset_RPF: non-negative integer or 'readcenter'\n",
    "    #min_rlen: non-negative integer or tuple \n",
    "    #    if tuple, applying minimum rlen for each position to the pos_5end_range, \n",
    "    #    assert len(minrlen) == pos_5end_range[1]-pos_5end_range[0])\n",
    "    fname = BAMDIR_cov2mapped %(assay_prefix,samplename)\n",
    "    bam = pysam.AlignmentFile(fname)\n",
    "    arr_f1_3p = np.zeros(len(SARSCOV2_genome))\n",
    "    arr_f2_5p = np.zeros(len(SARSCOV2_genome))\n",
    "    arr_RPF   = np.zeros(len(SARSCOV2_genome))\n",
    "    \n",
    "    assigned_ct_raw     = 0\n",
    "    assigned_ct_weighted= 0\n",
    "    contig = bam.references[0]\n",
    "    for read in bam.fetch(contig,fetch_range[0],fetch_range[1]):\n",
    "        cigartuples = read.cigartuples\n",
    "        pos_5end    = read.reference_start\n",
    "        rlen        = read.infer_read_length()\n",
    "        nh          = read.get_tag('NH')\n",
    "        assigned    = False\n",
    "        if norm_by_nh:\n",
    "            norm_ct = 1/nh\n",
    "        else:\n",
    "            norm_ct = 1\n",
    "            \n",
    "        if not ((pos_5end_range[0]<= pos_5end) and (pos_5end< pos_5end_range[1])):\n",
    "            continue\n",
    "            \n",
    "        if type(min_rlen) == int:\n",
    "            if rlen<min_rlen:\n",
    "                continue\n",
    "        else:\n",
    "            if rlen<min_rlen[pos_5end-pos_5end_range[0]]:\n",
    "                continue\n",
    "        reference_positions = read.get_reference_positions(full_length=True)    \n",
    "        reference_positions = [pos for pos in reference_positions if pos != None]\n",
    "        if offset_RPF == 'readcenter':\n",
    "            offset = rlen//2\n",
    "        else:\n",
    "            offset = offset_RPF\n",
    "            \n",
    "        refpos = reference_positions[offset]\n",
    "        \n",
    "        contain_junction = (3 in [i[0] for i in cigartuples])\n",
    "        cur_pos = pos_5end\n",
    "        f1_detected = False\n",
    "        f1_3p = None\n",
    "        f2_5p = None\n",
    "        \n",
    "        if (not contain_junction):\n",
    "            if junction_spanning_only:\n",
    "                continue\n",
    "            else: \n",
    "                pass\n",
    "        else: ##determination of f1_3p (3' end of the first fragment) and f2_5p (5' end of the other fragments) for junction_spanning reads\n",
    "            for idx,cigar_tp in enumerate(cigartuples):\n",
    "                operation, length = cigar_tp\n",
    "                if operation == 3: #skip\n",
    "                    f1_detected = True\n",
    "                    cur_pos += length\n",
    "                else:\n",
    "                    if operation == 4 or operation == 1:    # softclip or insertion\n",
    "                        pass\n",
    "                    elif operation == 0 or operation == 2:  # match/mismatch or deletion\n",
    "                        cur_pos += length\n",
    "                if operation == 0:\n",
    "                    if not f1_detected: #junction 1st fragment\n",
    "                        f1_3p = cur_pos\n",
    "\n",
    "                    else:\n",
    "                        f2_5p = cur_pos - length\n",
    "                        break\n",
    "\n",
    "            assert f1_3p != None\n",
    "            assert f2_5p != None\n",
    "            \n",
    "            if f2_5p_range == None:\n",
    "                arr_f2_5p[f2_5p] += norm_ct\n",
    "            else:\n",
    "                if (f2_5p_range[0]<= f2_5p) and (f2_5p< f2_5p_range[1]):\n",
    "                    arr_f2_5p[f2_5p] += norm_ct\n",
    "                else: ##roll back if the read doesn't meet f2_5p range\n",
    "                    continue\n",
    "            arr_f1_3p[f1_3p] += norm_ct\n",
    "        \n",
    "        arr_RPF[refpos]      += norm_ct\n",
    "        assigned_ct_raw      += 1\n",
    "        assigned_ct_weighted += norm_ct\n",
    "    \n",
    "    bam.close()\n",
    "    arr_include = np.ones(len(arr_RPF))\n",
    "    for s,e in regions_exclude:\n",
    "        arr_include[s:e] = 0 \n",
    "    arr_RPF = arr_RPF*arr_include\n",
    "    \n",
    "    total_mapped_reads = 0\n",
    "    viral_mapped_reads = 0\n",
    "    host_mapped_reads  = 0\n",
    "    if as_RPM:\n",
    "        #virus reads\n",
    "        bam = pysam.AlignmentFile(fname)\n",
    "        for read in bam.fetch():\n",
    "            viral_mapped_reads += (1/(read.get_tag('NH')))\n",
    "        bam.close()\n",
    "        #human reads\n",
    "        fname_host = BAMDIR_hostmapped %(assay_prefix,samplename)\n",
    "        bam = pysam.AlignmentFile(fname_host)\n",
    "        for read in bam.fetch():\n",
    "            host_mapped_reads += (1/(read.get_tag('NH')))\n",
    "        bam.close()\n",
    "        total_mapped_reads = host_mapped_reads + viral_mapped_reads\n",
    "        arr_f1_3p /= (total_mapped_reads / 1e+06)\n",
    "        arr_f2_5p /= (total_mapped_reads / 1e+06)\n",
    "        arr_RPF   /= (total_mapped_reads / 1e+06)\n",
    "    \n",
    "    report = [total_mapped_reads,viral_mapped_reads,host_mapped_reads, assigned_ct_raw, assigned_ct_weighted]\n",
    "    report = pd.Series(report)\n",
    "    report.index = ['total_mapped','viral_mapped','host_mapped','region_assigned(raw)','region_assigned(weighted)']\n",
    "    if verbose:\n",
    "        print(fname)\n",
    "        display(report)\n",
    "    return arr_f1_3p, arr_f2_5p, arr_RPF\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Figure 1e- QC analysis"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Metagene analysis (which includes analysis of read enrichment at ORF start) (Figure 1e)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "def calc_metagene_array(samplename,assay_prefix='RPFSeq', genome='SARSCOV2', offset_RPF = 12, \n",
    "                        flanking=10, align='5p', nreads_cutoff = 50, verbose = True):\n",
    "    if genome == 'SARSCOV2':\n",
    "        fname = BAMDIR_cov2mapped %(assay_prefix,samplename)\n",
    "        refdir= REFFLAT_SARSCOV2\n",
    "    elif genome in ['hg38','chlSab2']:\n",
    "        fname = BAMDIR_hostmapped %(assay_prefix,samplename)\n",
    "        if genome == 'hg38':\n",
    "            refdir = REFFLAT_hg38\n",
    "        else:\n",
    "            refdir = REFFLAT_chlSab2\n",
    "    else:\n",
    "        print('Wrong genome name:',genome)\n",
    "        return None\n",
    "    \n",
    "    chr_ref_dict = get_dict_refgenes(refdir) #{'%s%s'%(refgene.sChrID, refgene.sStrandDir): [gene,,]}\n",
    "    if genome == 'SARSCOV2': #modify annotation for flanking region\n",
    "        for sChrID_strand,refgene_list in chr_ref_dict.items():\n",
    "            for refgene in refgene_list:\n",
    "                refgene.nExonStartlist[0]  -= flanking\n",
    "                refgene.nExonEndlist[-1]   += flanking\n",
    "                \n",
    "                refgene.nExonlen += (flanking*2)\n",
    "                refgene.nU5len    = flanking\n",
    "                refgene.nU3len    = flanking\n",
    "        \n",
    "    bam      = pysam.AlignmentFile(fname)\n",
    "    arr_RPF_ORF_list = []\n",
    "    for sChrID_strand,refgene_list in chr_ref_dict.items():\n",
    "        if verbose:\n",
    "            print(sChrID_strand, time.ctime())\n",
    "        sChrID     = sChrID_strand[:-1]\n",
    "        sStrandDir = sChrID_strand[-1]\n",
    "        gene_negstrand = (sStrandDir == '-')\n",
    "        \n",
    "        for refgene in refgene_list:\n",
    "            if refgene.sGeneSym in [\"5'UTR\", \"3'UTR\", \"ORF1b\"]: #filter out those UTR annotations in SARSCOV2 refflat\n",
    "                continue\n",
    "            \n",
    "            exon_s_list = refgene.nExonStartlist\n",
    "            exon_e_list = refgene.nExonEndlist\n",
    "            \n",
    "            arr_RPF     = np.zeros(refgene.nExonlen,dtype=np.float32)\n",
    "            offset_arr  = 0\n",
    "            for exon_s, exon_e in zip(exon_s_list,exon_e_list):\n",
    "                for read in bam.fetch(sChrID, exon_s,exon_e):\n",
    "                    rlen        = read.infer_read_length()\n",
    "                    nh          = read.get_tag('NH')\n",
    "                    \n",
    "                    if read.is_reverse != gene_negstrand:\n",
    "                        continue\n",
    "                    if offset_RPF == 'readcenter':\n",
    "                        offset = rlen//2\n",
    "                    else:\n",
    "                        offset = offset_RPF\n",
    "                    reference_positions = read.get_reference_positions(full_length=False)    \n",
    "                    if None in reference_positions:\n",
    "                        print('None in read reference position?')\n",
    "                        print(reference_positions)\n",
    "                        print(read.cigarstring)\n",
    "                        return None\n",
    "                    if gene_negstrand:\n",
    "                        refpos   = reference_positions[-(offset+1)]\n",
    "                    else:\n",
    "                        refpos   = reference_positions[offset]\n",
    "                    \n",
    "                    nh = read.get_tag('NH')\n",
    "                    if (exon_s <= refpos) and (refpos < exon_e):\n",
    "                        arr_RPF[refpos - exon_s + offset_arr] += (1/nh)\n",
    "                offset_arr += (exon_e - exon_s)\n",
    "                \n",
    "            assert offset_arr == len(arr_RPF)\n",
    "            \n",
    "            if sChrID == '-':\n",
    "                arr_RPF = arr_RPF[::-1]\n",
    "            \n",
    "            flanking_5p = min(refgene.nU5len,flanking)\n",
    "            arr_RPF_ORF_flank = np.concatenate([np.zeros(flanking-flanking_5p),\n",
    "                                                arr_RPF[refgene.nU5len-flanking_5p:]])\n",
    "            \n",
    "            flanking_3p = min(refgene.nU3len,flanking)\n",
    "            if (refgene.nU3len-flanking_3p)>0:\n",
    "                arr_RPF_ORF_flank = arr_RPF_ORF_flank[:-(refgene.nU3len-flanking_3p)]\n",
    "            arr_RPF_ORF_flank = np.concatenate([arr_RPF_ORF_flank,\n",
    "                                                np.zeros(flanking-flanking_3p)])\n",
    "            \n",
    "            nreads_ORFmapped = arr_RPF_ORF_flank[flanking:-flanking].sum() #divide by ORF RPF signal sum\n",
    "            if nreads_ORFmapped < nreads_cutoff:\n",
    "                continue\n",
    "            arr_RPF_ORF_flank /= nreads_ORFmapped\n",
    "            \n",
    "            len_ORF = len(arr_RPF) - refgene.nU5len - refgene.nU3len\n",
    "            assert len_ORF+(2*flanking) == len(arr_RPF_ORF_flank)\n",
    "            arr_RPF_ORF_list.append(arr_RPF_ORF_flank)\n",
    "            \n",
    "    max_ORF_flank_len = max([len(arr) for arr in arr_RPF_ORF_list])\n",
    "    if align == '5p':\n",
    "        arr_RPF_ORF_list = [np.concatenate([arr,np.full(max_ORF_flank_len-len(arr), np.nan)]) for arr in arr_RPF_ORF_list]\n",
    "    elif align == '3p':\n",
    "        arr_RPF_ORF_list = [np.concatenate([np.full(max_ORF_flank_len-len(arr), np.nan),arr]) for arr in arr_RPF_ORF_list]\n",
    "        \n",
    "    arr_mean_freq = np.array(arr_RPF_ORF_list)\n",
    "    arr_mean_freq = np.nanmean(arr_mean_freq,axis=0)\n",
    "    bam.close()\n",
    "    return arr_mean_freq\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "def plot_metagene_analysis(hpi,assay_prefix='RPFSeq', sample_prefix_list=['RPF'],genome='SARSCOV2',\n",
    "                           offset_RPF = 12, flanking = 10, align = '5p', nreads_cutoff = 100, \n",
    "                           verbose=False, plot_range = 50,ylim=None,OutFigname = '', ax=None,show_fig=True):\n",
    "    if hpi == '48h':\n",
    "        reps = 3\n",
    "    else:\n",
    "        reps = 2\n",
    "        \n",
    "    arr_merged = []\n",
    "    for sample_prefix in sample_prefix_list:\n",
    "        for rep in range(1,reps+1):\n",
    "            samplename = f'{sample_prefix}_{hpi}_rep{rep}'\n",
    "            arr_mean_freq = calc_metagene_array(samplename, assay_prefix=assay_prefix, genome=genome, \n",
    "                                                offset_RPF = offset_RPF, flanking=flanking, align=align, \n",
    "                                                nreads_cutoff = nreads_cutoff, verbose = verbose)\n",
    "            if align == '5p':\n",
    "                arr_merged.append(arr_mean_freq[:plot_range])\n",
    "            elif align == '3p':\n",
    "                arr_merged.append(arr_mean_freq[-plot_range:])\n",
    "                \n",
    "    arr_merged = np.array(arr_merged).mean(axis=0)*100\n",
    "    if align == '5p':\n",
    "        x   = np.arange(-flanking,plot_range-flanking)\n",
    "        xticks      = [0,10,20,30]\n",
    "    elif align == '3p':\n",
    "        x   = np.arange(-(plot_range-flanking),flanking)\n",
    "        xticks      = [-30,-20,-10,0]\n",
    "        \n",
    "    if ax == None:\n",
    "        fig, ax = plt.subplots(figsize=(4,3))\n",
    "    if genome == 'SARSCOV2': #red\n",
    "        bar_colors = [['#ff6666','#fc8d59','#fdcc8a'][i%3] for i in x]\n",
    "    else:                    #green\n",
    "        bar_colors = [['#25ad62','#74c476','#bae4b3'][i%3] for i in x]\n",
    "    ax.bar(x,arr_merged,width=1,color=bar_colors)\n",
    "    ax.set_xticks(xticks)\n",
    "    ax.set_xlim(x[0],x[-1])\n",
    "    ax.set_xlabel('Relative position')\n",
    "    ax.set_ylabel('Relative fraction (%)')\n",
    "    if ylim!= None:\n",
    "        ax.set_ylim(ylim)\n",
    "    \n",
    "    if OutFigname != '':\n",
    "        plt.tight_layout()\n",
    "        plt.savefig(FIGDIR %OutFigname)\n",
    "        plt.close()\n",
    "        \n",
    "    elif show_fig:\n",
    "        plt.tight_layout()\n",
    "        plt.show()\n",
    "        plt.close()\n",
    "    \n",
    "    return arr_merged\n",
    "\n",
    "###################################\n",
    "def plot_merged_main(hpi='4h',ylim_list=[[]],OutFigname=''):\n",
    "    assay_prefix_list  = ['mRNASeq','RPFSeq','RPFSeq']\n",
    "    sample_prefix_list_list = [['RPFpaired','QTIpaired'],['RPF'],['QTI']]\n",
    "    genome_list        = ['hg38','SARSCOV2']\n",
    "    \n",
    "    fig, axes= plt.subplots(nrows=len(genome_list),\n",
    "                            ncols=len(sample_prefix_list_list), \n",
    "                            figsize=(3*len(sample_prefix_list_list),2*len(genome_list)))\n",
    "    df_source = pd.DataFrame(columns =['%s_%s' %(genome,sample_prefix_list[0]) for sample_prefix_list in sample_prefix_list_list \n",
    "                                       for genome in genome_list],\n",
    "                             index = np.arange(-10,40))\n",
    "    for idx_row,genome in enumerate(genome_list):\n",
    "        for idx_col,sample_prefix_list in enumerate(sample_prefix_list_list):\n",
    "            print(genome, sample_prefix_list, time.ctime())\n",
    "            assay_prefix = assay_prefix_list[idx_col]\n",
    "            ax   = axes[idx_row,idx_col]\n",
    "            ylim = ylim_list[idx_row][idx_col]\n",
    "            arr  = plot_metagene_analysis(hpi,assay_prefix=assay_prefix, sample_prefix_list=sample_prefix_list,\n",
    "                                   genome=genome, offset_RPF = 12, flanking = 10, align = '5p', nreads_cutoff = 50, \n",
    "                                   verbose=False, plot_range = 50,ylim=ylim, ax=ax, show_fig = False)\n",
    "            df_source['%s_%s' %(genome,sample_prefix_list[0])] = arr\n",
    "    plt.tight_layout()\n",
    "    plt.savefig(FIGDIR %(OutFigname))\n",
    "    plt.close()\n",
    "    df_source.to_csv(SOURCEDATADIR %(OutFigname.replace('pdf','tsv')), sep = '\\t')\n",
    "    return None"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "hg38 ['RPFpaired', 'QTIpaired'] Mon Sep 27 15:19:55 2021\n",
      "hg38 ['RPF'] Mon Sep 27 15:28:17 2021\n",
      "hg38 ['QTI'] Mon Sep 27 15:30:40 2021\n",
      "SARSCOV2 ['RPFpaired', 'QTIpaired'] Mon Sep 27 15:31:47 2021\n",
      "SARSCOV2 ['RPF'] Mon Sep 27 15:31:49 2021\n",
      "SARSCOV2 ['QTI'] Mon Sep 27 15:31:49 2021\n",
      "hg38 ['RPFpaired', 'QTIpaired'] Mon Sep 27 15:31:50 2021\n",
      "hg38 ['RPF'] Mon Sep 27 15:35:24 2021\n",
      "hg38 ['QTI'] Mon Sep 27 15:36:08 2021\n",
      "SARSCOV2 ['RPFpaired', 'QTIpaired'] Mon Sep 27 15:36:46 2021\n",
      "SARSCOV2 ['RPF'] Mon Sep 27 15:40:11 2021\n",
      "SARSCOV2 ['QTI'] Mon Sep 27 15:40:21 2021\n"
     ]
    }
   ],
   "source": [
    "ylim_list_4h = [[(0.0,1.0),(0.0,6.0), (0.0,6.0) ], #hg38\n",
    "                [(0.0,4.0),(0.0,20.0),(0.0,20.0)]] #SARSCOV2\n",
    "plot_merged_main(hpi='4h', ylim_list = ylim_list_4h,\n",
    "                 OutFigname='Fig1e_metagene_4h.pdf')\n",
    "ylim_list_36h = [[(0.0,1.0),(0.0,4.0),(0.0,4.0) ], #hg38\n",
    "                 [(0.0,4.0),(0.0,8.0),(0.0,8.0)]] #SARSCOV2\n",
    "plot_merged_main(hpi='36h', ylim_list = ylim_list_36h,\n",
    "                 OutFigname='Fig1e_metagene_36h.pdf')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Read length histogram (Supplementary Figure 1d)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "def hist_readlength_cmp(assay_prefix='RPFSeq',sample_prefix_list=['RPF'],hpi='48h',\n",
    "                        genome_list = ['SARSCOV2','host'],\n",
    "                        color_h = 'g', color_v = 'r',xlim=[18,40],\n",
    "                        merge_reps = True,OutF='',OutFigname='', ax = None, show_fig = False):\n",
    "    summary= []\n",
    "    bins   = np.arange(*xlim)\n",
    "    hist_group_keys = []\n",
    "    hist_group_dict = {}\n",
    "    reps = 3 if hpi == '48h' else 2\n",
    "    \n",
    "    for genome in genome_list:\n",
    "        if genome == 'SARSCOV2':\n",
    "            BAMDIR = BAMDIR_cov2mapped\n",
    "        else:\n",
    "            BAMDIR = BAMDIR_hostmapped\n",
    "        \n",
    "        for sample_prefix in sample_prefix_list:\n",
    "            if merge_reps:\n",
    "                hist_group_key = f'{genome}/{sample_prefix}_{hpi}'\n",
    "                hist_group_keys.append(hist_group_key)\n",
    "                hist_group_dict[hist_group_key] = []\n",
    "            for rep in range(1, 1+reps):\n",
    "                samplename = f'{sample_prefix}_{hpi}_rep{rep}'\n",
    "                bam = pysam.AlignmentFile(BAMDIR %(assay_prefix,samplename))    #%(RPFSeq'','RPF_2h_rep1')\n",
    "                rlen_list = []\n",
    "                for read in bam.fetch():\n",
    "                    rlen = read.infer_read_length()\n",
    "                    rlen_list.append(rlen)\n",
    "                bam.close()\n",
    "                if not merge_reps:\n",
    "                    hist_group_key = f'{genome}/{sample_prefix}_{hpi}_rep{rep}'\n",
    "                    hist_group_keys.append(hist_group_key)\n",
    "                    hist_group_dict[hist_group_key] = []\n",
    "                \n",
    "                hist_group_dict[hist_group_key] += rlen_list\n",
    "    \n",
    "    if ax == None:\n",
    "        fig, ax = plt.subplots(figsize=(3,3))\n",
    "    sub_df_source = pd.DataFrame(columns = hist_group_keys)\n",
    "    for hist_group_key in hist_group_keys:\n",
    "        rlen_list  = hist_group_dict[hist_group_key]\n",
    "        genome, _  = hist_group_key.split('/')\n",
    "        if genome == 'host':\n",
    "            color = color_h\n",
    "        else:\n",
    "            color = color_v\n",
    "        \n",
    "        rlen_arr = np.array(rlen_list)\n",
    "        n_reads = len(rlen_arr)\n",
    "        \n",
    "        summary.append([hist_group_key,n_reads,np.mean(rlen_arr),np.median(rlen_arr), np.std(rlen_arr)])        \n",
    "        n, _, _ = ax.hist(rlen_arr,bins=bins,histtype='step',label = hist_group_key,\n",
    "                          density=True,color=color,  align = 'left')\n",
    "        sub_df_source[hist_group_key] = n\n",
    "        \n",
    "    ax.set_xlabel('Read length (nt)')\n",
    "    ax.set_ylabel('Density')\n",
    "    ax.set_xlim(*xlim)\n",
    "    ax.set_xticks(np.arange(*xlim,2))\n",
    "    ax.set_ylim(0,0.55)\n",
    "    \n",
    "    if merge_reps:\n",
    "        if assay_prefix == 'mRNASeq':\n",
    "            ax.legend(ncol=1,loc='upper left')\n",
    "        else:\n",
    "            ax.legend(ncol=1,loc='upper right') \n",
    "    else: \n",
    "        pass\n",
    "    \n",
    "    if OutFigname != '':\n",
    "        plt.tight_layout()\n",
    "        plt.savefig(FIGDIR %OutFigname)\n",
    "        plt.close()\n",
    "        \n",
    "    elif show_fig:\n",
    "        plt.show()\n",
    "        display(summary_df)\n",
    "        plt.close()\n",
    "        \n",
    "    summary_df = pd.DataFrame(summary)\n",
    "    summary_df.columns = ['samplename','n','mean_len','median_len','std_len']\n",
    "    summary_df = summary_df.set_index('samplename')\n",
    "    summary_df.to_csv(RESULTDIR %OutF, sep = '\\t')\n",
    "    return sub_df_source\n",
    "\n",
    "def draw_readhist_multipanel(OutFigname='',transposed=True):\n",
    "    #(RPF, QTI) * (hg38, SARSCOV2) * (0~48h)\n",
    "    hpi_list     = ['%dh' %i for i in [4,12,16,24,36,48]]+['48hCaco2']\n",
    "    assay_prefix = 'RPFSeq'\n",
    "    color_h = '#25ad62'\n",
    "    color_v = '#ff6666'\n",
    "    xlim = [20,33]  \n",
    "    \n",
    "    if transposed:\n",
    "        fig, axes  = plt.subplots(ncols=len(hpi_list),nrows=2, figsize=(3*len(hpi_list),5))\n",
    "    else:\n",
    "        fig, axes  = plt.subplots(nrows=len(hpi_list),ncols=2, figsize=(6,2.5*len(hpi_list)))\n",
    "        \n",
    "    df_source = []\n",
    "    for idx_row,hpi in enumerate(hpi_list):\n",
    "        for idx_col,sample_prefix in enumerate(['RPF','QTI']):\n",
    "            print(hpi,sample_prefix,time.ctime())\n",
    "            if transposed:\n",
    "                ax = axes[idx_col,idx_row]\n",
    "            else:\n",
    "                ax = axes[idx_row,idx_col]\n",
    "            sub_df_source = hist_readlength_cmp(assay_prefix=assay_prefix,sample_prefix_list=[sample_prefix],hpi=hpi,\n",
    "                                genome_list = ['SARSCOV2','host'],\n",
    "                                color_h = color_h, color_v = color_v,xlim=xlim,\n",
    "                                merge_reps = True, OutF=f'read_hist_summary_{sample_prefix}_{hpi}.tsv',\n",
    "                                OutFigname='', ax=ax, show_fig=False)\n",
    "            df_source.append(sub_df_source)\n",
    "    df_source = pd.concat(df_source,axis=1)*100\n",
    "    \n",
    "    plt.tight_layout()\n",
    "    plt.savefig(FIGDIR %(OutFigname))\n",
    "    df_source.to_csv(SOURCEDATADIR %(OutFigname.replace('pdf','tsv')), sep = '\\t')\n",
    "    plt.close()\n",
    "    return None"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "4h RPF Mon Sep 27 15:40:27 2021\n",
      "4h QTI Mon Sep 27 15:40:39 2021\n",
      "12h RPF Mon Sep 27 15:40:45 2021\n",
      "12h QTI Mon Sep 27 15:40:49 2021\n",
      "16h RPF Mon Sep 27 15:40:51 2021\n",
      "16h QTI Mon Sep 27 15:40:55 2021\n",
      "24h RPF Mon Sep 27 15:40:56 2021\n",
      "24h QTI Mon Sep 27 15:40:59 2021\n",
      "36h RPF Mon Sep 27 15:41:00 2021\n",
      "36h QTI Mon Sep 27 15:41:02 2021\n",
      "48h RPF Mon Sep 27 15:41:03 2021\n",
      "48h QTI Mon Sep 27 15:41:11 2021\n",
      "48hCaco2 RPF Mon Sep 27 15:41:15 2021\n",
      "48hCaco2 QTI Mon Sep 27 15:41:18 2021\n"
     ]
    }
   ],
   "source": [
    "draw_readhist_multipanel(OutFigname = 'SFig1d_readhist_merged_transposed.pdf',transposed=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Triplet periodicity (Supplementary Figure 1e)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "def calc_triplet_periodicity(samplename,assay_prefix = 'RPFSeq',genome = 'hg38', offset=12, target_rlen = None,\n",
    "                             offset_from_5p = True, verbose=False, return_codon1pos1 = False):\n",
    "    \n",
    "    if genome == 'SARSCOV2':\n",
    "        BAMDIR = BAMDIR_cov2mapped\n",
    "    else:\n",
    "        BAMDIR = BAMDIR_hostmapped\n",
    "        \n",
    "    if genome == 'hg38':\n",
    "        refdir = REFFLAT_hg38 \n",
    "    elif genome == 'SARSCOV2':\n",
    "        refdir = REFFLAT_SARSCOV2 \n",
    "    elif genome == 'chlSab2':\n",
    "        refdir = REFFLAT_chlSab2\n",
    "    else:\n",
    "        return None\n",
    "    \n",
    "    chrdict = get_dict_refgenes(refdir)\n",
    "    \n",
    "    bam = pysam.AlignmentFile(BAMDIR %(assay_prefix,samplename), 'rb')\n",
    "    triplet_counts = np.zeros(3,dtype=np.float32)\n",
    "    codon1pos1 = 0.0\n",
    "    for sChrID_strand,refgene_list in chrdict.items():\n",
    "        if verbose:\n",
    "            print(sChrID_strand, time.ctime())\n",
    "        sChrID     = sChrID_strand[:-1]\n",
    "        sStrandDir = sChrID_strand[-1]\n",
    "        gene_negstrand = (sStrandDir == '-')\n",
    "        \n",
    "        for refgene in refgene_list:\n",
    "            if refgene.sGeneSym in [\"5'UTR\", \"3'UTR\"]: #filter out those UTR annotations in SARSCOV2 refflat\n",
    "                continue\n",
    "            len_ORF = refgene.nExonlen - refgene.nU5len - refgene.nU3len\n",
    "            if len_ORF%3 != 0:\n",
    "                continue\n",
    "            exon_s_list = refgene.nExonStartlist\n",
    "            exon_e_list = refgene.nExonEndlist\n",
    "            arr_RPF     = np.zeros(refgene.nExonlen,dtype=np.float32)\n",
    "            offset_arr  = 0\n",
    "            for exon_s, exon_e in zip(exon_s_list,exon_e_list):\n",
    "                for read in bam.fetch(sChrID, exon_s,exon_e):\n",
    "                    if target_rlen != None:\n",
    "                        rlen = read.infer_read_length()\n",
    "                        if rlen!=target_rlen:\n",
    "                            continue\n",
    "                    if read.is_reverse != gene_negstrand:\n",
    "                        continue\n",
    "                    reference_positions = read.get_reference_positions(full_length=False)    \n",
    "                    if None in reference_positions:\n",
    "                        print('None in read reference position?')\n",
    "                        print(reference_positions)\n",
    "                        print(read.cigarstring)\n",
    "                        return None\n",
    "                    if gene_negstrand:\n",
    "                        if offset_from_5p == True:\n",
    "                            refpos   = reference_positions[-(offset+1)]\n",
    "                        else:\n",
    "                            refpos   = reference_positions[offset]\n",
    "                    else:\n",
    "                        if offset_from_5p == True:\n",
    "                            refpos   = reference_positions[offset]\n",
    "                        else:\n",
    "                            refpos   = reference_positions[-(offset+1)]\n",
    "                    #print(ribo_5p,read.infer_query_length(),read.is_reverse,reference_positions)\n",
    "                    #return None\n",
    "                    \n",
    "                    nh = read.get_tag('NH')\n",
    "                    if (exon_s <= refpos) and (refpos < exon_e):\n",
    "                        arr_RPF[refpos-exon_s+offset_arr] += (1/nh)\n",
    "                offset_arr += (exon_e - exon_s)\n",
    "                \n",
    "            assert offset_arr == len(arr_RPF)\n",
    "            if sChrID == '-':\n",
    "                arr_RPF = arr_RPF[::-1]\n",
    "            arr_RPF_ORF = arr_RPF[refgene.nU5len:]\n",
    "            if refgene.nU3len>0: \n",
    "                arr_RPF_ORF = arr_RPF_ORF[:-refgene.nU3len]\n",
    "            codon1pos1 += arr_RPF_ORF[0]\n",
    "            \n",
    "            to_add = arr_RPF_ORF.reshape((len_ORF//3,3)).sum(axis=0)\n",
    "            triplet_counts += to_add\n",
    "            #print(refgene.sGeneSym,to_add/to_add.sum())\n",
    "\n",
    "    triplet_periodicity = triplet_counts/(triplet_counts.sum())\n",
    "    \n",
    "    if verbose:\n",
    "        print(samplename,genome, triplet_counts.sum(), triplet_periodicity, sep = '\\t')\n",
    "    bam.close()\n",
    "    if return_codon1pos1:\n",
    "        return triplet_counts, triplet_periodicity, codon1pos1\n",
    "    else:\n",
    "        return triplet_counts, triplet_periodicity\n",
    "\n",
    "\n",
    "def calc_plot_periodicity(assay_prefix = 'RPFSeq', sample_prefix='RPF',load_precalc=False,\n",
    "                          OutF='031521_RPF_triplet.merged.tsv', OutFigname=''):\n",
    "    hpi_list = ['%dh' %i for i in [4,12,16,24,36,48]]+['48hCaco2']\n",
    "    df_merged  = []\n",
    "    if load_precalc:\n",
    "        df_merged = pd.read_csv(RESULTDIR %(OutF),sep='\\t')\n",
    "    else:\n",
    "        for hpi in hpi_list:\n",
    "            print(hpi,time.ctime())\n",
    "            for genome in ['hg38','SARSCOV2']:\n",
    "                merged_counts = np.zeros(3,dtype=np.float32)\n",
    "                reps = 3 if (hpi == '48h') else 2\n",
    "                for rep in range(1, 1+reps):\n",
    "                    samplename = f'{sample_prefix}_{hpi}_rep{rep}'\n",
    "                    triplet_counts, triplet_periodicity = calc_triplet_periodicity(samplename, assay_prefix = assay_prefix, \n",
    "                                                                                   genome = genome, offset=12, target_rlen = None,\n",
    "                                                                                   offset_from_5p = True, verbose=False, \n",
    "                                                                                   return_codon1pos1 = False)\n",
    "                    merged_counts += triplet_counts\n",
    "\n",
    "                df_merged.append([hpi,genome,*merged_counts,*(merged_counts/merged_counts.sum()*100)])\n",
    "\n",
    "        df_merged  = pd.DataFrame(df_merged)\n",
    "        df_merged.columns = ['hpi','genome','pos1','pos2','pos3','pct1','pct2','pct3']\n",
    "        df_merged.to_csv(RESULTDIR %(OutF), sep='\\t',index=False)   \n",
    "    ###########\n",
    "    df_merged = df_merged.set_index('genome')\n",
    "    \n",
    "    df_merged = df_merged[df_merged['hpi'].isin(hpi_list)]\n",
    "    df_merged.index = df_merged.index.str.replace('hg38','host')\n",
    "    df_merged.index = df_merged.index.str.replace('chlSab2','host')\n",
    "    fig = plt.figure(figsize=(len(hpi_list)+2,4))\n",
    "    ax = fig.add_subplot(111)\n",
    "    x = np.arange(len(hpi_list))\n",
    "    bar_width=0.12\n",
    "    gap_hg38_SARSCOV2 = 0.06\n",
    "    xpos_offset = -bar_width*2.5 - gap_hg38_SARSCOV2*0.5 \n",
    "    colors = {}\n",
    "    \n",
    "    colors['host']     = ['#25ad62','#74c476','#bae4b3']\n",
    "    colors['SARSCOV2'] = ['#ff6666','#fc8d59','#fdcc8a']\n",
    "    \n",
    "    for genome in ['host','SARSCOV2']:\n",
    "        for i in range(3):\n",
    "            pct = f'pct{i+1}'\n",
    "            val = df_merged.loc[genome,pct]\n",
    "            ax.bar(x+xpos_offset,val,width = bar_width,label = f'{genome} nt {i+1}', color= colors[genome][i])\n",
    "            xpos_offset += bar_width\n",
    "        xpos_offset += gap_hg38_SARSCOV2\n",
    "    \n",
    "    \n",
    "    ax.set_xticks(x)\n",
    "    ax.set_xticklabels(hpi_list)\n",
    "    ax.set_xlabel('hpi')\n",
    "    ax.set_xlim((-1,len(hpi_list)+1))\n",
    "    ax.set_ylabel('Fraction of reads (%)')\n",
    "    ax.set_ylim(top=70)\n",
    "    ax.legend(loc='upper right')\n",
    "    \n",
    "    plt.tight_layout()\n",
    "    if OutFigname == '':\n",
    "        plt.show()\n",
    "    else:\n",
    "        plt.savefig(FIGDIR %OutFigname)\n",
    "        \n",
    "    plt.close()\n",
    "    \n",
    "    return None\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "4h Mon Sep 27 15:41:21 2021\n",
      "12h Mon Sep 27 15:43:28 2021\n",
      "16h Mon Sep 27 15:44:29 2021\n",
      "24h Mon Sep 27 15:45:28 2021\n",
      "36h Mon Sep 27 15:46:21 2021\n",
      "48h Mon Sep 27 15:47:11 2021\n",
      "48hCaco2 Mon Sep 27 15:49:47 2021\n",
      "4h Mon Sep 27 15:50:51 2021\n",
      "12h Mon Sep 27 15:51:52 2021\n",
      "16h Mon Sep 27 15:52:43 2021\n",
      "24h Mon Sep 27 15:53:22 2021\n",
      "36h Mon Sep 27 15:53:57 2021\n",
      "48h Mon Sep 27 15:54:37 2021\n",
      "48hCaco2 Mon Sep 27 15:56:12 2021\n"
     ]
    }
   ],
   "source": [
    "calc_plot_periodicity(assay_prefix = 'RPFSeq', sample_prefix='RPF',load_precalc=False,\n",
    "                          OutF='RPF_triplet.merged.tsv', OutFigname='SFig1e_triplet_RPF.pdf')\n",
    "calc_plot_periodicity(assay_prefix = 'RPFSeq', sample_prefix='QTI',load_precalc=False,\n",
    "                          OutF='QTI_triplet.merged.tsv', OutFigname='SFig1e_triplet_QTI.pdf')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Figure 2b- Quantification of viral ORF expression at mRNA-, RPF-, QTI-seq level\n",
    "- mRNA-seq: RPM of reads starting from the very first genomic location\n",
    "    - ORF10: aggregated all junctions within 100nt region upstream of ORF10\n",
    "- RPF-seq, QTI-seq, and TE level: read __Methods__ section of the paper"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Plotting function"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "def plot_timecourse_exp(exp_df, hpi_list =[], log10=True, ylim=None, psc = 0.0, grey_exceptS=False,\n",
    "                        ORFs_to_highlight=[], include_3UTR=False, OutFigname = ''):\n",
    "    \n",
    "    global ORF_palette\n",
    "    plt.style.use('default')\n",
    "    \n",
    "    fig = plt.figure(figsize=(4,3))\n",
    "    ORFlist = list(exp_df.index)\n",
    "    \n",
    "    exp_df = exp_df[hpi_list]\n",
    "    if include_3UTR:\n",
    "        assert '3\\'UTR' in ORFlist\n",
    "    else:\n",
    "        if '3\\'UTR' in ORFlist:\n",
    "            ORFlist.remove('3\\'UTR')\n",
    "    x = np.arange(len(exp_df.columns))\n",
    "    for ORF in ORFlist:\n",
    "        exp = exp_df.loc[ORF]\n",
    "        \n",
    "        if log10:\n",
    "            exp = np.log10(exp+psc) #psc: pseudocount\n",
    "        \n",
    "        if ORF in ORFs_to_highlight:\n",
    "            linewidth = 2\n",
    "        else:\n",
    "            linewidth = 1.5\n",
    "            \n",
    "        linestyle = '-'\n",
    "        if ORF == 'S_after': ## for S_after (incorporating RPF signals from TIS-L)\n",
    "            label    = 'S, including\\nTIS-L reads'\n",
    "        else:\n",
    "            if ('S_after' in ORFlist) and (ORF == 'S'):\n",
    "                linestyle='--'\n",
    "            label    = ORF\n",
    "        \n",
    "        if grey_exceptS:\n",
    "            color = '#e8e8e8'\n",
    "            if ORF in ['S','S_after']:\n",
    "                color = ORF_palette['S']\n",
    "        elif ORF == 'ORF10_NmRNA': #ORF10_NmRNA: calculation of ORF10 TE by dividing N mRNA level\n",
    "            linestyle='--'\n",
    "            color = ORF_palette['ORF10']\n",
    "        else:\n",
    "            color = ORF_palette[ORF]\n",
    "        plt.plot(x,exp,label=label,color=color,linewidth=linewidth,linestyle=linestyle)\n",
    "        plt.text(x[-1],exp[-1],ORF)\n",
    "    if ylim != None:\n",
    "        plt.ylim(*ylim)\n",
    "    plt.xlim(x.min(),x.max())\n",
    "    plt.legend(loc='upper left')\n",
    "    plt.xlabel('hpi')\n",
    "    \n",
    "    plt.xticks(x,exp_df.columns.str.replace('h',''))\n",
    "    \n",
    "    if log10:\n",
    "        plt.ylabel('$\\mathregular{log_{10}}$ (RPM+1)')\n",
    "    \n",
    "    if OutFigname != '':\n",
    "        plt.savefig(FIGDIR %OutFigname)\n",
    "        \n",
    "    else:\n",
    "        plt.show()\n",
    "    plt.close()\n",
    "    return None"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### mRNA-seq quantification"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "def calc_timecourse_mRNAexp(hpi_list, assay_prefix='mRNASeq', as_RPM=True, ORF10_range=(29457,29557),\n",
    "                           OutF = 'mRNA_quantification.tsv'):\n",
    "    global CANONICAL_sgRNA_junc_dict\n",
    "\n",
    "    exp_df = []\n",
    "    for hpi in hpi_list:\n",
    "        arr_f2_5p_list   = []\n",
    "        ORF1a_quant_list = [] #gRNA amount\n",
    "        reps = 3 if hpi == '48h' else 2\n",
    "        \n",
    "        for paired_assay in ['RPF','QTI']:\n",
    "            for rep in range(1,reps+1):\n",
    "                samplename = f'{paired_assay}paired_{hpi}_rep{rep}'\n",
    "                arr_f1_3p, arr_f2_5p, arr_r5p = get_readpos_arr(samplename,assay_prefix,norm_by_nh=True,\n",
    "                                                                pos_5end_range=(0,1), f2_5p_range=None, fetch_range=(0,2),\n",
    "                                                                offset_RPF = 0,junction_spanning_only=False,\n",
    "                                                                min_rlen=80,as_RPM=as_RPM)\n",
    "                arr_f2_5p_list.append(arr_f2_5p)\n",
    "                \n",
    "                ORF1a_quant = arr_r5p.sum() - arr_f2_5p.sum()\n",
    "                ORF1a_quant_list.append(ORF1a_quant)\n",
    "\n",
    "        avg_arr_f2_5p   = np.array(arr_f2_5p_list).mean(axis=0)\n",
    "        sgORFs = ['S','ORF3a','E','M','ORF6','ORF7a','ORF7b','ORF8','N']\n",
    "        result_list = [['ORF1a',np.mean(ORF1a_quant_list)]]\n",
    "        for ORF in sgORFs:\n",
    "            f2_5p = CANONICAL_sgRNA_junc_dict[ORF][1] #TRS junction position at body\n",
    "            result_list.append([ORF,avg_arr_f2_5p[f2_5p]])\n",
    "            \n",
    "        #Manually added ORF10 noncanonical sgRNA quantification\n",
    "        result_list.append(['ORF10',sum([avg_arr_f2_5p[f2_5p] for f2_5p in range(*ORF10_range)]) ] )\n",
    "        result_df = pd.DataFrame(result_list)\n",
    "        result_df.columns = ['ORF',hpi]\n",
    "        result_df = result_df.set_index('ORF')\n",
    "        print(hpi,time.ctime())\n",
    "        exp_df.append(result_df)\n",
    "    exp_df = pd.concat(exp_df,axis=1)\n",
    "    display(exp_df)\n",
    "    exp_df.to_csv(RESULTDIR %(OutF),sep='\\t')\n",
    "    return None\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0h Mon Sep 27 15:58:06 2021\n",
      "1h Mon Sep 27 15:59:18 2021\n",
      "2h Mon Sep 27 16:00:28 2021\n",
      "4h Mon Sep 27 16:01:23 2021\n",
      "12h Mon Sep 27 16:02:03 2021\n",
      "16h Mon Sep 27 16:02:33 2021\n",
      "24h Mon Sep 27 16:03:05 2021\n",
      "36h Mon Sep 27 16:03:39 2021\n",
      "48h Mon Sep 27 16:05:24 2021\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0h</th>\n",
       "      <th>1h</th>\n",
       "      <th>2h</th>\n",
       "      <th>4h</th>\n",
       "      <th>12h</th>\n",
       "      <th>16h</th>\n",
       "      <th>24h</th>\n",
       "      <th>36h</th>\n",
       "      <th>48h</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ORF</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ORF1a</th>\n",
       "      <td>0.107815</td>\n",
       "      <td>0.129749</td>\n",
       "      <td>0.119889</td>\n",
       "      <td>0.910100</td>\n",
       "      <td>16.132398</td>\n",
       "      <td>31.001447</td>\n",
       "      <td>158.753534</td>\n",
       "      <td>338.529870</td>\n",
       "      <td>683.669011</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.418520</td>\n",
       "      <td>9.322995</td>\n",
       "      <td>18.295904</td>\n",
       "      <td>85.856082</td>\n",
       "      <td>97.167328</td>\n",
       "      <td>163.876215</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ORF3a</th>\n",
       "      <td>0.044131</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.048968</td>\n",
       "      <td>2.015062</td>\n",
       "      <td>22.145708</td>\n",
       "      <td>38.519896</td>\n",
       "      <td>255.103649</td>\n",
       "      <td>321.403137</td>\n",
       "      <td>731.816001</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>E</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.322009</td>\n",
       "      <td>2.866600</td>\n",
       "      <td>5.840935</td>\n",
       "      <td>32.823472</td>\n",
       "      <td>47.721645</td>\n",
       "      <td>60.186040</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>M</th>\n",
       "      <td>0.022940</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.055068</td>\n",
       "      <td>7.030216</td>\n",
       "      <td>44.363508</td>\n",
       "      <td>97.459738</td>\n",
       "      <td>595.241464</td>\n",
       "      <td>755.120171</td>\n",
       "      <td>1263.557677</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ORF6</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.291927</td>\n",
       "      <td>2.072134</td>\n",
       "      <td>5.004241</td>\n",
       "      <td>25.800855</td>\n",
       "      <td>52.193750</td>\n",
       "      <td>95.940276</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ORF7a</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.075236</td>\n",
       "      <td>3.718929</td>\n",
       "      <td>29.267856</td>\n",
       "      <td>42.778169</td>\n",
       "      <td>290.442073</td>\n",
       "      <td>444.031032</td>\n",
       "      <td>529.155850</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ORF7b</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.030081</td>\n",
       "      <td>0.327514</td>\n",
       "      <td>0.949644</td>\n",
       "      <td>2.882017</td>\n",
       "      <td>6.472817</td>\n",
       "      <td>9.246202</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ORF8</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.024484</td>\n",
       "      <td>4.240098</td>\n",
       "      <td>27.619384</td>\n",
       "      <td>57.608939</td>\n",
       "      <td>249.959172</td>\n",
       "      <td>396.180132</td>\n",
       "      <td>970.996473</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>N</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.020811</td>\n",
       "      <td>0.079552</td>\n",
       "      <td>13.363990</td>\n",
       "      <td>51.541292</td>\n",
       "      <td>91.409176</td>\n",
       "      <td>505.518065</td>\n",
       "      <td>662.097469</td>\n",
       "      <td>1414.976093</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ORF10</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.046163</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.223243</td>\n",
       "      <td>0.501116</td>\n",
       "      <td>0.625909</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             0h        1h        2h         4h        12h        16h  \\\n",
       "ORF                                                                    \n",
       "ORF1a  0.107815  0.129749  0.119889   0.910100  16.132398  31.001447   \n",
       "S      0.000000  0.000000  0.000000   1.418520   9.322995  18.295904   \n",
       "ORF3a  0.044131  0.000000  0.048968   2.015062  22.145708  38.519896   \n",
       "E      0.000000  0.000000  0.000000   0.322009   2.866600   5.840935   \n",
       "M      0.022940  0.000000  0.055068   7.030216  44.363508  97.459738   \n",
       "ORF6   0.000000  0.000000  0.000000   0.291927   2.072134   5.004241   \n",
       "ORF7a  0.000000  0.000000  0.075236   3.718929  29.267856  42.778169   \n",
       "ORF7b  0.000000  0.000000  0.000000   0.030081   0.327514   0.949644   \n",
       "ORF8   0.000000  0.000000  0.024484   4.240098  27.619384  57.608939   \n",
       "N      0.000000  0.020811  0.079552  13.363990  51.541292  91.409176   \n",
       "ORF10  0.000000  0.000000  0.000000   0.000000   0.046163   0.000000   \n",
       "\n",
       "              24h         36h          48h  \n",
       "ORF                                         \n",
       "ORF1a  158.753534  338.529870   683.669011  \n",
       "S       85.856082   97.167328   163.876215  \n",
       "ORF3a  255.103649  321.403137   731.816001  \n",
       "E       32.823472   47.721645    60.186040  \n",
       "M      595.241464  755.120171  1263.557677  \n",
       "ORF6    25.800855   52.193750    95.940276  \n",
       "ORF7a  290.442073  444.031032   529.155850  \n",
       "ORF7b    2.882017    6.472817     9.246202  \n",
       "ORF8   249.959172  396.180132   970.996473  \n",
       "N      505.518065  662.097469  1414.976093  \n",
       "ORF10    0.223243    0.501116     0.625909  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "'''mRNA-seq quantification'''\n",
    "hpi_list   = ['%dh' %i for i in [0,1,2,4,12,16,24,36]]+['48h']\n",
    "calc_timecourse_mRNAexp(hpi_list, OutF = 'mRNA_quantification.tsv')\n",
    "\n",
    "hpi_list_timecourse = ['%dh' %i for i in [0,1,2,4,12,16,24,36]]\n",
    "mRNAexp_df = pd.read_csv(RESULTDIR %('mRNA_quantification.tsv'), sep = '\\t', index_col=0)\n",
    "plot_timecourse_exp(mRNAexp_df,hpi_list = hpi_list_timecourse,log10=True,ylim=(0,3), psc = 1.0,\n",
    "                    include_3UTR = False, OutFigname='Fig2b_mRNA.pdf')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### RPF-, QTI-seq quantification, and PRF(programmed frameshift) of ORF1ab"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "def measure_PRF_ORF1ab(samplename, assay_prefix = 'RPFSeq', offset_RPF=12, as_RPM=True):\n",
    "    ORF1a_range_nonovl = [265,13467]   #real end idx of ORF1a= 13483 \n",
    "    ORF1b_range_nonovl = [13483,21555] #real start idx of ORF1b= 13467 \n",
    "    \n",
    "    _, _, arr_RPF = get_readpos_arr(samplename,assay_prefix,norm_by_nh=True,\n",
    "                                            pos_5end_range=(0,29903), f2_5p_range=None, fetch_range=(0,29903),\n",
    "                                            offset_RPF = offset_RPF,junction_spanning_only=False,\n",
    "                                            min_rlen=0,as_RPM=as_RPM)\n",
    "    \n",
    "    density_ORF1a = arr_RPF[ORF1a_range_nonovl[0]:ORF1a_range_nonovl[1]].mean() #avg.RPM across the ORF\n",
    "    density_ORF1b = arr_RPF[ORF1b_range_nonovl[0]:ORF1b_range_nonovl[1]].mean()\n",
    "    density_ratio = density_ORF1b/density_ORF1a\n",
    "    return density_ORF1a, density_ORF1b, density_ratio\n",
    "\n",
    "def calc_timecourse_PRF_ORF1ab(hpi_list, assay_prefix = 'RPFSeq',sample_prefix='RPF', offset_RPF = 12,\n",
    "                               as_RPM = True, OutF = 'ORF1ab_PRF.tsv'):\n",
    "    PRF_df = []\n",
    "    for hpi in hpi_list:\n",
    "        reps = 3 if hpi == '48h' else 2\n",
    "        avg_ORF1a, avg_ORF1b, avg_ratio = 0.0, 0.0, 0.0    \n",
    "        for rep in range(1,reps+1):\n",
    "            samplename = f'{sample_prefix}_{hpi}_rep{rep}'\n",
    "            density_ORF1a,density_ORF1b,density_ratio = measure_PRF_ORF1ab(samplename, \n",
    "                                                                           assay_prefix = assay_prefix, \n",
    "                                                                           offset_RPF = offset_RPF, \n",
    "                                                                           as_RPM = as_RPM)\n",
    "            avg_ORF1a += density_ORF1a\n",
    "            avg_ORF1b += density_ORF1b\n",
    "            \n",
    "        avg_ORF1a = (avg_ORF1a*1000)/reps # -> RPKM\n",
    "        avg_ORF1b = (avg_ORF1b*1000)/reps # -> RPKM\n",
    "        avg_ratio = avg_ORF1b/avg_ORF1a\n",
    "        PRF_df.append([hpi,avg_ORF1a,avg_ORF1b,avg_ratio])\n",
    "        print(hpi,time.ctime())\n",
    "        \n",
    "    PRF_df = pd.DataFrame(PRF_df)\n",
    "    PRF_df.columns = ['hpi','density_ORF1a','density_ORF1b','PRFratio']\n",
    "    PRF_df = PRF_df.set_index('hpi')\n",
    "    display(PRF_df)\n",
    "    PRF_df.to_csv(RESULTDIR %(OutF),sep='\\t')\n",
    "    return None\n",
    "\n",
    "def calc_timecourse_RPFexp(hpi_list, assay_prefix = 'RPFSeq',sample_prefix='RPF', offset_RPF = 12,\n",
    "                           quant_range = (0,15), as_RPM = True, df_precalc_ORF1ab_PRFratio = None, \n",
    "                           OutF = 'RPF_quantification.tsv'):\n",
    "    \n",
    "    df_SARSCOV2_annot = get_df_SARSCOV2_annot(exc_UTR = False)\n",
    "    df_SARSCOV2_annot = df_SARSCOV2_annot.drop(index='5\\'UTR')\n",
    "    exp_df = pd.DataFrame(index=df_SARSCOV2_annot.index, columns = hpi_list)\n",
    "    for hpi in hpi_list:\n",
    "        arr_RPF_list  = []\n",
    "        reps = 3 if hpi == '48h' else 2\n",
    "        for rep in range(1,reps+1):\n",
    "            samplename = f'{sample_prefix}_{hpi}_rep{rep}'\n",
    "            _, _, arr_RPF = get_readpos_arr(samplename,assay_prefix,norm_by_nh=True,\n",
    "                                            pos_5end_range=(0,29903), f2_5p_range=None, fetch_range=(0,29903),\n",
    "                                            offset_RPF = offset_RPF,junction_spanning_only=False,\n",
    "                                            min_rlen=0,as_RPM=as_RPM)\n",
    "            arr_RPF_list.append(arr_RPF)\n",
    "        \n",
    "        avg_arr_RPF = np.array(arr_RPF_list).mean(axis=0)\n",
    "        \n",
    "        for ORF,row in df_SARSCOV2_annot.iterrows():\n",
    "            ORFstart = row['ORFstart']\n",
    "            exp = avg_arr_RPF[ORFstart+quant_range[0] : ORFstart+quant_range[1]].sum()\n",
    "            exp_df.at[ORF,hpi] = exp\n",
    "            \n",
    "        if type(df_precalc_ORF1ab_PRFratio) == pd.core.frame.DataFrame: #ORF1b estimation\n",
    "            PRFratio = df_precalc_ORF1ab_PRFratio.at[hpi,'PRFratio']\n",
    "            exp_df.at['ORF1b',hpi] = PRFratio*exp_df.at['ORF1a',hpi]\n",
    "        print(hpi,time.ctime())\n",
    "        \n",
    "    display(exp_df)\n",
    "    exp_df.to_csv(RESULTDIR %(OutF),sep='\\t')        \n",
    "    return None"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ORF1ab PRF ratio\n",
      "0h Mon Sep 27 16:05:35 2021\n",
      "1h Mon Sep 27 16:05:50 2021\n",
      "2h Mon Sep 27 16:06:09 2021\n",
      "4h Mon Sep 27 16:06:29 2021\n",
      "12h Mon Sep 27 16:06:35 2021\n",
      "16h Mon Sep 27 16:06:41 2021\n",
      "24h Mon Sep 27 16:06:47 2021\n",
      "36h Mon Sep 27 16:06:55 2021\n",
      "48h Mon Sep 27 16:07:51 2021\n",
      "48h0008U Mon Sep 27 16:08:09 2021\n",
      "24hVero Mon Sep 27 16:08:42 2021\n",
      "48hCaco2 Mon Sep 27 16:08:51 2021\n",
      "48hCalu3cured Mon Sep 27 16:08:53 2021\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>density_ORF1a</th>\n",
       "      <th>density_ORF1b</th>\n",
       "      <th>PRFratio</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>hpi</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0h</th>\n",
       "      <td>3.852245</td>\n",
       "      <td>3.799826</td>\n",
       "      <td>0.986393</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1h</th>\n",
       "      <td>2.114074</td>\n",
       "      <td>1.428437</td>\n",
       "      <td>0.675680</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2h</th>\n",
       "      <td>1.432021</td>\n",
       "      <td>0.969564</td>\n",
       "      <td>0.677060</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4h</th>\n",
       "      <td>3.758342</td>\n",
       "      <td>0.827316</td>\n",
       "      <td>0.220128</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12h</th>\n",
       "      <td>5.350346</td>\n",
       "      <td>3.793709</td>\n",
       "      <td>0.709059</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16h</th>\n",
       "      <td>32.746510</td>\n",
       "      <td>25.637633</td>\n",
       "      <td>0.782912</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24h</th>\n",
       "      <td>319.988442</td>\n",
       "      <td>216.949719</td>\n",
       "      <td>0.677992</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36h</th>\n",
       "      <td>2061.842526</td>\n",
       "      <td>1290.796059</td>\n",
       "      <td>0.626040</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48h</th>\n",
       "      <td>2513.888935</td>\n",
       "      <td>850.186167</td>\n",
       "      <td>0.338196</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48h0008U</th>\n",
       "      <td>6415.928753</td>\n",
       "      <td>4792.684596</td>\n",
       "      <td>0.746998</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24hVero</th>\n",
       "      <td>10782.642922</td>\n",
       "      <td>7683.328385</td>\n",
       "      <td>0.712564</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48hCaco2</th>\n",
       "      <td>747.096631</td>\n",
       "      <td>498.994181</td>\n",
       "      <td>0.667911</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48hCalu3cured</th>\n",
       "      <td>134.765600</td>\n",
       "      <td>75.962935</td>\n",
       "      <td>0.563667</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "               density_ORF1a  density_ORF1b  PRFratio\n",
       "hpi                                                  \n",
       "0h                  3.852245       3.799826  0.986393\n",
       "1h                  2.114074       1.428437  0.675680\n",
       "2h                  1.432021       0.969564  0.677060\n",
       "4h                  3.758342       0.827316  0.220128\n",
       "12h                 5.350346       3.793709  0.709059\n",
       "16h                32.746510      25.637633  0.782912\n",
       "24h               319.988442     216.949719  0.677992\n",
       "36h              2061.842526    1290.796059  0.626040\n",
       "48h              2513.888935     850.186167  0.338196\n",
       "48h0008U         6415.928753    4792.684596  0.746998\n",
       "24hVero         10782.642922    7683.328385  0.712564\n",
       "48hCaco2          747.096631     498.994181  0.667911\n",
       "48hCalu3cured     134.765600      75.962935  0.563667"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "RPF expression\n",
      "0h Mon Sep 27 16:09:02 2021\n",
      "1h Mon Sep 27 16:09:17 2021\n",
      "2h Mon Sep 27 16:09:36 2021\n",
      "4h Mon Sep 27 16:09:56 2021\n",
      "12h Mon Sep 27 16:10:01 2021\n",
      "16h Mon Sep 27 16:10:07 2021\n",
      "24h Mon Sep 27 16:10:14 2021\n",
      "36h Mon Sep 27 16:10:21 2021\n",
      "48h Mon Sep 27 16:11:16 2021\n",
      "48h0008U Mon Sep 27 16:11:34 2021\n",
      "24hVero Mon Sep 27 16:12:08 2021\n",
      "48hCaco2 Mon Sep 27 16:12:17 2021\n",
      "48hCalu3cured Mon Sep 27 16:12:19 2021\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0h</th>\n",
       "      <th>1h</th>\n",
       "      <th>2h</th>\n",
       "      <th>4h</th>\n",
       "      <th>12h</th>\n",
       "      <th>16h</th>\n",
       "      <th>24h</th>\n",
       "      <th>36h</th>\n",
       "      <th>48h</th>\n",
       "      <th>48h0008U</th>\n",
       "      <th>24hVero</th>\n",
       "      <th>48hCaco2</th>\n",
       "      <th>48hCalu3cured</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ORF</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ORF1a</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2.42076</td>\n",
       "      <td>3.84885</td>\n",
       "      <td>35.3779</td>\n",
       "      <td>105.819</td>\n",
       "      <td>4.90632</td>\n",
       "      <td>108.585</td>\n",
       "      <td>70.5798</td>\n",
       "      <td>23.6646</td>\n",
       "      <td>33.3965</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ORF1b</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1.71646</td>\n",
       "      <td>3.01331</td>\n",
       "      <td>23.9859</td>\n",
       "      <td>66.2468</td>\n",
       "      <td>1.6593</td>\n",
       "      <td>81.1128</td>\n",
       "      <td>50.2927</td>\n",
       "      <td>15.8058</td>\n",
       "      <td>18.8245</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0818989</td>\n",
       "      <td>0</td>\n",
       "      <td>0.578354</td>\n",
       "      <td>3.04492</td>\n",
       "      <td>8.63771</td>\n",
       "      <td>27.9509</td>\n",
       "      <td>35.2136</td>\n",
       "      <td>23.071</td>\n",
       "      <td>10.5037</td>\n",
       "      <td>14.5698</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ORF3a</th>\n",
       "      <td>0</td>\n",
       "      <td>0.101959</td>\n",
       "      <td>0.18881</td>\n",
       "      <td>6.69988</td>\n",
       "      <td>5.4988</td>\n",
       "      <td>20.3412</td>\n",
       "      <td>293.531</td>\n",
       "      <td>834.811</td>\n",
       "      <td>4562.89</td>\n",
       "      <td>891.422</td>\n",
       "      <td>225.611</td>\n",
       "      <td>439.96</td>\n",
       "      <td>313.551</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>E</th>\n",
       "      <td>0.140218</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>5.59895</td>\n",
       "      <td>2.7136</td>\n",
       "      <td>11.388</td>\n",
       "      <td>117.891</td>\n",
       "      <td>389.034</td>\n",
       "      <td>1660.7</td>\n",
       "      <td>148.952</td>\n",
       "      <td>580.207</td>\n",
       "      <td>110.149</td>\n",
       "      <td>72.2494</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>M</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0944049</td>\n",
       "      <td>7.26109</td>\n",
       "      <td>23.5016</td>\n",
       "      <td>80.4882</td>\n",
       "      <td>937.694</td>\n",
       "      <td>1889.31</td>\n",
       "      <td>957.258</td>\n",
       "      <td>69.1741</td>\n",
       "      <td>2138.78</td>\n",
       "      <td>700.772</td>\n",
       "      <td>466.458</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ORF6</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1.2617</td>\n",
       "      <td>2.12792</td>\n",
       "      <td>6.04287</td>\n",
       "      <td>76.4316</td>\n",
       "      <td>401.414</td>\n",
       "      <td>1954.27</td>\n",
       "      <td>102.718</td>\n",
       "      <td>82.857</td>\n",
       "      <td>94.3114</td>\n",
       "      <td>72.2494</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ORF7a</th>\n",
       "      <td>0</td>\n",
       "      <td>0.114928</td>\n",
       "      <td>0.18881</td>\n",
       "      <td>8.90474</td>\n",
       "      <td>7.84796</td>\n",
       "      <td>43.4177</td>\n",
       "      <td>665.603</td>\n",
       "      <td>2230.68</td>\n",
       "      <td>4047.84</td>\n",
       "      <td>1991.04</td>\n",
       "      <td>1924.78</td>\n",
       "      <td>339.878</td>\n",
       "      <td>592.119</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ORF7b</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2.4264</td>\n",
       "      <td>8.463</td>\n",
       "      <td>18.6267</td>\n",
       "      <td>140.374</td>\n",
       "      <td>413.08</td>\n",
       "      <td>1522.53</td>\n",
       "      <td>276.211</td>\n",
       "      <td>239.048</td>\n",
       "      <td>162.699</td>\n",
       "      <td>149.297</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ORF8</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0798371</td>\n",
       "      <td>0.333635</td>\n",
       "      <td>0.29284</td>\n",
       "      <td>2.11378</td>\n",
       "      <td>36.3118</td>\n",
       "      <td>97.863</td>\n",
       "      <td>1212.33</td>\n",
       "      <td>216.888</td>\n",
       "      <td>33.6666</td>\n",
       "      <td>48.9144</td>\n",
       "      <td>255.107</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>N</th>\n",
       "      <td>0.355777</td>\n",
       "      <td>0.906453</td>\n",
       "      <td>2.28728</td>\n",
       "      <td>96.6946</td>\n",
       "      <td>121.429</td>\n",
       "      <td>311.399</td>\n",
       "      <td>4363.38</td>\n",
       "      <td>8149.58</td>\n",
       "      <td>35038.6</td>\n",
       "      <td>1307.06</td>\n",
       "      <td>6902.09</td>\n",
       "      <td>5361.78</td>\n",
       "      <td>5959.11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ORF10</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.771121</td>\n",
       "      <td>0.777988</td>\n",
       "      <td>3.43441</td>\n",
       "      <td>19.1565</td>\n",
       "      <td>6.40157</td>\n",
       "      <td>368.749</td>\n",
       "      <td>80.4239</td>\n",
       "      <td>10.7647</td>\n",
       "      <td>3.49251</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3'UTR</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0944049</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.421339</td>\n",
       "      <td>1.68915</td>\n",
       "      <td>1.37597</td>\n",
       "      <td>310.878</td>\n",
       "      <td>3.36303</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             0h        1h         2h         4h       12h       16h       24h  \\\n",
       "ORF                                                                             \n",
       "ORF1a         0         0          0          0   2.42076   3.84885   35.3779   \n",
       "ORF1b         0         0          0          0   1.71646   3.01331   23.9859   \n",
       "S             0         0          0  0.0818989         0  0.578354   3.04492   \n",
       "ORF3a         0  0.101959    0.18881    6.69988    5.4988   20.3412   293.531   \n",
       "E      0.140218         0          0    5.59895    2.7136    11.388   117.891   \n",
       "M             0         0  0.0944049    7.26109   23.5016   80.4882   937.694   \n",
       "ORF6          0         0          0     1.2617   2.12792   6.04287   76.4316   \n",
       "ORF7a         0  0.114928    0.18881    8.90474   7.84796   43.4177   665.603   \n",
       "ORF7b         0         0          0     2.4264     8.463   18.6267   140.374   \n",
       "ORF8          0         0  0.0798371   0.333635   0.29284   2.11378   36.3118   \n",
       "N      0.355777  0.906453    2.28728    96.6946   121.429   311.399   4363.38   \n",
       "ORF10         0         0          0          0  0.771121  0.777988   3.43441   \n",
       "3'UTR         0         0  0.0944049          0         0         0  0.421339   \n",
       "\n",
       "           36h      48h 48h0008U  24hVero 48hCaco2 48hCalu3cured  \n",
       "ORF                                                               \n",
       "ORF1a  105.819  4.90632  108.585  70.5798  23.6646       33.3965  \n",
       "ORF1b  66.2468   1.6593  81.1128  50.2927  15.8058       18.8245  \n",
       "S      8.63771  27.9509  35.2136   23.071  10.5037       14.5698  \n",
       "ORF3a  834.811  4562.89  891.422  225.611   439.96       313.551  \n",
       "E      389.034   1660.7  148.952  580.207  110.149       72.2494  \n",
       "M      1889.31  957.258  69.1741  2138.78  700.772       466.458  \n",
       "ORF6   401.414  1954.27  102.718   82.857  94.3114       72.2494  \n",
       "ORF7a  2230.68  4047.84  1991.04  1924.78  339.878       592.119  \n",
       "ORF7b   413.08  1522.53  276.211  239.048  162.699       149.297  \n",
       "ORF8    97.863  1212.33  216.888  33.6666  48.9144       255.107  \n",
       "N      8149.58  35038.6  1307.06  6902.09  5361.78       5959.11  \n",
       "ORF10  19.1565  6.40157  368.749  80.4239  10.7647       3.49251  \n",
       "3'UTR  1.68915  1.37597  310.878  3.36303        0             0  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "'''RPF-seq quantification, wothout TIS-L S'''\n",
    "#ORF1ab programmed frameshift ratio calculation\n",
    "hpi_list   = ['%dh' %i for i in [0,1,2,4,12,16,24,36]]+['48h','48h0008U','24hVero','48hCaco2', '48hCalu3cured']\n",
    "print('ORF1ab PRF ratio')\n",
    "calc_timecourse_PRF_ORF1ab(hpi_list, assay_prefix = 'RPFSeq',sample_prefix='RPF', offset_RPF = 12,\n",
    "                           as_RPM = True, OutF = 'RPF_ORF1ab_PRF.tsv')\n",
    "\n",
    "#RPF RPM calculation\n",
    "print('RPF expression')\n",
    "hpi_list   = ['%dh' %i for i in [0,1,2,4,12,16,24,36]]+['48h','48h0008U','24hVero','48hCaco2','48hCalu3cured']\n",
    "df_precalc_ORF1ab_PRFratio = pd.read_csv(RESULTDIR %('RPF_ORF1ab_PRF.tsv'), sep = '\\t',index_col=0)\n",
    "calc_timecourse_RPFexp(hpi_list, assay_prefix = 'RPFSeq',sample_prefix='RPF', offset_RPF = 12,\n",
    "                       quant_range = (0,15), as_RPM = True, \n",
    "                       df_precalc_ORF1ab_PRFratio = df_precalc_ORF1ab_PRFratio, \n",
    "                       OutF = 'RPF_quantification_woTISL_S.tsv')\n",
    "#Plotting\n",
    "RPFexp_df = pd.read_csv(RESULTDIR %('RPF_quantification_woTISL_S.tsv'), sep = '\\t', index_col=0)\n",
    "hpi_list_timecourse   = ['%dh' %i for i in [0,1,2,4,12,16,24,36]]\n",
    "plot_timecourse_exp(RPFexp_df,hpi_list = hpi_list_timecourse, log10=True,\n",
    "                    ylim=(0.0,4.0),psc = 1.0, include_3UTR = True,\n",
    "                    OutFigname='Fig2b_RPF_woTISL_S.pdf')     "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "QTI expression\n",
      "0h Mon Sep 27 16:12:25 2021\n",
      "1h Mon Sep 27 16:12:31 2021\n",
      "2h Mon Sep 27 16:12:40 2021\n",
      "4h Mon Sep 27 16:12:48 2021\n",
      "12h Mon Sep 27 16:12:52 2021\n",
      "16h Mon Sep 27 16:12:55 2021\n",
      "24h Mon Sep 27 16:12:57 2021\n",
      "36h Mon Sep 27 16:13:01 2021\n",
      "48h Mon Sep 27 16:13:29 2021\n",
      "48hCaco2 Mon Sep 27 16:13:33 2021\n",
      "48hCalu3cured Mon Sep 27 16:13:33 2021\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0h</th>\n",
       "      <th>1h</th>\n",
       "      <th>2h</th>\n",
       "      <th>4h</th>\n",
       "      <th>12h</th>\n",
       "      <th>16h</th>\n",
       "      <th>24h</th>\n",
       "      <th>36h</th>\n",
       "      <th>48h</th>\n",
       "      <th>48hCaco2</th>\n",
       "      <th>48hCalu3cured</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ORF</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ORF1a</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.411109</td>\n",
       "      <td>2.0096</td>\n",
       "      <td>9.56013</td>\n",
       "      <td>124.238</td>\n",
       "      <td>71.1906</td>\n",
       "      <td>11.8725</td>\n",
       "      <td>61.1326</td>\n",
       "      <td>154.529</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ORF1b</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0904966</td>\n",
       "      <td>1.42493</td>\n",
       "      <td>7.48474</td>\n",
       "      <td>84.2321</td>\n",
       "      <td>44.5682</td>\n",
       "      <td>4.01522</td>\n",
       "      <td>40.8311</td>\n",
       "      <td>87.1027</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.745068</td>\n",
       "      <td>17.2914</td>\n",
       "      <td>13.14</td>\n",
       "      <td>31.8385</td>\n",
       "      <td>98.2594</td>\n",
       "      <td>40.5256</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ORF3a</th>\n",
       "      <td>0</td>\n",
       "      <td>1.5853</td>\n",
       "      <td>0.795824</td>\n",
       "      <td>2.92383</td>\n",
       "      <td>9.17342</td>\n",
       "      <td>95.8456</td>\n",
       "      <td>702.125</td>\n",
       "      <td>698.253</td>\n",
       "      <td>5640.28</td>\n",
       "      <td>757.216</td>\n",
       "      <td>335.779</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>E</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.198956</td>\n",
       "      <td>1.84999</td>\n",
       "      <td>5.28756</td>\n",
       "      <td>34.8355</td>\n",
       "      <td>316.102</td>\n",
       "      <td>268.594</td>\n",
       "      <td>2979.24</td>\n",
       "      <td>501.326</td>\n",
       "      <td>210.196</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>M</th>\n",
       "      <td>0</td>\n",
       "      <td>0.528433</td>\n",
       "      <td>0</td>\n",
       "      <td>0.822218</td>\n",
       "      <td>28.7267</td>\n",
       "      <td>117.367</td>\n",
       "      <td>873.418</td>\n",
       "      <td>1094.84</td>\n",
       "      <td>1339.11</td>\n",
       "      <td>565.607</td>\n",
       "      <td>450.674</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ORF6</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.205555</td>\n",
       "      <td>4.62083</td>\n",
       "      <td>14.7756</td>\n",
       "      <td>311.767</td>\n",
       "      <td>582.785</td>\n",
       "      <td>2268.38</td>\n",
       "      <td>395.435</td>\n",
       "      <td>150.745</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ORF7a</th>\n",
       "      <td>0</td>\n",
       "      <td>1.05687</td>\n",
       "      <td>0</td>\n",
       "      <td>3.50595</td>\n",
       "      <td>22.4963</td>\n",
       "      <td>121.683</td>\n",
       "      <td>1090.67</td>\n",
       "      <td>1532.49</td>\n",
       "      <td>6772.48</td>\n",
       "      <td>1579.03</td>\n",
       "      <td>976.829</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ORF7b</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.99478</td>\n",
       "      <td>1.64444</td>\n",
       "      <td>16.6071</td>\n",
       "      <td>155.814</td>\n",
       "      <td>1213.37</td>\n",
       "      <td>527.329</td>\n",
       "      <td>2560.76</td>\n",
       "      <td>1227.07</td>\n",
       "      <td>529.942</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ORF8</th>\n",
       "      <td>0</td>\n",
       "      <td>0.264216</td>\n",
       "      <td>0</td>\n",
       "      <td>0.205555</td>\n",
       "      <td>0.669868</td>\n",
       "      <td>9.4225</td>\n",
       "      <td>80.7841</td>\n",
       "      <td>94.4012</td>\n",
       "      <td>1578.42</td>\n",
       "      <td>128.079</td>\n",
       "      <td>217.766</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>N</th>\n",
       "      <td>0.554496</td>\n",
       "      <td>7.0726</td>\n",
       "      <td>1.95977</td>\n",
       "      <td>39.7429</td>\n",
       "      <td>217.21</td>\n",
       "      <td>736.903</td>\n",
       "      <td>7407.3</td>\n",
       "      <td>5087.3</td>\n",
       "      <td>51438.8</td>\n",
       "      <td>25103.1</td>\n",
       "      <td>8657.43</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ORF10</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.81388</td>\n",
       "      <td>12.637</td>\n",
       "      <td>22.665</td>\n",
       "      <td>22.9504</td>\n",
       "      <td>21.8179</td>\n",
       "      <td>4.8985</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3'UTR</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1.48045</td>\n",
       "      <td>3.18017</td>\n",
       "      <td>2.42775</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             0h        1h        2h         4h       12h       16h      24h  \\\n",
       "ORF                                                                           \n",
       "ORF1a         0         0         0   0.411109    2.0096   9.56013  124.238   \n",
       "ORF1b         0         0         0  0.0904966   1.42493   7.48474  84.2321   \n",
       "S             0         0         0          0         0  0.745068  17.2914   \n",
       "ORF3a         0    1.5853  0.795824    2.92383   9.17342   95.8456  702.125   \n",
       "E             0         0  0.198956    1.84999   5.28756   34.8355  316.102   \n",
       "M             0  0.528433         0   0.822218   28.7267   117.367  873.418   \n",
       "ORF6          0         0         0   0.205555   4.62083   14.7756  311.767   \n",
       "ORF7a         0   1.05687         0    3.50595   22.4963   121.683  1090.67   \n",
       "ORF7b         0         0   0.99478    1.64444   16.6071   155.814  1213.37   \n",
       "ORF8          0  0.264216         0   0.205555  0.669868    9.4225  80.7841   \n",
       "N      0.554496    7.0726   1.95977    39.7429    217.21   736.903   7407.3   \n",
       "ORF10         0         0         0          0         0   0.81388   12.637   \n",
       "3'UTR         0         0         0          0         0         0  1.48045   \n",
       "\n",
       "           36h      48h 48hCaco2 48hCalu3cured  \n",
       "ORF                                             \n",
       "ORF1a  71.1906  11.8725  61.1326       154.529  \n",
       "ORF1b  44.5682  4.01522  40.8311       87.1027  \n",
       "S        13.14  31.8385  98.2594       40.5256  \n",
       "ORF3a  698.253  5640.28  757.216       335.779  \n",
       "E      268.594  2979.24  501.326       210.196  \n",
       "M      1094.84  1339.11  565.607       450.674  \n",
       "ORF6   582.785  2268.38  395.435       150.745  \n",
       "ORF7a  1532.49  6772.48  1579.03       976.829  \n",
       "ORF7b  527.329  2560.76  1227.07       529.942  \n",
       "ORF8   94.4012  1578.42  128.079       217.766  \n",
       "N       5087.3  51438.8  25103.1       8657.43  \n",
       "ORF10   22.665  22.9504  21.8179        4.8985  \n",
       "3'UTR  3.18017  2.42775        0             0  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "'''QTI-seq quantification, without TIS-L S*'''\n",
    "#QTI RPM calculation\n",
    "print('QTI expression')\n",
    "hpi_list   = ['%dh' %i for i in [0,1,2,4,12,16,24,36]]+['48h','48hCaco2', '48hCalu3cured']\n",
    "df_precalc_ORF1ab_PRFratio = pd.read_csv(RESULTDIR %('RPF_ORF1ab_PRF.tsv'), sep = '\\t', index_col = 0)\n",
    "calc_timecourse_RPFexp(hpi_list, assay_prefix = 'RPFSeq',sample_prefix='QTI', offset_RPF = 12,\n",
    "                       quant_range = (0,15), as_RPM = True, df_precalc_ORF1ab_PRFratio = df_precalc_ORF1ab_PRFratio, \n",
    "                       OutF = 'QTI_quantification_woTISL_S.tsv')\n",
    "\n",
    "#Plotting\n",
    "RPFexp_df = pd.read_csv(RESULTDIR %('QTI_quantification_woTISL_S.tsv'), sep = '\\t', index_col=0)\n",
    "hpi_list_timecourse   = ['%dh' %i for i in [0,1,2,4,12,16,24,36]]\n",
    "plot_timecourse_exp(RPFexp_df,hpi_list = hpi_list_timecourse, log10=True,ylim=(0.0,4.0), \n",
    "                    psc = 1.0, include_3UTR = True,\n",
    "                    OutFigname='Fig2b_QTI_woTISL_S_3UTR.pdf')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Supplementary Figure 2j: ORF1ab PRF rate"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "'''Plotting ORF1ab PRF rate, calculated above( calc_timecourse_PRF_ORF1ab() )'''\n",
    "def plot_ORF1ab(hpi_list, InF='RPF_ORF1ab_PRF.tsv', OutFigname = ''):\n",
    "    df_precalc_ORF1ab_PRFratio = pd.read_csv(RESULTDIR %(InF), \n",
    "                                             sep = '\\t', index_col = 0)\n",
    "    df_results = df_precalc_ORF1ab_PRFratio.loc[hpi_list]\n",
    "    display(df_results)\n",
    "    fig = plt.figure(figsize=(len(hpi_list)*0.5,3))\n",
    "    ax = fig.add_subplot(111)\n",
    "    x = np.arange(len(hpi_list))\n",
    "    bar_width=0.8\n",
    "    \n",
    "    ax.bar(x,df_results['PRFratio']*100,width=bar_width)\n",
    "    ax.set_xticks(x)\n",
    "    ax.set_xticklabels(hpi_list)\n",
    "    ax.set_xlabel('hpi')\n",
    "    \n",
    "    ax.set_ylabel('ORF1b/ORF1a (%)')\n",
    "    ax.set_ylim(top=100)\n",
    "    \n",
    "    plt.tight_layout()\n",
    "    if OutFigname == '':\n",
    "        plt.show()\n",
    "    else:\n",
    "        plt.savefig(FIGDIR %OutFigname)\n",
    "    plt.close()\n",
    "    return None\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>density_ORF1a</th>\n",
       "      <th>density_ORF1b</th>\n",
       "      <th>PRFratio</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>hpi</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>4h</th>\n",
       "      <td>3.758342</td>\n",
       "      <td>0.827316</td>\n",
       "      <td>0.220128</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12h</th>\n",
       "      <td>5.350346</td>\n",
       "      <td>3.793709</td>\n",
       "      <td>0.709059</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16h</th>\n",
       "      <td>32.746510</td>\n",
       "      <td>25.637633</td>\n",
       "      <td>0.782912</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24h</th>\n",
       "      <td>319.988442</td>\n",
       "      <td>216.949719</td>\n",
       "      <td>0.677992</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36h</th>\n",
       "      <td>2061.842526</td>\n",
       "      <td>1290.796059</td>\n",
       "      <td>0.626040</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48h</th>\n",
       "      <td>2513.888935</td>\n",
       "      <td>850.186167</td>\n",
       "      <td>0.338196</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48hCaco2</th>\n",
       "      <td>747.096631</td>\n",
       "      <td>498.994181</td>\n",
       "      <td>0.667911</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          density_ORF1a  density_ORF1b  PRFratio\n",
       "hpi                                             \n",
       "4h             3.758342       0.827316  0.220128\n",
       "12h            5.350346       3.793709  0.709059\n",
       "16h           32.746510      25.637633  0.782912\n",
       "24h          319.988442     216.949719  0.677992\n",
       "36h         2061.842526    1290.796059  0.626040\n",
       "48h         2513.888935     850.186167  0.338196\n",
       "48hCaco2     747.096631     498.994181  0.667911"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "hpi_list   = ['%dh' %i for i in [4,12,16,24,36,48]]+['48hCaco2']\n",
    "plot_ORF1ab(hpi_list,InF='RPF_ORF1ab_PRF.tsv', \n",
    "            OutFigname = 'SFig2j_ORF1abPRF.pdf')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### TE (Translaion efficiency)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "def calc_timecourse_TE(InF_mRNA, InF_RPF, OutF_TE,hpi_list = [], psc = 1.0):\n",
    "    mRNAexp_df = pd.read_csv(RESULTDIR %(InF_mRNA),sep='\\t', index_col = 0)[hpi_list]\n",
    "    RPFexp_df  = pd.read_csv(RESULTDIR %(InF_RPF), sep='\\t', index_col = 0)[hpi_list]\n",
    "    TE_df      = pd.DataFrame(columns = RPFexp_df.columns)\n",
    "    \n",
    "    for ORF,RPF_row in RPFexp_df.iterrows():\n",
    "        if ORF == 'ORF1b':\n",
    "            mRNA_row = mRNAexp_df.loc['ORF1a']\n",
    "        elif ORF == 'S_after':\n",
    "            mRNA_row = mRNAexp_df.loc['S']\n",
    "        elif ORF == 'ORF7b': #ORF7a mRNA leaky scanning\n",
    "            mRNA_row = mRNAexp_df.loc['ORF7a']\n",
    "        elif ORF == '3\\'UTR': #N mRNA leaky scanning\n",
    "            mRNA_row = mRNAexp_df.loc['N']\n",
    "        elif ORF == 'ORF10':\n",
    "            mRNA_row = mRNAexp_df.loc[ORF]\n",
    "            TE_row = (RPF_row+psc)/(mRNA_row+psc)\n",
    "            TE_df.loc[ORF] = TE_row\n",
    "            \n",
    "            mRNA_row = mRNAexp_df.loc['N']\n",
    "            TE_row = (RPF_row+psc)/(mRNA_row+psc)\n",
    "            TE_df.loc['ORF10_NmRNA'] = TE_row\n",
    "            continue\n",
    "        else:\n",
    "            mRNA_row = mRNAexp_df.loc[ORF]\n",
    "            \n",
    "        TE_row = (RPF_row+psc)/(mRNA_row+psc)\n",
    "        TE_df.loc[ORF] = TE_row\n",
    "            \n",
    "    display(TE_df)\n",
    "    TE_df.to_csv(RESULTDIR %(OutF_TE),sep='\\t') \n",
    "    return None\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "RPF\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0h</th>\n",
       "      <th>1h</th>\n",
       "      <th>2h</th>\n",
       "      <th>4h</th>\n",
       "      <th>12h</th>\n",
       "      <th>16h</th>\n",
       "      <th>24h</th>\n",
       "      <th>36h</th>\n",
       "      <th>48h</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ORF1a</th>\n",
       "      <td>0.902678</td>\n",
       "      <td>0.885152</td>\n",
       "      <td>0.892946</td>\n",
       "      <td>0.523533</td>\n",
       "      <td>0.199666</td>\n",
       "      <td>0.151520</td>\n",
       "      <td>0.227713</td>\n",
       "      <td>0.314608</td>\n",
       "      <td>0.008627</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ORF1b</th>\n",
       "      <td>0.902678</td>\n",
       "      <td>0.885152</td>\n",
       "      <td>0.892946</td>\n",
       "      <td>0.523533</td>\n",
       "      <td>0.158557</td>\n",
       "      <td>0.125410</td>\n",
       "      <td>0.156403</td>\n",
       "      <td>0.198059</td>\n",
       "      <td>0.003884</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S</th>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.447339</td>\n",
       "      <td>0.096871</td>\n",
       "      <td>0.081797</td>\n",
       "      <td>0.046570</td>\n",
       "      <td>0.098176</td>\n",
       "      <td>0.175592</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ORF3a</th>\n",
       "      <td>0.957734</td>\n",
       "      <td>1.101959</td>\n",
       "      <td>1.133313</td>\n",
       "      <td>2.553803</td>\n",
       "      <td>0.280778</td>\n",
       "      <td>0.540013</td>\n",
       "      <td>1.150045</td>\n",
       "      <td>2.592439</td>\n",
       "      <td>6.227877</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>E</th>\n",
       "      <td>1.140218</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>4.991612</td>\n",
       "      <td>0.960431</td>\n",
       "      <td>1.810863</td>\n",
       "      <td>3.515035</td>\n",
       "      <td>8.005347</td>\n",
       "      <td>27.158108</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>M</th>\n",
       "      <td>0.977574</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.037284</td>\n",
       "      <td>1.028751</td>\n",
       "      <td>0.540118</td>\n",
       "      <td>0.827629</td>\n",
       "      <td>1.574352</td>\n",
       "      <td>2.500011</td>\n",
       "      <td>0.757781</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ORF6</th>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.750640</td>\n",
       "      <td>1.018159</td>\n",
       "      <td>1.172983</td>\n",
       "      <td>2.889146</td>\n",
       "      <td>7.565054</td>\n",
       "      <td>20.169854</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ORF7a</th>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.114928</td>\n",
       "      <td>1.105627</td>\n",
       "      <td>2.098938</td>\n",
       "      <td>0.292322</td>\n",
       "      <td>1.014608</td>\n",
       "      <td>2.287256</td>\n",
       "      <td>5.014661</td>\n",
       "      <td>7.637075</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ORF7b</th>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.930028</td>\n",
       "      <td>0.726098</td>\n",
       "      <td>0.312642</td>\n",
       "      <td>0.448322</td>\n",
       "      <td>0.485083</td>\n",
       "      <td>0.930452</td>\n",
       "      <td>2.873734</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ORF8</th>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.054030</td>\n",
       "      <td>0.254506</td>\n",
       "      <td>0.045174</td>\n",
       "      <td>0.053128</td>\n",
       "      <td>0.148677</td>\n",
       "      <td>0.248912</td>\n",
       "      <td>1.248284</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>N</th>\n",
       "      <td>1.355777</td>\n",
       "      <td>1.867586</td>\n",
       "      <td>3.045044</td>\n",
       "      <td>6.801352</td>\n",
       "      <td>2.330148</td>\n",
       "      <td>3.380608</td>\n",
       "      <td>8.616431</td>\n",
       "      <td>12.291682</td>\n",
       "      <td>24.745890</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ORF10</th>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.692968</td>\n",
       "      <td>1.777988</td>\n",
       "      <td>3.625129</td>\n",
       "      <td>13.427712</td>\n",
       "      <td>4.552269</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ORF10_NmRNA</th>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.979613</td>\n",
       "      <td>0.926310</td>\n",
       "      <td>0.069619</td>\n",
       "      <td>0.033709</td>\n",
       "      <td>0.019240</td>\n",
       "      <td>0.008755</td>\n",
       "      <td>0.030398</td>\n",
       "      <td>0.005227</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3'UTR</th>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.979613</td>\n",
       "      <td>1.013759</td>\n",
       "      <td>0.069619</td>\n",
       "      <td>0.019033</td>\n",
       "      <td>0.010821</td>\n",
       "      <td>0.002806</td>\n",
       "      <td>0.004055</td>\n",
       "      <td>0.001678</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                   0h        1h        2h        4h       12h       16h  \\\n",
       "ORF1a        0.902678  0.885152  0.892946  0.523533  0.199666  0.151520   \n",
       "ORF1b        0.902678  0.885152  0.892946  0.523533  0.158557  0.125410   \n",
       "S            1.000000  1.000000  1.000000  0.447339  0.096871  0.081797   \n",
       "ORF3a        0.957734  1.101959  1.133313  2.553803  0.280778  0.540013   \n",
       "E            1.140218  1.000000  1.000000  4.991612  0.960431  1.810863   \n",
       "M            0.977574  1.000000  1.037284  1.028751  0.540118  0.827629   \n",
       "ORF6         1.000000  1.000000  1.000000  1.750640  1.018159  1.172983   \n",
       "ORF7a        1.000000  1.114928  1.105627  2.098938  0.292322  1.014608   \n",
       "ORF7b        1.000000  1.000000  0.930028  0.726098  0.312642  0.448322   \n",
       "ORF8         1.000000  1.000000  1.054030  0.254506  0.045174  0.053128   \n",
       "N            1.355777  1.867586  3.045044  6.801352  2.330148  3.380608   \n",
       "ORF10        1.000000  1.000000  1.000000  1.000000  1.692968  1.777988   \n",
       "ORF10_NmRNA  1.000000  0.979613  0.926310  0.069619  0.033709  0.019240   \n",
       "3'UTR        1.000000  0.979613  1.013759  0.069619  0.019033  0.010821   \n",
       "\n",
       "                  24h        36h        48h  \n",
       "ORF1a        0.227713   0.314608   0.008627  \n",
       "ORF1b        0.156403   0.198059   0.003884  \n",
       "S            0.046570   0.098176   0.175592  \n",
       "ORF3a        1.150045   2.592439   6.227877  \n",
       "E            3.515035   8.005347  27.158108  \n",
       "M            1.574352   2.500011   0.757781  \n",
       "ORF6         2.889146   7.565054  20.169854  \n",
       "ORF7a        2.287256   5.014661   7.637075  \n",
       "ORF7b        0.485083   0.930452   2.873734  \n",
       "ORF8         0.148677   0.248912   1.248284  \n",
       "N            8.616431  12.291682  24.745890  \n",
       "ORF10        3.625129  13.427712   4.552269  \n",
       "ORF10_NmRNA  0.008755   0.030398   0.005227  \n",
       "3'UTR        0.002806   0.004055   0.001678  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "QTI\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0h</th>\n",
       "      <th>1h</th>\n",
       "      <th>2h</th>\n",
       "      <th>4h</th>\n",
       "      <th>12h</th>\n",
       "      <th>16h</th>\n",
       "      <th>24h</th>\n",
       "      <th>36h</th>\n",
       "      <th>48h</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ORF1a</th>\n",
       "      <td>0.902678</td>\n",
       "      <td>0.885152</td>\n",
       "      <td>0.892946</td>\n",
       "      <td>0.738762</td>\n",
       "      <td>0.175667</td>\n",
       "      <td>0.329989</td>\n",
       "      <td>0.783943</td>\n",
       "      <td>0.212619</td>\n",
       "      <td>0.018801</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ORF1b</th>\n",
       "      <td>0.902678</td>\n",
       "      <td>0.885152</td>\n",
       "      <td>0.892946</td>\n",
       "      <td>0.570911</td>\n",
       "      <td>0.141540</td>\n",
       "      <td>0.265136</td>\n",
       "      <td>0.533523</td>\n",
       "      <td>0.134210</td>\n",
       "      <td>0.007325</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S</th>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.413476</td>\n",
       "      <td>0.096871</td>\n",
       "      <td>0.090437</td>\n",
       "      <td>0.210594</td>\n",
       "      <td>0.144040</td>\n",
       "      <td>0.199171</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ORF3a</th>\n",
       "      <td>0.957734</td>\n",
       "      <td>2.585299</td>\n",
       "      <td>1.711991</td>\n",
       "      <td>1.301410</td>\n",
       "      <td>0.439538</td>\n",
       "      <td>2.450553</td>\n",
       "      <td>2.745471</td>\n",
       "      <td>2.168879</td>\n",
       "      <td>7.698090</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>E</th>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.198956</td>\n",
       "      <td>2.155804</td>\n",
       "      <td>1.626120</td>\n",
       "      <td>5.238390</td>\n",
       "      <td>9.375199</td>\n",
       "      <td>5.533343</td>\n",
       "      <td>48.707899</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>M</th>\n",
       "      <td>0.977574</td>\n",
       "      <td>1.528433</td>\n",
       "      <td>0.947806</td>\n",
       "      <td>0.226920</td>\n",
       "      <td>0.655299</td>\n",
       "      <td>1.202189</td>\n",
       "      <td>1.466550</td>\n",
       "      <td>1.449297</td>\n",
       "      <td>1.059745</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ORF6</th>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.933144</td>\n",
       "      <td>1.829618</td>\n",
       "      <td>2.627410</td>\n",
       "      <td>11.670036</td>\n",
       "      <td>10.974695</td>\n",
       "      <td>23.410133</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ORF7a</th>\n",
       "      <td>1.000000</td>\n",
       "      <td>2.056866</td>\n",
       "      <td>0.930028</td>\n",
       "      <td>0.954866</td>\n",
       "      <td>0.776277</td>\n",
       "      <td>2.802386</td>\n",
       "      <td>3.745768</td>\n",
       "      <td>3.445812</td>\n",
       "      <td>12.776390</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ORF7b</th>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.855202</td>\n",
       "      <td>0.560389</td>\n",
       "      <td>0.581709</td>\n",
       "      <td>3.582016</td>\n",
       "      <td>4.166759</td>\n",
       "      <td>1.187172</td>\n",
       "      <td>4.832080</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ORF8</th>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.264216</td>\n",
       "      <td>0.976101</td>\n",
       "      <td>0.230063</td>\n",
       "      <td>0.058347</td>\n",
       "      <td>0.177831</td>\n",
       "      <td>0.325886</td>\n",
       "      <td>0.240196</td>\n",
       "      <td>1.624924</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>N</th>\n",
       "      <td>1.554496</td>\n",
       "      <td>7.908023</td>\n",
       "      <td>2.741663</td>\n",
       "      <td>2.836461</td>\n",
       "      <td>4.153108</td>\n",
       "      <td>7.985171</td>\n",
       "      <td>14.625932</td>\n",
       "      <td>7.673535</td>\n",
       "      <td>36.328168</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ORF10</th>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.955874</td>\n",
       "      <td>1.813880</td>\n",
       "      <td>11.148275</td>\n",
       "      <td>15.764916</td>\n",
       "      <td>14.730471</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ORF10_NmRNA</th>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.979613</td>\n",
       "      <td>0.926310</td>\n",
       "      <td>0.069619</td>\n",
       "      <td>0.019033</td>\n",
       "      <td>0.019629</td>\n",
       "      <td>0.026923</td>\n",
       "      <td>0.035689</td>\n",
       "      <td>0.016914</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3'UTR</th>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.979613</td>\n",
       "      <td>0.926310</td>\n",
       "      <td>0.069619</td>\n",
       "      <td>0.019033</td>\n",
       "      <td>0.010821</td>\n",
       "      <td>0.004897</td>\n",
       "      <td>0.006304</td>\n",
       "      <td>0.002421</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                   0h        1h        2h        4h       12h       16h  \\\n",
       "ORF1a        0.902678  0.885152  0.892946  0.738762  0.175667  0.329989   \n",
       "ORF1b        0.902678  0.885152  0.892946  0.570911  0.141540  0.265136   \n",
       "S            1.000000  1.000000  1.000000  0.413476  0.096871  0.090437   \n",
       "ORF3a        0.957734  2.585299  1.711991  1.301410  0.439538  2.450553   \n",
       "E            1.000000  1.000000  1.198956  2.155804  1.626120  5.238390   \n",
       "M            0.977574  1.528433  0.947806  0.226920  0.655299  1.202189   \n",
       "ORF6         1.000000  1.000000  1.000000  0.933144  1.829618  2.627410   \n",
       "ORF7a        1.000000  2.056866  0.930028  0.954866  0.776277  2.802386   \n",
       "ORF7b        1.000000  1.000000  1.855202  0.560389  0.581709  3.582016   \n",
       "ORF8         1.000000  1.264216  0.976101  0.230063  0.058347  0.177831   \n",
       "N            1.554496  7.908023  2.741663  2.836461  4.153108  7.985171   \n",
       "ORF10        1.000000  1.000000  1.000000  1.000000  0.955874  1.813880   \n",
       "ORF10_NmRNA  1.000000  0.979613  0.926310  0.069619  0.019033  0.019629   \n",
       "3'UTR        1.000000  0.979613  0.926310  0.069619  0.019033  0.010821   \n",
       "\n",
       "                   24h        36h        48h  \n",
       "ORF1a         0.783943   0.212619   0.018801  \n",
       "ORF1b         0.533523   0.134210   0.007325  \n",
       "S             0.210594   0.144040   0.199171  \n",
       "ORF3a         2.745471   2.168879   7.698090  \n",
       "E             9.375199   5.533343  48.707899  \n",
       "M             1.466550   1.449297   1.059745  \n",
       "ORF6         11.670036  10.974695  23.410133  \n",
       "ORF7a         3.745768   3.445812  12.776390  \n",
       "ORF7b         4.166759   1.187172   4.832080  \n",
       "ORF8          0.325886   0.240196   1.624924  \n",
       "N            14.625932   7.673535  36.328168  \n",
       "ORF10        11.148275  15.764916  14.730471  \n",
       "ORF10_NmRNA   0.026923   0.035689   0.016914  \n",
       "3'UTR         0.004897   0.006304   0.002421  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "hpi_list   = ['%dh' %i for i in [0,1,2,4,12,16,24,36]]+['48h']\n",
    "hpi_list_timecourse = ['%dh' %i for i in [0,1,2,4,12,16,24,36]]\n",
    "for assay in ['RPF','QTI']:\n",
    "    print(assay)\n",
    "    calc_timecourse_TE('mRNA_quantification.tsv', \n",
    "                       f'{assay}_quantification_woTISL_S.tsv', \n",
    "                       f'TE_{assay}_quantification_woTISL_S.tsv',\n",
    "                       hpi_list = hpi_list, psc = 1.0)\n",
    "    TE_df = pd.read_csv(RESULTDIR %(f'TE_{assay}_quantification_woTISL_S.tsv'), \n",
    "                        sep = '\\t', index_col=0)\n",
    "    plot_timecourse_exp(TE_df,hpi_list = hpi_list_timecourse, log10=True,ylim=(-2.5,1.5), psc = 0.0,\n",
    "                        OutFigname=f'Fig2b_TE_{assay}_woTISL_S_3UTR.pdf')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Figure 3- Detection of TIS-L\n",
    "- Figure 3b, 3h, Extended Data Figures 3b, 4a: TIS-L fraction\n",
    "- Figure 3c, 3e, Extended Data Figures 3c, 4b-f: TIS-L long read split (for Sukjun)\n",
    "- Figure 3d, 3f, 3h, Extended Data Figures 3d, 4c, 4d: bar plots for TIS-L fraction or expression level\n",
    "- Figure 3g, Supplementary Figure 4e (timecourse expression after consideration of ORF S from TIS-L)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### TIS-L enrichment bar plots (Figure 3b, 3h, Extended Data Figures 3b, 4a, 5c, 6a,b)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "def calc_plot_TISL_enrichment(hpi, assay_prefix = 'RPFSeq',sample_prefix='RPF', offset_RPF = 12,\n",
    "                        quant_range = (58,59), plot_range = (39,100), as_RPM = True, \n",
    "                        figsize = (4,3), OutFigname = ''):\n",
    "    idx_TISL = 58\n",
    "    arr_RPF_list  = []\n",
    "    reps = 3 if hpi == '48h' else 2\n",
    "    for rep in range(1,reps+1):\n",
    "        samplename = f'{sample_prefix}_{hpi}_rep{rep}'\n",
    "        _, _, arr_RPF = get_readpos_arr(samplename,assay_prefix,norm_by_nh=True,\n",
    "                                        pos_5end_range=(0,29903), f2_5p_range=None, fetch_range=(0,29903),\n",
    "                                        offset_RPF = offset_RPF,junction_spanning_only=False,\n",
    "                                        min_rlen=0,as_RPM=as_RPM)\n",
    "\n",
    "        arr_RPF_list.append(arr_RPF)\n",
    "\n",
    "    avg_arr_RPF = np.array(arr_RPF_list).mean(axis=0)\n",
    "    TISL_mapped = avg_arr_RPF[quant_range[0]:quant_range[1]].sum()\n",
    "    total_mapped= avg_arr_RPF.sum()\n",
    "    TISL_pct    = TISL_mapped/total_mapped*100\n",
    "    report = [f'{sample_prefix}_{hpi}', total_mapped, TISL_pct]\n",
    "    print(*report, sep = '\\t')\n",
    "    \n",
    "    fig = plt.figure(figsize=figsize)\n",
    "    arr_window = avg_arr_RPF[plot_range[0]:plot_range[1]]\n",
    "    x   = np.arange(plot_range[0],plot_range[1])\n",
    "    bar_colors = [['darkblue','cornflowerblue','lightskyblue'][(i-idx_TISL)%3] for i in x]\n",
    "    plt.bar(x+1,arr_window,color=bar_colors)\n",
    "    plt.ylabel('RPM')\n",
    "    plt.text(70,arr_window[idx_TISL-plot_range[0]]/2, \n",
    "             '%.01f%% of total\\nSARS-CoV-2\\n%s-seq reads' %(TISL_pct,sample_prefix))\n",
    "    plt.xlim(plot_range[0]+1,plot_range[1])\n",
    "    \n",
    "    plt.tight_layout()\n",
    "    if OutFigname == '':\n",
    "        plt.show()\n",
    "    else:\n",
    "        plt.savefig(FIGDIR %OutFigname)\n",
    "        df_source = pd.DataFrame([arr_window], index=['RPM'],columns = np.arange(plot_range[0],plot_range[1])+1).T\n",
    "        df_source.to_csv(SOURCEDATADIR %(OutFigname.replace('pdf','tsv')), sep = '\\t')\n",
    "    plt.close()\n",
    "    \n",
    "    return None"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "RPF_48h\t340550.6026858094\t21.982331559945944\n",
      "QTI_48h\t421921.61884141556\t10.870565508348424\n",
      "RPF_48h\t340550.6026858094\t21.982331559945944\n",
      "RPF_48hCaco2\t97068.87031917248\t7.116556154314048\n",
      "RPF_16h\t7594.026120825337\t5.2092537432978485\n",
      "QTI_16h\t8150.962477589109\t14.002747645685435\n",
      "RPF_24h\t77153.2697142527\t8.579375504419998\n",
      "QTI_24h\t80499.15902695848\t13.105030149565774\n",
      "RPF_36h\t192586.06863679376\t6.45702079387735\n",
      "QTI_36h\t155506.56815578183\t5.32983150090252\n",
      "RPF_48hCalu3cured\t52736.809839021116\t11.187050845645382\n",
      "QTI_48hCalu3cured\t49326.75882488812\t12.452554050461785\n",
      "QTI_48h\t421921.61884141556\t10.870565508348424\n",
      "QTI_48hCaco2\t166457.8185324419\t6.391328284080052\n",
      "RPF_24h\t77153.2697142527\t8.579375504419998\n",
      "RPF_24hVero\t816914.4666214483\t0.16243046524531893\n"
     ]
    }
   ],
   "source": [
    "'''Fig.3b'''\n",
    "calc_plot_TISL_enrichment('48h', assay_prefix = 'RPFSeq',sample_prefix='RPF', offset_RPF = 12,\n",
    "                        quant_range = (58,59), plot_range = (39,100), as_RPM = True, \n",
    "                        figsize = (4,3), OutFigname = 'Fig3b_RPF48h.pdf')\n",
    "\n",
    "'''SFig.3b'''\n",
    "calc_plot_TISL_enrichment('48h', assay_prefix = 'RPFSeq',sample_prefix='QTI', offset_RPF = 12,\n",
    "                        quant_range = (58,59), plot_range = (39,100), as_RPM = True, \n",
    "                        figsize = (4,3), OutFigname = 'SFig3b_QTI48h.pdf')\n",
    "\n",
    "'''Fig.3h'''\n",
    "calc_plot_TISL_enrichment('48h', assay_prefix = 'RPFSeq',sample_prefix='RPF', offset_RPF = 12,\n",
    "                        quant_range = (58,59), plot_range = (49,80), as_RPM = True, \n",
    "                        figsize = (4,3), OutFigname = 'Fig3h_RPF48h.pdf')\n",
    "calc_plot_TISL_enrichment('48hCaco2', assay_prefix = 'RPFSeq',sample_prefix='RPF', offset_RPF = 12,\n",
    "                        quant_range = (58,59), plot_range = (49,80), as_RPM = True, \n",
    "                        figsize = (4,3), OutFigname = 'Fig3h_RPF48hCaco2.pdf')\n",
    "\n",
    "'''SFig.4a'''\n",
    "for hpi in ['16h','24h','36h']:\n",
    "    for sample_prefix in ['RPF','QTI']:\n",
    "        calc_plot_TISL_enrichment(hpi, assay_prefix = 'RPFSeq',sample_prefix=sample_prefix, offset_RPF = 12,\n",
    "                                quant_range = (58,59), plot_range = (39,100), as_RPM = True, \n",
    "                                figsize = (4,3), OutFigname = f'SFig4a_{sample_prefix}{hpi}.pdf')\n",
    "\n",
    "        \n",
    "'''SFig.5c'''\n",
    "calc_plot_TISL_enrichment('48hCalu3cured', assay_prefix = 'RPFSeq',sample_prefix='RPF', offset_RPF = 12,\n",
    "                        quant_range = (58,59), plot_range = (49,80), as_RPM = True, \n",
    "                        figsize = (4,3), OutFigname = 'SFig5c_RPF48hCalu3cured.pdf')\n",
    "calc_plot_TISL_enrichment('48hCalu3cured', assay_prefix = 'RPFSeq',sample_prefix='QTI', offset_RPF = 12,\n",
    "                        quant_range = (58,59), plot_range = (49,80), as_RPM = True, \n",
    "                        figsize = (4,3), OutFigname = 'SFig5c_QTI48hCalu3cured.pdf')\n",
    "'''SFig.6'''\n",
    "calc_plot_TISL_enrichment('48h', assay_prefix = 'RPFSeq',sample_prefix='QTI', offset_RPF = 12,\n",
    "                        quant_range = (58,59), plot_range = (49,80), as_RPM = True, \n",
    "                        figsize = (4,3), OutFigname = 'SFig6a_QTI48h.pdf')\n",
    "calc_plot_TISL_enrichment('48hCaco2', assay_prefix = 'RPFSeq',sample_prefix='QTI', offset_RPF = 12,\n",
    "                        quant_range = (58,59), plot_range = (49,80), as_RPM = True, \n",
    "                        figsize = (4,3), OutFigname = 'SFig6a_QTI48hCaco2.pdf')\n",
    "\n",
    "calc_plot_TISL_enrichment('24h', assay_prefix = 'RPFSeq',sample_prefix='RPF', offset_RPF = 12,\n",
    "                        quant_range = (58,59), plot_range = (39,100), as_RPM = True, \n",
    "                        figsize = (4,3), OutFigname = 'SFig6b_RPF24hCalu3.pdf')\n",
    "calc_plot_TISL_enrichment('24hVero', assay_prefix = 'RPFSeq',sample_prefix='RPF', offset_RPF = 12,\n",
    "                        quant_range = (58,59), plot_range = (39,100), as_RPM = True, \n",
    "                        figsize = (4,3), OutFigname = 'SFig6b_RPF24hVero.pdf')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### TIS-L long read split (Figure 3c,e, Supplementary Figures 3c,e,f, 4b, 6c)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_TISL_reads(ORF,hpi,assay_prefix='RPFSeq', sample_prefix='RPF', offset_RPF=12,\n",
    "                   pos_5end = 46, window_downstream = 6,min_rlen=[32,31,30,29,28,27]):\n",
    "    #collect TISL reads with matched condition\n",
    "    #pos_5end 46: 58(idx_TISl) - 12 (offset_RPF), (window_downstream = 6; 2 codons)\n",
    "    #pos_5end 39: 58(idx-TISL) - 13 (offset_RPF) - 6 (window_downstream = 12; 4 codons)\n",
    "    assert window_downstream == len(min_rlen)\n",
    "    sgRNA_junc_dict = CANONICAL_sgRNA_junc_dict.copy()\n",
    "    sgRNA_junc_dict['ORF1a'] = [75,75] ##no junction, CUG-translation gRNA ORF end\n",
    "    \n",
    "    reads_dict = {} #(read_5p,f1_3p,f2_5p,read_3p): count\n",
    "    reps = 3 if hpi == '48h' else 2\n",
    "    \n",
    "    ORF_f2_5p = sgRNA_junc_dict[ORF][1]\n",
    "    pos_5end_range = (pos_5end, pos_5end+window_downstream)\n",
    "    for rep in range(1,reps+1):\n",
    "        samplename = f'{sample_prefix}_{hpi}_rep{rep}'\n",
    "        bam    = pysam.AlignmentFile(BAMDIR_cov2mapped %(assay_prefix,samplename))\n",
    "        contig = bam.references[0]\n",
    "        for read in bam.fetch(contig, pos_5end_range[0], pos_5end_range[1]):\n",
    "            cigartuples = read.cigartuples\n",
    "            pos_5end    = read.reference_start\n",
    "            pos_3end    = read.reference_end\n",
    "            rlen = read.infer_read_length()\n",
    "            nh   = read.get_tag('NH')\n",
    "            if not ((pos_5end_range[0]<= pos_5end) and (pos_5end< pos_5end_range[1])):\n",
    "                continue\n",
    "            if rlen<min_rlen[pos_5end-pos_5end_range[0]]:\n",
    "                continue\n",
    "            contain_junction = (3 in [i[0] for i in cigartuples])\n",
    "            if (not contain_junction): #not a junction-spanning read\n",
    "                if ORF == 'ORF1a':\n",
    "                    try:\n",
    "                        reads_dict[(pos_5end,*sgRNA_junc_dict['ORF1a'],pos_3end)] += (1/nh)\n",
    "                    except KeyError:\n",
    "                        reads_dict[(pos_5end,*sgRNA_junc_dict['ORF1a'],pos_3end)] = (1/nh)\n",
    "                    continue\n",
    "                else:\n",
    "                    continue\n",
    "            cur_pos = pos_5end\n",
    "            f1_detected = False\n",
    "            f1_3p = None\n",
    "            f2_5p = None\n",
    "            for idx,cigar_tp in enumerate(cigartuples):\n",
    "                operation, length = cigar_tp\n",
    "                if operation == 3: #skip\n",
    "                    f1_detected = True\n",
    "                    cur_pos += length\n",
    "                else:\n",
    "                    if operation == 4 or operation == 1:    # softclip or insertion\n",
    "                        pass\n",
    "                    elif operation == 0 or operation == 2:  # match/mismatch or deletion\n",
    "                        cur_pos += length\n",
    "                if operation == 0:\n",
    "                    if not f1_detected: #junction 1st frag\n",
    "                        f1_3p = cur_pos\n",
    "\n",
    "                    else: #second fragment\n",
    "                        f2_5p = cur_pos - length\n",
    "                        break\n",
    "\n",
    "            assert f1_3p != None\n",
    "            assert f2_5p != None\n",
    "            if f2_5p == ORF_f2_5p:\n",
    "                try:\n",
    "                    reads_dict[(pos_5end,f1_3p,f2_5p,pos_3end)] += (1/nh)\n",
    "                except KeyError:\n",
    "                    reads_dict[(pos_5end,f1_3p,f2_5p,pos_3end)]  = (1/nh)\n",
    "            \n",
    "        bam.close()\n",
    "                    \n",
    "    reads_df = [[*key,value] for key,value in reads_dict.items()]\n",
    "    \n",
    "    if len(reads_df) == 0:\n",
    "        reads_df = pd.DataFrame(columns = ['read_5p','f1_3p','f2_5p','read_3p','count'])\n",
    "    else:\n",
    "        reads_df = pd.DataFrame(reads_df)\n",
    "        reads_df.columns = ['read_5p','f1_3p','f2_5p','read_3p','count']\n",
    "    reads_df = reads_df.sort_values('count', ascending=False)\n",
    "    return reads_df\n",
    "\n",
    "def plot_topreads_seqalign(ORF,reads_df,ax,n_allreads=0,window_start = 37,window_size=50,\n",
    "                           n_topreads=3, splitspace=11):\n",
    "    global SARSCOV2_genome\n",
    "    idx_TISL      = 58\n",
    "    idx_TRSLeader = 69\n",
    "    len_TRSLeader = 6\n",
    "    sgRNA_junc_dict = CANONICAL_sgRNA_junc_dict.copy()\n",
    "    sgRNA_junc_dict['ORF1a'] = [75,75] ##no junction, CUG-translation gRNA ORF end\n",
    "    \n",
    "    total_n      = reads_df['count'].sum()\n",
    "    junc_s, junc_e = sgRNA_junc_dict[ORF] #= f1_3p, f2_5p for read fragments\n",
    "    \n",
    "    top_reads_df = reads_df.iloc[:n_topreads]\n",
    "    if ORF == 'ORF1a':\n",
    "        seq_annot= (SARSCOV2_genome[window_start:junc_s]+\n",
    "                    SARSCOV2_genome[junc_e:junc_e+window_size-(junc_s-window_start)+splitspace]\n",
    "                   ).replace('T','U')\n",
    "        coordinate = [i+1 for i in range(window_start,junc_s)]\n",
    "        coordinate+= [i+1 for i in range(junc_e,junc_e+window_size-(junc_s-window_start)+splitspace)]\n",
    "    else:\n",
    "        seq_annot= (SARSCOV2_genome[window_start : idx_TRSLeader+5] +\n",
    "                    ' ' + SARSCOV2_genome[junc_e+(idx_TRSLeader-junc_s)-5:\n",
    "                                          junc_e+window_size-(junc_s-window_start)]\n",
    "                   ).replace('T','U')\n",
    "        coordinate = [i+1 for i in range(window_start,idx_TRSLeader+5)]\n",
    "        coordinate+= [-1 for i in range(1)]+[i+1 for i in range(junc_e+(idx_TRSLeader-junc_s)-5,\n",
    "                                                                junc_e+window_size-(junc_s-window_start))]\n",
    "    seq_list = []\n",
    "    n_reads_list = []\n",
    "    \n",
    "    for _,row in top_reads_df.iterrows():\n",
    "        read_s,f1_3p, f2_5p,read_e = row.iloc[:4].astype(int)\n",
    "        n_reads = row['count']\n",
    "        \n",
    "        if ORF == 'ORF1a':\n",
    "            seq = SARSCOV2_genome[read_s:f1_3p] + SARSCOV2_genome[f2_5p:read_e]\n",
    "            seq = seq+(' '*splitspace)\n",
    "            seq = ' '*(read_s-window_start) + seq + ' '*(window_size-(len(seq)-splitspace)-(read_s-window_start))\n",
    "            \n",
    "        else:\n",
    "            seq = SARSCOV2_genome[read_s:f1_3p] +(' '*splitspace)+ SARSCOV2_genome[f2_5p:read_e]\n",
    "            seq = ' '*(read_s-window_start) + seq + ' '*(window_size-(len(seq)-splitspace)-(read_s-window_start))\n",
    "        \n",
    "        seq_list.append(seq.replace('T','U'))\n",
    "        n_reads_list.append(n_reads)\n",
    "        \n",
    "    for ntidx,(nt,coor) in enumerate(zip(seq_annot,coordinate)):\n",
    "        color = 'black'\n",
    "        if (idx_TISL<=(ntidx+window_start) and \n",
    "            (ntidx+window_start)<(idx_TISL+3)): #TISL\n",
    "            color = 'green'\n",
    "        elif ((idx_TRSLeader+splitspace) <=(ntidx+window_start) and \n",
    "              (ntidx+window_start)<(idx_TRSLeader+splitspace+len_TRSLeader) and ORF!='ORF1a'): #TRS-Body\n",
    "                color='red'\n",
    "        elif (idx_TRSLeader <= (ntidx+window_start)   and \n",
    "              (ntidx+window_start)< (idx_TRSLeader+len_TRSLeader) and ORF=='ORF1a'): #TRS-Leader\n",
    "            color='red'\n",
    "            \n",
    "        if nt != ' ':\n",
    "            ax.text(ntidx/len(seq_annot)*0.9,0.1,nt,\n",
    "                     fontsize=12,color=color)\n",
    "        if coor%10 == 0:\n",
    "            ax.text(ntidx/len(seq_annot)*0.9,0,coor,\n",
    "                 fontsize=12,color='k')\n",
    "    \n",
    "    ax.axhline(1/(n_topreads+2),xmin=0,xmax=0.9,linestyle='-',color='k',linewidth=1)\n",
    "    for rowidx,(seq,n_reads) in enumerate(zip(seq_list,n_reads_list)):\n",
    "        for ntidx,nt in enumerate(seq):\n",
    "            color = 'black'\n",
    "            if (idx_TISL<=(ntidx+window_start) and \n",
    "                (ntidx+window_start)<(idx_TISL+3)): #TISL\n",
    "                color = 'green'\n",
    "            elif ((idx_TRSLeader+splitspace) <=(ntidx+window_start) and \n",
    "                  (ntidx+window_start)<(idx_TRSLeader+splitspace+len_TRSLeader) and ORF!='ORF1a'): #TRS-Body\n",
    "                    color='red'\n",
    "            elif (idx_TRSLeader <= (ntidx+window_start)   and \n",
    "                  (ntidx+window_start)< (idx_TRSLeader+len_TRSLeader) and ORF=='ORF1a'): #TRS-Leader\n",
    "                color='red'\n",
    "            if nt != ' ':\n",
    "                ax.text(ntidx/len(seq)*0.9,(n_topreads-rowidx+1)/(n_topreads+2),nt,\n",
    "                        fontsize=12,color=color)\n",
    "            \n",
    "        ax.text(0.9,(n_topreads-rowidx+1)/(n_topreads+2),np.ceil(n_reads).astype(np.int),\n",
    "                 fontsize=12)\n",
    "        \n",
    "        if ORF == 'ORF1a':\n",
    "            ax.text(int(len(seq)/2)/len(seq)*0.9,1/(n_topreads+2),'.',fontsize=12)\n",
    "            ax.text(int(len(seq)/2)/len(seq)*0.9,1/(n_topreads+2)+0.05,'.',fontsize=12)\n",
    "            ax.text(int(len(seq)/2)/len(seq)*0.9,1/(n_topreads+2)+0.1,'.',fontsize=12)\n",
    "            ax.text(int(len(seq)/2)/len(seq)*0.9,1.1,'...',fontsize=15)\n",
    "            ax.axhline(1,linestyle='--',color='r',linewidth=2)\n",
    "        else:\n",
    "            ax.text(int(len(seq)/4)/len(seq)*0.9,1/(n_topreads+2),'.',fontsize=12)\n",
    "            ax.text(int(len(seq)/4)/len(seq)*0.9,1/(n_topreads+2)+0.05,'.',fontsize=12)\n",
    "            ax.text(int(len(seq)/4)/len(seq)*0.9,1/(n_topreads+2)+0.1,'.',fontsize=12)\n",
    "            ax.text(int(len(seq)/4*3)/len(seq)*0.9,1/(n_topreads+2),'.',fontsize=12)\n",
    "            ax.text(int(len(seq)/4*3)/len(seq)*0.9,1/(n_topreads+2)+0.05,'.',fontsize=12)\n",
    "            ax.text(int(len(seq)/4*3)/len(seq)*0.9,1/(n_topreads+2)+0.1,'.',fontsize=12)\n",
    "            \n",
    "        ax.text(0.9,1/(n_topreads+2),'.',fontsize=12)\n",
    "        ax.text(0.9,1/(n_topreads+2)+0.05,'.',fontsize=12)\n",
    "        ax.text(0.9,1/(n_topreads+2)+0.1,'.',fontsize=12)\n",
    "    ax.text(1,0.4,'%s\\n%d (%.01f%%)' %(ORF,np.ceil(total_n).astype(np.int),total_n/n_allreads*100),fontsize=12)\n",
    "    ax.axis('off')\n",
    "    return None\n",
    "\n",
    "def plot_CUGalignments(assay_prefix = 'RPFSeq',sample_prefix = 'RPF',hpi = '36h', \n",
    "                       offset_RPF = 12, pos_5end = 46, window_downstream = 6, \n",
    "                      n_display_sgRNA = 3, n_topreads = 3, splitspace=11, OutFigname=''):\n",
    "    \n",
    "    reps = 3 if hpi == '48h' else 2\n",
    "    min_rlen = [78-i for i in range(pos_5end, pos_5end+window_downstream)]\n",
    "    ORFlist  = ['ORF1a','S','ORF3a','E','M','ORF6','ORF7a','ORF7b','ORF8','N']\n",
    "    \n",
    "    ORF_reads_list = []\n",
    "    for ORF in ORFlist:\n",
    "        reads_df = get_TISL_reads(ORF,hpi,assay_prefix=assay_prefix, \n",
    "                                 sample_prefix=sample_prefix, offset_RPF=offset_RPF, pos_5end = pos_5end,\n",
    "                                 window_downstream = window_downstream, min_rlen=min_rlen)\n",
    "        \n",
    "        ORF_reads_list.append([ORF,reads_df,reads_df['count'].sum()])\n",
    "    total = sum([i[2] for i in ORF_reads_list])\n",
    "    gRNA  = ORF_reads_list[0][2]\n",
    "    \n",
    "    ORF_reads_list_sorted = sorted(ORF_reads_list[1:],key = lambda x: x[2],reverse = True)\n",
    "    ORF_reads_list_sorted = ORF_reads_list_sorted[:n_display_sgRNA]\n",
    "    if not ('S' in [i[0] for i in ORF_reads_list_sorted]):\n",
    "        ORF_reads_list_sorted.append(ORF_reads_list[1])\n",
    "        \n",
    "    ORF1a          = ORF_reads_list[0] \n",
    "    ORF_reads_list_sorted.append(ORF1a)\n",
    "    \n",
    "    if 'U' in hpi: #low RNAs conc.\n",
    "        window_size = 60\n",
    "        window_start= 35\n",
    "        figwidth    = 12\n",
    "    else:\n",
    "        window_size = 40\n",
    "        window_start= 45\n",
    "        figwidth    = 9\n",
    "        \n",
    "    fig,axes = plt.subplots(nrows=len(ORF_reads_list_sorted),ncols=1,figsize=(figwidth,1.8*len(ORF_reads_list_sorted)))\n",
    "    for ax,(ORF,reads_df,totalct) in zip(axes,ORF_reads_list_sorted):\n",
    "        plot_topreads_seqalign(ORF,reads_df,ax,n_allreads=total,\n",
    "                               window_start = window_start,window_size=window_size,\n",
    "                               n_topreads=min(len(reads_df),n_topreads), splitspace=splitspace)\n",
    "    gRNA_int = np.ceil(gRNA).astype(int)\n",
    "    plt.suptitle('sgRNAs: %d (%.01f%%) gRNA: %d (%.01f%%)' %(total-gRNA_int,((total-gRNA)/total*100),\n",
    "                                                             gRNA_int,gRNA/total*100))\n",
    "    plt.tight_layout()\n",
    "    \n",
    "    if OutFigname != '':\n",
    "        plt.savefig(FIGDIR %OutFigname)\n",
    "    else:\n",
    "        plt.show()\n",
    "    plt.close()\n",
    "    return None"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [],
   "source": [
    "'''Fig.3c: RPF48h'''\n",
    "plot_CUGalignments(assay_prefix = 'RPFSeq',sample_prefix = 'RPF',hpi = '48h', \n",
    "                       offset_RPF = 12, pos_5end = 46, window_downstream = 6, \n",
    "                      n_display_sgRNA = 3, n_topreads = 3, OutFigname='Fig3c_RPF48h_split.pdf')\n",
    "\n",
    "'''Fig.3e: RPF48h, 0008U (low RNase I conc.)'''\n",
    "##CAUTION: used 13nt as RPF offset, based on the previous analysis on triplet periodicity\n",
    "plot_CUGalignments(assay_prefix = 'RPFSeq',sample_prefix = 'RPF',hpi = '48h0008U', \n",
    "                       offset_RPF = 13, pos_5end = 39, window_downstream = 12, \n",
    "                      n_display_sgRNA = 3, n_topreads = 3, OutFigname='Fig3e_RPF48h0008U_split.pdf')\n",
    "\n",
    "'''SFig.3c: QTI48h'''\n",
    "plot_CUGalignments(assay_prefix = 'RPFSeq',sample_prefix = 'QTI',hpi = '48h', \n",
    "                       offset_RPF = 12, pos_5end = 46, window_downstream = 6,\n",
    "                      n_display_sgRNA = 3, n_topreads = 3, OutFigname='SFig3c_QTI48h_split.pdf')\n",
    "\n",
    "'''SFig. 3e,f: exact TIS-L reads'''\n",
    "plot_CUGalignments(assay_prefix = 'RPFSeq',sample_prefix = 'RPF',hpi = '48h', \n",
    "                       offset_RPF = 12, pos_5end = 46, window_downstream = 1, \n",
    "                      n_display_sgRNA = 3, n_topreads = 3, OutFigname='SFig3e_RPF48h_exactTIS-L.pdf')\n",
    "##CAUTION: used 13nt as RPF offset, based on the previous analysis on triplet periodicity\n",
    "plot_CUGalignments(assay_prefix = 'RPFSeq',sample_prefix = 'RPF',hpi = '48h0008U', \n",
    "                       offset_RPF = 13, pos_5end = 45, window_downstream = 1, \n",
    "                      n_display_sgRNA = 3, n_topreads = 3, OutFigname='SFig3f_RPF48h0008U_exactTIS-L.pdf')\n",
    "\n",
    "'''SFig.4b: RPF16-36h'''\n",
    "for tmp_hpi in ['16h','24h','36h']:\n",
    "    plot_CUGalignments(assay_prefix = 'RPFSeq',sample_prefix = 'RPF',hpi = tmp_hpi, \n",
    "                           offset_RPF = 12, pos_5end = 46, window_downstream = 6,\n",
    "                          n_display_sgRNA = 3, n_topreads = 3, OutFigname=f'SFig4b_RPF{tmp_hpi}_split.pdf')\n",
    "\n",
    "'''SFig.6c: Vero and Caco2'''\n",
    "plot_CUGalignments(assay_prefix = 'RPFSeq',sample_prefix = 'RPF',hpi = '24hVero', \n",
    "                       offset_RPF = 12, pos_5end = 46, window_downstream = 6,\n",
    "                      n_display_sgRNA = 3, n_topreads = 3, OutFigname='SFig6c_RPF24hVero_split.pdf')\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Bar plots for TIS-L fraction or expression level (Figure 3d,f,h, Supplementary Figures 3d, 4c,d, 5d, 6a,b)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "def calc_TISL_ORF_fraction(hpi='48h',assay_prefix='RPFSeq',sample_prefix='RPF',offset_RPF = 12, \n",
    "                             pos_5end = 46,window_downstream=6,OutF = 'RPF48h_fraction.tsv'):\n",
    "    global CANONICAL_sgRNA_junc_dict\n",
    "    idx_TISL = 58\n",
    "    \n",
    "    pos_5end_range = (pos_5end, pos_5end+window_downstream)\n",
    "    min_rlen = [78-i for i in range(*pos_5end_range)]\n",
    "    \n",
    "    arr_f2_5p_list  = []\n",
    "    arr_RPF_list    = []\n",
    "    reps=3 if hpi == '48h' else 2\n",
    "    for rep in range(1,reps+1):\n",
    "        samplename = f'{sample_prefix}_{hpi}_rep{rep}'\n",
    "        _, arr_f2_5p, arr_RPF = get_readpos_arr(samplename,assay_prefix,norm_by_nh=True,\n",
    "                                                pos_5end_range=pos_5end_range, f2_5p_range=None, fetch_range=pos_5end_range,\n",
    "                                                offset_RPF = offset_RPF,junction_spanning_only=False,\n",
    "                                                min_rlen=min_rlen,as_RPM=False)\n",
    "        arr_f2_5p_list.append(arr_f2_5p)\n",
    "        arr_RPF_list.append(arr_RPF)\n",
    "        \n",
    "    sum_arr_f2_5p = np.array(arr_f2_5p_list).sum(axis=0)\n",
    "    sum_arr_RPF   = np.array(arr_RPF_list).sum(axis=0)\n",
    "    \n",
    "    TISLreads_df  = [['ORF1a',sum_arr_RPF.sum()-sum_arr_f2_5p.sum()]]\n",
    "    ORFs = ['S','ORF3a','E','M','ORF6','ORF7a','ORF7b','ORF8','N']\n",
    "    for ORF in ORFs:\n",
    "        f2_5p = CANONICAL_sgRNA_junc_dict[ORF][1]\n",
    "        TISLreads_df.append([ORF,sum_arr_f2_5p[f2_5p]])\n",
    "        \n",
    "    TISLreads_df = pd.DataFrame(TISLreads_df)\n",
    "    TISLreads_df.columns = ['ORF','n_TISLreads']\n",
    "    TISLreads_df = TISLreads_df.set_index('ORF')\n",
    "    TISLreads_df['Percentage'] = TISLreads_df['n_TISLreads']/(TISLreads_df['n_TISLreads'].sum())*100\n",
    "    \n",
    "    TISLreads_df.to_csv(RESULTDIR %OutF, sep = '\\t')\n",
    "    return None\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Calculation of prerequisite data: relative fraction of each ORF for TIS-L reads\n",
    "\n",
    "#Calu-3\n",
    "tmp_hpi_list = ['%dh' %i for i in [0,1,2,4,12,16,24,36,48]]\n",
    "for tmp_hpi in tmp_hpi_list:\n",
    "    for tmp_sample_prefix in ['RPF','QTI']:\n",
    "        calc_TISL_ORF_fraction(hpi=tmp_hpi,assay_prefix='RPFSeq',sample_prefix=tmp_sample_prefix,offset_RPF = 12, \n",
    "                               pos_5end = 46,window_downstream=6,\n",
    "                               OutF = f'{tmp_sample_prefix}{tmp_hpi}_fraction.tsv')\n",
    "#Vero\n",
    "for tmp_hpi in ['24hVero']:\n",
    "    for tmp_sample_prefix in ['RPF']:\n",
    "        calc_TISL_ORF_fraction(hpi=tmp_hpi,assay_prefix='RPFSeq',sample_prefix=tmp_sample_prefix,offset_RPF = 12, \n",
    "                               pos_5end = 46,window_downstream=6,\n",
    "                               OutF = f'{tmp_sample_prefix}{tmp_hpi}_fraction.tsv')\n",
    "#low Rnase conc.\n",
    "for tmp_hpi in ['48h0008U']:\n",
    "    for tmp_sample_prefix in ['RPF']:\n",
    "        calc_TISL_ORF_fraction(hpi=tmp_hpi,assay_prefix='RPFSeq',sample_prefix=tmp_sample_prefix,offset_RPF = 13, \n",
    "                               pos_5end = 39, window_downstream=12,\n",
    "                               OutF = f'{tmp_sample_prefix}{tmp_hpi}_fraction.tsv')\n",
    "#Caco-2\n",
    "for tmp_hpi in ['48hCaco2']:\n",
    "    for tmp_sample_prefix in ['RPF','QTI']:\n",
    "        calc_TISL_ORF_fraction(hpi=tmp_hpi,assay_prefix='RPFSeq',sample_prefix=tmp_sample_prefix,offset_RPF = 12, \n",
    "                               pos_5end = 46,window_downstream=6,\n",
    "                               OutF = f'{tmp_sample_prefix}{tmp_hpi}_fraction.tsv')\n",
    "#Calu-3, cured sample\n",
    "for tmp_hpi in ['48hCalu3cured']:\n",
    "    for tmp_sample_prefix in ['RPF','QTI']:\n",
    "        calc_TISL_ORF_fraction(hpi=tmp_hpi,assay_prefix='RPFSeq',sample_prefix=tmp_sample_prefix,offset_RPF = 12, \n",
    "                               pos_5end = 46,window_downstream=6,\n",
    "                               OutF = f'{tmp_sample_prefix}{tmp_hpi}_fraction.tsv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [],
   "source": [
    "#For Fig. 3d, 3h, etc (relative fractions)\n",
    "def barplot_fraction(sample_prefix='RPF', hpi_list = ['48h','48h0008U'], label_list=['5U','0.008U'],\n",
    "                    hatch_list=['','///'],ylim=(0,45),rotation_x=0,OutFigname=''):\n",
    "    global ORF_palette\n",
    "    merged_df = []\n",
    "    \n",
    "    for hpi in hpi_list:\n",
    "        TISLreads_df = pd.read_csv(RESULTDIR %(f'{sample_prefix}{hpi}_fraction.tsv'),\n",
    "                                   sep = '\\t', index_col = 0)\n",
    "        pct = TISLreads_df['Percentage']\n",
    "        pct.name = hpi\n",
    "        merged_df.append(pct)\n",
    "        \n",
    "    merged_df = pd.concat(merged_df, axis=1)\n",
    "    colors    = [ORF_palette[ORF] for ORF in merged_df.index]\n",
    "    \n",
    "    fig,ax = plt.subplots(figsize=(6,4))\n",
    "    width  = 0.8/len(hpi_list)\n",
    "    x_offset_list = [-0.4+width*(i+1/2) for i in range(len(hpi_list))]\n",
    "    \n",
    "    for hpi,label,hatch,x_offset in zip(hpi_list,label_list,hatch_list,x_offset_list):\n",
    "        ax.bar(np.arange(len(merged_df))+x_offset, merged_df[hpi], width=width, linewidth=0.5,\n",
    "               edgecolor='k',hatch=hatch,color = colors,label = label)\n",
    "    \n",
    "    ax.set_xticks(np.arange(len(merged_df)))\n",
    "    ax.set_xticklabels(merged_df.index,rotation=rotation_x)\n",
    "    ax.set_xlim(-0.5,len(merged_df)-0.5)\n",
    "    ax.set_ylim(*ylim)\n",
    "    ax.set_ylabel('Relative fraction (%)')\n",
    "    gRNA_text = 'gRNA\\n'\n",
    "    sgRNA_text= 'sgRNA\\n'\n",
    "    for hpi,label in zip(hpi_list,label_list):\n",
    "        gRNA_text  += '%-8s %.01f%%\\n' %(label+':',merged_df.loc['ORF1a',hpi])\n",
    "        sgRNA_text += '%-8s %.01f%%\\n' %(label+':',100-merged_df.loc['ORF1a',hpi])\n",
    "    \n",
    "    ax.text(-0.4,ylim[1]+3,gRNA_text)\n",
    "    ax.text(4,ylim[1]+3,sgRNA_text)\n",
    "    ax.axvline(0.5,color='k',linewidth=1.5,linestyle='--')\n",
    "    \n",
    "    plt.legend()\n",
    "    plt.tight_layout()\n",
    "    if OutFigname != '':\n",
    "        plt.savefig(FIGDIR %OutFigname)\n",
    "    else:\n",
    "        plt.show()\n",
    "    plt.close()\n",
    "    \n",
    "    return None\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [],
   "source": [
    "'''Fig. 3d: RPF 48h (5U vs. 0.008U)'''\n",
    "barplot_fraction(sample_prefix='RPF', hpi_list = ['48h','48h0008U'], label_list=['5U','0.008U'],\n",
    "                 hatch_list=['','///'],ylim=(0,45),rotation_x=90,\n",
    "                 OutFigname='Fig3d_RPF48h_5U0008U.pdf')\n",
    "\n",
    "'''Fig. 3h right panel: Calu vs. Caco'''\n",
    "barplot_fraction(sample_prefix='RPF', hpi_list = ['48h','48hCaco2'], label_list=['Calu-3','Caco-2'],\n",
    "                 hatch_list=['','///'],ylim=(0,45),rotation_x=90,\n",
    "                 OutFigname='Fig3h_CaluCaco.pdf')\n",
    "\n",
    "'''SFig.3d : QTI 48h'''\n",
    "barplot_fraction(sample_prefix='QTI', hpi_list = ['48h'], label_list=['QTI 48h'],\n",
    "                 hatch_list=[''],ylim=(0,45),rotation_x=90,\n",
    "                 OutFigname='SFig3d_QTI48h.pdf')\n",
    "\n",
    "'''SFig. 4c: Calu RPF,QTI 16, 24, 36h'''\n",
    "barplot_fraction(sample_prefix='RPF', hpi_list = ['16h','24h','36h'], label_list=['16 hpi','24 hpi','36 hpi'],\n",
    "                 hatch_list=['','/','///'],ylim=(0,55),rotation_x=90,\n",
    "                 OutFigname='SFig4c_RPF_162436h.pdf')\n",
    "barplot_fraction(sample_prefix='QTI', hpi_list = ['16h','24h','36h'], label_list=['16 hpi','24 hpi','36 hpi'],\n",
    "                 hatch_list=['','/','///'],ylim=(0,55),rotation_x=90,\n",
    "                 OutFigname='SFig4c_QTI_162436h.pdf')\n",
    "\n",
    "'''SFig.5d: before and after treatment of mycoplasma'''\n",
    "barplot_fraction(sample_prefix='RPF', hpi_list = ['48h','48hCalu3cured'], label_list=['previous','Mycoplasma cured'],\n",
    "                 hatch_list=['','/','///'],ylim=(0,60),rotation_x=90,\n",
    "                 OutFigname='SFig5d_RPF48h_mycoplasma_cmp.pdf')\n",
    "barplot_fraction(sample_prefix='QTI', hpi_list = ['48h','48hCalu3cured'], label_list=['previous','Mycoplasma cured'],\n",
    "                 hatch_list=['','/','///'],ylim=(0,60),rotation_x=90,\n",
    "                 OutFigname='SFig5d_QTI48h_mycoplasma_cmp.pdf')\n",
    "\n",
    "'''SFig.6a,b: Calu vs. Caco (QTI),  Calu vs. Vero'''\n",
    "barplot_fraction(sample_prefix='QTI', hpi_list = ['48h','48hCaco2'], label_list=['Calu-3','Caco-2'],\n",
    "                 hatch_list=['','///'],ylim=(0,45),rotation_x=90,\n",
    "                 OutFigname='SFig6a_CaluCaco_QTI.pdf')\n",
    "barplot_fraction(sample_prefix='RPF', hpi_list = ['24h','24hVero'], label_list=['Calu-3','Vero'],\n",
    "                 hatch_list=['','///'],ylim=(0,45),rotation_x=90,\n",
    "                 OutFigname='SFig6b_CaluVero.pdf')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [],
   "source": [
    "#For Fig. 3f, SFig. 4d (comparison b/w annotated ORF translation and ORF from TIS-L for each gene)\n",
    "def calc_barplot_TISL_exp(sample_prefix='RPF', hpi = '36h',offset_RPF = 12,quant_range=(58,64),\n",
    "                          hatch_list=['','///'],ylim=(0,5),rotation_x=0,OutFigname=''):\n",
    "    global ORF_palette\n",
    "    ORF_list = ['ORF1a','S','ORF3a','E','M','ORF6','ORF7a','ORF7b','ORF8','N','ORF10']\n",
    "    \n",
    "    exp_df = pd.read_csv(RESULTDIR %(f'{sample_prefix}_quantification_woTISL_S.tsv'),\n",
    "                         sep = '\\t', index_col = 0)\n",
    "    \n",
    "    merged_df = pd.DataFrame(index=ORF_list)\n",
    "    merged_df['annot_RPM'] = exp_df.loc[ORF_list,hpi]\n",
    "    \n",
    "    '''\n",
    "    Calculate read enrichment at TISL and \n",
    "    distribute them in proportion to the precalculated relative fractions\n",
    "    '''\n",
    "    TISLreads_df = pd.read_csv(RESULTDIR %(f'{sample_prefix}{hpi}_fraction.tsv'),\n",
    "                               sep = '\\t', index_col = 0)\n",
    "    TISL_pct = TISLreads_df['Percentage']\n",
    "    arr_RPF_list  = []\n",
    "    reps = 3 if hpi == '48h' else 2\n",
    "    for rep in range(1,reps+1):\n",
    "        samplename = f'{sample_prefix}_{hpi}_rep{rep}'\n",
    "        _, _, arr_RPF = get_readpos_arr(samplename,'RPFSeq',norm_by_nh=True,\n",
    "                                        pos_5end_range=(0,29903), f2_5p_range=None, fetch_range=quant_range,\n",
    "                                        offset_RPF = offset_RPF,junction_spanning_only=False,\n",
    "                                        min_rlen=0,as_RPM=True)\n",
    "\n",
    "        arr_RPF_list.append(arr_RPF)\n",
    "\n",
    "    avg_arr_RPF = np.array(arr_RPF_list).mean(axis=0)\n",
    "    TISL_mapped = avg_arr_RPF[quant_range[0]:quant_range[1]].sum()\n",
    "    TISL_RPM    = TISL_pct*TISL_mapped/100\n",
    "    merged_df['TISL_RPM'] = TISL_RPM\n",
    "    merged_df   = merged_df.fillna(0.0)\n",
    "    ######################################################    \n",
    "    \n",
    "    display(merged_df)\n",
    "    \n",
    "    colors    = [ORF_palette[ORF] for ORF in merged_df.index]\n",
    "    cond_list = merged_df.columns\n",
    "    label_list= ['Canonical AUG','TIS-L']\n",
    "    df_source = pd.DataFrame(columns = label_list, index = merged_df.index)\n",
    "    \n",
    "    fig,ax = plt.subplots(figsize=(6,4))\n",
    "    width  = 0.8/len(cond_list)\n",
    "    x_offset_list = [-0.4+width*(i+1/2) for i in range(len(cond_list))]\n",
    "    \n",
    "    for cond,label,hatch,x_offset in zip(cond_list,label_list,hatch_list,x_offset_list):\n",
    "        vals = np.log10(merged_df[cond]+1)\n",
    "        ax.bar(np.arange(len(merged_df))+x_offset, vals, width=width, linewidth=0.5,\n",
    "               edgecolor='k',hatch=hatch,color = colors,label = label)\n",
    "        df_source[label] = vals\n",
    "    \n",
    "    ax.set_xticks(np.arange(len(merged_df)))\n",
    "    ax.set_xticklabels(merged_df.index, rotation=rotation_x)\n",
    "    ax.set_xlim(-0.5,len(merged_df)-0.5)\n",
    "    ax.set_ylim(*ylim)\n",
    "    ax.set_ylabel('log10(RPM+1)')\n",
    "    \n",
    "    ax.axvline(0.5,color='k',linewidth=1.5,linestyle='--')\n",
    "    \n",
    "    plt.legend()\n",
    "    plt.tight_layout()\n",
    "    if OutFigname != '':\n",
    "        plt.savefig(FIGDIR %OutFigname)\n",
    "        df_source.to_csv(SOURCEDATADIR %(OutFigname.replace('pdf','tsv')), sep = '\\t')\n",
    "    else:\n",
    "        plt.show()\n",
    "    plt.close()\n",
    "    \n",
    "    return None\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>annot_RPM</th>\n",
       "      <th>TISL_RPM</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ORF1a</th>\n",
       "      <td>105.818750</td>\n",
       "      <td>671.103372</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S</th>\n",
       "      <td>8.637710</td>\n",
       "      <td>964.120337</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ORF3a</th>\n",
       "      <td>834.810537</td>\n",
       "      <td>233.153284</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>E</th>\n",
       "      <td>389.033652</td>\n",
       "      <td>233.153284</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>M</th>\n",
       "      <td>1889.308844</td>\n",
       "      <td>198.495363</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ORF6</th>\n",
       "      <td>401.413564</td>\n",
       "      <td>9767.232169</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ORF7a</th>\n",
       "      <td>2230.679694</td>\n",
       "      <td>2284.272040</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ORF7b</th>\n",
       "      <td>413.079963</td>\n",
       "      <td>28.356480</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ORF8</th>\n",
       "      <td>97.863007</td>\n",
       "      <td>2397.697962</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>N</th>\n",
       "      <td>8149.582970</td>\n",
       "      <td>3223.186616</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ORF10</th>\n",
       "      <td>19.156548</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         annot_RPM     TISL_RPM\n",
       "ORF1a   105.818750   671.103372\n",
       "S         8.637710   964.120337\n",
       "ORF3a   834.810537   233.153284\n",
       "E       389.033652   233.153284\n",
       "M      1889.308844   198.495363\n",
       "ORF6    401.413564  9767.232169\n",
       "ORF7a  2230.679694  2284.272040\n",
       "ORF7b   413.079963    28.356480\n",
       "ORF8     97.863007  2397.697962\n",
       "N      8149.582970  3223.186616\n",
       "ORF10    19.156548     0.000000"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>annot_RPM</th>\n",
       "      <th>TISL_RPM</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ORF1a</th>\n",
       "      <td>71.190628</td>\n",
       "      <td>462.138466</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S</th>\n",
       "      <td>13.140028</td>\n",
       "      <td>145.938463</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ORF3a</th>\n",
       "      <td>698.253459</td>\n",
       "      <td>243.230772</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>E</th>\n",
       "      <td>268.593588</td>\n",
       "      <td>243.230772</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>M</th>\n",
       "      <td>1094.842478</td>\n",
       "      <td>218.907695</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ORF6</th>\n",
       "      <td>582.785177</td>\n",
       "      <td>4961.907746</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ORF7a</th>\n",
       "      <td>1532.493368</td>\n",
       "      <td>1970.169252</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ORF7b</th>\n",
       "      <td>527.328549</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ORF8</th>\n",
       "      <td>94.401192</td>\n",
       "      <td>2043.138483</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>N</th>\n",
       "      <td>5087.301753</td>\n",
       "      <td>1751.261557</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ORF10</th>\n",
       "      <td>22.664962</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         annot_RPM     TISL_RPM\n",
       "ORF1a    71.190628   462.138466\n",
       "S        13.140028   145.938463\n",
       "ORF3a   698.253459   243.230772\n",
       "E       268.593588   243.230772\n",
       "M      1094.842478   218.907695\n",
       "ORF6    582.785177  4961.907746\n",
       "ORF7a  1532.493368  1970.169252\n",
       "ORF7b   527.328549     0.000000\n",
       "ORF8     94.401192  2043.138483\n",
       "N      5087.301753  1751.261557\n",
       "ORF10    22.664962     0.000000"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "'''Fig.3f: RPF'''\n",
    "calc_barplot_TISL_exp(sample_prefix='RPF', hpi = '36h',offset_RPF = 12,quant_range=(58,64),\n",
    "                      hatch_list=['','///'],ylim=(0,5),rotation_x=90,\n",
    "                      OutFigname='Fig3f_RPF_annot_TISL.pdf')\n",
    "'''SFig.4d: QTI'''\n",
    "calc_barplot_TISL_exp(sample_prefix='QTI', hpi = '36h',offset_RPF = 12,quant_range=(58,64),\n",
    "                      hatch_list=['','///'],ylim=(0,5),rotation_x=90,\n",
    "                      OutFigname='SFig4d_QTI_annot_TISL.pdf')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [],
   "source": [
    "def calc_plot_timecourse_exp_TISL_S(sample_prefix='RPF',hpi_list=[], offset_RPF = 12,window_downstream=15,\n",
    "                                    OutF = '', OutFigname=''):\n",
    "    global ORF_palette, CANONICAL_sgRNA_junc_dict\n",
    "    idx_TISL     = 58\n",
    "    idx_f1_3p_S,idx_f2_5p_S  = CANONICAL_sgRNA_junc_dict['S']\n",
    "    #print(idx_f1_3p_S,idx_f2_5p_S)\n",
    "    exp_df = pd.read_csv(RESULTDIR %(f'{sample_prefix}_quantification_woTISL_S.tsv'),\n",
    "                         sep = '\\t', index_col = 0)\n",
    "    \n",
    "    exp_df = exp_df[hpi_list]\n",
    "    \n",
    "    S_after_list = []\n",
    "    \n",
    "    \n",
    "    for hpi in hpi_list:\n",
    "        print(hpi,time.ctime())\n",
    "        '''\n",
    "        Calculate read enrichment at TISL and \n",
    "        distribute them in proportion to the precalculated relative fractions \n",
    "        (used 48h fraction for early time points)\n",
    "        '''\n",
    "        if hpi in ['0h','1h','2h','4h','12h','16h']:\n",
    "            TISLreads_df = pd.read_csv(RESULTDIR %(f'{sample_prefix}48h_fraction.tsv'),\n",
    "                                       sep = '\\t', index_col = 0)\n",
    "        else:\n",
    "            TISLreads_df = pd.read_csv(RESULTDIR %(f'{sample_prefix}{hpi}_fraction.tsv'),\n",
    "                                       sep = '\\t', index_col = 0)\n",
    "        TISL_pct = TISLreads_df.loc['S','Percentage']\n",
    "        \n",
    "        arr_RPF_list = []\n",
    "        reps=3 if hpi == '48h' else 2\n",
    "        for rep in range(1,reps+1):\n",
    "            samplename = f'{sample_prefix}_{hpi}_rep{rep}'\n",
    "            _, _, arr_RPF = get_readpos_arr(samplename,'RPFSeq',norm_by_nh=True,\n",
    "                                            pos_5end_range=(0,29903), f2_5p_range=None, fetch_range=(0,29903),\n",
    "                                            offset_RPF = offset_RPF,junction_spanning_only=False,\n",
    "                                            min_rlen=0,as_RPM=True)\n",
    "\n",
    "            arr_RPF_list.append(arr_RPF)\n",
    "\n",
    "        avg_arr_RPF   = np.array(arr_RPF_list).mean(axis=0)\n",
    "        window_leader = min(window_downstream,idx_f1_3p_S-idx_TISL)\n",
    "        window_body   = window_downstream-window_leader\n",
    "        \n",
    "        S_TISL    = avg_arr_RPF[idx_TISL:idx_TISL+window_leader].sum()*TISL_pct/100\n",
    "        S_TISL   += avg_arr_RPF[idx_f2_5p_S:idx_f2_5p_S+window_body].sum()\n",
    "        \n",
    "        \n",
    "        S_after_list.append(exp_df.loc['S',hpi]+S_TISL)\n",
    "    \n",
    "    exp_df.loc['S_after'] = S_after_list\n",
    "    display(exp_df)\n",
    "    exp_df.to_csv(RESULTDIR %OutF, sep='\\t')\n",
    "    plot_timecourse_exp(exp_df, hpi_list=hpi_list, log10=True,ylim=(0,4),psc = 1.0,grey_exceptS=True,\n",
    "                        ORFs_to_highlight=[], OutFigname = OutFigname)\n",
    "    \n",
    "    return None\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0h Mon Sep 27 16:21:56 2021\n",
      "1h Mon Sep 27 16:22:05 2021\n",
      "2h Mon Sep 27 16:22:20 2021\n",
      "4h Mon Sep 27 16:22:39 2021\n",
      "12h Mon Sep 27 16:22:59 2021\n",
      "16h Mon Sep 27 16:23:05 2021\n",
      "24h Mon Sep 27 16:23:11 2021\n",
      "36h Mon Sep 27 16:23:17 2021\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0h</th>\n",
       "      <th>1h</th>\n",
       "      <th>2h</th>\n",
       "      <th>4h</th>\n",
       "      <th>12h</th>\n",
       "      <th>16h</th>\n",
       "      <th>24h</th>\n",
       "      <th>36h</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ORF</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ORF1a</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>2.420761</td>\n",
       "      <td>3.848847</td>\n",
       "      <td>35.377896</td>\n",
       "      <td>105.818750</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ORF1b</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.716461</td>\n",
       "      <td>3.013308</td>\n",
       "      <td>23.985943</td>\n",
       "      <td>66.246779</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.081899</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.578354</td>\n",
       "      <td>3.044923</td>\n",
       "      <td>8.637710</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ORF3a</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.101959</td>\n",
       "      <td>0.188810</td>\n",
       "      <td>6.699876</td>\n",
       "      <td>5.498802</td>\n",
       "      <td>20.341246</td>\n",
       "      <td>293.530734</td>\n",
       "      <td>834.810537</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>E</th>\n",
       "      <td>0.140218</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>5.598954</td>\n",
       "      <td>2.713601</td>\n",
       "      <td>11.387999</td>\n",
       "      <td>117.890693</td>\n",
       "      <td>389.033652</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>M</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.094405</td>\n",
       "      <td>7.261089</td>\n",
       "      <td>23.501650</td>\n",
       "      <td>80.488151</td>\n",
       "      <td>937.694018</td>\n",
       "      <td>1889.308844</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ORF6</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.261700</td>\n",
       "      <td>2.127921</td>\n",
       "      <td>6.042872</td>\n",
       "      <td>76.431588</td>\n",
       "      <td>401.413564</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ORF7a</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.114928</td>\n",
       "      <td>0.188810</td>\n",
       "      <td>8.904739</td>\n",
       "      <td>7.847964</td>\n",
       "      <td>43.417664</td>\n",
       "      <td>665.602721</td>\n",
       "      <td>2230.679694</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ORF7b</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>2.426403</td>\n",
       "      <td>8.463002</td>\n",
       "      <td>18.626729</td>\n",
       "      <td>140.373620</td>\n",
       "      <td>413.079963</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ORF8</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.079837</td>\n",
       "      <td>0.333635</td>\n",
       "      <td>0.292840</td>\n",
       "      <td>2.113784</td>\n",
       "      <td>36.311806</td>\n",
       "      <td>97.863007</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>N</th>\n",
       "      <td>0.355777</td>\n",
       "      <td>0.906453</td>\n",
       "      <td>2.287283</td>\n",
       "      <td>96.694553</td>\n",
       "      <td>121.429009</td>\n",
       "      <td>311.399222</td>\n",
       "      <td>4363.377817</td>\n",
       "      <td>8149.582970</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ORF10</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.771121</td>\n",
       "      <td>0.777988</td>\n",
       "      <td>3.434412</td>\n",
       "      <td>19.156548</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3'UTR</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.094405</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.421339</td>\n",
       "      <td>1.689145</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S_after</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.002801</td>\n",
       "      <td>0.079981</td>\n",
       "      <td>3.786104</td>\n",
       "      <td>5.498166</td>\n",
       "      <td>19.779416</td>\n",
       "      <td>974.391297</td>\n",
       "      <td>1003.397056</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "               0h        1h        2h         4h         12h         16h  \\\n",
       "ORF                                                                        \n",
       "ORF1a    0.000000  0.000000  0.000000   0.000000    2.420761    3.848847   \n",
       "ORF1b    0.000000  0.000000  0.000000   0.000000    1.716461    3.013308   \n",
       "S        0.000000  0.000000  0.000000   0.081899    0.000000    0.578354   \n",
       "ORF3a    0.000000  0.101959  0.188810   6.699876    5.498802   20.341246   \n",
       "E        0.140218  0.000000  0.000000   5.598954    2.713601   11.387999   \n",
       "M        0.000000  0.000000  0.094405   7.261089   23.501650   80.488151   \n",
       "ORF6     0.000000  0.000000  0.000000   1.261700    2.127921    6.042872   \n",
       "ORF7a    0.000000  0.114928  0.188810   8.904739    7.847964   43.417664   \n",
       "ORF7b    0.000000  0.000000  0.000000   2.426403    8.463002   18.626729   \n",
       "ORF8     0.000000  0.000000  0.079837   0.333635    0.292840    2.113784   \n",
       "N        0.355777  0.906453  2.287283  96.694553  121.429009  311.399222   \n",
       "ORF10    0.000000  0.000000  0.000000   0.000000    0.771121    0.777988   \n",
       "3'UTR    0.000000  0.000000  0.094405   0.000000    0.000000    0.000000   \n",
       "S_after  0.000000  0.002801  0.079981   3.786104    5.498166   19.779416   \n",
       "\n",
       "                 24h          36h  \n",
       "ORF                                \n",
       "ORF1a      35.377896   105.818750  \n",
       "ORF1b      23.985943    66.246779  \n",
       "S           3.044923     8.637710  \n",
       "ORF3a     293.530734   834.810537  \n",
       "E         117.890693   389.033652  \n",
       "M         937.694018  1889.308844  \n",
       "ORF6       76.431588   401.413564  \n",
       "ORF7a     665.602721  2230.679694  \n",
       "ORF7b     140.373620   413.079963  \n",
       "ORF8       36.311806    97.863007  \n",
       "N        4363.377817  8149.582970  \n",
       "ORF10       3.434412    19.156548  \n",
       "3'UTR       0.421339     1.689145  \n",
       "S_after   974.391297  1003.397056  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0h Mon Sep 27 16:23:25 2021\n",
      "1h Mon Sep 27 16:23:31 2021\n",
      "2h Mon Sep 27 16:23:38 2021\n",
      "4h Mon Sep 27 16:23:47 2021\n",
      "12h Mon Sep 27 16:23:55 2021\n",
      "16h Mon Sep 27 16:23:59 2021\n",
      "24h Mon Sep 27 16:24:01 2021\n",
      "36h Mon Sep 27 16:24:04 2021\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0h</th>\n",
       "      <th>1h</th>\n",
       "      <th>2h</th>\n",
       "      <th>4h</th>\n",
       "      <th>12h</th>\n",
       "      <th>16h</th>\n",
       "      <th>24h</th>\n",
       "      <th>36h</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ORF</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ORF1a</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.411109</td>\n",
       "      <td>2.009605</td>\n",
       "      <td>9.560125</td>\n",
       "      <td>124.237608</td>\n",
       "      <td>71.190628</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ORF1b</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.090497</td>\n",
       "      <td>1.424927</td>\n",
       "      <td>7.484736</td>\n",
       "      <td>84.232149</td>\n",
       "      <td>44.568186</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.745068</td>\n",
       "      <td>17.291390</td>\n",
       "      <td>13.140028</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ORF3a</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.585299</td>\n",
       "      <td>0.795824</td>\n",
       "      <td>2.923833</td>\n",
       "      <td>9.173421</td>\n",
       "      <td>95.845618</td>\n",
       "      <td>702.125246</td>\n",
       "      <td>698.253459</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>E</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.198956</td>\n",
       "      <td>1.849991</td>\n",
       "      <td>5.287556</td>\n",
       "      <td>34.835483</td>\n",
       "      <td>316.101780</td>\n",
       "      <td>268.593588</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>M</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.528433</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.822218</td>\n",
       "      <td>28.726653</td>\n",
       "      <td>117.367168</td>\n",
       "      <td>873.417625</td>\n",
       "      <td>1094.842478</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ORF6</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.205555</td>\n",
       "      <td>4.620833</td>\n",
       "      <td>14.775599</td>\n",
       "      <td>311.766944</td>\n",
       "      <td>582.785177</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ORF7a</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.056866</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>3.505945</td>\n",
       "      <td>22.496250</td>\n",
       "      <td>121.683348</td>\n",
       "      <td>1090.674377</td>\n",
       "      <td>1532.493368</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ORF7b</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.994780</td>\n",
       "      <td>1.644437</td>\n",
       "      <td>16.607071</td>\n",
       "      <td>155.814106</td>\n",
       "      <td>1213.368826</td>\n",
       "      <td>527.328549</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ORF8</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.264216</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.205555</td>\n",
       "      <td>0.669868</td>\n",
       "      <td>9.422500</td>\n",
       "      <td>80.784137</td>\n",
       "      <td>94.401192</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>N</th>\n",
       "      <td>0.554496</td>\n",
       "      <td>7.072600</td>\n",
       "      <td>1.959768</td>\n",
       "      <td>39.742895</td>\n",
       "      <td>217.209642</td>\n",
       "      <td>736.903077</td>\n",
       "      <td>7407.298612</td>\n",
       "      <td>5087.301753</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ORF10</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.813880</td>\n",
       "      <td>12.637045</td>\n",
       "      <td>22.664962</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3'UTR</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.480446</td>\n",
       "      <td>3.180169</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S_after</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.072626</td>\n",
       "      <td>0.016889</td>\n",
       "      <td>0.381628</td>\n",
       "      <td>2.147722</td>\n",
       "      <td>14.540974</td>\n",
       "      <td>26.690357</td>\n",
       "      <td>169.129407</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "               0h        1h        2h         4h         12h         16h  \\\n",
       "ORF                                                                        \n",
       "ORF1a    0.000000  0.000000  0.000000   0.411109    2.009605    9.560125   \n",
       "ORF1b    0.000000  0.000000  0.000000   0.090497    1.424927    7.484736   \n",
       "S        0.000000  0.000000  0.000000   0.000000    0.000000    0.745068   \n",
       "ORF3a    0.000000  1.585299  0.795824   2.923833    9.173421   95.845618   \n",
       "E        0.000000  0.000000  0.198956   1.849991    5.287556   34.835483   \n",
       "M        0.000000  0.528433  0.000000   0.822218   28.726653  117.367168   \n",
       "ORF6     0.000000  0.000000  0.000000   0.205555    4.620833   14.775599   \n",
       "ORF7a    0.000000  1.056866  0.000000   3.505945   22.496250  121.683348   \n",
       "ORF7b    0.000000  0.000000  0.994780   1.644437   16.607071  155.814106   \n",
       "ORF8     0.000000  0.264216  0.000000   0.205555    0.669868    9.422500   \n",
       "N        0.554496  7.072600  1.959768  39.742895  217.209642  736.903077   \n",
       "ORF10    0.000000  0.000000  0.000000   0.000000    0.000000    0.813880   \n",
       "3'UTR    0.000000  0.000000  0.000000   0.000000    0.000000    0.000000   \n",
       "S_after  0.000000  0.072626  0.016889   0.381628    2.147722   14.540974   \n",
       "\n",
       "                 24h          36h  \n",
       "ORF                                \n",
       "ORF1a     124.237608    71.190628  \n",
       "ORF1b      84.232149    44.568186  \n",
       "S          17.291390    13.140028  \n",
       "ORF3a     702.125246   698.253459  \n",
       "E         316.101780   268.593588  \n",
       "M         873.417625  1094.842478  \n",
       "ORF6      311.766944   582.785177  \n",
       "ORF7a    1090.674377  1532.493368  \n",
       "ORF7b    1213.368826   527.328549  \n",
       "ORF8       80.784137    94.401192  \n",
       "N        7407.298612  5087.301753  \n",
       "ORF10      12.637045    22.664962  \n",
       "3'UTR       1.480446     3.180169  \n",
       "S_after    26.690357   169.129407  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "'''Fig.3g, SFig.4e (left): RPF, QTI with TISL_S'''\n",
    "tmp_hpi_list = ['%dh' %i for i in [0,1,2,4,12,16,24,36]]\n",
    "\n",
    "tmp_sample_prefix = 'RPF'\n",
    "calc_plot_timecourse_exp_TISL_S(sample_prefix=tmp_sample_prefix,hpi_list=tmp_hpi_list, offset_RPF = 12,\n",
    "                                window_downstream=15,\n",
    "                                OutF = f'{tmp_sample_prefix}_quantification_TISL_S.tsv',\n",
    "                                OutFigname=f'Fig3g_{tmp_sample_prefix}.pdf')\n",
    "tmp_sample_prefix = 'QTI'\n",
    "calc_plot_timecourse_exp_TISL_S(sample_prefix=tmp_sample_prefix,hpi_list=tmp_hpi_list, offset_RPF = 12,\n",
    "                                window_downstream=15,\n",
    "                                OutF = f'{tmp_sample_prefix}_quantification_TISL_S.tsv',\n",
    "                                OutFigname=f'SFig4e_{tmp_sample_prefix}.pdf')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [],
   "source": [
    "def calc_plot_timecourse_TE_TISL_S(InF_mRNA, InF_RPF, OutF_TE, hpi_list = [], OutFigname=''):\n",
    "    calc_timecourse_TE(InF_mRNA, InF_RPF, OutF_TE, hpi_list = hpi_list, psc = 1.0)\n",
    "    \n",
    "    TE_df = pd.read_csv(RESULTDIR %(OutF_TE), sep = '\\t', index_col=0)\n",
    "    plot_timecourse_exp(TE_df, hpi_list=hpi_list, log10=True,ylim=(-1.5,1.5), psc = 0.0, grey_exceptS=True,\n",
    "                        OutFigname=OutFigname)\n",
    "    \n",
    "    return None"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0h</th>\n",
       "      <th>1h</th>\n",
       "      <th>2h</th>\n",
       "      <th>4h</th>\n",
       "      <th>12h</th>\n",
       "      <th>16h</th>\n",
       "      <th>24h</th>\n",
       "      <th>36h</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ORF1a</th>\n",
       "      <td>0.902678</td>\n",
       "      <td>0.885152</td>\n",
       "      <td>0.892946</td>\n",
       "      <td>0.523533</td>\n",
       "      <td>0.199666</td>\n",
       "      <td>0.151520</td>\n",
       "      <td>0.227713</td>\n",
       "      <td>0.314608</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ORF1b</th>\n",
       "      <td>0.902678</td>\n",
       "      <td>0.885152</td>\n",
       "      <td>0.892946</td>\n",
       "      <td>0.523533</td>\n",
       "      <td>0.158557</td>\n",
       "      <td>0.125410</td>\n",
       "      <td>0.156403</td>\n",
       "      <td>0.198059</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S</th>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.447339</td>\n",
       "      <td>0.096871</td>\n",
       "      <td>0.081797</td>\n",
       "      <td>0.046570</td>\n",
       "      <td>0.098176</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ORF3a</th>\n",
       "      <td>0.957734</td>\n",
       "      <td>1.101959</td>\n",
       "      <td>1.133313</td>\n",
       "      <td>2.553803</td>\n",
       "      <td>0.280778</td>\n",
       "      <td>0.540013</td>\n",
       "      <td>1.150045</td>\n",
       "      <td>2.592439</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>E</th>\n",
       "      <td>1.140218</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>4.991612</td>\n",
       "      <td>0.960431</td>\n",
       "      <td>1.810863</td>\n",
       "      <td>3.515035</td>\n",
       "      <td>8.005347</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>M</th>\n",
       "      <td>0.977574</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.037284</td>\n",
       "      <td>1.028751</td>\n",
       "      <td>0.540118</td>\n",
       "      <td>0.827629</td>\n",
       "      <td>1.574352</td>\n",
       "      <td>2.500011</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ORF6</th>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.750640</td>\n",
       "      <td>1.018159</td>\n",
       "      <td>1.172983</td>\n",
       "      <td>2.889146</td>\n",
       "      <td>7.565054</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ORF7a</th>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.114928</td>\n",
       "      <td>1.105627</td>\n",
       "      <td>2.098938</td>\n",
       "      <td>0.292322</td>\n",
       "      <td>1.014608</td>\n",
       "      <td>2.287256</td>\n",
       "      <td>5.014661</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ORF7b</th>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.930028</td>\n",
       "      <td>0.726098</td>\n",
       "      <td>0.312642</td>\n",
       "      <td>0.448322</td>\n",
       "      <td>0.485083</td>\n",
       "      <td>0.930452</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ORF8</th>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.054030</td>\n",
       "      <td>0.254506</td>\n",
       "      <td>0.045174</td>\n",
       "      <td>0.053128</td>\n",
       "      <td>0.148677</td>\n",
       "      <td>0.248912</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>N</th>\n",
       "      <td>1.355777</td>\n",
       "      <td>1.867586</td>\n",
       "      <td>3.045044</td>\n",
       "      <td>6.801352</td>\n",
       "      <td>2.330148</td>\n",
       "      <td>3.380608</td>\n",
       "      <td>8.616431</td>\n",
       "      <td>12.291682</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ORF10</th>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.692968</td>\n",
       "      <td>1.777988</td>\n",
       "      <td>3.625129</td>\n",
       "      <td>13.427712</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ORF10_NmRNA</th>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.979613</td>\n",
       "      <td>0.926310</td>\n",
       "      <td>0.069619</td>\n",
       "      <td>0.033709</td>\n",
       "      <td>0.019240</td>\n",
       "      <td>0.008755</td>\n",
       "      <td>0.030398</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3'UTR</th>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.979613</td>\n",
       "      <td>1.013759</td>\n",
       "      <td>0.069619</td>\n",
       "      <td>0.019033</td>\n",
       "      <td>0.010821</td>\n",
       "      <td>0.002806</td>\n",
       "      <td>0.004055</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S_after</th>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.002801</td>\n",
       "      <td>1.079981</td>\n",
       "      <td>1.978940</td>\n",
       "      <td>0.629485</td>\n",
       "      <td>1.076882</td>\n",
       "      <td>11.229971</td>\n",
       "      <td>10.231480</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                   0h        1h        2h        4h       12h       16h  \\\n",
       "ORF1a        0.902678  0.885152  0.892946  0.523533  0.199666  0.151520   \n",
       "ORF1b        0.902678  0.885152  0.892946  0.523533  0.158557  0.125410   \n",
       "S            1.000000  1.000000  1.000000  0.447339  0.096871  0.081797   \n",
       "ORF3a        0.957734  1.101959  1.133313  2.553803  0.280778  0.540013   \n",
       "E            1.140218  1.000000  1.000000  4.991612  0.960431  1.810863   \n",
       "M            0.977574  1.000000  1.037284  1.028751  0.540118  0.827629   \n",
       "ORF6         1.000000  1.000000  1.000000  1.750640  1.018159  1.172983   \n",
       "ORF7a        1.000000  1.114928  1.105627  2.098938  0.292322  1.014608   \n",
       "ORF7b        1.000000  1.000000  0.930028  0.726098  0.312642  0.448322   \n",
       "ORF8         1.000000  1.000000  1.054030  0.254506  0.045174  0.053128   \n",
       "N            1.355777  1.867586  3.045044  6.801352  2.330148  3.380608   \n",
       "ORF10        1.000000  1.000000  1.000000  1.000000  1.692968  1.777988   \n",
       "ORF10_NmRNA  1.000000  0.979613  0.926310  0.069619  0.033709  0.019240   \n",
       "3'UTR        1.000000  0.979613  1.013759  0.069619  0.019033  0.010821   \n",
       "S_after      1.000000  1.002801  1.079981  1.978940  0.629485  1.076882   \n",
       "\n",
       "                   24h        36h  \n",
       "ORF1a         0.227713   0.314608  \n",
       "ORF1b         0.156403   0.198059  \n",
       "S             0.046570   0.098176  \n",
       "ORF3a         1.150045   2.592439  \n",
       "E             3.515035   8.005347  \n",
       "M             1.574352   2.500011  \n",
       "ORF6          2.889146   7.565054  \n",
       "ORF7a         2.287256   5.014661  \n",
       "ORF7b         0.485083   0.930452  \n",
       "ORF8          0.148677   0.248912  \n",
       "N             8.616431  12.291682  \n",
       "ORF10         3.625129  13.427712  \n",
       "ORF10_NmRNA   0.008755   0.030398  \n",
       "3'UTR         0.002806   0.004055  \n",
       "S_after      11.229971  10.231480  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0h</th>\n",
       "      <th>1h</th>\n",
       "      <th>2h</th>\n",
       "      <th>4h</th>\n",
       "      <th>12h</th>\n",
       "      <th>16h</th>\n",
       "      <th>24h</th>\n",
       "      <th>36h</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ORF1a</th>\n",
       "      <td>0.902678</td>\n",
       "      <td>0.885152</td>\n",
       "      <td>0.892946</td>\n",
       "      <td>0.738762</td>\n",
       "      <td>0.175667</td>\n",
       "      <td>0.329989</td>\n",
       "      <td>0.783943</td>\n",
       "      <td>0.212619</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ORF1b</th>\n",
       "      <td>0.902678</td>\n",
       "      <td>0.885152</td>\n",
       "      <td>0.892946</td>\n",
       "      <td>0.570911</td>\n",
       "      <td>0.141540</td>\n",
       "      <td>0.265136</td>\n",
       "      <td>0.533523</td>\n",
       "      <td>0.134210</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S</th>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.413476</td>\n",
       "      <td>0.096871</td>\n",
       "      <td>0.090437</td>\n",
       "      <td>0.210594</td>\n",
       "      <td>0.144040</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ORF3a</th>\n",
       "      <td>0.957734</td>\n",
       "      <td>2.585299</td>\n",
       "      <td>1.711991</td>\n",
       "      <td>1.301410</td>\n",
       "      <td>0.439538</td>\n",
       "      <td>2.450553</td>\n",
       "      <td>2.745471</td>\n",
       "      <td>2.168879</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>E</th>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.198956</td>\n",
       "      <td>2.155804</td>\n",
       "      <td>1.626120</td>\n",
       "      <td>5.238390</td>\n",
       "      <td>9.375199</td>\n",
       "      <td>5.533343</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>M</th>\n",
       "      <td>0.977574</td>\n",
       "      <td>1.528433</td>\n",
       "      <td>0.947806</td>\n",
       "      <td>0.226920</td>\n",
       "      <td>0.655299</td>\n",
       "      <td>1.202189</td>\n",
       "      <td>1.466550</td>\n",
       "      <td>1.449297</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ORF6</th>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.933144</td>\n",
       "      <td>1.829618</td>\n",
       "      <td>2.627410</td>\n",
       "      <td>11.670036</td>\n",
       "      <td>10.974695</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ORF7a</th>\n",
       "      <td>1.000000</td>\n",
       "      <td>2.056866</td>\n",
       "      <td>0.930028</td>\n",
       "      <td>0.954866</td>\n",
       "      <td>0.776277</td>\n",
       "      <td>2.802386</td>\n",
       "      <td>3.745768</td>\n",
       "      <td>3.445812</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ORF7b</th>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.855202</td>\n",
       "      <td>0.560389</td>\n",
       "      <td>0.581709</td>\n",
       "      <td>3.582016</td>\n",
       "      <td>4.166759</td>\n",
       "      <td>1.187172</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ORF8</th>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.264216</td>\n",
       "      <td>0.976101</td>\n",
       "      <td>0.230063</td>\n",
       "      <td>0.058347</td>\n",
       "      <td>0.177831</td>\n",
       "      <td>0.325886</td>\n",
       "      <td>0.240196</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>N</th>\n",
       "      <td>1.554496</td>\n",
       "      <td>7.908023</td>\n",
       "      <td>2.741663</td>\n",
       "      <td>2.836461</td>\n",
       "      <td>4.153108</td>\n",
       "      <td>7.985171</td>\n",
       "      <td>14.625932</td>\n",
       "      <td>7.673535</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ORF10</th>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.955874</td>\n",
       "      <td>1.813880</td>\n",
       "      <td>11.148275</td>\n",
       "      <td>15.764916</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ORF10_NmRNA</th>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.979613</td>\n",
       "      <td>0.926310</td>\n",
       "      <td>0.069619</td>\n",
       "      <td>0.019033</td>\n",
       "      <td>0.019629</td>\n",
       "      <td>0.026923</td>\n",
       "      <td>0.035689</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3'UTR</th>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.979613</td>\n",
       "      <td>0.926310</td>\n",
       "      <td>0.069619</td>\n",
       "      <td>0.019033</td>\n",
       "      <td>0.010821</td>\n",
       "      <td>0.004897</td>\n",
       "      <td>0.006304</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S_after</th>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.072626</td>\n",
       "      <td>1.016889</td>\n",
       "      <td>0.571270</td>\n",
       "      <td>0.304923</td>\n",
       "      <td>0.805403</td>\n",
       "      <td>0.318807</td>\n",
       "      <td>1.733055</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                   0h        1h        2h        4h       12h       16h  \\\n",
       "ORF1a        0.902678  0.885152  0.892946  0.738762  0.175667  0.329989   \n",
       "ORF1b        0.902678  0.885152  0.892946  0.570911  0.141540  0.265136   \n",
       "S            1.000000  1.000000  1.000000  0.413476  0.096871  0.090437   \n",
       "ORF3a        0.957734  2.585299  1.711991  1.301410  0.439538  2.450553   \n",
       "E            1.000000  1.000000  1.198956  2.155804  1.626120  5.238390   \n",
       "M            0.977574  1.528433  0.947806  0.226920  0.655299  1.202189   \n",
       "ORF6         1.000000  1.000000  1.000000  0.933144  1.829618  2.627410   \n",
       "ORF7a        1.000000  2.056866  0.930028  0.954866  0.776277  2.802386   \n",
       "ORF7b        1.000000  1.000000  1.855202  0.560389  0.581709  3.582016   \n",
       "ORF8         1.000000  1.264216  0.976101  0.230063  0.058347  0.177831   \n",
       "N            1.554496  7.908023  2.741663  2.836461  4.153108  7.985171   \n",
       "ORF10        1.000000  1.000000  1.000000  1.000000  0.955874  1.813880   \n",
       "ORF10_NmRNA  1.000000  0.979613  0.926310  0.069619  0.019033  0.019629   \n",
       "3'UTR        1.000000  0.979613  0.926310  0.069619  0.019033  0.010821   \n",
       "S_after      1.000000  1.072626  1.016889  0.571270  0.304923  0.805403   \n",
       "\n",
       "                   24h        36h  \n",
       "ORF1a         0.783943   0.212619  \n",
       "ORF1b         0.533523   0.134210  \n",
       "S             0.210594   0.144040  \n",
       "ORF3a         2.745471   2.168879  \n",
       "E             9.375199   5.533343  \n",
       "M             1.466550   1.449297  \n",
       "ORF6         11.670036  10.974695  \n",
       "ORF7a         3.745768   3.445812  \n",
       "ORF7b         4.166759   1.187172  \n",
       "ORF8          0.325886   0.240196  \n",
       "N            14.625932   7.673535  \n",
       "ORF10        11.148275  15.764916  \n",
       "ORF10_NmRNA   0.026923   0.035689  \n",
       "3'UTR         0.004897   0.006304  \n",
       "S_after       0.318807   1.733055  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "'''Fig.3g, SFig.4e (right): RPF, QTI TE with TISL_S'''\n",
    "tmp_hpi_list = ['%dh' %i for i in [0,1,2,4,12,16,24,36]]\n",
    "\n",
    "tmp_sample_prefix = 'RPF'\n",
    "calc_plot_timecourse_TE_TISL_S('mRNA_quantification.tsv',  \n",
    "                               f'{tmp_sample_prefix}_quantification_TISL_S.tsv', \n",
    "                               f'TE_{tmp_sample_prefix}_quantification_TISL_S.tsv',\n",
    "                               hpi_list = tmp_hpi_list,\n",
    "                               OutFigname= f'Fig3g_TE_{tmp_sample_prefix}_TISL_S.pdf')\n",
    "tmp_sample_prefix = 'QTI'\n",
    "calc_plot_timecourse_TE_TISL_S('mRNA_quantification.tsv',  \n",
    "                               f'{tmp_sample_prefix}_quantification_TISL_S.tsv', \n",
    "                               f'TE_{tmp_sample_prefix}_quantification_TISL_S.tsv',\n",
    "                               hpi_list = tmp_hpi_list,\n",
    "                               OutFigname= f'SFig4e_TE_{tmp_sample_prefix}_TISL_S.pdf')\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "toc-hr-collapsed": true
   },
   "source": [
    "### Comparison b/w Calu and Caco or cured Calu annotated SARS-CoV-2 viral ORF expression levels (Supplementary Figure 5a)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [],
   "source": [
    "def calc_barplot_cmp_annot_exp(sample_prefix='RPF', hpi_list = ['48h','48hCaco2'], label_list=['Calu-3','Caco-2'],\n",
    "                    hatch_list=['','///'], ylim=(0,5),as_pct = False,rotation_x=0,OutFigname=''):\n",
    "    global ORF_palette\n",
    "    merged_df = []\n",
    "    exp_df = pd.read_csv(RESULTDIR %(f'{sample_prefix}_quantification_woTISL_S.tsv'),\n",
    "                         sep = '\\t', index_col = 0)\n",
    "    for hpi in hpi_list:\n",
    "        exp = exp_df[hpi]\n",
    "        if as_pct:\n",
    "            exp = exp/exp.sum()*100\n",
    "        merged_df.append(exp)\n",
    "        \n",
    "    merged_df = pd.concat(merged_df, axis=1)\n",
    "    colors    = [ORF_palette[ORF] for ORF in merged_df.index]\n",
    "    \n",
    "    fig,ax = plt.subplots(figsize=(6,4))\n",
    "    width  = 0.8/len(hpi_list)\n",
    "    x_offset_list = [-0.4+width*(i+1/2) for i in range(len(hpi_list))]\n",
    "    \n",
    "    for hpi,label,hatch,x_offset in zip(hpi_list,label_list,hatch_list,x_offset_list):\n",
    "        if as_pct:\n",
    "            ax.bar(np.arange(len(merged_df))+x_offset, merged_df[hpi], width=width, linewidth=0.5,\n",
    "                   edgecolor='k',hatch=hatch,color = colors,label = label)\n",
    "        else:\n",
    "            ax.bar(np.arange(len(merged_df))+x_offset, np.log10(merged_df[hpi]+1), width=width, linewidth=0.5,\n",
    "                   edgecolor='k',hatch=hatch,color = colors,label = label)\n",
    "    \n",
    "    ax.set_xticks(np.arange(len(merged_df)))\n",
    "    ax.set_xticklabels(merged_df.index,rotation=rotation_x)\n",
    "    ax.set_xlim(-0.5,len(merged_df)-0.5)\n",
    "    ax.set_ylim(*ylim)\n",
    "    if as_pct:\n",
    "        ax.set_ylabel('Relative fraction (%)')\n",
    "    else:\n",
    "        ax.set_ylabel('log10(RPM+1)')\n",
    "        \n",
    "    ax.axvline(1.5,color='k',linewidth=1.5,linestyle='--')\n",
    "    plt.legend()\n",
    "    plt.tight_layout()\n",
    "    if OutFigname != '':\n",
    "        plt.savefig(FIGDIR %OutFigname)\n",
    "    else:\n",
    "        plt.show()\n",
    "    plt.close()\n",
    "    \n",
    "    return None\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [],
   "source": [
    "for sample_prefix in ['RPF','QTI']:\n",
    "    calc_barplot_cmp_annot_exp(sample_prefix=sample_prefix, hpi_list = ['48h','48hCalu3cured'], \n",
    "                               label_list=['Before treatment','After treatment'],\n",
    "                               hatch_list=['','///'],ylim=(0,80),rotation_x=90,as_pct = True,\n",
    "                               OutFigname = f'SFig5a_{sample_prefix}48h_mycoplasma_annotORFs_cmp_pct.pdf')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Figure 4- RPF-seq reads mapped on TIS-L and annotated ORFs\n",
    "- Figure 4b-e, Supplementary Figure 7a-f: RPF-seq\n",
    "- Supplementary Figure 8a-j: QTI-seq\n",
    "- Supplementary Figures 7g, 8k: ORF7b\n",
    "- Supplementary Figure 9: ORF10"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Figure 4b-e, Supplementary Figure 7a-f: RPF-seq\n",
    "### Supplementary Figure 8a-j: QTI-seq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [],
   "source": [
    "def plot_sgRNA(ORF = 'N',hpi_list = ['16h','24h','36h'], s = 49,\n",
    "               assay_prefix = 'RPFSeq', sample_prefix = 'RPF', as_RPM = True, offset_RPF = 12,\n",
    "               as_log10= True, ylim= 4, OutFigname=''):\n",
    "    global CANONICAL_sgRNA_junc_dict, SARSCOV2_genome\n",
    "    sgRNA_junc_dict = CANONICAL_sgRNA_junc_dict.copy()\n",
    "    sgRNA_junc_dict['ORF1a'] = [75,75] ##no junction, CUG-translation gRNA ORF end\n",
    "    idx_TISL = 58\n",
    "    contig_len  =  len(SARSCOV2_genome)\n",
    "    grid     = GridSpec(nrows=2*(1+len(hpi_list)),ncols=1)\n",
    "    fig      = plt.figure(figsize=(10,1.5*(1+len(hpi_list))))\n",
    "    thickness= 5\n",
    "    \n",
    "    junc_5p, junc_3p  = sgRNA_junc_dict[ORF]\n",
    "    \n",
    "    df_SARSCOV2_annot = get_df_SARSCOV2_annot(exc_UTR = True)\n",
    "    ORFstart          = df_SARSCOV2_annot.at[ORF,'ORFstart']\n",
    "    ORFstart_shifted  = (ORFstart-junc_3p)+(junc_5p-s)\n",
    "    \n",
    "    window_size       = 50*(ORFstart_shifted//50+1)\n",
    "    if ORF in ['ORF3a','E']:\n",
    "        window_size  += 50\n",
    "    elif ORF == 'M':\n",
    "        window_size  += 100\n",
    "    \n",
    "    if window_size == 50:\n",
    "        ticklabel_width = 10\n",
    "    else:\n",
    "        ticklabel_width = 20\n",
    "    \n",
    "    e = ORFstart + (window_size-ORFstart_shifted)\n",
    "    \n",
    "    df_source = []\n",
    "    for hpi_idx,hpi in enumerate(hpi_list):\n",
    "        arr_RPF_list  = []\n",
    "        if hpi == '48h':\n",
    "            reps=3\n",
    "        else:\n",
    "            reps=2\n",
    "        for rep in range(1,reps+1):\n",
    "            samplename = f'{sample_prefix}_{hpi}_rep{rep}'\n",
    "            _, _, arr_RPF = get_readpos_arr(samplename,assay_prefix,norm_by_nh=True,\n",
    "                                            pos_5end_range=(0,29903), f2_5p_range=None, fetch_range=(0,29903),\n",
    "                                            offset_RPF = offset_RPF,junction_spanning_only=False,\n",
    "                                            min_rlen=0,as_RPM=as_RPM)\n",
    "\n",
    "            arr_RPF_list.append(arr_RPF)\n",
    "        \n",
    "        avg_arr_RPF  = np.array(arr_RPF_list).mean(axis=0)\n",
    "        \n",
    "        TISLreads_df = pd.read_csv(RESULTDIR %(f'{sample_prefix}{hpi}_fraction.tsv'),\n",
    "                            sep = '\\t', index_col = 0)\n",
    "        TISL_pct     = TISLreads_df.loc[ORF,'Percentage']\n",
    "        arr_RPF_fused= np.concatenate((avg_arr_RPF[s:junc_5p]*TISL_pct/100, avg_arr_RPF[junc_3p:e]))\n",
    "        \n",
    "        if as_log10:\n",
    "            psc = 1\n",
    "            arr_RPF_fused = np.log10(arr_RPF_fused+psc)\n",
    "        \n",
    "        x = np.arange(len(arr_RPF_fused))\n",
    "        tmp_xticklabels = np.arange(s,junc_5p)+1 #1-base coordinate\n",
    "        xticks      = [i for i,nt in enumerate(tmp_xticklabels) if nt%ticklabel_width == 0]\n",
    "        xticklabels = tmp_xticklabels[xticks]\n",
    "        \n",
    "        tmp_xticklabels = np.arange(junc_3p,e)+1 #1-base coordinate\n",
    "        tmp_xticks      = [i for i,nt in enumerate(tmp_xticklabels) if nt%ticklabel_width == 0]\n",
    "        tmp_xticklabels = tmp_xticklabels[tmp_xticks]\n",
    "        xticks = xticks+[i+junc_5p-s for i in tmp_xticks]\n",
    "        xticklabels= np.concatenate((xticklabels,tmp_xticklabels))\n",
    "\n",
    "        bar_colors = [['darkblue','cornflowerblue','lightskyblue'][(i-ORFstart_shifted)%3] for i in x]\n",
    "\n",
    "        ax1 = fig.add_subplot(grid[hpi_idx*2:hpi_idx*2+2,0])\n",
    "        if ORF != 'ORF1a':\n",
    "            ax1.axvline(junc_5p-s-0.5,linestyle='--',color='k')\n",
    "            if hpi_idx == 0:\n",
    "                #ax1.text(0,ylim+0.1,'Estimated RPM')\n",
    "                #ax1.text(junc_5p-s,ylim+0.1,'Measured RPM')\n",
    "                ax1.text(junc_5p-s,ylim+0.1,'Junction')\n",
    "        ax1.bar(x,arr_RPF_fused,color = bar_colors)\n",
    "        \n",
    "        ax1.set_xlim((-0.5,len(arr_RPF_fused)-0.5))\n",
    "        ax1.text(len(arr_RPF_fused)*0.95,ylim*0.9,hpi.replace('h', ' hpi'))\n",
    "        ax1.set_xticks(xticks)\n",
    "        if hpi_idx==(len(hpi_list)-1):\n",
    "            ax1.set_xticklabels(xticklabels)\n",
    "            #ax1.set_xlabel('Genomic position (nt)')\n",
    "        else:\n",
    "            ax1.set_xticklabels(['' for x in range(len(xticks))])\n",
    "\n",
    "        if as_log10:\n",
    "            #ax1.set_ylabel('$\\mathregular{log_{10}}$ (Avg. RPM+1)')\n",
    "            if ylim != None:\n",
    "                ax1.set_ylim(0,ylim)\n",
    "        else:\n",
    "            ax1.set_ylabel('Avg. RPM')\n",
    "        df_source.append(arr_RPF_fused)\n",
    "        \n",
    "    df_source = pd.DataFrame(df_source)\n",
    "    df_source.index   = hpi_list\n",
    "    df_source.columns = np.concatenate([np.arange(s,junc_5p), np.arange(junc_3p,e)])+1\n",
    "    df_source = df_source.T\n",
    "    df_source.to_csv(SOURCEDATADIR %OutFigname.replace('.pdf','.tsv'), sep = '\\t')    \n",
    "\n",
    "    ######################\n",
    "    ax2 = fig.add_subplot(grid[len(hpi_list)*2:len(hpi_list)*2+2,0])\n",
    "\n",
    "    annot_image = vis_cov2annotation(thickness)\n",
    "    annot_image = np.concatenate((annot_image,np.full((thickness*2,contig_len,3),255)),axis=0)\n",
    "    annot_image = np.concatenate((annot_image[:,s:junc_5p],annot_image[:,junc_3p:e]),axis=1)\n",
    "    annot_texts = get_annot_textlist(s,junc_5p,thickness)\n",
    "    annot_texts+= [[i[0],i[1],i[2]+junc_5p-s] for i in get_annot_textlist(junc_3p,e,thickness)]\n",
    "    for txt,ypos,xpos in annot_texts:\n",
    "        if txt == '5\\'UTR':\n",
    "            color='white'\n",
    "        else:\n",
    "            color='k'\n",
    "        ax2.text(xpos,ypos,txt,size=12,color=color)\n",
    "    ax2.text(1,thickness/2,'Leader',size=12,color='white')\n",
    "\n",
    "\n",
    "    trs_list = get_trs_b_list(s,junc_5p)\n",
    "    trs_list+= [i+junc_5p-s for i in get_trs_b_list(junc_3p,e)]\n",
    "    for trs_pos in trs_list:\n",
    "        if ORF == 'ORF1a':\n",
    "            TRS = 'TRS-L'\n",
    "        else:\n",
    "            TRS = 'TRS-B'\n",
    "        ax2.text(trs_pos,int(thickness*1.2),TRS,size=12)\n",
    "        annot_image[int(thickness*0.6):int(thickness*1.0),trs_pos:trs_pos+6] = (255,215,0)\n",
    "\n",
    "    ax2.text(58-s,int(thickness*1.2),'CUG',size=12)\n",
    "    annot_image[int(thickness*0.6):int(thickness*1.0),58-s:61-s] = (255,87,51)\n",
    "\n",
    "    ax2.text(ORFstart_shifted,int(thickness*1.2),'AUG',size=12)\n",
    "    annot_image[int(thickness*0.6):int(thickness*1.0),ORFstart_shifted:ORFstart_shifted+3] = (0,255,0)\n",
    "    txt = SARSCOV2_genome[idx_TISL:junc_5p]+SARSCOV2_genome[junc_3p:]\n",
    "\n",
    "    for i in range(0,len(txt),3):\n",
    "        codon = txt[i:i+3]\n",
    "        if codon in ['TGA','TAA','TAG']:\n",
    "            putative_ORF_end = i+3\n",
    "            break\n",
    "    ax2.text(58-s,thickness*2.5,'ORF from TIS-L (%da.a)' %((putative_ORF_end-3)/3),size=12)\n",
    "    annot_image[int(thickness*2):int(thickness*3),idx_TISL-s:idx_TISL-s+putative_ORF_end] = (3,181,148)\n",
    "\n",
    "    ax2.imshow(annot_image,aspect='auto')\n",
    "    if ORF != 'ORF1a':\n",
    "        ax2.axvline(junc_5p-s-0.5,linestyle='--',color='k')\n",
    "        \n",
    "    ###nucleotide text###\n",
    "    nt_fused  = (SARSCOV2_genome[s:junc_5p]+SARSCOV2_genome[junc_3p:e]).replace('T','U')\n",
    "    nt_font_size = min(10,1000//window_size)\n",
    "    for i in range(len(arr_RPF_fused)):\n",
    "        ax2.text(i-0.5,0,nt_fused[i],fontsize=nt_font_size)\n",
    "    #####################\n",
    "    \n",
    "    ax2.set_xlim((-0.5,len(arr_RPF_fused)-0.5))\n",
    "    ax2.set_xticks([])\n",
    "    ax2.set_yticks([0,2,14.5])\n",
    "    #ax2.set_yticklabels(['','sgRNA',''])\n",
    "    ax2.set_yticklabels([])\n",
    "    ax2.grid(False)\n",
    "    #####################\n",
    "    plt.tight_layout()\n",
    "    if OutFigname != '':\n",
    "        plt.savefig(FIGDIR %OutFigname)\n",
    "    else:\n",
    "        plt.show()\n",
    "    plt.close()\n",
    "    return None\n",
    "\n",
    "def vis_cov2annotation(thickness = 5):\n",
    "    global ORF_palette\n",
    "    thick_unit = thickness//5\n",
    "    df_SARSCOV2_annot = get_df_SARSCOV2_annot(exc_UTR = False)\n",
    "    \n",
    "    contig_len  =  29903\n",
    "    arr = np.full((thickness,contig_len,3),255) #all white\n",
    "    arr[0*thick_unit:5*thick_unit] = 0 ##black line at center\n",
    "    \n",
    "    for ORFname,row in df_SARSCOV2_annot.iterrows():\n",
    "        ORF_s,ORF_e     = row['ORFstart'], row['ORFend']\n",
    "        color   = ORF_palette[ORFname]\n",
    "        color_int = []\n",
    "        for i in range(1,len(color),2):\n",
    "            color_int.append(int(color[i:i+2],16))\n",
    "        thick_range=range(thickness)\n",
    "        arr[thick_range,ORF_s:ORF_e] = color_int\n",
    "    \n",
    "    return arr\n",
    "\n",
    "def get_annot_textlist(s,e,thickness):\n",
    "    textlist = []\n",
    "    df_SARSCOV2_annot = get_df_SARSCOV2_annot(exc_UTR = False)\n",
    "    \n",
    "    for ORFname,row in df_SARSCOV2_annot.iterrows():\n",
    "        ORF_s,ORF_e     = row['ORFstart'], row['ORFend']\n",
    "        if 'UTR' in ORFname:\n",
    "            pass\n",
    "        else:\n",
    "            pass\n",
    "        if s<=ORF_s and ORF_s<e:\n",
    "            if ORF_e <= e: #whole ORF included in the window\n",
    "                textlist.append([ORFname,thickness//2, (ORF_s+ORF_e)//2 -s-2]) #ORFname, textposition within window\n",
    "            else:\n",
    "                textlist.append([ORFname,thickness//2, (ORF_s+e)//2 - s-2])\n",
    "        elif s<ORF_e and ORF_e<=e:\n",
    "            textlist.append(    [ORFname,thickness//2, (s+ORF_e)//2 - s-2])\n",
    "        else:\n",
    "            pass\n",
    "        \n",
    "    return textlist\n",
    "\n",
    "def get_trs_b_list(s,e):\n",
    "    global SARSCOV2_genome\n",
    "    cov2_fa = SARSCOV2_genome[:]\n",
    "    cov2_fa = cov2_fa[s:e].upper().replace('T','U')\n",
    "    trs_b_list = []\n",
    "    for i in re.finditer('ACGAAC',cov2_fa):\n",
    "        trs_b_list.append(i.start())\n",
    "    return trs_b_list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "RPF ORF1a Mon Sep 27 16:24:12 2021\n",
      "RPF S Mon Sep 27 16:24:34 2021\n",
      "RPF ORF3a Mon Sep 27 16:24:54 2021\n",
      "RPF E Mon Sep 27 16:25:15 2021\n",
      "RPF M Mon Sep 27 16:25:36 2021\n",
      "RPF ORF6 Mon Sep 27 16:25:58 2021\n",
      "RPF ORF7a Mon Sep 27 16:26:19 2021\n",
      "RPF ORF7b Mon Sep 27 16:26:39 2021\n",
      "RPF ORF8 Mon Sep 27 16:27:00 2021\n",
      "RPF N Mon Sep 27 16:27:20 2021\n",
      "QTI ORF1a Mon Sep 27 16:27:41 2021\n",
      "QTI S Mon Sep 27 16:27:51 2021\n",
      "QTI ORF3a Mon Sep 27 16:28:00 2021\n",
      "QTI E Mon Sep 27 16:28:10 2021\n",
      "QTI M Mon Sep 27 16:28:19 2021\n",
      "QTI ORF6 Mon Sep 27 16:28:29 2021\n",
      "QTI ORF7a Mon Sep 27 16:28:39 2021\n",
      "QTI ORF7b Mon Sep 27 16:28:48 2021\n",
      "QTI ORF8 Mon Sep 27 16:28:57 2021\n",
      "QTI N Mon Sep 27 16:29:06 2021\n"
     ]
    }
   ],
   "source": [
    "#Main plots for Fig. 4 and SFigs. 7,8\n",
    "tmp_ylim_dict = {'ORF1a':3, 'S': 3,'ORF3a':4,'E':2.5,'M':3,  'ORF6':4,'ORF7a':4,'ORF7b':2,'ORF8':3.5,'N':4}\n",
    "for tmp_sample_prefix in ['RPF','QTI']:\n",
    "    for ORF, tmp_ylim in tmp_ylim_dict.items():\n",
    "        print(tmp_sample_prefix, ORF, time.ctime())\n",
    "        plot_sgRNA(ORF = ORF, hpi_list = ['16h','24h','36h'], s = 49,\n",
    "                   assay_prefix = 'RPFSeq', sample_prefix = tmp_sample_prefix, as_RPM = True, offset_RPF = 12,\n",
    "                   as_log10= True, ylim = tmp_ylim, OutFigname=f'Fig4_{tmp_sample_prefix}_{ORF}.pdf')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [],
   "source": [
    "def plot_ORF7b_cmp(assay_prefix = 'RPFSeq',sample_prefix = 'RPF',hpi = '36h', ylim=3.5, s = 49,\n",
    "                   offset_RPF = 12, as_RPM=True, OutFigname = ''):\n",
    "    global CANONICAL_sgRNA_junc_dict, SARSCOV2_genome\n",
    "    ORF      = 'ORF7b'\n",
    "    idx_TISL = 58\n",
    "    sgRNA_junc_dict = CANONICAL_sgRNA_junc_dict.copy()\n",
    "    junc_5p,junc_3p = sgRNA_junc_dict[ORF]\n",
    "    \n",
    "    contig_len  =  len(SARSCOV2_genome)\n",
    "    df_SARSCOV2_annot = get_df_SARSCOV2_annot(exc_UTR = False)\n",
    "    \n",
    "    TISLreads_df = pd.read_csv(RESULTDIR %(f'{sample_prefix}{hpi}_fraction.tsv'),\n",
    "                               sep = '\\t', index_col = 0)\n",
    "    TISL_pct     = TISLreads_df.loc[ORF,'Percentage']    \n",
    "    \n",
    "    grid = GridSpec(nrows=11,ncols=1)\n",
    "    fig  = plt.figure(figsize=(10,6.67*0.75))\n",
    "    \n",
    "    as_log10=True\n",
    "    \n",
    "    ORFstart        = df_SARSCOV2_annot.at[ORF,'ORFstart']\n",
    "    ORFstart_shifted= (ORFstart-junc_3p)+(junc_5p-s)\n",
    "    window_size     = 50*(ORFstart_shifted//50+1)\n",
    "    e               = ORFstart + (window_size-ORFstart_shifted)\n",
    "    \n",
    "    arr_RPF_list    = []\n",
    "    reps            = 2\n",
    "    for rep in range(1,reps+1):\n",
    "        samplename    = f'{sample_prefix}_{hpi}_rep{rep}'\n",
    "        _, _, arr_RPF = get_readpos_arr(samplename,assay_prefix,norm_by_nh=True,\n",
    "                                        pos_5end_range=(0,29903), f2_5p_range=None, fetch_range=(0,29903),\n",
    "                                        offset_RPF = offset_RPF,junction_spanning_only=False,\n",
    "                                        min_rlen=0,as_RPM=as_RPM)\n",
    "        arr_RPF_list.append(arr_RPF)\n",
    "\n",
    "    avg_arr_RPF  = np.array(arr_RPF_list).mean(axis=0)\n",
    "    \n",
    "    arr_RPF_raw     = avg_arr_RPF[e-window_size:e]\n",
    "    arr_RPF_fused   = np.concatenate((avg_arr_RPF[s:junc_5p]*TISL_pct/100, avg_arr_RPF[junc_3p:e]))\n",
    "    \n",
    "    if window_size == 50:\n",
    "        ticklabel_width = 10\n",
    "    else:\n",
    "        ticklabel_width = 20\n",
    "    df_source = pd.DataFrame(columns = ['position_ORF7a_sgRNA','logRPM_ORF7a_sgRNA','position_ORF7b_sgRNA','logRPM_ORF7b_sgRNA'])\n",
    "    df_source['position_ORF7a_sgRNA'] = np.arange(e-window_size, e)+1\n",
    "    df_source['position_ORF7b_sgRNA'] = np.concatenate([np.arange(s,junc_5p),np.arange(junc_3p,e)])+1\n",
    "    #######RPF, non-fused########\n",
    "    arr = arr_RPF_raw\n",
    "    if as_log10:\n",
    "        psc = 1\n",
    "        arr = np.log10(arr+psc)\n",
    "    df_source['logRPM_ORF7a_sgRNA'] = arr\n",
    "    thickness       = 5\n",
    "    \n",
    "    x = np.arange(len(arr))\n",
    "    bar_colors = [['darkblue','cornflowerblue','lightskyblue'][(i-ORFstart_shifted)%3] for i in x]\n",
    "    ax1 = fig.add_subplot(grid[0:3,0])\n",
    "    ax1.bar(x,arr,color = bar_colors)\n",
    "    ax1.set_xlim((-0.5,len(arr)-0.5))\n",
    "\n",
    "\n",
    "    tmp_xticklabels = np.arange(e-window_size,e)+1\n",
    "    xticks = [i for i,nt in enumerate(tmp_xticklabels) if nt%10 == 0]\n",
    "    xticklabels = tmp_xticklabels[xticks]\n",
    "    ax1.set_xticks(xticks)\n",
    "    ax1.set_xticklabels(xticklabels)\n",
    "    if as_log10:\n",
    "        ax1.set_ylabel('$\\mathregular{log_{10}}$ (Avg. RPM+1)')\n",
    "    else:\n",
    "        ax1.set_ylabel('Avg. RPM')\n",
    "    if ylim != None:\n",
    "        ax1.set_ylim(0,ylim)\n",
    "    #########################\n",
    "    #######Annotation of raw genome########\n",
    "    ax2 = fig.add_subplot(grid[3:5,0])\n",
    "    \n",
    "    annot_image = vis_cov2annotation(thickness)\n",
    "    annot_image = np.concatenate((annot_image,np.full((thickness,contig_len,3),255)),axis=0)[:,e-window_size:e]\n",
    "    annot_texts = get_annot_textlist(e-window_size,e,thickness)\n",
    "    for txt,ypos,xpos in annot_texts:\n",
    "        ax2.text(xpos,ypos,txt,size=12)\n",
    "    ax2.text(ORFstart_shifted,int(thickness*1.2),'AUG',size=12)\n",
    "    annot_image[int(thickness*0.6):int(thickness*1.0),ORFstart_shifted:ORFstart_shifted+3] = (0,255,0)\n",
    "    \n",
    "    ###nucleotide text###\n",
    "    nt_raw  = (SARSCOV2_genome[e-window_size:e]).replace('T','U')\n",
    "    nt_font_size = min(10,1000//window_size)\n",
    "    for i in range(len(arr_RPF_raw)):\n",
    "        ax2.text(i-0.5,0,nt_raw[i],fontsize=nt_font_size)\n",
    "    #####################\n",
    "    \n",
    "    ax2.imshow(annot_image,aspect='auto')\n",
    "\n",
    "    ax2.set_xlim((-0.5,len(arr)-0.5))\n",
    "    ax2.set_xticks([])\n",
    "    ax2.set_yticks([0,2,9.5])\n",
    "    ax2.set_yticklabels(['','',''])\n",
    "    ax2.grid(False)\n",
    "    #####################\n",
    "    #######RPF, fused########\n",
    "    arr = arr_RPF_fused\n",
    "    if as_log10:\n",
    "        psc = 1\n",
    "        arr = np.log10(arr+psc)\n",
    "    df_source['logRPM_ORF7b_sgRNA'] = arr\n",
    "    thickness       = 5\n",
    "    x = np.arange(len(arr))\n",
    "\n",
    "    bar_colors = [['darkblue','cornflowerblue','lightskyblue'][(i-ORFstart_shifted)%3] for i in x]\n",
    "    ax1 = fig.add_subplot(grid[5:8,0])\n",
    "    ax1.bar(x,arr,color = bar_colors)\n",
    "    ax1.set_xlim((-0.5,len(arr)-0.5))\n",
    "    ax1.set_xticks(xticks)\n",
    "\n",
    "    tmp_xticklabels = np.arange(s,junc_5p)+1 #1-base coordinate\n",
    "    xticks      = [i for i,nt in enumerate(tmp_xticklabels) if nt%10 == 0]\n",
    "    xticklabels = tmp_xticklabels[xticks]\n",
    "    tmp_xticklabels = np.arange(junc_3p,e)+1 #1-base coordinate\n",
    "    tmp_xticks      = [i for i,nt in enumerate(tmp_xticklabels) if nt%10 == 0]\n",
    "    tmp_xticklabels = tmp_xticklabels[tmp_xticks]\n",
    "    xticks = xticks+[i+junc_5p-s for i in tmp_xticks]\n",
    "    xticklabels= np.concatenate((xticklabels,tmp_xticklabels))\n",
    "    ax1.set_xticklabels(xticklabels)\n",
    "\n",
    "    if ORF != 'ORF1a':\n",
    "        ax1.axvline(junc_5p-s-0.5,linestyle='--',color='k')\n",
    "        ax1.text(junc_5p-s,ylim+0.1,'Junction')\n",
    "    if as_log10:\n",
    "        ax1.set_ylabel('log10 (Avg. RPM+1)')\n",
    "    else:\n",
    "        ax1.set_ylabel('Avg. RPM')\n",
    "    if ylim != None:\n",
    "        ax1.set_ylim(0,ylim)\n",
    "    ######################\n",
    "\n",
    "    ax2 = fig.add_subplot(grid[8:11,0])\n",
    "\n",
    "    annot_image = vis_cov2annotation(thickness)\n",
    "    annot_image = np.concatenate((annot_image,np.full((thickness*2,contig_len,3),255)),axis=0)\n",
    "    annot_image = np.concatenate((annot_image[:,s:junc_5p],annot_image[:,junc_3p:e]),axis=1)\n",
    "    annot_texts = get_annot_textlist(s,junc_5p,thickness)\n",
    "    annot_texts+= [[i[0],i[1],i[2]+junc_5p-s] for i in get_annot_textlist(junc_3p,e,thickness)]\n",
    "    for txt,ypos,xpos in annot_texts:\n",
    "        ax2.text(xpos,ypos,txt,size=12)\n",
    "    ax2.text(1,thickness/2,'Leader',size=12,color='white')\n",
    "\n",
    "\n",
    "    trs_list = get_trs_b_list(s,junc_5p)\n",
    "    trs_list+= [i+junc_5p-s for i in get_trs_b_list(junc_3p,e)]\n",
    "    for trs_pos in trs_list:\n",
    "        if ORF == 'ORF1a':\n",
    "            TRS = 'TRS-L'\n",
    "        else:\n",
    "            TRS = 'TRS-B'\n",
    "        ax2.text(trs_pos,int(thickness*1.6),TRS,size=12)\n",
    "        annot_image[int(thickness*1):int(thickness*1.5),trs_pos:trs_pos+6] = (255,215,0)\n",
    "\n",
    "    ax2.text(58-s,int(thickness*1.2),'CUG',size=12)\n",
    "    annot_image[int(thickness*0.6):int(thickness*1.0),58-s:61-s] = (255,87,51)\n",
    "\n",
    "    txt = SARSCOV2_genome[idx_TISL:junc_5p]+SARSCOV2_genome[junc_3p:]\n",
    "\n",
    "    for i in range(0,len(txt),3):\n",
    "        codon = txt[i:i+3]\n",
    "        if codon in ['TGA','TAA','TAG']:\n",
    "            putative_ORF_end = i+3\n",
    "            break\n",
    "\n",
    "    ax2.text(58-s,thickness*2.5,'ORF from TIS-L (%da.a)' %((putative_ORF_end-3)/3),size=12)\n",
    "    annot_image[int(thickness*2):int(thickness*3),idx_TISL-s:idx_TISL-s+putative_ORF_end] = (3,181,148)\n",
    "\n",
    "    ax2.imshow(annot_image,aspect='auto')\n",
    "    if ORF != 'ORF1a':\n",
    "        ax2.axvline(junc_5p-s-0.5,linestyle='--',color='k')\n",
    "\n",
    "    ###nucleotide text###\n",
    "    nt_fused  = (SARSCOV2_genome[s:junc_5p]+SARSCOV2_genome[junc_3p:e]).replace('T','U')\n",
    "    nt_font_size = min(10,1000//window_size)\n",
    "    for i in range(len(arr_RPF_fused)):\n",
    "        ax2.text(i-0.5,0,nt_fused[i],fontsize=nt_font_size)\n",
    "    #####################\n",
    "    \n",
    "    ax2.set_xlim((-0.5,len(arr_RPF_fused)-0.5))\n",
    "    ax2.set_xticks([])\n",
    "    ax2.set_yticks([0,2,14.5])\n",
    "    ax2.set_yticklabels(['','',''])\n",
    "    ax2.grid(False)\n",
    "    #####################\n",
    "    if OutFigname == '':\n",
    "        plt.show()\n",
    "    else:\n",
    "        plt.savefig(FIGDIR %OutFigname)\n",
    "        df_source.to_csv(SOURCEDATADIR %OutFigname.replace('.pdf','.tsv'), sep = '\\t', index = False)\n",
    "    plt.close()\n",
    "    return None\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [],
   "source": [
    "plot_ORF7b_cmp(sample_prefix = 'RPF',hpi = '36h', ylim=3.5, s = 49,\n",
    "               OutFigname = 'SFig7g_RPF_ORF7b.pdf')\n",
    "plot_ORF7b_cmp(sample_prefix = 'QTI',hpi = '36h', ylim=3.5, s = 49,\n",
    "               OutFigname = 'SFig8k_QTI_ORF7b.pdf')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### ORF10 sgRNA production and ORF10 translation (SFig. 9)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_junctionread_pairs(assay_prefix='mRNASeq',sample_prefix='RPFpaired',hpi='48h', rep=1,\n",
    "                           norm_by_nh=True,\n",
    "                           pos_5end_range=(0,265), f2_5p_range=None, fetch_range=(0,265),\n",
    "                           min_rlen=(32,31,30,29,28,27),as_RPM=True):\n",
    "    #Mainly for analyzing mRNA-seq data, but designed as it can be applied to RPF-seq data analysis\n",
    "    samplename = f'{sample_prefix}_{hpi}_rep{rep}'\n",
    "    fname = BAMDIR_cov2mapped %(assay_prefix,samplename)\n",
    "    bam = pysam.AlignmentFile(fname)\n",
    "    \n",
    "    contig = bam.references[0]\n",
    "    jpair_count_dict = defaultdict(float) #{(junction s, e): weighted count}\n",
    "    for read in bam.fetch(contig,fetch_range[0],fetch_range[1]):\n",
    "        cigartuples = read.cigartuples\n",
    "        pos_5end    = read.reference_start\n",
    "        rlen        = read.infer_read_length()\n",
    "        nh          = read.get_tag('NH')\n",
    "        assigned    = False\n",
    "        if norm_by_nh:\n",
    "            norm_ct = 1/nh\n",
    "        else:\n",
    "            norm_ct = 1\n",
    "        ####read filtering####\n",
    "        contain_junction = (3 in [i[0] for i in cigartuples])\n",
    "        if not contain_junction:\n",
    "            continue\n",
    "            \n",
    "        if not ((pos_5end_range[0]<= pos_5end) and (pos_5end< pos_5end_range[1])):\n",
    "            continue\n",
    "            \n",
    "        if type(min_rlen) == int:\n",
    "            if rlen<min_rlen:\n",
    "                continue\n",
    "        else:\n",
    "            if rlen<min_rlen[pos_5end-pos_5end_range[0]]:\n",
    "                continue\n",
    "        \n",
    "        #############################\n",
    "        cur_pos = pos_5end\n",
    "        f1_detected = False\n",
    "        f1_3p = None\n",
    "        f2_5p = None\n",
    "        for idx,cigar_tp in enumerate(cigartuples):\n",
    "            operation, length = cigar_tp\n",
    "            if operation == 3: #skip\n",
    "                f1_detected = True\n",
    "                cur_pos += length\n",
    "            else:\n",
    "                if operation == 4 or operation == 1:    # softclip or insertion\n",
    "                    pass\n",
    "                elif operation == 0 or operation == 2:  # match/mismatch or deletion\n",
    "                    cur_pos += length\n",
    "            if operation == 0:\n",
    "                if not f1_detected: #junction 1st frag\n",
    "                    f1_3p = cur_pos\n",
    "                else: #second fragment\n",
    "                    f2_5p = cur_pos - length\n",
    "                    break\n",
    "        assert f1_3p != None\n",
    "        assert f2_5p != None\n",
    "        \n",
    "        if f2_5p_range != None:\n",
    "            if not ((f2_5p_range[0]<= f2_5p) and (f2_5p< f2_5p_range[1])):\n",
    "                continue\n",
    "        jpair_count_dict[(f1_3p,f2_5p)] += norm_ct\n",
    "        \n",
    "    bam.close()\n",
    "    jpair_count_dict = dict(jpair_count_dict) #defaultdict to dict\n",
    "    if as_RPM:\n",
    "        viral_mapped_reads = 0\n",
    "        host_mapped_reads  = 0\n",
    "        #virus reads\n",
    "        bam = pysam.AlignmentFile(fname)\n",
    "        for read in bam.fetch():\n",
    "            viral_mapped_reads += (1/(read.get_tag('NH')))\n",
    "        bam.close()\n",
    "        #human reads\n",
    "        fname_hg38 = fname.replace('_SARSCOV2align','_hostalign')\n",
    "        bam = pysam.AlignmentFile(fname_hg38)\n",
    "        for read in bam.fetch():\n",
    "            host_mapped_reads += (1/(read.get_tag('NH')))\n",
    "        bam.close()\n",
    "        total_mapped_reads = host_mapped_reads + viral_mapped_reads\n",
    "        for jpair in jpair_count_dict: \n",
    "            jpair_count_dict[jpair] /= (total_mapped_reads / 1e+06)\n",
    "        \n",
    "    return jpair_count_dict\n",
    "\n",
    "def junctionpair_analysis(assay_prefix='mRNASeq',sample_prefix_list=['RPFpaired','QTIpaired'],hpi='48h',\n",
    "                      pos_5end_range=(0,265), f2_5p_range=(28240,28280), fetch_range=(28240,28280),\n",
    "                      min_rlen=1):\n",
    "    \n",
    "    reps = 3 if hpi == '48h' else 2\n",
    "    jpair_dict_list = []\n",
    "    for sample_prefix in sample_prefix_list:\n",
    "        for rep in range(1,reps+1):\n",
    "            jpair_count_dict = get_junctionread_pairs(assay_prefix=assay_prefix,sample_prefix=sample_prefix,\n",
    "                                                      hpi=hpi, rep=rep,norm_by_nh=True,\n",
    "                                                      pos_5end_range=pos_5end_range, f2_5p_range=f2_5p_range, \n",
    "                                                      fetch_range=fetch_range, min_rlen=min_rlen, as_RPM=True)\n",
    "            \n",
    "            jpair_dict_list.append(jpair_count_dict)\n",
    "    ndict = len(jpair_dict_list)\n",
    "    total_jpairs = set([jpair for jpair_count_dict in jpair_dict_list for jpair in jpair_count_dict.keys() ])\n",
    "    \n",
    "    avg_jpair_count_list = []\n",
    "    for jpair in total_jpairs:\n",
    "        count = 0\n",
    "        for jpair_count_dict in jpair_dict_list:\n",
    "            try:\n",
    "                count += jpair_count_dict[jpair]\n",
    "            except KeyError:\n",
    "                pass\n",
    "        count /= ndict\n",
    "        avg_jpair_count_list.append([*jpair,count]) #junction start, end, count\n",
    "    avg_jpair_count_list = sorted(avg_jpair_count_list,key = lambda x: x[2],reverse = True)\n",
    "    \n",
    "    return avg_jpair_count_list\n",
    "\n",
    "def plot_ORF10_junctions(hpi,f2_5p_range=(29457,29607),U5_display_range=(58,88),display_top=10,\n",
    "                         OutFigname=''): #-100,+50\n",
    "    canonical_AUG = 29557 #ORF10 start idx \n",
    "    idx_TISL = 58\n",
    "    ticklabel_width = 20\n",
    "    avg_jpair_count_list = junctionpair_analysis(assay_prefix='mRNASeq',sample_prefix_list=['RPFpaired','QTIpaired'],hpi=hpi,\n",
    "                                                 pos_5end_range=(0,265), f2_5p_range=f2_5p_range, \n",
    "                                                 fetch_range=f2_5p_range,min_rlen=1)\n",
    "    if len(avg_jpair_count_list) == 0:\n",
    "        first_jpair= [np.nan, np.nan, np.nan]\n",
    "        first_jpair_pct = np.nan\n",
    "    else:\n",
    "        total_jpairs    = sum([i[2] for i in avg_jpair_count_list]) #junction start, end, count\n",
    "        first_jpair     = avg_jpair_count_list[0]\n",
    "        first_jpair_pct = first_jpair[2]/total_jpairs*100\n",
    "        \n",
    "    TISL_frame_list = [0,0,0]\n",
    "    for f1_3p,f2_5p,RPM in avg_jpair_count_list:\n",
    "        dist_annotAUG_TISL = (canonical_AUG-f2_5p)+(f1_3p-idx_TISL)\n",
    "        TISL_frame_list[dist_annotAUG_TISL%3] += RPM\n",
    "    TISL_frame_list = [RPM/total_jpairs*100 for RPM in TISL_frame_list]\n",
    "    \n",
    "    \n",
    "    \n",
    "    f1_3p_arr = np.zeros(U5_display_range[1]-U5_display_range[0],dtype=np.float)\n",
    "    f2_5p_arr = np.zeros(f2_5p_range[1]     -f2_5p_range[0],     dtype=np.float)\n",
    "    \n",
    "    rank_junc = 1\n",
    "    for f1_3p,f2_5p,RPM in avg_jpair_count_list[:display_top]:\n",
    "        f1_3p_arr[f1_3p-U5_display_range[0]] += RPM\n",
    "        f2_5p_arr[f2_5p-f2_5p_range[0]]      += RPM\n",
    "        rank_junc +=1\n",
    "        \n",
    "    fig, ((ax1, ax2), (ax3, ax4)) = plt.subplots(2,2, sharey='row', figsize=(8,4),\n",
    "                                 gridspec_kw={'width_ratios': [len(f1_3p_arr),len(f2_5p_arr)],\n",
    "                                              'height_ratios':[3,1]})\n",
    "    \n",
    "    ax1.bar( np.arange(*U5_display_range),f1_3p_arr)   \n",
    "    ax1.spines['right'].set_visible(False)\n",
    "    ax1.set_ylabel('RPM')\n",
    "    ax1.set_xlim(U5_display_range[0]-0.5, U5_display_range[1]-0.5)\n",
    "    tmp_xticklabels = np.arange(U5_display_range[0],U5_display_range[1])\n",
    "    ##assigned value at the last 'position' because f1_3p is end index of [s,e)\n",
    "    xticks      = [nt for nt in range(*U5_display_range) if nt%ticklabel_width == 0]\n",
    "    xticklabels = xticks\n",
    "    ax1.set_xticks(xticks)\n",
    "    ax1.set_xticklabels(xticklabels)\n",
    "    \n",
    "    ax2.bar(np.arange(*f2_5p_range),  f2_5p_arr)\n",
    "    ax2.spines['left'].set_visible(False)\n",
    "    ax2.set_xlabel('Genomic position (nt)')\n",
    "    ax2.set_xlim(f2_5p_range[0]-0.5,      f2_5p_range[1]-0.5)\n",
    "    tmp_xticklabels = np.arange(f2_5p_range[0],f2_5p_range[1])+1 #1-base coordinate\n",
    "    xticks      = [i for i,nt in enumerate(tmp_xticklabels) if nt%ticklabel_width == 0]\n",
    "    xticklabels = tmp_xticklabels[xticks]\n",
    "    xticks      = xticklabels-1\n",
    "    \n",
    "    ax2.set_xticks(xticks)\n",
    "    ax2.set_xticklabels(xticklabels)\n",
    "    \n",
    "    ###Annotation: 5UTR###\n",
    "    thickness   = 5\n",
    "    contig_len  = len(SARSCOV2_genome)\n",
    "    annot_image = vis_cov2annotation(thickness)\n",
    "    annot_image = np.concatenate((annot_image,np.full((thickness*2,contig_len,3),255)),axis=0)\n",
    "    annot_image = annot_image[:,U5_display_range[0]:U5_display_range[1]]\n",
    "    annot_texts = get_annot_textlist(U5_display_range[0],U5_display_range[1],thickness)\n",
    "    for txt,ypos,xpos in annot_texts:\n",
    "        if txt == '5\\'UTR':\n",
    "            color='white'\n",
    "        else:\n",
    "            color='k'\n",
    "        ax3.text(xpos,ypos,txt,size=12,color=color)\n",
    "    ax3.text(1,thickness/2,'Leader',size=12,color='white')\n",
    "\n",
    "\n",
    "    trs_list = get_trs_b_list(U5_display_range[0], U5_display_range[1])\n",
    "    \n",
    "    for trs_pos in trs_list:\n",
    "        TRS = 'TRS-L'\n",
    "        ax3.text(trs_pos,int(thickness*1.2),TRS,size=12)\n",
    "        annot_image[int(thickness*0.6):int(thickness*1.0),trs_pos:trs_pos+6] = (255,215,0)\n",
    "\n",
    "    ax3.text(58-U5_display_range[0],int(thickness*1.2),'CUG',size=12)\n",
    "    annot_image[int(thickness*0.6):int(thickness*1.0),\n",
    "                58-U5_display_range[0]:61-U5_display_range[0]] = (255,87,51)\n",
    "    \n",
    "    ax3.imshow(annot_image,aspect='auto')\n",
    "        \n",
    "    ###nucleotide text###\n",
    "    nt_raw  = SARSCOV2_genome[U5_display_range[0]:U5_display_range[1]].replace('T','U')\n",
    "    nt_font_size = min(10,1000//(f2_5p_range[1]-f2_5p_range[0]))\n",
    "    for i in range(U5_display_range[1]-U5_display_range[0]):\n",
    "        ax3.text(i-0.5,0,nt_raw[i],fontsize=nt_font_size)\n",
    "    #####################\n",
    "    \n",
    "    ax3.set_xlim((-0.5,U5_display_range[1]-U5_display_range[0]-0.5))\n",
    "    ax3.set_xticks([])\n",
    "    ax3.set_yticks([0,2,14.5])\n",
    "    ax3.set_yticklabels([])\n",
    "    ax3.grid(False)\n",
    "    #################################################\n",
    "    ###Annotation: ORF10###\n",
    "    thickness   = 5\n",
    "    contig_len  = len(SARSCOV2_genome)\n",
    "    annot_image = vis_cov2annotation(thickness)\n",
    "    annot_image = np.concatenate((annot_image,np.full((thickness*2,contig_len,3),255)),axis=0)\n",
    "    annot_image = annot_image[:,f2_5p_range[0]:f2_5p_range[1]]\n",
    "    annot_texts = get_annot_textlist(f2_5p_range[0],f2_5p_range[1],thickness)\n",
    "    for txt,ypos,xpos in annot_texts:\n",
    "        if txt == '5\\'UTR':\n",
    "            color='white'\n",
    "        else:\n",
    "            color='k'\n",
    "        ax4.text(xpos,ypos,txt,size=12,color=color)\n",
    "    \n",
    "    ax4.imshow(annot_image,aspect='auto')\n",
    "        \n",
    "    ###nucleotide text###\n",
    "    nt_raw  = SARSCOV2_genome[f2_5p_range[0]:f2_5p_range[1]].replace('T','U')\n",
    "    nt_font_size = min(10,1000//(f2_5p_range[1]-f2_5p_range[0]))\n",
    "    for i in range(f2_5p_range[1]-f2_5p_range[0]):\n",
    "        ax4.text(i-0.5,0,nt_raw[i],fontsize=nt_font_size)\n",
    "    #####################\n",
    "    \n",
    "    ax4.set_xlim((-0.5,f2_5p_range[1]-f2_5p_range[0]-0.5))\n",
    "    ax4.set_xticks([])\n",
    "    ax4.set_yticks([0,2,14.5])\n",
    "    ax4.set_yticklabels([])\n",
    "    ax4.grid(False)\n",
    "    #################################################\n",
    "    plt.tight_layout()\n",
    "    for f1_3p,f2_5p,RPM in avg_jpair_count_list[:display_top]:\n",
    "        # 1. Get transformation operators for axis and figure\n",
    "        ax1tr = ax1.transData # ax1 -> Display\n",
    "        ax2tr = ax2.transData # ax2 -> Display\n",
    "        figtr = fig.transFigure.inverted() # Display -> Figure\n",
    "        # 2. Transform arrow start point from axis 0 to figure coordinates\n",
    "        ptB = figtr.transform(ax1tr.transform((f1_3p, f1_3p_arr[f1_3p-U5_display_range[0]]) ))\n",
    "        # 3. Transform arrow end point from axis 1 to figure coordinates\n",
    "        ptE = figtr.transform(ax2tr.transform((f2_5p, f2_5p_arr[f2_5p-f2_5p_range[0]])))\n",
    "        # 4. Create the patch\n",
    "        arrow = mpl.patches.FancyArrowPatch(\n",
    "            ptB, ptE, transform=fig.transFigure,  # Place arrow in figure coord system\n",
    "            fc = \"k\", connectionstyle=\"arc3,rad=-0.2\", arrowstyle='simple', alpha = 1,\n",
    "            mutation_scale = 20\n",
    "        )\n",
    "        # 5. Add patch to list of objects to draw onto the figure\n",
    "        fig.patches.append(arrow)\n",
    "    \n",
    "    if OutFigname == '':\n",
    "        plt.show()\n",
    "    else:\n",
    "        plt.savefig(FIGDIR %OutFigname)\n",
    "    plt.close()\n",
    "    return None"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [],
   "source": [
    "plot_ORF10_junctions('36h',f2_5p_range=(29457,29607),display_top=5, OutFigname='SFig9_ORF10sgRNA_36h.pdf')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [],
   "source": [
    "def plot_ORF10_RPF(hpi_list = ['16h','24h','36h'],\n",
    "               assay_prefix = 'RPFSeq', sample_prefix = 'RPF', as_RPM = True, offset_RPF = 12,\n",
    "               as_log10= True, ylim= 2.5, OutFigname=''):\n",
    "    global CANONICAL_sgRNA_junc_dict, SARSCOV2_genome\n",
    "    ORF = 'ORF10'\n",
    "    sgRNA_junc_dict = CANONICAL_sgRNA_junc_dict.copy()\n",
    "    \n",
    "    idx_TISL = 58\n",
    "    contig_len  =  len(SARSCOV2_genome)\n",
    "    grid     = GridSpec(nrows=11,ncols=1)\n",
    "    fig      = plt.figure(figsize=(10,1.5*(1+len(hpi_list))))\n",
    "    thickness= 5\n",
    "    \n",
    "    df_SARSCOV2_annot = get_df_SARSCOV2_annot(exc_UTR = True)\n",
    "    ORFstart          = df_SARSCOV2_annot.at[ORF,'ORFstart']\n",
    "    \n",
    "    window_size       = 100\n",
    "    ticklabel_width   = 20\n",
    "    e                 = ORFstart + 50\n",
    "    s                 = e-window_size\n",
    "    ORFstart_shifted  = ORFstart-s\n",
    "    \n",
    "    df_source = pd.DataFrame(columns = hpi_list, index = np.arange(s,e)+1)\n",
    "    \n",
    "    for hpi_idx,hpi in enumerate(hpi_list):\n",
    "        arr_RPF_list  = []\n",
    "        reps = 3 if hpi == '48h' else 2\n",
    "        for rep in range(1,reps+1):\n",
    "            samplename = f'{sample_prefix}_{hpi}_rep{rep}'\n",
    "            _, _, arr_RPF = get_readpos_arr(samplename,assay_prefix,norm_by_nh=True,\n",
    "                                            pos_5end_range=(0,29903), f2_5p_range=None, fetch_range=(0,29903),\n",
    "                                            offset_RPF = offset_RPF,junction_spanning_only=False,\n",
    "                                            min_rlen=0,as_RPM=as_RPM)\n",
    "            arr_RPF_list.append(arr_RPF)\n",
    "        \n",
    "        avg_arr_RPF  = np.array(arr_RPF_list).mean(axis=0)\n",
    "        arr_RPF_raw  = avg_arr_RPF[e-window_size:e]\n",
    "        df_source[hpi] = arr_RPF_raw\n",
    "        if as_log10:\n",
    "            psc = 1.0\n",
    "            arr_RPF_raw = np.log10(arr_RPF_raw+psc)\n",
    "        \n",
    "        x = np.arange(len(arr_RPF_raw))\n",
    "        tmp_xticklabels = np.arange(e-window_size,e)+1\n",
    "        xticks = [i for i,nt in enumerate(tmp_xticklabels) if nt%ticklabel_width == 0]\n",
    "        xticklabels = tmp_xticklabels[xticks]\n",
    "\n",
    "        bar_colors = [['darkblue','cornflowerblue','lightskyblue'][(i-ORFstart_shifted)%3] for i in x]\n",
    "\n",
    "        ax1 = fig.add_subplot(grid[hpi_idx*3:(hpi_idx+1)*3,0])\n",
    "        \n",
    "        ax1.bar(x,arr_RPF_raw,color = bar_colors)\n",
    "        \n",
    "        ax1.set_xlim((-0.5,len(arr_RPF_raw)-0.5))\n",
    "        ax1.text(len(arr_RPF_raw)*0.95,ylim*0.9,hpi.replace('h', ' hpi'))\n",
    "        ax1.set_xticks(xticks)\n",
    "        if hpi_idx==(len(hpi_list)-1):\n",
    "            ax1.set_xticklabels(xticklabels)\n",
    "            #ax1.set_xlabel('Genomic position (nt)')\n",
    "        else:\n",
    "            ax1.set_xticklabels(['' for x in range(len(xticks))])\n",
    "\n",
    "        if as_log10:\n",
    "            if ylim != None:\n",
    "                ax1.set_ylim(0,ylim)\n",
    "        else:\n",
    "            ax1.set_ylabel('Avg. RPM')\n",
    "    ######################\n",
    "    ax2 = fig.add_subplot(grid[9:11,0])\n",
    "\n",
    "    annot_image = vis_cov2annotation(thickness)\n",
    "    annot_image = np.concatenate((annot_image,np.full((thickness,contig_len,3),255)),axis=0)[:,e-window_size:e]\n",
    "    annot_texts = get_annot_textlist(e-window_size,e,thickness)\n",
    "    for txt,ypos,xpos in annot_texts:\n",
    "        ax2.text(xpos,ypos,txt,size=12)\n",
    "    trs_list = get_trs_b_list(e-window_size,e)\n",
    "    for trs_pos in trs_list:\n",
    "        if ORF == 'ORF1a':\n",
    "            TRS = 'TRS-L'\n",
    "        else:\n",
    "            TRS = 'TRS-B'\n",
    "        ax2.text(trs_pos,int(thickness*1.6),TRS,size=12)\n",
    "        annot_image[int(thickness*1):int(thickness*1.5),trs_pos:trs_pos+6] = (255,215,0)\n",
    "    ax2.text(ORFstart_shifted,int(thickness*1.2),'AUG',size=12)\n",
    "    annot_image[int(thickness*0.6):int(thickness*1.0),ORFstart_shifted:ORFstart_shifted+3] = (0,255,0)\n",
    "    \n",
    "    ###nucleotide text###\n",
    "    nt_raw  = (SARSCOV2_genome[e-window_size:e]).replace('T','U')\n",
    "    nt_font_size = min(10,1000//window_size)\n",
    "    for i in range(len(arr_RPF_raw)):\n",
    "        ax2.text(i-0.5,0,nt_raw[i],fontsize=nt_font_size)\n",
    "    #####################\n",
    "    \n",
    "    ax2.imshow(annot_image,aspect='auto')\n",
    "\n",
    "    ax2.set_xlim((-0.5,len(arr_RPF_raw)-0.5))\n",
    "    ax2.set_xticks([])\n",
    "    ax2.set_yticks([0,2,9.5])\n",
    "    ax2.set_yticklabels([])\n",
    "    ax2.grid(False)\n",
    "    #####################\n",
    "    plt.tight_layout()\n",
    "    if OutFigname != '':\n",
    "        plt.savefig(FIGDIR %OutFigname)\n",
    "        df_source.to_csv(SOURCEDATADIR %OutFigname.replace('.pdf','.tsv'), sep = '\\t')\n",
    "    else:\n",
    "        plt.show()\n",
    "    plt.close()\n",
    "    return None\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [],
   "source": [
    "plot_ORF10_RPF(hpi_list = ['16h','24h','36h'],\n",
    "               assay_prefix = 'RPFSeq', sample_prefix = 'RPF', as_RPM = True, offset_RPF = 12,\n",
    "               as_log10= True, ylim= 2.5, OutFigname='SFig9b_RPF_ORF10.pdf')\n",
    "plot_ORF10_RPF(hpi_list = ['16h','24h','36h'],\n",
    "               assay_prefix = 'RPFSeq', sample_prefix = 'QTI', as_RPM = True, offset_RPF = 12,\n",
    "               as_log10= True, ylim= 2.5, OutFigname='SFig9c_QTI_ORF10.pdf')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [],
   "source": [
    "def triplet_periodicity_ORF(samplename,assay_prefix = 'RPFSeq',genome = 'SARSCOV2', ORF = 'ORF10',\n",
    "                            offset=12, target_rlen = None,\n",
    "                            offset_from_5p = True, verbose=False, return_codon1pos1 = False):\n",
    "    \n",
    "    if genome == 'SARSCOV2':\n",
    "        BAMDIR = BAMDIR_cov2mapped\n",
    "    else:\n",
    "        BAMDIR = BAMDIR_hostmapped\n",
    "        \n",
    "    if genome == 'hg38':\n",
    "        refdir = REFFLAT_hg38 \n",
    "    elif genome == 'SARSCOV2':\n",
    "        refdir = REFFLAT_SARSCOV2 \n",
    "    elif genome == 'chlSab2':\n",
    "        refdir = REFFLAT_chlSab2\n",
    "    else:\n",
    "        return None\n",
    "    \n",
    "    chrdict = get_dict_refgenes(refdir)\n",
    "    \n",
    "    bam = pysam.AlignmentFile(BAMDIR %(assay_prefix,samplename), 'rb')\n",
    "    triplet_counts = np.zeros(3,dtype=np.float32)\n",
    "    codon1pos1 = 0.0\n",
    "    for sChrID_strand,refgene_list in chrdict.items():\n",
    "        refgene_list = [refgene for refgene in refgene_list if refgene.sGeneSym == ORF]\n",
    "        assert len(refgene_list)<2\n",
    "        sChrID     = sChrID_strand[:-1]\n",
    "        sStrandDir = sChrID_strand[-1]\n",
    "        gene_negstrand = (sStrandDir == '-')\n",
    "        \n",
    "        for refgene in refgene_list:\n",
    "            len_ORF = refgene.nExonlen - refgene.nU5len - refgene.nU3len\n",
    "            if len_ORF%3 != 0:\n",
    "                continue\n",
    "            exon_s_list = refgene.nExonStartlist\n",
    "            exon_e_list = refgene.nExonEndlist\n",
    "            arr_RPF     = np.zeros(refgene.nExonlen,dtype=np.float32)\n",
    "            offset_arr  = 0\n",
    "            for exon_s, exon_e in zip(exon_s_list,exon_e_list):\n",
    "                for read in bam.fetch(sChrID, exon_s,exon_e):\n",
    "                    if target_rlen != None:\n",
    "                        rlen = read.infer_read_length()\n",
    "                        if rlen!=target_rlen:\n",
    "                            continue\n",
    "                    if read.is_reverse != gene_negstrand:\n",
    "                        continue\n",
    "                    reference_positions = read.get_reference_positions(full_length=False)    \n",
    "                    if None in reference_positions:\n",
    "                        print('None in read reference position?')\n",
    "                        print(reference_positions)\n",
    "                        print(read.cigarstring)\n",
    "                        return None\n",
    "                    if gene_negstrand:\n",
    "                        if offset_from_5p == True:\n",
    "                            refpos   = reference_positions[-(offset+1)]\n",
    "                        else:\n",
    "                            refpos   = reference_positions[offset]\n",
    "                    else:\n",
    "                        if offset_from_5p == True:\n",
    "                            refpos   = reference_positions[offset]\n",
    "                        else:\n",
    "                            refpos   = reference_positions[-(offset+1)]\n",
    "                    \n",
    "                    nh = read.get_tag('NH')\n",
    "                    if (exon_s <= refpos) and (refpos < exon_e):\n",
    "                        arr_RPF[refpos-exon_s+offset_arr] += (1/nh)\n",
    "                offset_arr += (exon_e - exon_s)\n",
    "                \n",
    "            assert offset_arr == len(arr_RPF)\n",
    "            if sChrID == '-':\n",
    "                arr_RPF = arr_RPF[::-1]\n",
    "            arr_RPF_ORF = arr_RPF[refgene.nU5len:]\n",
    "            if refgene.nU3len>0: \n",
    "                arr_RPF_ORF = arr_RPF_ORF[:-refgene.nU3len]\n",
    "            codon1pos1 += arr_RPF_ORF[0]\n",
    "            \n",
    "            to_add = arr_RPF_ORF.reshape((len_ORF//3,3)).sum(axis=0)\n",
    "            triplet_counts += to_add\n",
    "\n",
    "    triplet_periodicity = triplet_counts/(triplet_counts.sum())\n",
    "    \n",
    "    if verbose:\n",
    "        print(samplename,genome, triplet_counts.sum(), triplet_periodicity, sep = '\\t')\n",
    "    bam.close()\n",
    "    if return_codon1pos1:\n",
    "        return triplet_counts, triplet_periodicity, codon1pos1\n",
    "    else:\n",
    "        return triplet_counts, triplet_periodicity\n",
    "    \n",
    "def calc_plot_periodicity_ORF(assay_prefix = 'RPFSeq', sample_prefix='RPF', load_precalc=False, ORF='ORF10',\n",
    "                              OutF='RPF_ORF10_triplet.merged.tsv', OutFigname=''):\n",
    "    hpi_list = ['%dh' %i for i in [24,36,48]]+['48hCaco2'] ##read sum >100\n",
    "    df_merged  = []\n",
    "    if load_precalc:\n",
    "        df_merged = pd.read_csv(RESULTDIR %(OutF),sep='\\t')\n",
    "    else:\n",
    "        for hpi in hpi_list:\n",
    "            print(hpi,time.ctime())\n",
    "            for genome in ['SARSCOV2']:\n",
    "                merged_counts = np.zeros(3,dtype=np.float32)\n",
    "                reps = 3 if (hpi == '48h') else 2\n",
    "                for rep in range(1, 1+reps):\n",
    "                    samplename = f'{sample_prefix}_{hpi}_rep{rep}'\n",
    "                    triplet_counts, triplet_periodicity = triplet_periodicity_ORF(samplename, assay_prefix = assay_prefix, ORF = ORF,\n",
    "                                                                                  genome = genome, offset=12, target_rlen = None,\n",
    "                                                                                  offset_from_5p = True, verbose=True, \n",
    "                                                                                  return_codon1pos1 = False)\n",
    "                    merged_counts += triplet_counts\n",
    "\n",
    "                df_merged.append([hpi,genome,*merged_counts,*(merged_counts/merged_counts.sum()*100)])\n",
    "\n",
    "        df_merged  = pd.DataFrame(df_merged)\n",
    "        df_merged.columns = ['hpi','genome','pos1','pos2','pos3','pct1','pct2','pct3']\n",
    "        df_merged.to_csv(RESULTDIR %(OutF), sep='\\t',index=False)   \n",
    "    ###########\n",
    "    df_merged = df_merged.set_index('genome')\n",
    "    \n",
    "    df_merged = df_merged[df_merged['hpi'].isin(hpi_list)]\n",
    "    df_merged.index = df_merged.index.str.replace('hg38','host')\n",
    "    df_merged.index = df_merged.index.str.replace('chlSab2','host')\n",
    "    fig = plt.figure(figsize=(len(hpi_list)+2,4))\n",
    "    ax = fig.add_subplot(111)\n",
    "    x = np.arange(len(hpi_list))\n",
    "    bar_width=0.8/3\n",
    "    xpos_offset = -bar_width*1.5\n",
    "    colors = {}\n",
    "    \n",
    "    colors['SARSCOV2'] = ['#ff6666','#fc8d59','#fdcc8a']\n",
    "    \n",
    "    for genome in ['SARSCOV2']:\n",
    "        for i in range(3):\n",
    "            pct = f'pct{i+1}'\n",
    "            val = df_merged.loc[genome,pct]\n",
    "            ax.bar(x+xpos_offset,val,width = bar_width,label = f'{genome} nt {i+1}', color= colors[genome][i])\n",
    "            xpos_offset += bar_width\n",
    "    \n",
    "    ax.set_xticks(x)\n",
    "    ax.set_xticklabels(hpi_list)\n",
    "    ax.set_xlabel('hpi')\n",
    "    ax.set_xlim((-1,len(hpi_list)))\n",
    "    ax.set_ylabel('Fraction of reads (%)')\n",
    "    ax.set_ylim(top=70)\n",
    "    \n",
    "    plt.tight_layout()\n",
    "    if OutFigname == '':\n",
    "        plt.show()\n",
    "    else:\n",
    "        plt.savefig(FIGDIR %OutFigname)\n",
    "    plt.close()\n",
    "    \n",
    "    return None"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "24h Mon Sep 27 16:30:41 2021\n",
      "RPF_24h_rep1\tSARSCOV2\t141.0\t[0.65957445 0.17730497 0.16312057]\n",
      "RPF_24h_rep2\tSARSCOV2\t108.0\t[0.6944444  0.18518518 0.12037037]\n",
      "36h Mon Sep 27 16:30:41 2021\n",
      "RPF_36h_rep1\tSARSCOV2\t309.0\t[0.68608415 0.16828479 0.14563107]\n",
      "RPF_36h_rep2\tSARSCOV2\t317.0\t[0.55520505 0.24921136 0.1955836 ]\n",
      "48h Mon Sep 27 16:30:41 2021\n",
      "RPF_48h_rep1\tSARSCOV2\t315.0\t[0.53968257 0.2825397  0.17777778]\n",
      "RPF_48h_rep2\tSARSCOV2\t446.0\t[0.5560538  0.25336322 0.19058296]\n",
      "RPF_48h_rep3\tSARSCOV2\t332.0\t[0.57831323 0.28915662 0.13253012]\n",
      "48hCaco2 Mon Sep 27 16:30:42 2021\n",
      "RPF_48hCaco2_rep1\tSARSCOV2\t224.0\t[0.54910713 0.27232143 0.17857143]\n",
      "RPF_48hCaco2_rep2\tSARSCOV2\t170.0\t[0.5470588  0.24705882 0.20588236]\n",
      "24h Mon Sep 27 16:30:42 2021\n",
      "QTI_24h_rep1\tSARSCOV2\t66.0\t[0.6515151  0.24242425 0.10606061]\n",
      "QTI_24h_rep2\tSARSCOV2\t65.0\t[0.52307695 0.23076923 0.24615385]\n",
      "36h Mon Sep 27 16:30:42 2021\n",
      "QTI_36h_rep1\tSARSCOV2\t97.0\t[0.52577317 0.20618556 0.26804122]\n",
      "QTI_36h_rep2\tSARSCOV2\t127.0\t[0.43307087 0.2913386  0.27559054]\n",
      "48h Mon Sep 27 16:30:42 2021\n",
      "QTI_48h_rep1\tSARSCOV2\t490.0\t[0.4489796  0.30816326 0.24285714]\n",
      "QTI_48h_rep2\tSARSCOV2\t572.0\t[0.4090909  0.37587413 0.21503496]\n",
      "QTI_48h_rep3\tSARSCOV2\t165.0\t[0.3272727  0.35757574 0.3151515 ]\n",
      "48hCaco2 Mon Sep 27 16:30:43 2021\n",
      "QTI_48hCaco2_rep1\tSARSCOV2\t189.0\t[0.6296296  0.20634921 0.16402116]\n",
      "QTI_48hCaco2_rep2\tSARSCOV2\t140.0\t[0.47857141 0.23571429 0.2857143 ]\n"
     ]
    }
   ],
   "source": [
    "calc_plot_periodicity_ORF(assay_prefix = 'RPFSeq', sample_prefix='RPF', load_precalc=False, ORF='ORF10',\n",
    "                              OutF='RPF_ORF10_triplet.merged.tsv', OutFigname='SFig9d_ORF10_triplet_RPF.pdf')\n",
    "calc_plot_periodicity_ORF(assay_prefix = 'RPFSeq', sample_prefix='QTI', load_precalc=False, ORF='ORF10',\n",
    "                              OutF='QTI_ORF10_triplet.merged.tsv', OutFigname='SFig9d_ORF10_triplet_QTI.pdf')"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Doyeon-tf2",
   "language": "python",
   "name": "doyeon-tf2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.10"
  },
  "toc-autonumbering": false,
  "toc-showmarkdowntxt": false
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
